<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Adding new datasets into AusTraits • austraits.build</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Adding new datasets into AusTraits">
<meta property="og:description" content="austraits.build">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">austraits.build</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.9.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
    home
  </a>
</li>
<li>
  <a href="../articles/austraits.build.html">Get started</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fas fa-book"></span>
     
    Details
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/austraits_overview.html">Overview of `AusTraits` workflow</a>
    </li>
    <li>
      <a href="../articles/austraits.build.html">Getting started with austraits.build</a>
    </li>
    <li>
      <a href="../articles/austraits_database_structure.html">Structure of `AusTraits` data compilation</a>
    </li>
    <li>
      <a href="../articles/trait_definitions.html">Definitions for traits in `AusTraits`</a>
    </li>
    <li>
      <a href="../articles/austraits_file_structure.html">Repository file structure</a>
    </li>
    <li>
      <a href="../articles/contributing_data.html">Contributing data to `AusTraits`</a>
    </li>
    <li>
      <a href="../articles/adding_data.html">Adding new datasets into `AusTraits`</a>
    </li>
    <li>
      <a href="../articles/Docker.html">Docker for reproducible compute environment</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">
    <span class="fa fa-list"></span>
     
    Functions
  </a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/traitecoevo/austraits.build" class="external-link">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Adding new datasets into AusTraits</h1>
            
            <h4 data-toc-skip class="date">2023-02-09</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/traitecoevo/austraits.build/blob/HEAD/vignettes/adding_data.Rmd" class="external-link"><code>vignettes/adding_data.Rmd</code></a></small>
      <div class="hidden name"><code>adding_data.Rmd</code></div>

    </div>

    
    
<p>This vignette explains the protocol for adding a new study to
AusTraits. Before starting this, you should read more about</p>
<ul>
<li>an <a href="http://traitecoevo.github.io/austraits.build/articles/austraits_overview.html">overview
of AusTraits</a>,</li>
<li>the instructions provided to <a href="http://traitecoevo.github.io/austraits.build/articles/contributing_data.html">data
contributors</a>,</li>
<li>the <a href="http://traitecoevo.github.io/austraits.build/articles/austraits_database_structure.html">structure
of the compiled AusTraits database</a>,</li>
<li>the <a href="http://traitecoevo.github.io/austraits.build/articles/austraits_file_structure.html">structure
of the raw data files</a>, and</li>
<li>
<a href="http://traitecoevo.github.io/austraits.build/articles/austraits.build.html">how
to build AusTraits</a>.</li>
</ul>
<p>It is important that all steps are followed so that our automated
workflow proceeds without problems.</p>
<div class="section level3">
<h3 id="an-overview-of-the-main-steps">An overview of the main steps<a class="anchor" aria-label="anchor" href="#an-overview-of-the-main-steps"></a>
</h3>
<ol style="list-style-type: decimal">
<li>Clone the <code>austraits.build</code> repository from github</li>
<li>Create a new branch in the repo, named for the new
<code>dataset_id</code> in <code>author_year</code> format,
e.g. <code>Gallagher_2014</code>.</li>
<li>Create a new folder within the folder <code>data</code> with the
name <code>dataset_id</code>, e.g. <code>Gallagher_2014</code>.</li>
<li>Prepare the file <code>data.csv</code> and place it within the new
folder (<a href="#csv_file">details</a> here).</li>
<li>Prepare the file <code>metadata.yml</code> and place it within the
new folder (<a href="#metadata_file">details</a> here).</li>
<li>Add the new study into the build framework and rebuild AusTraits, by
running <code><a href="../reference/build_setup_pipeline.html">build_setup_pipeline()</a></code>.</li>
</ol>
<p>This step updates the file <code>remake.yml</code> with appropriate
rules for the new dataset; similarly if you remove datasets, do the
same. (At this stage, <a href="https://github.com/richfitz/remake" class="external-link">remake</a> offers no looping
constructs, so for now we generate the remake file using: <a href="https://github.com/edwindj/whisker" class="external-link">whisker</a>.)</p>
<p>You can then rebuild AusTraits, including your dataset.</p>
<ol start="7" style="list-style-type: decimal">
<li>Run tests and quality checks on the newly added dataset and correct
the <code>data.csv</code> and <code>metadata.yml</code> files as
necessary (<a href="#quality_checks">details</a> here).</li>
<li>Generate and proofread a report on the data. In particular, check
that numeric trait values fall within a logical range relative to other
studies, and that individual trait observations are not unnecessarily
excluded because their trait values are unsupported.</li>
<li>Return to step 6 if changes are made to the <code>data.csv</code> or
<code>metadata.yml</code> files.</li>
<li>Push the GitHub branch.</li>
</ol>
<p>It may help to download one of the <a href="https://github.com/traitecoevo/austraits.build/tree/master/data" class="external-link">existing
datasets</a> to use as a template for your own files and a guide on
required content. You should look at the files in the <a href="https://github.com/traitecoevo/austraits.build/tree/master/config" class="external-link">config
folder</a>, particularly the <code>definitions</code> file for the list
of traits we cover and the supported trait values for each trait. The
GitHub repository also hosts a compiled <a href="http://traitecoevo.github.io/austraits.build/articles/trait_definitions.html">trait
definitions table</a>.</p>
<p>The remainder of this vignette provides incredibly detailed
instructions for steps 4-8 above. It is intended for anyone wishing to
add datasets to either AusTraits itself or to use the austraits.build
workflow to create a separate database.</p>
</div>
<div class="section level2">
<h2 id="getting-started">Getting started<a class="anchor" aria-label="anchor" href="#getting-started"></a>
</h2>
<p>The <code>austraits.build</code> repository includes a selection of
functions that help build the repository. To use these, you’ll need to
make them available.</p>
<p>The easiest way to load the functions into your workspace is to run
the following (from within the repository)</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/source.html" class="external-link">source</a></span><span class="op">(</span><span class="st">"scripts/custom.R"</span><span class="op">)</span>            <span class="co"># source functions written for use within custom_R_code</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://traitecoevo.github.io/austraits.build/">austraits.build</a></span><span class="op">)</span>              <span class="co"># open the austraits.build package that provides the pipeline to build AusTraits</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="inputting-data">Inputting data<a class="anchor" aria-label="anchor" href="#inputting-data"></a>
</h2>
<div class="section level3">
<h3 id="add-a-new-folder">Add a new folder<a class="anchor" aria-label="anchor" href="#add-a-new-folder"></a>
</h3>
<p>Add a new folder within the <code>data</code> folder. Its name should
be the study’s <code>dataset_id</code>, the core organising unit behind
AusTraits.</p>
<p>The preferred format for <code>dataset_id</code> is the surname of
the first author of any corresponding publication, followed by the year,
as <code>surname_year</code>. E.g. <code>Falster_2005</code>. Wherever
there are multiple studies with the same id, we add a suffix
<code>_2</code>, <code>_3</code> etc. E.g.<code>Falster_2005</code>,
<code>Falster_2005_2</code>.</p>
</div>
<div class="section level3">
<h3 id="csv_file">Constructing the <code>data.csv</code> file<a class="anchor" aria-label="anchor" href="#csv_file"></a>
</h3>
<p>All data for a study (<code>dataset_id</code>) must be merged into a
single spreadsheet: <code>data.csv</code>. All accompanying metadata is
read in through the <code>metadata.yml</code> file. Some information
must be input explicitly through the <code>data.csv</code> or
<code>metdata.yml</code> file, while other information can be entered
via either file; this is explicitly indicated for each element.</p>
<ol style="list-style-type: decimal">
<li>
<strong>Required columns:</strong> Columns within the
<code>data.csv</code> file must include <code>taxon name</code>,
<code>location_name</code> (if there are multiple locations),
<code>contexts</code> (if appropriate), and <code>collection_date</code>
(if appropriate). The <code>data.csv</code> file can either be in a wide
format (1 column for each trait, with <code>trait name</code> as the
column header) or long format (a single column for all
<code>trait values</code> and additional columns for
<code>trait name</code> and <code>units</code>)</li>
</ol>
<ol style="list-style-type: lower-alpha">
<li><p>For all field studies, ensure there is a column for
<code>location_name</code>. If all measurements were made at a single
location, a <code>location_name</code> column can easily be mutated
using <a href="#custom_R">custom_R_code</a> within the metadata.yml
file. See sections <a href="#adding_locations">adding locations</a> and
<a href="#adding_contexts">adding contexts</a> below for more
information on compiling location and context data.</p></li>
<li><p>If available, be sure to include a column with
<code>collection date</code>. If possible, provide in
<code>yyyy-mm-dd</code> (e.g. 2020-03-05) format or, if the day of the
month isn’t known, as <code>yyyy-mm</code> (e.g. 2020-03). However, any
format is allowed and the column can be parsed to the proper yyyy-mm-dd
format using custom_R_code. If the same <code>collection date</code>
applies to the entire study it can be added directly into the
metadata.yml file.</p></li>
<li><p>If applicable, ensure there are columns for an context
properties, including experimental treatments, specific differences in
method, a stratified sampling scheme within a plot, or sampling season.
Additional context columns could be added through
<code>custom_R_code</code> or keyed in where traits are added, but it is
best to include a column in the data.csv file whenever possible. The
protocol for adding context properties to the metadata file is under <a href="#adding_contexts">adding contexts</a></p></li>
</ol>
<ol start="2" style="list-style-type: decimal">
<li>
<strong>Summarising data:</strong> Data submitted by a contributor
should be in the rawest form possible; always request data with
individual measurements over location/species means. Some studies make
replicate measurements on an individual at a single point in time. For
these studies, individual means need to be calculated, as AusTraits does
not include multiple measurements per individual. The raw values are
preserved in the contributor’s raw data files. Be sure to calculate the
number of replicates that contributed to each mean value.</li>
</ol>
<p>When there is just a single row of values to summarise, use:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_csv</a></span><span class="op">(</span><span class="st">"data/dataset_id/raw/raw_data.csv"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>leaf_area_replicates <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">individual</span>, <span class="va">`species name`</span>, <span class="va">location</span>, <span class="va">context</span>, <span class="va">etc</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span></span>
<span>    leaf_area_mean <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">leaf_area</span><span class="op">)</span>,</span>
<span>    leaf_area_replicates <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">leaf_area_replicates</span><span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">ungroup</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>(Make sure you <code>group_by</code> all categorical variables you
want to retain, for only columns that are grouping variables will be
kept)</p>
<p>When you want to take the mean of a series of continuous variables,
use:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_csv</a></span><span class="op">(</span><span class="st">"data/dataset_id/raw/raw_data.csv"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>replicates <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">individual</span>, <span class="va">`species name`</span>, <span class="va">location</span>, <span class="va">context</span>, <span class="va">etc</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/across.html" class="external-link">across</a></span><span class="op">(</span></span>
<span>      <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">leaf_area</span>, <span class="va">`leaf N`</span><span class="op">)</span>, .fns <span class="op">=</span> <span class="va">mean</span>,</span>
<span>      <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">replicates</span><span class="op">)</span>, .fns <span class="op">=</span> <span class="va">sum</span>,</span>
<span>      <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">growth_form</span>, <span class="va">`photosynthetic pathway`</span><span class="op">)</span>, .fns <span class="op">=</span> <span class="va">first</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">ungroup</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<ul>
<li><p>Categorical variables not included as grouping variables will
return <code>NA</code></p></li>
<li><p>This allows you to retain character variables, but can be tedious
with many columns. Generally use the function <code>first</code> for
categorical variables - it simply retains the trait value in the first
column. In the rare case when rows in a particular grouping have
different categorical values, more complex manipulations are
required.</p></li>
<li><p>You can identify runs of columns by column number/position. For
instance <code>c(5:25), .fns = mean</code> or
<code>c(leaf_area:leaf_N), .fns = mean</code></p></li>
</ul>
<ol start="3" style="list-style-type: decimal">
<li>
<strong>Merging multiple spreadsheets:</strong> If multiple
spreadsheets of data are submitted these must be merged together.</li>
</ol>
<ol style="list-style-type: lower-alpha">
<li>If the spreadsheets include different trait measurements made on the
same individual (or location means for the same species), they are best
merged using <code>full_join</code>, specifying all conditions that need
to be matched across spreadsheets (e.g. individual, species, location,
context). Ensure the column names are identical between spreadsheets or
specify columns that need to be matched.</li>
</ol>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_csv</a></span><span class="op">(</span><span class="st">"data/dataset_id/raw/data_file_1.csv"</span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">data_1</span></span>
<span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_csv</a></span><span class="op">(</span><span class="st">"data/dataset_id/raw/data_file_2.csv"</span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">data_2</span></span>
<span><span class="va">data_1</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">full_join</a></span><span class="op">(</span><span class="va">data_2</span>, by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Individual"</span>, <span class="st">"Taxon"</span>, <span class="st">"Location"</span>, <span class="st">"Context"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<ol start="2" style="list-style-type: lower-alpha">
<li>If the spreadsheets include trait measurements for different
individuals (or possibly data at different scales - such as individual
level data for some traits and species means for other traits), they are
best merged using <code>bind_rows</code>. Ensure the column names for
taxon name, location name, context, individual, and collection date are
identical between spreadsheets. If there are data for the same traits in
both spreadsheets, make sure those column headers are identical as
well.</li>
</ol>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_csv</a></span><span class="op">(</span><span class="st">"data/dataset_id/raw/data_file_1.csv"</span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">data_1</span></span>
<span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_csv</a></span><span class="op">(</span><span class="st">"data/dataset_id/raw/data_file_2.csv"</span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">data_2</span></span>
<span><span class="va">data_1</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind.html" class="external-link">bind_rows</a></span><span class="op">(</span><span class="va">data_2</span><span class="op">)</span></span></code></pre></div>
<ol start="4" style="list-style-type: decimal">
<li>
<strong>Taxon names:</strong> Taxon names need to be complete names.
If the main data file includes code names, with a key as a separate
file, they need to be merged:</li>
</ol>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_csv</a></span><span class="op">(</span><span class="st">"data/dataset_id/raw/species_key.csv"</span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">species_key</span></span>
<span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_csv</a></span><span class="op">(</span><span class="st">"data/dataset_id/raw/data_file.csv"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span><span class="va">species_key</span>, by <span class="op">=</span> <span class="st">"code"</span><span class="op">)</span></span></code></pre></div>
<div class="section level4">
<h4 id="unexpected-hangups">Unexpected hangups<a class="anchor" aria-label="anchor" href="#unexpected-hangups"></a>
</h4>
<ul>
<li>When Excel saves an <code>.xls</code> file as a <code>.csv</code>
file it only preserves the number of significant figures that are
displayed on the screen. This means that if, for some reason, a column
has been set to display a very low number of significant figures or a
column is very narrow, data quality is lost.<br>
</li>
<li>If you’re reading a file into R where there are lots of blanks at
the beginning of a column of numeric data, the defaults for
<code>read_csv</code> fail to register the column as numeric. It is
fixed by adding the argument <code>guess_max</code>:</li>
</ul>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_csv</a></span><span class="op">(</span><span class="st">"data/dataset_id/raw/raw_data.csv"</span>, guess_max <span class="op">=</span> <span class="fl">10000</span><span class="op">)</span></span></code></pre></div>
<p>This checks 10,000 rows of data before declaring the column is
non-numeric. The value can be set even higher…</p>
</div>
</div>
<div class="section level3">
<h3 id="metadata_file">Constructing the <code>metadata.yml</code> file<a class="anchor" aria-label="anchor" href="#metadata_file"></a>
</h3>
<p>One way to construct the <code>metadata.yml</code> file is to use one
of the existing files and modify yours to follow the same format. As a
start, check out some examples from <a href="https://github.com/traitecoevo/austraits.build/tree/master/data" class="external-link">existing
studies in AusTraits</a>, e.g. <a href="https://github.com/traitecoevo/austraits.build/tree/master/data/Angevin_2011/metadata.yml" class="external-link">Angevin_2010</a>
or <a href="https://github.com/traitecoevo/austraits.build/tree/master/data/Wright_2009/metadata.yml" class="external-link">Wright_2009</a>.</p>
<p>Note, when editing the <code>metadata.yml</code>, edits should be
made in a proper text editor (Microsoft word tends to mess up the
formatting). For example, Rstudio, textmate, sublime text, and Visual
Studio Code are all good editors.</p>
<p>To assist you in constructing the <code>metadata.yml</code> file, we
have developed functions to help fill in the different sections of the
file. You can then manually edit the file further to fill in missing
details.</p>
<p>First run the following to make the functions available</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://traitecoevo.github.io/austraits.build/">austraits.build</a></span><span class="op">)</span></span></code></pre></div>
<p>The functions for populating the metadata file all begin with
<code>metadata_</code>. A list of the available functions is
automatically generated within the <code>man/</code> folder within the
<code>austraits.build</code> directory.</p>
<div class="section level4">
<h4 id="creating-a-template">Creating a template<a class="anchor" aria-label="anchor" href="#creating-a-template"></a>
</h4>
<p>Create a basic template for the <code>metadata.yml</code> file for
your study. Note, it requires you to have already created a file
<code>data.csv</code> in the folder
<code>data/your_dataset_id</code>.</p>
<p>Let’s imagine you’re entering a study called
<code>Yang_2028</code></p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">current_study</span> <span class="op">&lt;-</span> <span class="st">"Yang_2028"</span></span>
<span></span>
<span><span class="fu"><a href="../reference/metadata_create_template.html">metadata_create_template</a></span><span class="op">(</span><span class="va">current_study</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># or simply</span></span>
<span></span>
<span><span class="fu"><a href="../reference/metadata_create_template.html">metadata_create_template</a></span><span class="op">(</span><span class="st">"Yang_2028"</span><span class="op">)</span></span></code></pre></div>
<p>The function will ask a series of questions and then create a
relatively empty file <code>data/your_dataset_id/metadata.yml</code>.
The key questions are:</p>
<ul>
<li><p>Is the data long vs wide? A wide dataset has each variable
(i.e. trait ) as a column. A long dataset has a single row containing
all trait values and additional columns specifying <code>units</code>
and <code>trait_name</code>.</p></li>
<li><p>Select column for <code>taxon_name</code><br></p></li>
<li><p>Select column for <code>trait_name</code> (long datasets
only)</p></li>
<li><p>Select column for <code>trait values</code> (long datasets
only)</p></li>
<li><p>Select column for <code>location_name</code></p></li>
<li><p>Select column for <code>individual_id</code> (a column that links
measurements on the same individual)</p></li>
<li><p>Select column for <code>collection_date</code></p></li>
</ul>
<p>If your <code>data.csv</code> file does not yet have a
<code>location_name</code> column, this information can later be added
manually.</p>
</div>
<div class="section level4">
<h4 id="adding-a-source">Adding a source<a class="anchor" aria-label="anchor" href="#adding-a-source"></a>
</h4>
<p>Three functions are available to help with entering citation details
for the source data.</p>
<p>The function <code>metadata_create_template</code> creates a template
for the primary source with default fields for a journal article, which
you can then edit manually.</p>
<p>If you have a <code>doi</code> for your study, use the function:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/metadata_add_source_doi.html">metadata_add_source_doi</a></span><span class="op">(</span>dataset_id <span class="op">=</span> <span class="va">current_study</span>, doi <span class="op">=</span> <span class="st">"doi"</span><span class="op">)</span></span></code></pre></div>
<p>and the different elements within the source will automatically be
generated. Double check the information added to ensure:<br>
1. The title is in <code>sentence case</code><br>
2. Overall, the information isn’t in <code>all caps</code> (information
from a few journals is read in like this)<br>
3. Pages numbers are present and added as, for example,
<code>123 -- 134</code> ; note the <code>--</code> between page
numbers</p>
<p>By default, details are added as the primary source. If multiple
sources are linked to a single <code>dataset_id</code>, you can specify
a source as <code>secondary</code>. Attempting to add a second primary
source will overwrite the information already input.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/metadata_add_source_doi.html">metadata_add_source_doi</a></span><span class="op">(</span><span class="va">dataset_id</span>, <span class="va">doi</span>, type <span class="op">=</span> <span class="st">"secondary"</span><span class="op">)</span></span></code></pre></div>
<ul>
<li>Secondary sources will be assigned the same dataset_id as the
primary source. Manually edit the <code>key</code> in the metadata.yml
file to be the appropriate <code>author_yyyy</code> code for the
secondary reference. Sequential qualifiers can be used if necessary
(e.g. <code>author_yyyy_2</code>)</li>
<li>A “secondary” source might be either a second research output from
the main dataset (truly a <code>secondary</code> source) or the original
source of some data compiled for a metaanalysis (an
<code>original</code> source). After adding a second source, you must
manually change the source’s header to beginning with either
<code>secondary</code> or <code>original</code>, as is appropriate.</li>
<li>If there are many sources to add (i.e. for datasets compiled for
metaanalyses), after you add each reference, go to the
<code>metadata.yml</code> file and manually change the source’s header
from <code>original</code> to <code>original_01</code> (and then
<code>original_02</code>, etc.; or <code>secondary_01</code> ;
<code>secondary_02</code>). See <a href="https://github.com/traitecoevo/austraits.build/blob/master/data/Richards_2008/metadata.yml" class="external-link">Richards_2008</a>
for an example of a complex source list.</li>
</ul>
<p>Alternatively, if you have reference details saved in a bibtex file
called <code>myref.bib</code> you can use the function</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/metadata_add_source_doi.html">metadata_add_source_doi</a></span><span class="op">(</span><span class="va">dataset_id</span>, file <span class="op">=</span> <span class="st">"myref.bib"</span><span class="op">)</span></span></code></pre></div>
<p>(These options require the packages <a href="https://github.com/ropensci/rcrossref" class="external-link">rcrossref</a> and <a href="https://github.com/ropensci/RefManageR/" class="external-link">RefManageR</a> to be
installed.)</p>
<p>For a book, the proper format is:</p>
<pre><code>source:
  primary:
      key: Cooper_2013
      bibtype: Book
      year: 2013
      author: Wendy Cooper and William T. Cooper
      title: Australian rainforest fruits
      publisher: CSIRO Publishing
      pages: 272</code></pre>
<p>For an online resource, the proper format is:</p>
<pre><code>source:
  primary:
    key: TMAG_2009
    bibtype: Online
    author: '{Tasmanian Herbarium}'
    year: 2009
    title: Flora of Tasmania Online
    publisher: Tasmanian Museum &amp; Art Gallery (Hobart)
    url: http://www.tmag.tas.gov.au/floratasmania</code></pre>
<p>For a thesis, the proper format is:</p>
<pre><code>source:
  primary:   
      key: Kanowski_2000   
      bibtype: Thesis  
      year: 1999  
      author: John Kanowski  
      title: Ecological determinants of the distribution and abundance of the folivorous
        marsupials endemic to the rainforests of the Atherton uplands, north Queensland.  
      type: PhD  
      institution: James Cook University, Townsville</code></pre>
<p>For an unpublished dataset, the proper format is:</p>
<pre><code>source:
  primary:
    key: Ooi_2018
    bibtype: Unpublished
    year: 2018
    author: Mark K. J. Ooi
    title: "Unpublished data: Herbivory survey within Royal National Park, University
      of New South Wales"</code></pre>
<p>If you manually add information, note that if there is a colon (:) or
apostrophe (’) in a reference, the text for that line must be in quotes
(“).</p>
</div>
<div class="section level4">
<h4 id="adding-contributors">Adding contributors<a class="anchor" aria-label="anchor" href="#adding-contributors"></a>
</h4>
<p>The skeletal <code>metadata.yml</code> file created by the function
<code>metadata_create_template</code> includes a template for entering
details about data contributors. Edit this manually, duplicating if
details for multiple people are required.</p>
<ul>
<li>Authorship is extended to anyone who played a key intellectual role
in the experimental design and data collection. Most studies have 1-3
authors. For each author, please provide a <strong>last_name</strong>,
<strong>given_name</strong>, <strong>institutional affiliation</strong>,
<strong>email address</strong>, and their <strong>ORCID</strong> (if
available). Nominate a single contributor to be the dataset’s point of
contact; this person’s email will not be listed in the metadata file,
but is the person future AusTraits users are likely to seek out if they
have questions.</li>
<li>Additional field assistants can be listed under
<code>assistants:</code>
</li>
<li>The AusTraits data entry person is listed under
<code>austraits_curators:</code>
</li>
</ul>
<p>For example, in Roderick_2002</p>
<pre><code>contributors:
  data_collectors:
  - last_name: Roderick
    given_name: Michael
    ORCID: 0000-0002-3630-7739
    affiliation: The Australian National University, Australia
    additional_role: contact
  assistants: Michelle Cochrane
  austraits_curators: Elizabeth Wenk</code></pre>
</div>
<div class="section level4">
<h4 id="custom_R">Custom R code<a class="anchor" aria-label="anchor" href="#custom_R"></a>
</h4>
<p>For many studies there are changes we want to make to a dataset
before the data.csv file is read into AusTraits. These most often
include applying a function to transform data, a function to filter
data, or a function to replace a contributor’s “measurement missing”
placeholder symbol with <code>NA</code>. In each case it is appropriate
to leave the rawer data in <code>data.csv</code>.</p>
<div class="section level5">
<h5 id="background">Background<a class="anchor" aria-label="anchor" href="#background"></a>
</h5>
<p>In each case we want to make some custom modifications to a
particular dataset before the common pipeline of operations gets
applied. To make this possible, the workflow allows for some custom R
code to be run as a first step in the processing pipeline. That pipeline
(the function <code>process_custom_code</code> called within <a href="https://github.com/traitecoevo/austraits.build/blob/master/R/process.R" class="external-link"><code>dataset_process</code></a>)
looks like this:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_csv</a></span><span class="op">(</span><span class="va">filename_data_raw</span>, col_types <span class="op">=</span> <span class="fu"><a href="https://readr.tidyverse.org/reference/cols.html" class="external-link">cols</a></span><span class="op">(</span><span class="op">)</span>, guess_max <span class="op">=</span> <span class="fl">1e5</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/process_custom_code.html">process_custom_code</a></span><span class="op">(</span><span class="va">metadata</span><span class="op">[[</span><span class="st">"dataset"</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="st">"custom_R_code"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/process_parse_data.html">process_parse_data</a></span><span class="op">(</span><span class="va">dataset_id</span>, <span class="va">metadata</span><span class="op">)</span></span></code></pre></div>
<p>Note the second line. This is where the custom code gets applied,
right after the file is loaded.</p>
</div>
<div class="section level5">
<h5 id="summary">Summary<a class="anchor" aria-label="anchor" href="#summary"></a>
</h5>
<ul>
<li>source the file containing functions the AusTraits team have
explicitly developed to use within the custom_R_code field:
<code>source("scripts/custom.R")</code>
</li>
<li>assume a single object called <code>data</code>, and apply whatever
fixes are needed<br>
</li>
<li>use functions from the packages <a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a> or <a href="https://tidyr.tidyverse.org" class="external-link">tiydr</a>, like <code>mutate</code>,
<code>rename</code>, etc, and otherwise avoid external packages<br>
</li>
<li>alternatively, use the functions we’ve created explicitly for
pre-processing data that were sourced through the file
<code>custom.R</code>. In consultation with AusTraits team leaders you
can add functions to this file.</li>
<li>be fully self-contained (we’re not going to use any of the other
remake machinery here)<br>
</li>
<li>use pipes to weave together a single statement, where possible.
(Otherwise you’ll need a semi colons <code>;</code> at the end of each
statement).<br>
</li>
<li>place a single apostrophe (’) at the start and end of your custom R
code; this allows you to add line breaks between pipes.</li>
</ul>
</div>
<div class="section level5">
<h5 id="examples-of-appropriate-use-of-custom-r-code">Examples of appropriate use of custom R code<a class="anchor" aria-label="anchor" href="#examples-of-appropriate-use-of-custom-r-code"></a>
</h5>
<ol style="list-style-type: decimal">
<li>
<p>Most sources from herbaria record <code>flowering_time</code> and
<code>fruiting_time</code> as a span of months, while AusTraits codes
these variables as a sequence of 12 N’s and Y’s for the 12 months. A
series of functions make this conversion in custom_R_code. These
include:</p>
<ul>
<li>‘<code>format_flowering_months</code>’ (Create flowering times from
start to end pair)<br>
</li>
<li>‘<code>convert_month_range_string_to_binary</code>’ (Converts
flowering and fruiting month ranges to 12 element character strings of
binary data)<br>
</li>
<li>‘<code>convert_month_range_vec_to_binary</code>’ (Convert vectors of
month range to 12 element character strings of binary data)<br>
</li>
<li>‘<code>collapse_multirow_phenology_data_to_binary_vec</code>’
(Converts multirow phenology data to a 12 digit binary string)</li>
</ul>
</li>
<li><p>Many datasets from herbaria record traits like
<code>leaf_length</code>, <code>leaf_width</code>,
<code>seed_length</code>, etc. as a range (e.g. <code>2-8</code>). The
function <code>separate_range</code> separates this data into a pair of
columns with <code>minimum</code> and <code>maximum</code> values,
required to properly align units</p></li>
<li><p>Duplicate values within a study need to be filtered out.</p></li>
</ol>
<p>If a species-level measurement has been entered for all
within-location replicates, you need to filter out the duplicates. This
is true for both numeric and categorical values.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">Species</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span></span>
<span>      <span class="fu"><a href="https://dplyr.tidyverse.org/reference/across.html" class="external-link">across</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">`leaf_percentN`</span>, <span class="va">`plant growth form`</span><span class="op">)</span>, <span class="va">replace_duplicates_with_NA</span><span class="op">)</span></span>
<span>      <span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">ungroup</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>Note: You would use <code>group_by(Species, Location)</code> if there
are unique values at the species x location level.</p>
<ol start="4" style="list-style-type: decimal">
<li><p>Values that were sourced from a different study need to be
filtered out. See <a href="#duplicates_between_studies">Duplicates
between studies</a> below - functions to automate this process are in
progress.</p></li>
<li><p>Author has represented missing data values with a symbol, such as
<code>0</code> :</p></li>
</ol>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/across.html" class="external-link">across</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">`height (cm)`</span>, <span class="va">`leaf area (mm2)`</span><span class="op">)</span>, <span class="op">~</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/na_if.html" class="external-link">na_if</a></span><span class="op">(</span><span class="va">.</span>, <span class="fl">0</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<ol start="6" style="list-style-type: decimal">
<li>If a subset of data in a column are also <code>values</code> for a
second trait in AusTraits, some data values can be duplicated in a
second temporary column. In the example below, some data in the
contributor’s <code>fruit_type</code> column <strong>also</strong> apply
to the trait <code>fruit_fleshiness</code> in AusTraits:</li>
</ol>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>fruit_fleshiness <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">`fruit type`</span> <span class="op">==</span> <span class="st">"pome"</span>, <span class="st">"fleshy"</span>, <span class="cn">NA</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<ol start="7" style="list-style-type: decimal">
<li>If a subset of data in a column are <em>instead</em>
<code>values</code> for a second trait in AusTraits, some data values
can be moved to a second column (second trait), using the function
‘<code>move_values_to_new_trait</code>’. In the example below, some data
in the contributor’s <code>growth_form</code> column
<strong>only</strong> apply to the trait <code>parasitic</code> in
AusTraits. Note you need to create a blank variable to move the trait
values to.</li>
</ol>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>new_trait <span class="op">=</span> <span class="va">NA_character</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">move_values_to_new_trait</span><span class="op">(</span></span>
<span>    original_trait<span class="op">=</span> <span class="st">"growth form"</span>, </span>
<span>    new_trait <span class="op">=</span> <span class="st">"parasitic"</span>,</span>
<span>    original_values <span class="op">=</span> <span class="st">"parasitic"</span>,</span>
<span>    values_for_new_trait <span class="op">=</span> <span class="st">"parasitic"</span>,</span>
<span>    values_to_keep <span class="op">=</span> <span class="st">"NA"</span><span class="op">)</span></span></code></pre></div>
<p>or</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>dispersal_appendage <span class="op">=</span> <span class="va">NA.char</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">move_values_to_new_trait</span><span class="op">(</span></span>
<span>    <span class="st">"fruits"</span>, <span class="st">"dispersal_appendage"</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"dry &amp; winged"</span>, <span class="st">"enclosed in aril"</span><span class="op">)</span>, </span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"wings"</span>, <span class="st">"aril"</span><span class="op">)</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"NA"</span>, <span class="st">"enclosed"</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<ul>
<li>Note, the parameter <code>values_to_keep</code> currently doesn’t
accept <code>NA</code> ; this bug is known and will be fixed.</li>
</ul>
<ol start="8" style="list-style-type: decimal">
<li>If the <code>data.csv</code> file includes raw data that you want to
manipulate into a <code>trait</code>, or the contributor presents the
data in a different formulation than AusTraits:</li>
</ol>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>root_mass_fraction <span class="op">=</span> <span class="va">`root mass`</span> <span class="op">/</span> <span class="op">(</span><span class="va">`root mass`</span> <span class="op">+</span> <span class="va">`shoot mass`</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<ol start="9" style="list-style-type: decimal">
<li>You can do manipulations, such as adding a column with
<code>locations</code> or manipulating <code>location names</code>. This
is only recommended for studies with a single (or few) location, where
manually adding the location data to the <code>metadata.yml</code> file
is fast, since in precludes automatically propagating location data into
metadata (see <a href="#Adding%20location%20details">Adding location
details</a>). As an example, see <code>Blackman_2010</code>:</li>
</ol>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span></span>
<span>    location_name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">location_name</span> <span class="op">==</span> <span class="st">"Mt Field"</span> <span class="op">&amp;</span> <span class="va">habitat</span> <span class="op">==</span> <span class="st">"Montane rainforest"</span>, <span class="st">"Mt Field_wet"</span>, <span class="va">location_name</span><span class="op">)</span>,</span>
<span>    location_name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">location_name</span> <span class="op">==</span> <span class="st">"Mt Field"</span> <span class="op">&amp;</span> <span class="va">habitat</span> <span class="op">==</span> <span class="st">"Dry sclerophyll"</span>, <span class="st">"Mt Field_dry"</span>, <span class="va">location_name</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<ol start="10" style="list-style-type: decimal">
<li>You can generate <code>observation_numbers</code> for sequential
measurements on the same individual</li>
</ol>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">Tree</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>observation_number <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/ranking.html" class="external-link">row_number</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">ungroup</a></span><span class="op">(</span><span class="op">)</span> </span></code></pre></div>
<ol start="11" style="list-style-type: decimal">
<li>You can generate<code>measurement_remarks</code> from more cryptic
notes</li>
</ol>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span></span>
<span>        measurement_remarks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">material</span> <span class="op">==</span> <span class="st">"FRESH"</span>,<span class="st">"fresh leaves (indicating amount of leaf moisture)"</span>, <span class="cn">NA</span><span class="op">)</span>,</span>
<span>        measurement_remarks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">material</span> <span class="op">==</span> <span class="st">"DRIED"</span>,<span class="st">"dry leaves (indicating amount of leaf moisture)"</span>, <span class="va">measurement_remarks</span><span class="op">)</span>,</span>
<span>        measurement_remarks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">material</span> <span class="op">==</span> <span class="st">"SENESCED"</span>,<span class="st">"senesced leaves (indicating amount of leaf moisture)"</span>, <span class="va">measurement_remarks</span><span class="op">)</span>,</span>
<span>  <span class="op">)</span></span></code></pre></div>
<ol start="12" style="list-style-type: decimal">
<li>You can reformat <code>collection_dates</code> supplied into the
<code>yyyy-mm-dd</code> format, or add a date column</li>
</ol>
<p>Converting from any <code>mdy</code> format to
<code>yyyy-mm-dd</code> (e.g. <code>Dec 3 2015</code> to
<code>2015-12-03</code>)</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>Date <span class="op">=</span> <span class="va">Date</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://lubridate.tidyverse.org/reference/ymd.html" class="external-link">mdy</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Converting from any <code>dmy</code> format to
<code>yyyy-mm-dd</code> (e.g. <code>3-12-2015</code> to
<code>2015-12-03</code>)</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>Date <span class="op">=</span> <span class="va">Date</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://lubridate.tidyverse.org/reference/ymd.html" class="external-link">dmy</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Converting from a <code>mmm-yyyy</code> (string) format to
<code>yyyy-mm</code> (e.g. <code>Dec 2015</code> to
<code>2015-12</code>)</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>Date <span class="op">=</span> <span class="fu"><a href="https://lubridate.tidyverse.org/reference/parse_date_time.html" class="external-link">parse_date_time</a></span><span class="op">(</span><span class="va">Date</span>, orders <span class="op">=</span> <span class="st">"my"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/as.Date.html" class="external-link">format.Date</a></span><span class="op">(</span><span class="st">"%Y-%m"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Converting from a <code>mdy</code> format to <code>yyyy-mm</code>
(e.g. Excel has reinterpreted the data as full dates
<code>12-01-2015</code> but the resolution should be “month”
<code>2015-12</code>)</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>Date <span class="op">=</span> <span class="fu"><a href="https://lubridate.tidyverse.org/reference/parse_date_time.html" class="external-link">parse_date_time</a></span><span class="op">(</span><span class="va">Date</span>, orders <span class="op">=</span> <span class="st">"mdy"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/as.Date.html" class="external-link">format.Date</a></span><span class="op">(</span><span class="st">"%Y-%m"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>A particularly complicated example where some dates are presented as
<code>yyyy-mm</code> and others as <code>yyyy-mm-dd</code></p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span></span>
<span>      weird_date <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_detect.html" class="external-link">str_detect</a></span><span class="op">(</span><span class="va">gathering_date</span>, <span class="st">"^[0-9]{4}"</span><span class="op">)</span>, <span class="va">gathering_date</span>, <span class="cn">NA</span><span class="op">)</span>,</span>
<span>      gathering_date <span class="op">=</span> <span class="va">gathering_date</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://lubridate.tidyverse.org/reference/ymd.html" class="external-link">mdy</a></span><span class="op">(</span>quiet <span class="op">=</span> <span class="cn">T</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>      gathering_date <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/coalesce.html" class="external-link">coalesce</a></span><span class="op">(</span><span class="va">gathering_date</span>, <span class="va">weird_date</span><span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">weird_date</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level5">
<h5 id="testing-your-custom-r-code">Testing your custom R code<a class="anchor" aria-label="anchor" href="#testing-your-custom-r-code"></a>
</h5>
<p>After you’ve added the custom R code to a file, check that it has
completed the intended data frame manipulation:</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/metadata_check_custom_R_code.html">metadata_check_custom_R_code</a></span><span class="op">(</span><span class="st">"Blackman_2010"</span><span class="op">)</span></span></code></pre></div>
<p>You could alternatively read the data.csv file into R and run the
code line by line.</p>
</div>
</div>
<div class="section level4">
<h4 id="metadata_dataset">Fill in metadata$dataset<a class="anchor" aria-label="anchor" href="#metadata_dataset"></a>
</h4>
<p>The <code>dataset</code> section is a mix of fields that are filled
in automatically during <code><a href="../reference/metadata_create_template.html">metadata_create_template()</a></code> and
fields that need to be manually filled in.</p>
<ul>
<li>
<strong>individual_id</strong> Individual_id is one of the fields
that can be read in during <code>metadata_create_template</code>.
However, you may instead mutate your own <code>individual_id</code>
using <code>custom_R_code</code> and add it in manually. For a wide
dataset individual_id is required anytime there are multiple rows of
data for the same individual and you want to keep these linked. This
field should only be included if it is required.</li>
</ul>
<p><strong>WARNING</strong> If you have an entry
<code>individual_id: unknown</code> this assigns all rows of data to an
individual named “unknown” and the entire dataset will be assumed to be
from a single individual. This is why it is essential to omit this field
if there isn’t an actual row of data being read in.</p>
<ul>
<li><p><strong>collection_date</strong> If this is not read in as a
specified column, it needs to be filled in manually as
<code>start date/end date</code> in yyyy-mm-dd, yyyy-mm, or yyyy format,
depending on the relevant resolution. If the collection dates are
unknown, write <code>unknown/publication year</code></p></li>
<li><p><strong>description:</strong> 1-2 sentence description of the
study’s goals. The abstract of a manuscript usually includes some good
sentences/phrases to borrow.</p></li>
<li><p><strong>basis_of_record:</strong> Allowable values include:
field, field_experiment, captive_cultivated, lab, preserved_specimen,
and literature. See the top of
<code>system.file("support", "austraits.build_schema.yml", package = "austraits.build"</code>
or <a href="http://traitecoevo.github.io/austraits.build/articles/austraits_database_structure.html">database
structure vignette</a> for definitions of these accepted basis_of_record
values. This field can also be read in from a column or can be specified
at the location or trait level, as described below. Entries under
<code>metadata$locations</code> or <code>metadata$traits</code> (which
apply to only that specific location or trait) override the global value
entered under <code>metadata$dataset</code>.</p></li>
<li><p><strong>life_stage:</strong> Allowable values include: adult,
sapling, seedling, juvenile. This field can also be read in from a
column or can be specified at the location or trait level, as described
below. Entries under <code>metadata$locations</code> or
<code>metadata$traits</code> (which apply to only that specific location
or trait) override the global value entered under
<code>metadata$dataset</code>.</p></li>
<li><p><strong>sampling_strategy:</strong> Often a quite long
description of the sampling strategy, extracted verbatim from a
manuscript.</p></li>
<li><p><strong>original_file:</strong> The name of the file initially
submitted to AusTraits and archived in a Google Drive folder and usually
in the dataset folder, in a subfolder named <code>raw</code>.</p></li>
<li><p><strong>notes:</strong> Notes about the study and processing of
data, especially if there were complications or if some data is
suspected duplicates with another study and were filtered out.</p></li>
</ul>
<p>There are also fields that will only be used for a subset of
datasets:</p>
<ul>
<li><p><strong>measurement_remarks</strong>: Measurement remarks is a
field to capture a miscellaneous notes column. This should be
information that is not captured by “methods” (which is fixed to a
single value for a trait). It can be read in for the whole dataset, or
entered under <a href="#add_traits">dataset$traits</a> if the remarks
only apply to specific traits.</p></li>
<li><p><strong>entity_type</strong>, <strong>value_type</strong>,
<strong>replicates</strong>, and <strong>basis_of_value</strong> are
standardly added to each trait, but a fixed value or column could be
read in under <code>metadata$dataset</code></p></li>
</ul>
</div>
<div class="section level4">
<h4 id="add_traits">Add traits<a class="anchor" aria-label="anchor" href="#add_traits"></a>
</h4>
<p>Begin by automatically adding all traits to your skeletal
<code>metadata.yml</code> file:</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/metadata_add_traits.html">metadata_add_traits</a></span><span class="op">(</span><span class="va">current_study</span><span class="op">)</span></span></code></pre></div>
<p>You will be asked to indicate the columns you wish to keep as
distinct traits. Include all columns with trait data.</p>
<p>This automatically propagates each trait selected into
<code>metadata.yml</code> as follows where <code>var_in</code> is the
name of a column in the <code>data.csv</code> file (for wide datasets)
or a unique trait name values in the <code>trait_name</code> column (for
a long dataset):</p>
<pre><code>- var_in: leaf area (mm2)  
  unit_in: .na  
  trait_name: .na
  entity_type: .na
  value_type: .na 
  basis_of_value: .na
  replicates: .na  
  methods: .na  </code></pre>
<p>The trait details then need to be filled in manually.</p>
<ul>
<li><p><strong>units</strong>: fill in the units specified by the author
- such as mm2. If you’re uncertain about the syntax/format used for some
more complex units, look through the traits definition file
(<code>config/traits.yml</code>) or the file showing unit conversions
(<code>config/unit_conversions.csv</code>). For categorical variables,
leave this as <code>.na</code>.</p></li>
<li><p><strong>trait_name</strong>: This is the appropriate trait name
from <code>config/traits.yml</code>. If no appropriate trait exists in
AusTraits, a new trait can often be added - just ensure it is a
<code>trait</code> where data will be comparable across studies and has
been measured for a fair number (~&gt;50) species. For currently
unsupported traits, we leave this as <code>.na</code> but then fill in
the rest of the data and flag this study as having a potential new
trait. Then in the future, when this trait is added to the
<code>traits.yml</code> file, the data can be read into AusTraits by
simply replacing the <code>.na</code> with a trait name.</p></li>
<li><p><strong>entity_type</strong>: Entity types indicate the
taxonomic/ecological hierarchical level corresponding to the trait
value. Entity types can be individual, population, species, genus,
family or order. Metapopulation-level measurements are coded as
<code>population</code> and infraspecific taxon-level measurements are
coded as <code>species</code>. See the top of
<code>system.file("support", "austraits.build_schema.yml", package = "austraits.build")</code>
for definitions of these accepted entity types. <em>Note: entity_type is
about the hierarchical level to which the trait measurement refers; this
is separate from the taxonomic resolution of the entity’s
name.</em></p></li>
<li><p><strong>value_type</strong>: Allowable value types are mean,
minimum, maximum, mode, range, raw, and bin. See the top of
<code>system.file("support", "austraits.build_schema.yml", package = "austraits.build")</code>
for definitions of these accepted value types. All categorical traits
are generally scored as being a <code>mode</code>, the most commonly
observed value. Note that for values that are <code>bins</code>, the two
numbers are separated by a double-hyphen, <code>1 -- 10</code>.</p></li>
<li><p><strong>basis_of_value</strong>: Basis of value indicates how a
value was determined. Allowable terms are measurement, expert_score,
model_derived, and literature. See the top of
<code>system.file("support", "austraits.build_schema.yml", package = "austraits.build")</code>
for definitions of these accepted value types, but in general most
categorical traits are values that have been scored by an expert
(expert_score) and more numeric trait values are measurements.</p></li>
<li><p><strong>replicates</strong>: Fill in with the appropriate value.
For categorical variables, leave this as <code>.na</code>. If there is a
column that specifies replicate number, you can list the column name in
the field.</p></li>
<li>
<p><strong>methods</strong>: This information can usually be copied
verbatim from a manuscript. In general, methods sections extracted from
pdfs include “special characters” (non-UTF-8 characters). Non-English
alphabet characters are recognised (e.g. é, ö) and should remain
unchanged. Other characters will be re-formatted during the study input
process, so double check that degree symbols (º), en-dashes (–),
em-dashes (—), and curly quotes (‘,’,“,”) have been maintained or
reformatted with a suitable alternative. Greek letters and some other
characters are replaced with their Unicode equivalent
(e.g. &lt;U+03A8&gt; replaces Psi (Ψ)); for these it is best to replace
the symbol with an interpretable English-character equivalent.</p>
<ul>
<li>Note with methods, if the identical methods apply to a string of
traits, for the first trait use the following syntax, where the
<code>&amp;leaf_length_method</code> notation assigns the remaining text
in the field as the <code>leaf_length_method</code>.</li>
</ul>
<pre><code>  methods: &amp;leaf_length_method All measurements were from dry herbarium collections, with leaf and bracteole measurements taken from the largest of these structures on each specimen.</code></pre>
<p>Then for the next trait that uses this method you can just include.
At the end of processing you can read/write the yml file and this will
fill in the assigned text throughout.</p>
<pre><code>  methods: *leaf_length_method</code></pre>
</li>
</ul>
<p>In addition to the automatically propagated fields, there are a
number of optional fields you can add if appropriate.</p>
<ul>
<li><p><strong>life_stage</strong> If all measurements in a dataset were
made on plants of the same <code>life stage</code> a global value should
be entered under <a href="#metadata_dataset">metadata$dataset</a>.
However if different traits were measured at different life stages or
different rows of data represent measurements at different life stages
you can specify a unique life stage for each trait or indicate a column
where this information is stored.</p></li>
<li><p><strong>basis_of_record</strong> If all measurements in a dataset
represent the same <code>basis_of_record</code> a global value should be
entered under <a href="#metadata_dataset">metadata$dataset</a>. However
if different traits have different basis_of_record values or different
rows of data represent different basis_of_record values you can specify
a unique basis_of_record value for each trait or indicate a column where
this information is stored.</p></li>
<li><p><strong>measurement_remarks</strong>: Measurement remarks is a
field to indicate miscellaneous comments. If these comments only apply
to specific trait(s), this field should be specified with those trait’s
metadata sections. This meant to be information that is not captured by
“methods” (which is fixed to a single value for a trait).</p></li>
<li><p><strong>method_context</strong> If different columns in a wide
data.csv file indicate measurements on the same trait using different
methods, this needs to be designated. At the bottom of the trait’s
metadata, add a <code>method_context_name</code> field
(e.g. <code>method_context</code> words well). Write a word or short
phrase that indicates which method context applies to that trait (data
column). For instance, one trait might have
<code>method_context: fully expanded leaves</code> and a second entry
with the same trait name and method might have
<code>method_context: leaves still expanding</code>. The method context
details must also be added to the <a href="#adding_contexts">contexts</a> section.</p></li>
<li><p><strong>temporal_context</strong> If different columns in a wide
data.csv file indicate measurements on the same trait, on the same
individuals at different points in time, this needs to be designated. At
the bottom of the trait’s metadata, add a
<code>temporal_context_name</code> field
(e.g. <code>temporal_context</code> words well). Write a word or short
phrase that indicates which temporal context applies to that trait (data
column). For instance, one trait might have
<code>temporal_context: dry season</code> and a second entry with the
same trait name and method might have
<code>temporal_context: after rain</code>. The temporal context details
must also be added to the <a href="#adding_contexts">contexts</a>
section.</p></li>
</ul>
</div>
<div class="section level4">
<h4 id="adding_locations">Adding location details<a class="anchor" aria-label="anchor" href="#adding_locations"></a>
</h4>
<p>Location data includes location names, latitude/longitude
coordinates, verbal location descriptions, and any additional
abiotic/biotic location variables provided by the contributor (or in the
accompanying manuscript). For studies with more than a few locations, it
is most efficient to create a table of this data that is automatically
read into the <code>metadata.yml</code> file.</p>
<ol style="list-style-type: decimal">
<li><p>Location names must be identical (including syntax, case) to
those in <code>data.csv</code></p></li>
<li><p>Column headers for latitude and longitude data must read
<code>latitude (deg)</code> and <code>longitude (deg)</code></p></li>
<li><p>Latitude and longitude must be in decimal degrees
(i.e. -46.5832). There are many online converters to convert from
<code>degrees,minutes,seconds</code> format or <code>UTM</code>. Or use
the following formula:
<code>decimel_degrees = degrees + (minutes/60) + (seconds/3600)</code></p></li>
<li><p>If there is a column with a general vegetation description
(i.e. <code>rainforest</code>, <code>coastal heath</code> it should be
titled <code>description</code>)</p></li>
<li><p>Although location properties are not restricted to a controlled
vocabulary, newly added studies should use the same location property
syntax as others whenever possible, to allow future discoverability. To
generate a list of already used under <code>location_property</code>,
use:</p></li>
</ol>
<pre><code><span><span class="va">austraits</span><span class="op">$</span><span class="va">locations</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/distinct.html" class="external-link">distinct</a></span><span class="op">(</span><span class="va">location_property</span><span class="op">)</span></span></code></pre>
<p>A few contributors provide a standalone file of all location data.
Otherwise, the following sequence works well:</p>
<ol style="list-style-type: decimal">
<li>Identify all location names in the data.csv file. The following code
extracts a list of location names and any other columns in the data file
that include location-specific information:</li>
</ol>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_csv</a></span><span class="op">(</span><span class="st">"data/dataset_id/data.csv"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/distinct.html" class="external-link">distinct</a></span><span class="op">(</span><span class="va">location</span>, .keep_all <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="co"># the argument `.keep_all` ensures columns aren't dropped</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">location</span>, <span class="va">rainfall</span>, <span class="va">lat</span>, <span class="va">lon</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="co"># list of relevant columns to keep</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html" class="external-link">rename</a></span><span class="op">(</span>`latitude (deg)` <span class="op">=</span> <span class="va">lat</span>, `longitude (deg)` <span class="op">=</span> <span class="va">long</span><span class="op">)</span>  <span class="co"># rename columns to how you want them to appear in the metadata file. Faster to do it once here than repeatedly in the metadata file</span></span>
<span>  <span class="fu"><a href="https://readr.tidyverse.org/reference/write_delim.html" class="external-link">write_csv</a></span><span class="op">(</span><span class="st">"data/dataset_id/raw/location_data.csv"</span><span class="op">)</span></span></code></pre></div>
<ol start="2" style="list-style-type: decimal">
<li><p>Open the spreadsheet in Excel (or any editor of your choice) and
manually add any additional data from the manuscript. Save as a .csv
file.</p></li>
<li><p>Open in R</p></li>
</ol>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_csv</a></span><span class="op">(</span><span class="st">"data/dataset_id/raw/location_data.csv"</span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">location_data</span></span></code></pre></div>
<p>As an example of what the location table should look like:</p>
<ol start="4" style="list-style-type: decimal">
<li>This location data can then be read into
<code>metadata.yml</code>:</li>
</ol>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/metadata_add_locations.html">metadata_add_locations</a></span><span class="op">(</span><span class="va">current_study</span>, <span class="va">site_data</span><span class="op">)</span></span></code></pre></div>
<p>You are first prompted to identify the column with the location name
and then to list all columns that contain location data. This
automatically fills in the location component on the metadata file.</p>
<p>It is possible that you will want to specify <code>life_stage</code>
or <code>basis_of_record</code> at the location_level. You can later
manually add these fields to some or all locations.</p>
<p>(During processing location_id’s are automatically generated and
paired with each location_name.)</p>
</div>
<div class="section level4">
<h4 id="adding_contexts">Context details<a class="anchor" aria-label="anchor" href="#adding_contexts"></a>
</h4>
<p>The dictionary definition of a context is <em>the situation within
which something exists or happens, and that can help explain it</em>.
This is exactly what <code>context_properties</code> are in AusTraits,
ancillary information that is important to explaining and understanding
a trait value.</p>
<p>AusTraits recognises 5 categories of contexts: -
<strong>treatment</strong> contexts are experimental treatments applied
to individuals, such as soil nutrient manipulations, growing
temperatures, or CO2 enchancement. - <strong>plot</strong> contexts are
either blocks/plots within an experimental design or a variable that has
been measured <em>within</em> a location and measurements have been
stratified across this variable. Topographic position within a location
is an example of this. - <strong>temporal</strong> contexts relate to
repeat measurements on the same entity (individual, population, or
species) across time. They may simply be number observations or might be
explicitly linked to growing season or time of day. -
<strong>method</strong> contexts indicate that the same trait has been
measured on the same entity (individual, population or species) using
multiple methods. These might be samples from different canopy light
environments, different leaf ages, or sapwood samples from different
branch diameters. - <strong>entity_contexts</strong> capture ancillary
information about the entity (individual, population or species) that
helps explain the measured trait values. This might be the entity’s sex,
caste (for social insects), or host plant (for insects).</p>
<p>Context properties are not restricted to a controlled vocabulary.
However, newly added studies should use the same context property syntax
as others whenever possible, to allow future discoverability. To
generate a list of terms already used under
<code>context_property</code>, use:</p>
<pre><code><span><span class="va">austraits</span><span class="op">$</span><span class="va">contexts</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/distinct.html" class="external-link">distinct</a></span><span class="op">(</span><span class="va">context_property</span><span class="op">)</span></span></code></pre>
<p>The AusTraits workflow can handle as many context properties as is
required. These are most easily read with the dedicated function</p>
<pre><code><span><span class="fu"><a href="../reference/metadata_add_contexts.html">metadata_add_contexts</a></span><span class="op">(</span><span class="va">dataset_id</span><span class="op">)</span></span></code></pre>
<p>The function first displays a list of all data columns (from the
data.csv file) and prompts you to select those that are context
properties. For each column you are asked to indicate its
<code>category</code> (those described above). You are shown a list of
unique values present in the data column and asked if these require any
substitutions. This function adds the following information to the
section <code>metadata$contexts</code> (example from
<code>Crous_2013</code>)</p>
<pre><code>- context_property: unknown
  category: temporal
  var_in: month
  values:
  - find: AUG
    value: unknown
    description: unknown
  - find: DEC
    value: unknown
    description: unknown
  - find: FEB
    value: unknown
    description: unknown
- context_property: unknown
  category: treatment
  var_in: Temp-trt
  values:
  - value: ambient
    description: unknown
  - value: elevated
    description: unknown
- context_property: unknown
  category: treatment
  var_in: CO2_Treat
  values:
  - find: ambient CO2
    value: unknown
    description: unknown
  - find: added CO2
    value: unknown
    description: unknown</code></pre>
<p>You must then manually fill in the fields designated as
<code>unknown</code>. You are permitted to omit the
<code>description</code> field if the context_property value itself
provides sufficient description.</p>
<p>If there are additional context properties that were designated in
the traits section, these will have to be added manually, as this
information is not captured in a column. A final output might be:</p>
<pre><code>- context_property: sampling season
  category: temporal
  var_in: month
  values:
  - find: AUG
    value: August
    description: August (late winter)
  - find: DEC
    value: December
    description: December (early summer)
  - find: FEB
    value: February
    description: February (late summer)
- context_property: temperature treatment
  category: treatment
  var_in: Temp-trt
  values:
  - value: ambient
    description: Plants grown at ambient temperatures; Jan average max = 29.4 dec
      C / July average min = 3.2 dec C.
  - value: elevated
    description: Plants grown 3 deg C above ambient temperatures.
- context_property: CO2 treatment
  category: treatment
  var_in: CO2_Treat
  values:
  - find: ambient CO2
    value: 400 ppm
    description: Plants grown at ambient CO2 (400 ppm).
  - find: added CO2
    value: 640 ppm
    description: Plants grown at elevated CO2 (640 ppm); 240 ppm above ambient.
- context_property: measurement temperature
  category: method
  var_in: method_context          #this field would be included in the relevant traits
  values:
  - value: 20°C                   # this value would be keyed in through the relevant traits
    description: Measurement made at 20°C
  - value: 25°C
    description: Measurement made at 25°C</code></pre>
</div>
<div class="section level4">
<h4 id="using-substitutions">Using substitutions<a class="anchor" aria-label="anchor" href="#using-substitutions"></a>
</h4>
<p>It is very unlikely that a contributor will use categorical trait
values that are entirely identical to those in the
<code>traits.yml</code> file. You need to add substitutions for those
that do not exactly align to match the wording and syntax supported by
AusTraits. Combinations of multiple trait values are allowed - simply
list them, space delimited (e.g. <code>shrub tree</code> for a species
whose growth form includes both)</p>
<p>Single substitutions can be added by running:</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/metadata_add_substitution.html">metadata_add_substitution</a></span><span class="op">(</span><span class="va">current_study</span>, <span class="st">"trait_name"</span>, <span class="st">"find"</span>, <span class="st">"replace"</span><span class="op">)</span></span></code></pre></div>
<p>where <code>trait_name</code> is the AusTraits defined trait name,
<code>find</code> is the trait value used in the data.csv file and
<code>replace</code> is the trait value supported by AusTraits.</p>
<p>If you have many substitutions to add, the following may be more
efficient:</p>
<ul>
<li><p>Add a single substitution via the function and then copy and
paste the lines many times in the metadata.yml file, changing the
relevant fields</p></li>
<li><p>Create a spreadsheet with a list of all <code>trait_name</code>
by <code>trait_value</code> combinations requiring substitutions. The
spreadsheet would have four columns with headers
<code>dataset_id</code>, <code>trait_name</code>, <code>find</code> and
<code>replace</code>. This table can be read directly into the
<code>metadata.yml</code> file using the function
<code>metadata_add_substitutions_table</code>. This is described below
under <a href="#adding_many_substitutions">Adding many
substitutions</a>.</p></li>
</ul>
<div class="section level5">
<h5 id="excluded-data">Excluded data<a class="anchor" aria-label="anchor" href="#excluded-data"></a>
</h5>
<p>This section of the <code>metadata.yml</code> file provides the
capacity to explicitly exclude specific trait values or taxon names.
These are values that are in the <code>data.csv</code> file but should
be excluded from AusTraits.</p>
<p>It includes three elements:<br>
- <strong>variable</strong>: A variable from the traits table, typically
<code>taxon_name</code>, <code>location_name</code> or
<code>context_name</code><br>
- <strong>find</strong>: Value of variable to remove<br>
- <strong>reason</strong>: Records why the data was removed,
e.g. <code>exotic</code></p>
<p>Multiple, comma-delimited values can be added under
<code>find</code>.</p>
<p>For example, in Munroe_2019:</p>
<div class="sourceCode" id="cb47"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a>exclude_observations<span class="sc">:</span></span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a><span class="sc">-</span> variable<span class="sc">:</span> taxon_name</span>
<span id="cb47-3"><a href="#cb47-3" aria-hidden="true" tabindex="-1"></a>  find<span class="sc">:</span> Campylopus introflexus, Dicranoloma menziesii, Philonotis tenuis, Polytrichastrum</span>
<span id="cb47-4"><a href="#cb47-4" aria-hidden="true" tabindex="-1"></a>    alpinum, Polytrichum juniperinum, Sphagnum cristatum</span>
<span id="cb47-5"><a href="#cb47-5" aria-hidden="true" tabindex="-1"></a>  reason<span class="sc">:</span> <span class="fu">moss</span> (E Wenk, <span class="dv">2020</span>.<span class="fl">06.18</span>)</span>
<span id="cb47-6"><a href="#cb47-6" aria-hidden="true" tabindex="-1"></a><span class="sc">-</span> variable<span class="sc">:</span> taxon_name</span>
<span id="cb47-7"><a href="#cb47-7" aria-hidden="true" tabindex="-1"></a>  find<span class="sc">:</span> Xanthoparmelia semiviridis</span>
<span id="cb47-8"><a href="#cb47-8" aria-hidden="true" tabindex="-1"></a>  reason<span class="sc">:</span> <span class="fu">lichen</span> (E Wenk, <span class="dv">2020</span>.<span class="fl">06.18</span>)</span></code></pre></div>
</div>
<div class="section level5">
<h5 id="questions">Questions<a class="anchor" aria-label="anchor" href="#questions"></a>
</h5>
<p>The final section of the <code>metadata.yml</code> file is titled
<code>questions</code>. This is a location to:</p>
<ol style="list-style-type: decimal">
<li>Ask the data contributor targeted questions about their study. When
you generate a report (described below) these questions will appear at
the top of the report.
<ul>
<li>Preface the first question you have with <code>contributor:</code>
(indented once), and additional questions with <code>question2:</code>,
etc.<br>
</li>
<li>Ask contributors about missing metadata<br>
</li>
<li>Point contributors attention to odd data distributions, to make sure
they look at those traits extra carefully.</li>
<li>Let contributors know if you’re uncertain about their units or if
you transformed the data in a fairly major way.<br>
</li>
<li>Ask the contributors if you’re uncertain you aligned their trait
names correctly.</li>
</ul>
</li>
<li>This is a place to list any trait data that are not yet
<code>traits</code> supported by AusTraits. Use the following syntax,
indented once: <code>additional_traits:</code>, followed by a list of
traits.</li>
</ol>
</div>
<div class="section level5">
<h5 id="hooray-you-now-have-a-fully-propagated-metadata-yml-file">Hooray! You now have a fully propagated metadata.yml file!<a class="anchor" aria-label="anchor" href="#hooray-you-now-have-a-fully-propagated-metadata-yml-file"></a>
</h5>
<p>Next is making sure it has captured all the data exactly as you’ve
intended.</p>
</div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="quality_checks">Quality checks<a class="anchor" aria-label="anchor" href="#quality_checks"></a>
</h2>
<p>Before starting the quality checks, it is helpful to assign a
variable, <code>current_study</code>:</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">current_study</span> <span class="op">&lt;-</span> <span class="st">"Wright_2001"</span></span></code></pre></div>
<p>This lets you have a list of tests you run for each study and you
just have to reassign a new <code>dataset_id</code> to
<code>current_study</code>.</p>
<p>It is best to run tests and fix formatting first.</p>
<div class="section level3">
<h3 id="clear-formatting">Clear formatting<a class="anchor" aria-label="anchor" href="#clear-formatting"></a>
</h3>
<p>The <code>clear formatting</code> code below reads and re-writes the
yaml file. This is the same process that is repeated when running
functions that automatically add substitutions or check taxonomy.
Running it first ensures that any formatting issues introduced (or
fixed) during the read/write process are identified and solved
first.</p>
<p>For instance, the <code>write_metadata</code> function inserts line
breaks every 80 characters and reworks other line breaks (except in
custom_R_code). It also reformats special characters in the text,
substituting in its accepted format for degree symbols, en-dashes,
em-dashes and quotes, and substituting in Unicode codes for more obscure
symbols.</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">f</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="st">"data"</span>, <span class="va">current_study</span>, <span class="st">"metadata.yml"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/read_metadata.html">read_metadata</a></span><span class="op">(</span><span class="va">f</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../reference/write_metadata.html">write_metadata</a></span><span class="op">(</span><span class="va">f</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="running_tests">Running tests<a class="anchor" aria-label="anchor" href="#running_tests"></a>
</h3>
<p>Begin by running some automated tests to ensure the dataset meets the
required set up. The tests run through a collection of pre-specified
checks on the files for each study. The output alerts you to possible
issues needing to be fixed, by comparing the data in the files with the
expected structure and allowed values, as specified in the schema and
definitions.</p>
<p>Certain special characters may show up as errors and need to be
manually adjusted in the <code>metadata.yml</code> file</p>
<p>The tests also identify mismatches between the location names in the
data.csv file vs. metadata.yml file (same for context), unsupported
trait names, etc.</p>
<p>To run the tests, the variable <code>dataset_ids</code> must be
defined in the global namespace, containing a vector of ids to check.
For example:</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load relevant functions</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://traitecoevo.github.io/austraits.build/">austraits.build</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Tests run test on one study</span></span>
<span><span class="va">dataset_ids</span> <span class="op">&lt;-</span> <span class="st">"Bragg_2002"</span></span>
<span><span class="fu"><a href="../reference/dataset_test.html">dataset_test</a></span><span class="op">(</span><span class="va">dataset_ids</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Tests run test on all studies</span></span>
<span><span class="va">dataset_ids</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">dir</a></span><span class="op">(</span><span class="st">"data"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/dataset_test.html">dataset_test</a></span><span class="op">(</span><span class="va">dataset_ids</span><span class="op">)</span></span></code></pre></div>
<p>Fix as many errors as you can and then rerun
<code><a href="../reference/dataset_test.html">dataset_test()</a></code> repeatedly until no errors remain.</p>
<p>See below for suggestions on how to implement large numbers of <a href="#adding_many_substitutions">trait value substitutions</a>.</p>
</div>
<div class="section level3">
<h3 id="rebuild-austraits">Rebuild AusTraits<a class="anchor" aria-label="anchor" href="#rebuild-austraits"></a>
</h3>
<p>Now incorporate the new study into AusTraits:</p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/build_setup_pipeline.html">build_setup_pipeline</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">austraits</span> <span class="op">&lt;-</span> <span class="fu">remake</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/remake/man/make.html" class="external-link">make</a></span><span class="op">(</span><span class="st">"austraits"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="check-excluded-data">Check excluded data<a class="anchor" aria-label="anchor" href="#check-excluded-data"></a>
</h3>
<p>AusTraits automatically excludes data for a number of reasons. These
are available in the frame <code>excluded_data</code>.</p>
<p>When you are finished running quality checks, no data should be
excluded due to <em>Missing unit conversion</em> and <em>Unsupported
trait</em>.</p>
<p>A few values may be legitimately excluded due to other errors, but
check each entry.</p>
<p>The best way to view excluded data for a study is:</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">austraits</span><span class="op">$</span><span class="va">excluded_data</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span></span>
<span>    <span class="va">dataset_id</span> <span class="op">==</span> <span class="va">current_study</span>,</span>
<span>    <span class="va">error</span> <span class="op">!=</span> <span class="st">"Observation excluded in metadata"</span></span>
<span>  <span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/View.html" class="external-link">View</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>Missing values (blank cells, cells with NA) are not included in the
<code>excluded_data</code> table, because they are assumed to be
legitimate blanks. If you want to confirm this, you need to temporarily
change the default arguments for the internal function
<code>dataset_process</code> where it is called within the
<code>remake.yml</code> file. For instance, the default,</p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>      <span class="fu"><a href="../reference/dataset_process.html">dataset_process</a></span><span class="op">(</span><span class="st">"data/Ahrens_2019/data.csv"</span>,</span>
<span>                  <span class="va">Ahrens_2019_config</span>,</span>
<span>                  <span class="va">schema</span></span>
<span>                 <span class="op">)</span></span></code></pre></div>
<p>needs to be changed to:</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>      <span class="fu"><a href="../reference/dataset_process.html">dataset_process</a></span><span class="op">(</span><span class="st">"data/Ahrens_2019/data.csv"</span>,</span>
<span>                  <span class="va">Ahrens_2019_config</span>,</span>
<span>                  <span class="va">schema</span>,</span>
<span>                  filter_missing_values <span class="op">=</span> <span class="cn">FALSE</span></span>
<span>                 <span class="op">)</span></span></code></pre></div>
<div class="section level4">
<h4 id="reasons-for-data-to-be-excluded">Reasons for data to be excluded<a class="anchor" aria-label="anchor" href="#reasons-for-data-to-be-excluded"></a>
</h4>
<p>Possible reasons for excluding a trait value includes:</p>
<ul>
<li><p><strong>Missing species name</strong>: Species name is missing
from data.csv file for a given row of data. This usually occurs when
there are stray characters in the data.csv file below the data – delete
these rows.</p></li>
<li><p><strong>Missing unit conversion</strong>: Value was present but
appropriate unit conversion was missing. This requires that you add a
new unit conversion to the file
<code>config/unit_conversions.csv</code>. Add additional conversions
near similar unit conversions already in the file for easier searching
in the future.</p></li>
<li><p><strong>Observation excluded in metadata</strong>: Specific
values, usually certain taxon names can be excluded in the metadata.
This is generally used when a study includes a number of non-native and
non-naturalised species that need to be excluded. These should be
intentional exclusions, as they have been added by you.</p></li>
<li><p><strong>Time contains non-number</strong>: Indicates a problem
with the value entered into the traits <code>flowering_time</code> and
<code>fruiting_time</code>. (<em>Note to AusTraits custodians: This
error should no longer appear - will retain for now as a
placeholder.</em>)</p></li>
<li><p><strong>Unsupported trait</strong>: <code>trait_name</code> not
listed in <code>config/traits.yml</code>, under <code>traits</code>.
Double check you have used the correct spelling/exact syntax for the
<code>trait_name</code>, adding a new <code>trait</code> to the
<code>traits.yml</code> file if appropriate. If there is a trait that is
currently unsupported by AusTraits, leave <code>trait_name: .na</code>.
Do not fill in an arbitrary name.</p></li>
<li><p><strong>Unsupported trait value</strong>: This error, referencing
categorical traits, means that the <code>value</code> for a trait is not
included in the list of supported trait values for that trait in
<code>config/traits.yml</code>. See <a href="#adding_many_substitutions">adding many substitutions</a> if there
are many trait values requiring substitutions. If appropriate, add
another trait value to the <code>traits.yml</code> file, but confer with
other curators, as the lists of trait values have been carefully agreed
upon through workshop sessions.</p></li>
<li><p><strong>Value does not convert to numeric</strong>: Is there a
strange character in the file preventing easy conversion? This error is
rare and generally justified.</p></li>
<li><p><strong>Value out of allowable range</strong>: This error,
referencing numeric traits, means that the trait value, after unit
conversions, falls outside of the allowable range specified for that
trait in <code>config/traits.yml</code>. Sometimes the AusTraits range
is too narrow and other times the author’s value is truly an outlier
that should be excluded. Look closely at these and adjust the range in
<code>config/traits.yml</code> if justified. Generally, don’t change the
range until you’ve create a report for the study and confirmed that the
general cloud of data aligns with other studies as excepted. Most
frequently the units or unit conversion is what is incorrect.</p></li>
</ul>
<p>You can also ask how many of each error type are present for a
study:</p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">austraits</span><span class="op">$</span><span class="va">excluded_data</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">dataset_id</span> <span class="op">==</span> <span class="st">"Cheal_2017"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="va">error</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; &lt; table of extent 0 &gt;</span></span></code></pre></div>
<p>Or produce a table of error type by trait:</p>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">austraits</span><span class="op">$</span><span class="va">excluded_data</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span></span>
<span>    <span class="va">dataset_id</span> <span class="op">==</span> <span class="st">"Cheal_2017"</span>,</span>
<span>  <span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">trait_name</span>, <span class="va">error</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; &lt; table of extent 0 x 0 &gt;</span></span></code></pre></div>
<p>Note, most studies have no excluded data. This study is an extreme
example!</p>
</div>
</div>
<div class="section level3">
<h3 id="adding_many_substitutions">Adding many substitutions<a class="anchor" aria-label="anchor" href="#adding_many_substitutions"></a>
</h3>
<p>For categorical traits, if you want to create a list of all values
that require substitutions:</p>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">austraits</span><span class="op">$</span><span class="va">excluded_data</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span></span>
<span>    <span class="va">dataset_id</span> <span class="op">==</span> <span class="va">current_study</span>,</span>
<span>    <span class="va">error</span> <span class="op">==</span> <span class="st">"Unsupported trait value"</span></span>
<span>  <span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/distinct.html" class="external-link">distinct</a></span><span class="op">(</span><span class="va">dataset_id</span>, <span class="va">trait_name</span>, <span class="va">value</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html" class="external-link">rename</a></span><span class="op">(</span>find <span class="op">=</span> <span class="va">value</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">dataset_id</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://readr.tidyverse.org/reference/write_delim.html" class="external-link">write_csv</a></span><span class="op">(</span><span class="st">"data/dataset_id/raw/substitutions_required.csv"</span><span class="op">)</span></span></code></pre></div>
<p>For studies with a small number of substitutions, add them
individually using:</p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/metadata_add_substitution.html">metadata_add_substitution</a></span><span class="op">(</span><span class="va">dataset_id</span>, <span class="va">trait_name</span>, <span class="va">find</span>, <span class="va">replace</span><span class="op">)</span></span></code></pre></div>
<p>For studies with large number of substitutions required, you can add
an additional column to this table, <code>replace</code>, and fill in
all the correct trait values. Then read the list of substitutions
directly into the metadata file:</p>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">substitutions_to_add</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_csv</a></span><span class="op">(</span><span class="st">"data/dataset_id/raw/substitutions_required_after_editing.csv"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/metadata_add_substitutions_list.html">metadata_add_substitutions_list</a></span><span class="op">(</span><span class="va">dataset_id</span>, <span class="va">substitutions_to_add</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="add-taxonomic-updates">Add taxonomic updates<a class="anchor" aria-label="anchor" href="#add-taxonomic-updates"></a>
</h3>
<p>The function <code>add_taxonomic_updates</code> allows you to
manually align submitted taxon names (the <code>original_name</code>)
with the taxon names in the taxonomic resource.</p>
<div class="sourceCode" id="cb60"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a>  metadata_add_taxonomic_change <span class="ot">&lt;-</span> <span class="cf">function</span>(dataset_id, find, replace, reason, taxonomic_resolution)</span></code></pre></div>
<ul>
<li>
<strong>find</strong> is the name in the taxon name column in the
dataset</li>
<li>
<strong>replace</strong> is the equivalent taxon name in the
taxonomic resource</li>
<li>
<strong>reason</strong> provides information about why the taxonomic
update is required</li>
<li>
<strong>taxonomic_resolution</strong> indicates the most specific
taxon rank that the name in <code>replace</code> aligns to</li>
</ul>
<p>As examples:</p>
<p>A simple fix correcting a minor typo to align with an accepted taxon
name:</p>
<div class="sourceCode" id="cb61"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a>taxonomic_updates<span class="sc">:</span></span>
<span id="cb61-2"><a href="#cb61-2" aria-hidden="true" tabindex="-1"></a><span class="sc">-</span> find<span class="sc">:</span> Drummondita rubroviridis</span>
<span id="cb61-3"><a href="#cb61-3" aria-hidden="true" tabindex="-1"></a>  replace<span class="sc">:</span> Drummondita rubriviridis</span>
<span id="cb61-4"><a href="#cb61-4" aria-hidden="true" tabindex="-1"></a>  reason<span class="sc">:</span> match_07_fuzzy. Fuzzy alignment with accepted canonical name <span class="cf">in</span> <span class="fu">APC</span> (<span class="dv">2022-11-21</span>)</span>
<span id="cb61-5"><a href="#cb61-5" aria-hidden="true" tabindex="-1"></a>  taxonomic_resolution<span class="sc">:</span> Species</span></code></pre></div>
<p>An example of a taxon name that can only be aligned to genus. The
<code>taxonomic_resolution</code> is therefore specified as
<code>genus</code>. The portion of the name that can be aligned to the
taxonomic resource must be before the square brackets. Any information
within the square brackets is important for uniquely identifying this
entry within AusTraits, but does not provide additional taxonomic
information.</p>
<div class="sourceCode" id="cb62"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a><span class="sc">-</span> find<span class="sc">:</span> Acacia ancistrophylla<span class="sc">/</span>sclerophylla</span>
<span id="cb62-2"><a href="#cb62-2" aria-hidden="true" tabindex="-1"></a>  replace<span class="sc">:</span> Acacia sp. [Acacia ancistrophylla<span class="sc">/</span>sclerophylla; White_2020]</span>
<span id="cb62-3"><a href="#cb62-3" aria-hidden="true" tabindex="-1"></a>  reason<span class="sc">:</span> match_04. Rewording taxon where <span class="st">`</span><span class="at">/</span><span class="st">`</span> indicates uncertain species identification</span>
<span id="cb62-4"><a href="#cb62-4" aria-hidden="true" tabindex="-1"></a>    to align with <span class="st">`</span><span class="at">APC accepted</span><span class="st">`</span> <span class="fu">genus</span> (<span class="dv">2022-11-10</span>)</span>
<span id="cb62-5"><a href="#cb62-5" aria-hidden="true" tabindex="-1"></a>  taxonomic_resolution<span class="sc">:</span> genus</span></code></pre></div>
<p>A taxonomic update that aligns a name to the most similar taxon_name
within a taxonomic resource (the APC), but this is a taxonomic synonym
and the austraits workflow will update it to its currently accepted name
(since this is documented within the <code>taxon_list.csv</code>
file):</p>
<div class="sourceCode" id="cb63"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a><span class="sc">-</span> find<span class="sc">:</span> <span class="fu">Polyalthia</span> (Wyvur)</span>
<span id="cb63-2"><a href="#cb63-2" aria-hidden="true" tabindex="-1"></a>  replace<span class="sc">:</span> Polyalthia <span class="fu">sp.</span> (Wyvuri B.P.Hyland RFK2632)</span>
<span id="cb63-3"><a href="#cb63-3" aria-hidden="true" tabindex="-1"></a>  reason<span class="sc">:</span> match_15_fuzzy. Fuzzy match alignment with species<span class="sc">-</span>level canonical name</span>
<span id="cb63-4"><a href="#cb63-4" aria-hidden="true" tabindex="-1"></a>    <span class="cf">in</span> <span class="st">`</span><span class="at">APC known</span><span class="st">`</span> when everything except first <span class="dv">2</span> words <span class="fu">ignored</span> (<span class="dv">2022-11-10</span>)</span>
<span id="cb63-5"><a href="#cb63-5" aria-hidden="true" tabindex="-1"></a>  taxonomic_resolution<span class="sc">:</span> Species</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="check-if-austraits-pivots-wider">Check if AusTraits pivots wider<a class="anchor" aria-label="anchor" href="#check-if-austraits-pivots-wider"></a>
</h3>
<p>AusTraits users want to be able to “pivot” between long and wide
formats. Each row of data should have a unique combination of the
following fields: <code>trait_name</code>, <code>dataset_id</code>,
<code>observation_id</code>, <code>source_id</code>,
<code>taxon_name</code>, <code>population_id</code>,
<code>individual_id</code>, <code>temporal_id</code>,
<code>method_id</code>, <code>value_type</code>, and
<code>original_name</code></p>
<p>Therefore, the dataset should be able to pivot wider and the
following code should have a <code>1</code> in every cell.</p>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">austraits</span><span class="op">$</span><span class="va">traits</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">dataset_id</span>, <span class="va">trait_name</span>, <span class="va">value</span>, <span class="va">observation_id</span>, <span class="va">source_id</span>, <span class="va">taxon_name</span>, <span class="va">population_id</span>, <span class="va">individual_id</span>, <span class="va">temporal_id</span>, <span class="va">method_id</span>, <span class="va">value_type</span>, <span class="va">original_name</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_wider.html" class="external-link">pivot_wider</a></span><span class="op">(</span>names_from <span class="op">=</span> <span class="va">trait_name</span>, values_from <span class="op">=</span> <span class="va">value</span>, values_fn <span class="op">=</span> <span class="va">length</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/View.html" class="external-link">View</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>If AusTraits fails to <code>pivot_wider</code>, likely problems are:
- <strong>Not all context information has been captured.</strong> For
instance, is it possible that you have two columns with data for the
same trait, measured using different methods? In this case you need to
add a <code>method_context</code> to both the relevant <a href="#add_traits">traits</a> and to the <a href="#adding_contexts">contexts section</a>. - <strong>There are
multiple observations per entity.</strong> In a number of large studies
which, in theory, include a single observation per species, have a few
scattered instances of a second row of trait values with the same taxon
name. They might be true duplicates and can be removed or perhaps they
are indeed some alternate values. In this case the following
custom_R_code works:</p>
<pre><code><span><span class="st">' data %&gt;%</span></span>
<span><span class="st">    group_by(taxon_name) %&gt;%</span></span>
<span><span class="st">      mutate(observation_number = dplyr::row_number()) %&gt;%</span></span>
<span><span class="st">    ungroup()'</span></span></code></pre>
<p>Then add <code>observation_number</code> as a context with
<code>category: temporal</code></p>
</div>
<div class="section level3">
<h3 id="check-for-duplicates">Check for duplicates<a class="anchor" aria-label="anchor" href="#check-for-duplicates"></a>
</h3>
<p>AusTraits strives to have no duplicate entries for numeric
(continuous) trait measurements. That is, each value in AusTraits should
represent a unique measurement, rather than a measurement sourced from
another study.</p>
<p>When you receive/solicit a dataset, ask the data contributor if all
data submitted was collected for the specific study and if they suspect
other studies from their lab/colleagues may also have contributed any of
this data.</p>
<p>In addition, there are tests to check for duplicates within and
across dataset_ids.</p>
<p><a href="#dataframe_of_duplicates">To check for duplicates:</a></p>
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">austraits_deduped</span> <span class="op">&lt;-</span> <span class="fu">remove_suspected_duplicates</span><span class="op">(</span><span class="va">austraits</span><span class="op">)</span></span>
<span><span class="va">duplicates_for_dataset_id</span> <span class="op">&lt;-</span></span>
<span>  <span class="va">austraits_deduped</span><span class="op">$</span><span class="va">excluded_data</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span></span>
<span>    <span class="va">dataset_id</span> <span class="op">==</span> <span class="va">current_study</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<div class="section level4">
<h4 id="duplicates_within_study">Duplicates within the study<a class="anchor" aria-label="anchor" href="#duplicates_within_study"></a>
</h4>
<ol style="list-style-type: decimal">
<li><p>First sort <code>duplicates_for_dataset_id</code> by the column
<code>error</code> and scan for duplicates within the study (these will
be entries under error that begin with the same <code>dataset_id</code>
as the dataset being processed)</p></li>
<li><p>For legitimately identical measurements, do nothing. For
instance, if %N has been measured on 50 replicates of a species and is
reported to the nearest 0.01% it is quite likely there will be a few
identical values within the study.</p></li>
<li><p>If a species-level measurement has been entered for all
within-location replicates, you need to filter out the duplicates. This
is true for both numeric and categorical values. Enter the following
code as <code>custom_R_code</code> in the dataset’s metadata
file:</p></li>
</ol>
<div class="sourceCode" id="cb67"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a>data <span class="sc">%&gt;%</span></span>
<span id="cb67-2"><a href="#cb67-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(Species) <span class="sc">%&gt;%</span></span>
<span id="cb67-3"><a href="#cb67-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(</span>
<span id="cb67-4"><a href="#cb67-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">across</span>(</span>
<span id="cb67-5"><a href="#cb67-5" aria-hidden="true" tabindex="-1"></a>      <span class="fu">c</span>(leaf_percentN, <span class="st">`</span><span class="at">plant growth form</span><span class="st">`</span>), replace_duplicates_with_NA)</span>
<span id="cb67-6"><a href="#cb67-6" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb67-7"><a href="#cb67-7" aria-hidden="true" tabindex="-1"></a>  <span class="er">)</span> <span class="sc">%&gt;%</span></span>
<span id="cb67-8"><a href="#cb67-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ungroup</span>()</span></code></pre></div>
<p>Note: Using custom R code instead of filtering the values in the
data.csv file itself ensures the relevant trait values are still
associated with each line of data in the data.csv file, but only read
into AusTraits a single time. Note: You would use
<code>group_by(Species, Location)</code> if there are unique values at
the species x location level.</p>
</div>
<div class="section level4">
<h4 id="duplicates_between_studies">Duplicates between studies<a class="anchor" aria-label="anchor" href="#duplicates_between_studies"></a>
</h4>
<p>AusTraits does not attempt to filter out duplicates in categorical
traits between studies. The commonly duplicated traits like
<code>life_form</code>, <code>plant_growth_form</code>,
<code>photosynthetic_pathway</code>, <code>fire_response</code>, etc.
are legitimately duplicated and if the occasional study reported a
different <code>plant_growth_form</code> or <code>fire_response</code>
it would be important to have documented that one trait value was much
more common than another. Such categorical trait values may have been
sourced from reference material or measured/identified by this research
team.</p>
<p>Identifying duplicates in numeric traits between studies can be
difficult, but it is essential that we attempt to filter out all
duplicate occurrences of the same measurement. Some common patterns of
duplication include:</p>
<ol style="list-style-type: decimal">
<li><p>For a single trait, if there are a large number of values
duplicated in a specific other dataset_id (i.e. the <code>error</code>
repeatedly starts with the same <code>dataset_id</code>), be suspicious.
Before contacting the author, check the metadata for the two datasets,
especially authors and study locations, to see if it is likely these are
data values that have been jointly collected and shared across studies.
Similar location names/locations, identical university affiliations, or
similar lists of traits being measured are good clues.</p></li>
<li><p><code>plant_height</code>, <code>leaf_length</code>,
<code>leaf_width</code>, <code>seed_length</code>,
<code>seed_width</code> and <code>seed_mass</code> are the numeric
variables that are most frequently sourced from reference material
(e.g. floras, herbarium collections, reference books, Kew seed database,
etc.)</p></li>
<li><p>The following datasets are flagged in AusTraits as
<code>reference</code> studies and are the source of most duplicates for
the variables listed above: <code>Kew_2019_1</code>,
<code>Kew_2019_2</code>, <code>Kew_2019_3</code>,
<code>Kew_2019_4</code>, <code>Kew_2019_5</code>,
<code>Kew_2019_6</code>, <code>ANBG_2019</code>,
<code>GrassBase_2014</code>, <code>CPBR_2002</code>,
<code>NTH_2014</code>,<code>RBGK_2014</code>, <code>NHNSW_2016</code>,
<code>RBGSYD__2014_2</code>, <code>RBGSYD_2014</code>,
<code>TMAG_2009</code>, <code>WAH_1998</code>,
<code>WAH_2016</code>,<code>Brock_1993</code>, <code>Barlow_1981</code>,
<code>Hyland_2003</code>, <code>Cooper_2013</code></p></li>
</ol>
<p>Data from these studies are assumed to be the source, and the other
study with the value is assumed to have sourced it from the above study.
We recognise this is not always accurate, especially for compilations
within <code>Kew_2019_1</code>, Kew’s seed mass database. Whenever we
input a raw dataset that is also part of the Kew compilation, we filter
that contributors data from <code>Kew_2019_1</code>.</p>
<ol start="4" style="list-style-type: decimal">
<li><p>Data for <code>wood_density</code> is also often sourced from
other studies, most commonly <code>Ilic_2000</code> or
<code>Zanne_2009</code>.</p></li>
<li><p>Data from a number of studies from <code>Leishman</code> and
<code>Wright</code> have been extensively shared within the trait
ecology community, especially through TRY</p></li>
</ol>
<p>If the dataset you are processing has a number of numeric trait
duplicates that follow one of the <code>patterns of duplication</code>
listed, the duplicates should be filtered out. Any other data explicitly
indicated in the manuscript as sourced should also be filtered out. Most
difficult are studies that have partially sourced data, often from many
small studies, and partially collected new data, but not identified the
source of each value.</p>
<p>Filtering duplicate data is a three-step process. In brief:</p>
<ol style="list-style-type: decimal">
<li>Identify traits and studies with duplicates you believe should be
removed.<br>
</li>
<li>Add additional columns to <code>data.csv</code>, identifying certain
<code>trait_values</code> as duplicates.<br>
</li>
<li>Add <code>custom R code</code> that filters out identified
duplicates when the study is merged into AusTraits.</li>
</ol>
<div class="section level5">
<h5 id="identify-traits-and-studies">Identify traits and studies<a class="anchor" aria-label="anchor" href="#identify-traits-and-studies"></a>
</h5>
<ol style="list-style-type: decimal">
<li>Either in R or Excel, manipulate <a href="#dataframe_of_duplicates"><code>duplicates_for_dataset_id</code></a>
to remove rows that you believe are legitimate duplicates, including
duplicates values due to replicate measurements within a single study
and stray duplicates across studies that likely true, incidental
duplicate values. Carefully consider which datasets and traits to
include/exclude from the filter.</li>
</ol>
<p>As an example:</p>
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Note, this code will be replaced by a function in the future.</span></span>
<span><span class="va">duplicates_to_filter</span> <span class="op">&lt;-</span></span>
<span>  <span class="va">duplicates_for_dataset_id</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span></span>
<span>    dataset_with_duplicate <span class="op">=</span></span>
<span>      <span class="va">error</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"Duplicate of "</span>, <span class="st">""</span>, <span class="va">.</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"[[:alnum:]]$"</span>, <span class="st">""</span>, <span class="va">.</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"[[:punct:]]$"</span>, <span class="st">""</span>, <span class="va">.</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">dataset_with_duplicate</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Ilic_2000"</span>, <span class="st">"Zanne_2009"</span>, <span class="st">"Kew_2019_1"</span>, <span class="st">"Barlow_1981"</span>, <span class="st">"NTH_2014"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">trait_name</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"wood_density"</span>, <span class="st">"seed_mass"</span>, <span class="st">"leaf_length"</span>, <span class="st">"leaf_width"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<ol start="2" style="list-style-type: decimal">
<li>Use the following code to add columns to <code>data.csv</code> that
identify specific values as duplicates:</li>
</ol>
<div class="sourceCode" id="cb69"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Note, this code will be replaced by a function in the future.</span></span>
<span><span class="va">wood_density_duplicates</span> <span class="op">&lt;-</span></span>
<span>  <span class="va">duplicates_to_filter</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">trait_name</span> <span class="op">==</span> <span class="st">"wood_density"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">error</span>, <span class="va">original_name</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html" class="external-link">rename</a></span><span class="op">(</span>wood_density_duplicate <span class="op">=</span> <span class="va">error</span><span class="op">)</span></span>
<span></span>
<span><span class="va">seed_mass_duplicates</span> <span class="op">&lt;-</span></span>
<span>  <span class="va">duplicates_to_filter</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">trait_name</span> <span class="op">==</span> <span class="st">"seed_width"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">error</span>, <span class="va">original_name</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html" class="external-link">rename</a></span><span class="op">(</span>seed_mass_duplicate <span class="op">=</span> <span class="va">error</span><span class="op">)</span></span>
<span></span>
<span><span class="va">leaf_width_min_duplicates</span> <span class="op">&lt;-</span></span>
<span>  <span class="va">duplicates_to_filter</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">trait_name</span> <span class="op">==</span> <span class="st">"leaf_width"</span>, <span class="va">value_type</span> <span class="op">==</span> <span class="st">"expert_min"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">error</span>, <span class="va">original_name</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html" class="external-link">rename</a></span><span class="op">(</span>leaf_width_min_duplicate <span class="op">=</span> <span class="va">error</span><span class="op">)</span></span>
<span></span>
<span><span class="va">leaf_width_max_duplicates</span> <span class="op">&lt;-</span></span>
<span>  <span class="va">duplicates_to_filter</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">trait_name</span> <span class="op">==</span> <span class="st">"leaf_width"</span>, <span class="va">value_type</span> <span class="op">==</span> <span class="st">"expert_max"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">error</span>, <span class="va">original_name</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html" class="external-link">rename</a></span><span class="op">(</span>leaf_width_max_duplicate <span class="op">=</span> <span class="va">error</span><span class="op">)</span></span>
<span></span>
<span><span class="va">leaf_length_min_duplicates</span> <span class="op">&lt;-</span></span>
<span>  <span class="va">duplicates_to_filter</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">trait_name</span> <span class="op">==</span> <span class="st">"leaf_length"</span>, <span class="va">value_type</span> <span class="op">==</span> <span class="st">"expert_min"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">error</span>, <span class="va">original_name</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html" class="external-link">rename</a></span><span class="op">(</span>leaf_length_min_duplicate <span class="op">=</span> <span class="va">error</span><span class="op">)</span></span>
<span></span>
<span><span class="va">leaf_length_max_duplicates</span> <span class="op">&lt;-</span></span>
<span>  <span class="va">duplicates_to_filter</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">trait_name</span> <span class="op">==</span> <span class="st">"leaf_length"</span>, <span class="va">value_type</span> <span class="op">==</span> <span class="st">"expert_max"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">error</span>, <span class="va">original_name</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html" class="external-link">rename</a></span><span class="op">(</span>leaf_length_max_duplicate <span class="op">=</span> <span class="va">error</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_csv</a></span><span class="op">(</span><span class="st">"data/dataset_id/data.csv"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span><span class="va">wood_density_duplicates</span>, by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"column_with_taxon_name"</span> <span class="op">=</span> <span class="st">"original_name"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span><span class="va">seed_mass_duplicates</span>, by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"column_with_taxon_name"</span> <span class="op">=</span> <span class="st">"original_name"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span><span class="va">leaf_width_min_duplicates</span>, by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"column_with_taxon_name"</span> <span class="op">=</span> <span class="st">"original_name"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span><span class="va">leaf_width_max_duplicates</span>, by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"column_with_taxon_name"</span> <span class="op">=</span> <span class="st">"original_name"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span><span class="va">leaf_length_min_duplicates</span>, by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"column_with_taxon_name"</span> <span class="op">=</span> <span class="st">"original_name"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span><span class="va">leaf_length_max_duplicates</span>, by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"column_with_taxon_name"</span> <span class="op">=</span> <span class="st">"original_name"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://readr.tidyverse.org/reference/write_delim.html" class="external-link">write_csv</a></span><span class="op">(</span><span class="st">"data/dataset_id/data.csv"</span><span class="op">)</span></span></code></pre></div>
<ol start="3" style="list-style-type: decimal">
<li>For the above example, then add the following code to custom R code,
removing the duplicate values from the data columns (by setting them as
<code>NA</code>) as the dataset is read into AusTraits.</li>
</ol>
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span></span>
<span>    `wood density` <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">wood_density_duplicate</span><span class="op">)</span>, <span class="va">`wood density`</span>, <span class="cn">NA</span><span class="op">)</span>,</span>
<span>    `seed mass (mg)` <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">seed_mass_duplicate</span><span class="op">)</span>, <span class="va">`seed mass (mg)`</span>, <span class="cn">NA</span><span class="op">)</span>,</span>
<span>    `leaf width minimum (mm)` <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">leaf_width_min_duplicate</span><span class="op">)</span>, <span class="va">`leaf width minimum (mm)`</span>, <span class="cn">NA</span><span class="op">)</span>,</span>
<span>    `leaf width maximum (mm)` <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">leaf_width_max_duplicate</span><span class="op">)</span>, <span class="va">`leaf width maximum (mm)`</span>, <span class="cn">NA</span><span class="op">)</span>,</span>
<span>    `leaf length minimum (mm)` <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">leaf_length_min_duplicate</span><span class="op">)</span>, <span class="va">`leaf length minimum (mm)`</span>, <span class="cn">NA</span><span class="op">)</span>,</span>
<span>    `leaf length maximum (mm)` <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">leaf_length_max_duplicate</span><span class="op">)</span>, <span class="va">`leaf length maximum (mm)`</span>, <span class="cn">NA</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p>Difficulties:</p>
<ul>
<li>This method only identifies values as duplicates if they have the
same number of significant figures.</li>
<li>More complex matching may reveal further duplicates. For seed mass
in particular, some studies likely source values from the Kew database
and then round these values. They may similarly source several values
from Kew and then include the mean in their dataset. If their methods or
correspondence with the contributor suggests the values were sourced
from Kew (or another lab, papers, etc.) it is best to filter out all
values, EXCEPT species that are not yet represented in AusTraits for the
trait in question.</li>
</ul>
</div>
</div>
</div>
<div class="section level3">
<h3 id="build-study-report">Build study report<a class="anchor" aria-label="anchor" href="#build-study-report"></a>
</h3>
<p>The report is located in the <code>export</code> folder.</p>
<p>Check the study report to ensure:</p>
<ul>
<li>All possible metadata fields were filled in<br>
</li>
<li>The locations plot sensibly on the map of Australia<br>
</li>
<li>For numeric traits, the trait values plot sensibly relative to other
studies<br>
</li>
<li>The list of unknown/unmatched species doesn’t include names you
think should be recognised/aligned</li>
</ul>
<p>If necessary, cycle back through earlier steps to fix any errors,
rebuilding the study report as necessary</p>
<p>At the very end, re-clear formatting, re-run tests, rebuild
AusTraits, rebuild report.</p>
<p>If you’re uncertain, also recheck excluded data and duplicates before
these final steps.</p>
<div class="sourceCode" id="cb71"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">f</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="st">"data"</span>, <span class="va">current_study</span>, <span class="st">"metadata.yml"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/read_metadata.html">read_metadata</a></span><span class="op">(</span><span class="va">f</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../reference/write_metadata.html">write_metadata</a></span><span class="op">(</span><span class="va">f</span><span class="op">)</span></span>
<span></span>
<span><span class="va">dataset_ids</span> <span class="op">&lt;-</span> <span class="va">current_study</span></span>
<span><span class="fu">austraits_run_tests</span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">austraits</span> <span class="op">&lt;-</span> <span class="fu">remake</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/remake/man/make.html" class="external-link">make</a></span><span class="op">(</span><span class="st">"austraits"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/dataset_report.html">dataset_report</a></span><span class="op">(</span><span class="va">current_study</span>, overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>To generate a report for a collection of studies:</p>
<div class="sourceCode" id="cb72"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">dataset_reports</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Falster_2005_1"</span>, <span class="st">"Wright_2002"</span><span class="op">)</span>, overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>Or for all studies:</p>
<div class="sourceCode" id="cb73"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">dataset_reports</span><span class="op">(</span>overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>Add the argument <code>overwrite=TRUE</code> if you already have a
copy of a specific report stored in your computer and want to replace it
with a newer version.</p>
<p>(Reports are written in <a href="https://rstudio.github.io/rmarkdown/" class="external-link">Rmarkdown</a> and generated
via the <a href="https://cran.r-project.org/web/packages/knitr/" class="external-link">knitr</a> package.
The template is stored in <code>scripts/report_study.html</code>).</p>
</div>
</div>
<div class="section level2">
<h2 id="working-with-our-github-repository">Working with our GitHub repository<a class="anchor" aria-label="anchor" href="#working-with-our-github-repository"></a>
</h2>
<p>By far our preferred way of contributing is for you to contribute
files directly into the repository and then send a <a href="https://help.github.com/articles/using-pull-requests/" class="external-link">pull
request</a> with your input. You can do this by</p>
<ul>
<li>(for approved maintainers of austraits.build) Creating a branch,
or</li>
<li>(for others) forking the database in github</li>
</ul>
<p>In short,</p>
<ol style="list-style-type: decimal">
<li>Create a Git branch for your new work, either within the AusTraits
repo (if you are an approved contributor) or as a <a href="https://help.github.com/en/github/getting-started-with-github/fork-a-repo" class="external-link">fork
of the repo</a>.</li>
<li>Make commits and push these up onto the branch.</li>
<li>Make sure everything runs fine before you send a pull request.</li>
<li>When you’re ready to merge in the new features,</li>
</ol>
<p>Before you make a substantial pull request, you should always <a href="https://github.com/traitecoevo/austraits.build/issues" class="external-link">file an
issue</a> and make sure someone from the team agrees that it’s worth
pursuing the problem. If you’ve found a bug, create an associated issue
and illustrate the bug with a minimal <a href="https://www.tidyverse.org/help/#reprex" class="external-link">reprex</a> illustrating
the issue.</p>
<p>If this is not possible, you could email the relevant files (see
above) to the AusTraits email: <a href="mailto:austraits.database@gmail.com" class="email">austraits.database@gmail.com</a></p>
<div class="section level3">
<h3 id="merging-a-pull-request">Merging a pull request<a class="anchor" aria-label="anchor" href="#merging-a-pull-request"></a>
</h3>
<p>There are multiple ways to merge a pull request, including using
GitHub’s built-in options for merging and squashing. When merging a PR,
we ideally want</p>
<ul>
<li>a single commit</li>
<li>to attribute the work to the original author</li>
<li>to run various checks along the way</li>
</ul>
<p>There are two ways to do this. For both, you need to be an approved
maintainer.</p>
<div class="section level4">
<h4 id="merging-in-your-own-pr">Merging in your own PR<a class="anchor" aria-label="anchor" href="#merging-in-your-own-pr"></a>
</h4>
<p>You can merge in your own PR after you’ve had someone else review
it.</p>
<ol style="list-style-type: decimal">
<li>Send the PR</li>
<li>Tag someone to review</li>
<li>Once ready, merge into main choosing “Squash &amp; Merge”, using an
informative commit message.</li>
</ol>
</div>
<div class="section level4">
<h4 id="merging-someone-elses-pr">Merging someone else’s PR<a class="anchor" aria-label="anchor" href="#merging-someone-elses-pr"></a>
</h4>
<p>When merging in someone else’s PR, the built-in options aren’t ideal,
as they either take all of the commits on a branch (ugh, messy), OR make
the commit under the name of the person merging the request.</p>
<p>The workflow below describes how to merge a pull request from the
command line, with a single commit &amp; attributing the work to the
original author. Lets assume a branch of name
<code>Smith_1995</code>.</p>
<p>First, from the master branch in the repo, run the following:</p>
<pre><code>git merge --squash origin/Smith_1995</code></pre>
<p>Then in R</p>
<p>Now back in the terminal</p>
<pre><code>git add .
git commit</code></pre>
<p>Add a commit message, referencing relevant pull requests and issues,
e.g.</p>
<pre><code>Smith_1995: Import new data

For #224, closes #286</code></pre>
<p>And finally, amend the commit author, to reference the person who did
all the work!</p>
<pre><code>git commit --amend --author "XXX &lt;XXX@gmail.com&gt;"</code></pre>
</div>
</div>
<div class="section level3">
<h3 id="commit-messages">Commit messages<a class="anchor" aria-label="anchor" href="#commit-messages"></a>
</h3>
<p>Informative commit messages are ideal. Where possible, these should
reference the issue being addressed. They should clearly describe the
work done and value added to AusTraits in a few, clear, bulleted
points.</p>
</div>
<div class="section level3">
<h3 id="version-updating-making-a-new-release">Version updating &amp; Making a new release<a class="anchor" aria-label="anchor" href="#version-updating-making-a-new-release"></a>
</h3>
<p>Releases of the dataset are snapshots that are archived and available
for use.</p>
<p>We use semantic versioning to label our versions. As discussed in <a href="http://doi.org/10.1093/gigascience/giz035" class="external-link">Falster et al 2019</a>,
semantic versioning can apply to datasets as well as code.</p>
<p>The version number will have 3 components for actual releases, and 4
for development versions. The structure is
<code>major.minor.patch.dev</code>, where <code>dev</code> is at least
9000. The <code>dev</code> component provides a visual signal that this
is a development version. So, if the current version is 0.9.1.9000, the
release be 0.9.2, 0.10.0 or 1.0.0.</p>
<p>Our approach to incrementing version numbers is</p>
<ul>
<li>
<code>major</code>: increment when you make changes to the structure
that are likely incompatible with any code written to work with previous
versions.</li>
<li>
<code>minor</code>: increment to communicate any changes to the
structure that are likely to be compatible with any code written to work
with the previous versions (i.e., allows code to run without error).
Such changes might involve adding new data within the existing
structure, so that the previous dataset version exists as a subset of
the new version. For tabular data, this includes adding columns or rows.
On the other hand, removing data should constitute a major version
because records previously relied on may no longer exist.</li>
<li>
<code>patch</code>: Increment to communicate correction of errors in
the actual data, without any changes to the structure. Such changes are
unlikely to break or change analyses written with the previous version
in a substantial way.</li>
</ul>
<p><img src="figures/giz035fig2.png"></p>
<p><strong>Figure:</strong> Semantic versioning communicates to users
the types of changes that have occurred between successive versions of
an evolving dataset, using a tri-digit label where increments in a
number indicate major, minor, and patch-level changes, respectively.
From <a href="http://doi.org/10.1093/gigascience/giz035" class="external-link">Falster et al
2019</a>, (CC-BY).</p>
<p>The process of making a release is as follows. Note that
corresponding releases and versions are needed in both
<code>austraits</code> and <code>austraits.build</code>:</p>
<ol style="list-style-type: decimal">
<li><p>Update the version number in the DESCRIPTION file, using
`</p></li>
<li><p>Compile <code>austraits.build</code>.</p></li>
<li><p>Update the documentation.</p></li>
<li><p>Commit and push to github.</p></li>
<li><p>Make a release on github, adding version number</p></li>
<li><p>Prepare for the next version by updating version
numbers.</p></li>
</ol>
</div>
</div>
<div class="section level2">
<h2 id="other">Other<a class="anchor" aria-label="anchor" href="#other"></a>
</h2>
<div class="section level3">
<h3 id="file-types">File types<a class="anchor" aria-label="anchor" href="#file-types"></a>
</h3>
<div class="section level4">
<h4 id="csv">CSV<a class="anchor" aria-label="anchor" href="#csv"></a>
</h4>
<p>A comma-separated values (CSV) file is a delimited text file that
uses a comma to separate values. Each line of the file is a data record.
Each record consists of one or more fields, separated by commas. This is
a comma format for storing tables of data in a simple text file. You can
edit it an Excel or in a text editor. For more, see <a href="https://en.wikipedia.org/wiki/Comma-separated_values" class="external-link">here</a>.</p>
</div>
<div class="section level4">
<h4 id="yaml">YAML files<a class="anchor" aria-label="anchor" href="#yaml"></a>
</h4>
<p>The <code>yml</code> file extension (pronounced “YAML”) <a href="https://en.wikipedia.org/wiki/YAML" class="external-link">is a type structured data
file</a>, that is both human and machine readable. You can edit it in
any text editor, or in Rstudio. Generally, yml is used in situations
where a table does not suit because of variable lengths and/or nested
structures. It has the advantage over a spreadsheet in that the nested
“headers” can have variable numbers of categories. The data under each
of the hierarchical headings are easily extracted by R.</p>
</div>
</div>
<div class="section level3">
<h3 id="extracting-data-from-pdf-tables">Extracting data from PDF tables<a class="anchor" aria-label="anchor" href="#extracting-data-from-pdf-tables"></a>
</h3>
<p>If you encounter a PDF table of data and need to extract values, this
can be achieved with the <a href="https://github.com/tabulapdf/tabula-java/" class="external-link"><code>tabula-java</code>
tool</a>. There’s actually an R wrapper (called <a href="https://github.com/ropensci/tabulizer" class="external-link"><code>tabulizer</code></a>),
but we haven’t succeeded in getting this running. However, it’s easy
enough to run the java tool at the command line on OSX.</p>
<ol style="list-style-type: decimal">
<li><p><a href="https://github.com/tabulapdf/tabula-java/releases" class="external-link">Download latest
release of <code>tabula-java</code></a> and save the file in your
path</p></li>
<li><p>Run</p></li>
</ol>
<pre><code>java -jar tabula-1.0.3-jar-with-dependencies.jar my_table.pdf -o my_data.csv</code></pre>
<p>This should output the data from the table in
<code>my_table.pdf</code> into the csv <code>my_data.csv</code></p>
<ol start="3" style="list-style-type: decimal">
<li>Clean up in Excel. check especially that the locations of white
spaces are correct.</li>
</ol>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Daniel Falster, Elizabeth Wenk, Rachael Gallagher, ARDC, ARC.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
