<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Structure of AusTraits data compilation • austraits.build</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Structure of AusTraits data compilation">
<meta property="og:description" content="austraits.build">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">austraits.build</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.9.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
    home
  </a>
</li>
<li>
  <a href="../articles/austraits.build.html">Get started</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fas fa-book"></span>
     
    Details
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/austraits_overview.html">Overview of `AusTraits` workflow</a>
    </li>
    <li>
      <a href="../articles/austraits.build.html">Getting started with austraits.build</a>
    </li>
    <li>
      <a href="../articles/austraits_database_structure.html">Structure of `AusTraits` data compilation</a>
    </li>
    <li>
      <a href="../articles/definitions.html">Definitions for traits in `AusTraits`</a>
    </li>
    <li>
      <a href="../articles/austraits_file_structure.html">Repository file structure</a>
    </li>
    <li>
      <a href="../articles/contributing_data.html">Contributing data to `AusTraits`</a>
    </li>
    <li>
      <a href="../articles/adding_data.html">Adding new datasets into `AusTraits`</a>
    </li>
    <li>
      <a href="../articles/Docker.html">Docker for reproducible compute environment</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">
    <span class="fa fa-list"></span>
     
    Functions
  </a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/traitecoevo/austraits.build" class="external-link">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Structure of AusTraits data compilation</h1>
            
            <h4 data-toc-skip class="date">2022-11-27</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/traitecoevo/austraits.build/blob/HEAD/vignettes/austraits_database_structure.Rmd" class="external-link"><code>vignettes/austraits_database_structure.Rmd</code></a></small>
      <div class="hidden name"><code>austraits_database_structure.Rmd</code></div>

    </div>

    
    
<p>This document describes the structure of the AusTraits compilation,
corresponding to <strong>Version 3.0.2</strong> of the dataset.</p>
<p>Note that the information provided below is based on the information
provided within the file <code>austraits.build_schema.yml</code>, which
can be accessed by running <code>get_schema</code> or
<code>system.file("support", "austraits.build_schema.yml", package = "austraits.build")</code>.</p>
<p>AusTraits is essentially a series of linked components, which cross
link against each other::</p>
<pre><code>austraits
├── traits
├── locations
├── contexts
├── methods
├── excluded_data
├── taxonomic_updates
├── taxa
├── contributors
├── sources
├── definitions
├── schema
├── metadata
└── build_info</code></pre>
<p>These include all the data and contextual information submitted with
each contributed dataset.</p>
<div class="section level2">
<h2 id="components">Components<a class="anchor" aria-label="anchor" href="#components"></a>
</h2>
<p>The core components are defined as follows.</p>
<div class="section level3">
<h3 id="traits">traits<a class="anchor" aria-label="anchor" href="#traits"></a>
</h3>
<p><strong>Description:</strong> A table containing measurements of
traits.</p>
<p><strong>Content:</strong></p>
<table class="table table table-striped table-hover table-condensed table-responsive" style="margin-left:30px; width: auto !important; ">
<thead><tr>
<th style="text-align:left;">
key
</th>
<th style="text-align:left;">
value
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
dataset_id
</td>
<td style="text-align:left;">
Primary identifier for each study contributed to AusTraits; most often
these are scientific papers, books, or online resources. By default this
should be the name of the first author and year of publication,
e.g. <code>Falster_2005</code>.
</td>
</tr>
<tr>
<td style="text-align:left;">
taxon_name
</td>
<td style="text-align:left;">
Scientific name of the taxon on which traits were sampled, without
authorship. When possible, this is the currently accepted (botanical) or
valid (zoological) scientific name, but might also be a higher taxonomic
level.
</td>
</tr>
<tr>
<td style="text-align:left;">
observation_id
</td>
<td style="text-align:left;">
A unique integral identifier for the observation, where an observation
is all measurements made on an individual at a single point in time. It
is important for joining traits coming from the same
<code>observation_id</code>. Within each dataset, observation_id’s are
unique combinations of <code>taxon_name</code>,
<code>population_id</code>, <code>individual_id</code>, and
<code>temporal_id</code>.
</td>
</tr>
<tr>
<td style="text-align:left;">
trait_name
</td>
<td style="text-align:left;">
Name of the trait sampled. Allowable values specified in the table
<code>definitions</code>.
</td>
</tr>
<tr>
<td style="text-align:left;">
value
</td>
<td style="text-align:left;">
The measured value of a trait, location property or context property.
</td>
</tr>
<tr>
<td style="text-align:left;">
unit
</td>
<td style="text-align:left;">
Units of the sampled trait value after aligning with AusTraits
standards.
</td>
</tr>
<tr>
<td style="text-align:left;">
entity_type
</td>
<td style="text-align:left;">
A categorical variable specifying the entity corresponding to the trait
values recorded.
</td>
</tr>
<tr>
<td style="text-align:left;">
value_type
</td>
<td style="text-align:left;">
A categorical variable describing the statistical nature of the trait
value recorded.
</td>
</tr>
<tr>
<td style="text-align:left;">
basis_of_value
</td>
<td style="text-align:left;">
A categorical variable describing how the trait value was obtained.
</td>
</tr>
<tr>
<td style="text-align:left;">
replicates
</td>
<td style="text-align:left;">
Number of replicate measurements that comprise a recorded trait
measurement. A numeric value (or range) is ideal and appropriate if the
value type is a <code>mean</code>, <code>median</code>, <code>min</code>
or <code>max</code>. For these value types, if replication is unknown
the entry should be <code>unknown</code>. If the value type is
<code>raw_value</code> the replicate value should be 1. If the trait is
categorical or the value indicates a measurement for an entire species
(or other taxon) replicate value should be <code>.na</code>.
</td>
</tr>
<tr>
<td style="text-align:left;">
basis_of_record
</td>
<td style="text-align:left;">
A categorical variable specifying from which kind of specimen traits
were recorded.
</td>
</tr>
<tr>
<td style="text-align:left;">
life_stage
</td>
<td style="text-align:left;">
A field to indicate the life stage or age class of the entity measured.
Standard values are <code>adult</code>, <code>sapling</code>,
<code>seedling</code> and <code>juvenile</code>.
</td>
</tr>
<tr>
<td style="text-align:left;">
population_id
</td>
<td style="text-align:left;">
A unique integer identifier for a population, where a population is
defined as individuals growing in the same location (location_id
/location_name) and plot (plot_id, a context category) and being
subjected to the same treatment (treatment_id, a context category).
</td>
</tr>
<tr>
<td style="text-align:left;">
individual_id
</td>
<td style="text-align:left;">
A unique integer identifier for an individual, with individuals numbered
sequentially within each dataset by taxon by population grouping. Most
often each row of data represents an individual, but in some datasets
trait data collected on a single individual is presented across multiple
rows of data, such as if the same trait is measured using different
methods or the same individual is measured repeatedly across time.
</td>
</tr>
<tr>
<td style="text-align:left;">
temporal_id
</td>
<td style="text-align:left;">
A unique integer identifier assigned where repeat observations are made
on the same individual (or population, or taxon) across time. The
identifier links to specific information in the context table.
</td>
</tr>
<tr>
<td style="text-align:left;">
source_id
</td>
<td style="text-align:left;">
For datasets that are compilations, an identifier for the original data
source.
</td>
</tr>
<tr>
<td style="text-align:left;">
location_id
</td>
<td style="text-align:left;">
A unique integer identifier for a location, with locations numbered
sequentially within a dataset. The identifier links to specific
information in the location table.
</td>
</tr>
<tr>
<td style="text-align:left;">
entity_context_id
</td>
<td style="text-align:left;">
A unique integer identifier indicating specific contextual properties of
an individual, possibly including the individual’s sex or caste (for
social insects).
</td>
</tr>
<tr>
<td style="text-align:left;">
plot_id
</td>
<td style="text-align:left;">
A unique integer identifier for a plot, where a plot is a distinct
collection of organisms within a single geographic location, such as
plants growing on different aspects or blocks in an experiment. The
identifier links to specific information in the context table.
</td>
</tr>
<tr>
<td style="text-align:left;">
treatment_id
</td>
<td style="text-align:left;">
A unique integer identifier for a treatment, where a treatment is any
experimental manipulation to an organism’s growing/living conditions.
The identifier links to specific information in the context table.
</td>
</tr>
<tr>
<td style="text-align:left;">
collection_date
</td>
<td style="text-align:left;">
Date sample was taken, in the format <code>yyyy-mm-dd</code>,
<code>yyyy-mm</code> or <code>yyyy</code>, depending on the resoluton
specified. Alternatively an overall range for the study can be
indicating, with the starting and ending sample date sepatated by a
<code>/</code>, as in 2010-10/2011-03
</td>
</tr>
<tr>
<td style="text-align:left;">
measurement_remarks
</td>
<td style="text-align:left;">
Brief comments or notes accompanying the trait measurement.
</td>
</tr>
<tr>
<td style="text-align:left;">
method_id
</td>
<td style="text-align:left;">
A unique integer identifier indicating a trait is measured multiple
times on the same entity, with different methods used for each entry.
This field is only used if a single trait is measured using multiple
methods within the same dataset. The identifier links to specific
information in the context table.
</td>
</tr>
<tr>
<td style="text-align:left;">
original_name
</td>
<td style="text-align:left;">
Name given to taxon in the original data supplied by the authors.
</td>
</tr>
</tbody>
</table>
<p><strong>Details:</strong></p>
<div class="section level5">
<h5 id="entity-type">Entity type<a class="anchor" aria-label="anchor" href="#entity-type"></a>
</h5>
<p>An entity is the <code>feature of interest</code>, indicating what a
trait value applies to. While an entity can be just a component of an
organism, within the scope of AusTraits, an <code>individual</code> is
the finest scale entity that can be documented. The same study might
measure some traits at a population-level
(<code>entity = population</code>) and others at an individual-level
(<code>entity = individual</code>).</p>
<p>In detail:</p>
<ul>
<li>
<code>entity_type</code> is a categorical variable specifying the
entity corresponding to the trait values recorded. Possible values
are:</li>
</ul>
<table class="table table table-striped table-hover table-condensed table-responsive" style="margin-left:30px; width: auto !important; ">
<thead><tr>
<th style="text-align:left;">
key
</th>
<th style="text-align:left;">
value
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
individual
</td>
<td style="text-align:left;">
Value comes from a single individual.
</td>
</tr>
<tr>
<td style="text-align:left;">
population
</td>
<td style="text-align:left;">
Value represents a summary statistic from multiple individuals at a
single location.
</td>
</tr>
<tr>
<td style="text-align:left;">
metapopulation
</td>
<td style="text-align:left;">
Value represents a summary statistic from individuals of the taxon
across multiple locations.
</td>
</tr>
<tr>
<td style="text-align:left;">
species
</td>
<td style="text-align:left;">
Value represents a summary statistic for a species or infraspecific
taxon across its range or as estimated by an expert based on their
knowledge of the taxon. Data fitting this category include estimates
from reference books that represent a taxon’s entire range and values
for categorical variables obtained from a reference book or identified
by an expert.
</td>
</tr>
<tr>
<td style="text-align:left;">
genus
</td>
<td style="text-align:left;">
Value represents a summary statistic or expert score for an entire
genus.
</td>
</tr>
<tr>
<td style="text-align:left;">
family
</td>
<td style="text-align:left;">
Value represents a summary statistic or expert score for an entire
family.
</td>
</tr>
<tr>
<td style="text-align:left;">
order
</td>
<td style="text-align:left;">
Value represents a summary statistic or expert score for an entire
order.
</td>
</tr>
</tbody>
</table>
</div>
<div class="section level5">
<h5 id="identifiers">Identifiers<a class="anchor" aria-label="anchor" href="#identifiers"></a>
</h5>
<p>The traits table includes 12 identifiers, <code>dataset_id</code>,
<code>observation_id</code>, <code>taxon_name</code>,
<code>population_id</code>, <code>individual_id</code>,
<code>temporal_id</code>, <code>source_id</code>,
<code>location_id</code>, <code>entity_context_id</code>,
<code>plot_id</code>, <code>treatment_id</code>, and
<code>method_id</code>.</p>
<p><code>dataset_id</code>, <code>source_id</code> and
<code>taxon_name</code> have easy-to-interpret values. The others are
simply integral identifiers that link groups of measurements and are
automatically generated through the AusTraits workflow
(<code>individual_id</code> can be assigned in the metadata file or
automatically generated.)</p>
<p>To expand on the definitions provided above,</p>
<ul>
<li><p><code>observation_id</code> links measurements made on the same
entity (individual, population, or species) at a single point in
time.</p></li>
<li><p><code>population_id</code> indicates entites that share a common
<code>location_id</code>, <code>plot_id</code>, and
<code>treatment_id</code>. It is used to align measurements and
observation_id’s for <code>individuals</code> versus
<code>populations</code> (i.e. distinct <code>entity_types</code>) that
share a common <code>population_id</code>. It is numbered sequentially
within a dataset.</p></li>
<li><p><code>individual_id</code> indicates a unique organisms. It is
numbered sequentially within a dataset by population. Multiple
observations on the same organism across time (with distinct
<code>observation_id</code> values), share a common
<code>individual_id</code>.</p></li>
<li><p><code>temporal_id</code> indicates a distinct point in time and
is used only if there are repeat measurements on a population or
individual across time. The identifier links to context properties
(&amp; their associated information) in the <code>contexts</code> table
for context properties of type <code>temporal</code>.</p></li>
<li><p><code>source_id</code> is applied if not all data within a single
dataset (<code>dataset_id</code>) is from the same source, such as when
a dataset represents a compilation for a meta-analysis.</p></li>
<li><p><code>location_id</code> links to a distinct
<code>location_name</code> and associated
<code>location_properties</code> in the <code>location</code>
table.</p></li>
<li><p><code>entity_context_id</code> links to information in the
<code>contexts</code> table for context properties (&amp; associated
values/descriptions) with category <code>entity_context</code>.
<code>Entity_contexts</code> include organism sex, organism caste and
any other features of an entity that needs to be documented.</p></li>
<li><p><code>plot_id</code> links to information in the
<code>contexts</code> table for context properties (&amp; associated
values/descriptions) with category <code>plot</code>.
<code>Plot contexts</code> include both blocks/plots within an
experimental design as well as any stratified variation within a
location that needs to be documented (e.g. slope position).</p></li>
<li><p><code>treatment_id</code>links to information in the
<code>contexts</code> table for context properties (&amp; associated
values/descriptions) with category <code>treatment</code>.
<code>Treatment contexts</code> are experimental manipulations applied
to groups of individuals.</p></li>
<li><p><code>method_id</code>links to information in the
<code>contexts</code> table for context properties (&amp; associated
values/descriptions) with category <code>method</code>. A
<code>method context</code> indicates that the same trait was measured
on or across individuals using different methods.</p></li>
</ul>
<p>As well, <code>measurement_remarks</code> is used to document brief
comments or notes accompanying the trait measurement.</p>
</div>
<div class="section level5">
<h5 id="life-stage-basis-of-record">Life stage, basis of record<a class="anchor" aria-label="anchor" href="#life-stage-basis-of-record"></a>
</h5>
<ul>
<li><p><code>life_stage</code> is a field to indicate the life stage or
age class of the entity measured. standard values are
<code>adult</code>, <code>sapling</code>, <code>seedling</code> and
<code>juvenile</code>..</p></li>
<li><p><code>basis_of_record</code> is a categorical variable specifying
from which kind of specimen traits were recorded. Possible values
are:</p></li>
</ul>
<table class="table table table-striped table-hover table-condensed table-responsive" style="margin-left:30px; width: auto !important; ">
<thead><tr>
<th style="text-align:left;">
key
</th>
<th style="text-align:left;">
value
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
field
</td>
<td style="text-align:left;">
Traits were recorded on entities living naturally in the field.
</td>
</tr>
<tr>
<td style="text-align:left;">
field_experiment
</td>
<td style="text-align:left;">
Traits were recorded on entities living under experimentally manipulated
conditions in the field.
</td>
</tr>
<tr>
<td style="text-align:left;">
captive_cultivated
</td>
<td style="text-align:left;">
Traits were recorded on entities living in a common garden, arboretum,
or botanical or zoological garden.
</td>
</tr>
<tr>
<td style="text-align:left;">
lab
</td>
<td style="text-align:left;">
Traits were recorded on entities growing in a lab, glasshouse or growth
chamber.
</td>
</tr>
<tr>
<td style="text-align:left;">
preserved_specimen
</td>
<td style="text-align:left;">
Traits were recorded from specimens preserved in a collection, eg.
herbarium or museum
</td>
</tr>
<tr>
<td style="text-align:left;">
literature
</td>
<td style="text-align:left;">
Traits were sourced from values reported in the literature, and where
the basis of record is not otherwise known.
</td>
</tr>
</tbody>
</table>
</div>
<div class="section level5">
<h5 id="value_types">Values, Value types, basis of value<a class="anchor" aria-label="anchor" href="#value_types"></a>
</h5>
<p>Each record in the table of trait data has an associated
<code>value</code>, <code>value_type</code>, and
<code>basis_of_value</code>.</p>
<p><code>Value type:</code> A trait’s <code>value_type</code> is either
<code>numeric</code> or <code>categorical</code>. - For traits with
numerical values, the recorded value has been converted into
standardised units and the AusTraits workflow has confirmed the value
can be converted into a number and lies within the allowable range. -
For categorical variables, records have been aligned through
substitutions to values listed as allowable values (terms) in a trait’s
definition. * we use <code>_</code> for multi-word terms,
e.g. <code>semi_deciduous</code><br>
* we use a space for situations where two values co-occur for the same
entity. For instance, a flora might indicate that a plant species can be
either annual or biennial, in which case the trait is scored as
<code>annual biennial</code>.</p>
<p>Each trait measurement has an associated <code>value_type</code>,
which is a categorical variable describing the statistical nature of the
trait value recorded. Possible values are:</p>
<table class="table table table-striped table-hover table-condensed table-responsive" style="margin-left:30px; width: auto !important; ">
<thead><tr>
<th style="text-align:left;">
key
</th>
<th style="text-align:left;">
value
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
raw
</td>
<td style="text-align:left;">
Value recorded for an entity.
</td>
</tr>
<tr>
<td style="text-align:left;">
minimum
</td>
<td style="text-align:left;">
Value is the minimum of values recorded for an entity.
</td>
</tr>
<tr>
<td style="text-align:left;">
mean
</td>
<td style="text-align:left;">
Value is the median of values recorded for an entity.
</td>
</tr>
<tr>
<td style="text-align:left;">
median
</td>
<td style="text-align:left;">
Value is the median of values recorded for an entity.
</td>
</tr>
<tr>
<td style="text-align:left;">
maximum
</td>
<td style="text-align:left;">
Value is the maximum of values recorded for an entity.
</td>
</tr>
<tr>
<td style="text-align:left;">
mode
</td>
<td style="text-align:left;">
Value is the mode of values recorded for an entity. This is the
appropriate value type for a categorical trait value.
</td>
</tr>
<tr>
<td style="text-align:left;">
range
</td>
<td style="text-align:left;">
Value is a range of values recorded for an entity.
</td>
</tr>
<tr>
<td style="text-align:left;">
bin
</td>
<td style="text-align:left;">
Value for an entity falls within specified limits.
</td>
</tr>
<tr>
<td style="text-align:left;">
unknown
</td>
<td style="text-align:left;">
Not currently known.
</td>
</tr>
</tbody>
</table>
<p>Each trait measurement also has an associated
<code>basis_of_value</code>, which is a categorical variable describing
how the trait value was obtained. Possible values are:</p>
<table class="table table table-striped table-hover table-condensed table-responsive" style="margin-left:30px; width: auto !important; ">
<thead><tr>
<th style="text-align:left;">
key
</th>
<th style="text-align:left;">
value
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
measurement
</td>
<td style="text-align:left;">
Value is the result of a measurement(s) made on a specimen(s).
</td>
</tr>
<tr>
<td style="text-align:left;">
expert_score
</td>
<td style="text-align:left;">
Value has been estimated by an expert based on their knowledge of the
entity.
</td>
</tr>
<tr>
<td style="text-align:left;">
model_derived
</td>
<td style="text-align:left;">
Value is derived from a statistical model, for example via gap-filling.
</td>
</tr>
<tr>
<td style="text-align:left;">
unknown
</td>
<td style="text-align:left;">
Not currently known.
</td>
</tr>
</tbody>
</table>
<p>AusTraits does not include intra-individual observations made at a
single point in time. When multiple measurements per individual are
submitted to AusTraits, we take the mean of the values and record the
value_type as <code>mean</code> and indicate under replicates the number
of measurements made.</p>
</div>
</div>
<div class="section level3">
<h3 id="locations">locations<a class="anchor" aria-label="anchor" href="#locations"></a>
</h3>
<p><strong>Description:</strong> A table containing observations of
location/site characteristics associated with information in
<code>traits</code>. Cross referencing between the two dataframes is
possible using combinations of the variables <code>dataset_id</code>,
<code>location_name</code>.</p>
<p><strong>Content:</strong></p>
<table class="table table table-striped table-hover table-condensed table-responsive" style="margin-left:30px; width: auto !important; ">
<thead><tr>
<th style="text-align:left;">
key
</th>
<th style="text-align:left;">
value
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
dataset_id
</td>
<td style="text-align:left;">
Primary identifier for each study contributed to AusTraits; most often
these are scientific papers, books, or online resources. By default this
should be the name of the first author and year of publication,
e.g. <code>Falster_2005</code>.
</td>
</tr>
<tr>
<td style="text-align:left;">
location_id
</td>
<td style="text-align:left;">
A unique integer identifier for a location, with locations numbered
sequentially within a dataset. The identifier links to specific
information in the location table.
</td>
</tr>
<tr>
<td style="text-align:left;">
location_name
</td>
<td style="text-align:left;">
location name
</td>
</tr>
<tr>
<td style="text-align:left;">
location_property
</td>
<td style="text-align:left;">
The location characteristic being recorded. The name should include
units of measurement, e.g. <code>MAT (C)</code>. Ideally we have at
least the following variables for each location,
<code>longitude (deg)</code>, <code>latitude (deg)</code>,
<code>description</code>.
</td>
</tr>
<tr>
<td style="text-align:left;">
value
</td>
<td style="text-align:left;">
The measured value of a location property.
</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="contexts">contexts<a class="anchor" aria-label="anchor" href="#contexts"></a>
</h3>
<p><strong>Description:</strong> A table containing observations of
contextual characteristics associated with information in
<code>traits</code>. Cross referencing between the two dataframes is
possible using combinations of the variables <code>dataset_id</code>,
<code>link_id</code>, and <code>link_vals</code>.</p>
<p><strong>Content:</strong></p>
<table class="table table table-striped table-hover table-condensed table-responsive" style="margin-left:30px; width: auto !important; ">
<thead><tr>
<th style="text-align:left;">
key
</th>
<th style="text-align:left;">
value
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
dataset_id
</td>
<td style="text-align:left;">
Primary identifier for each study contributed to AusTraits; most often
these are scientific papers, books, or online resources. By default this
should be the name of the first author and year of publication,
e.g. <code>Falster_2005</code>.
</td>
</tr>
<tr>
<td style="text-align:left;">
context_property
</td>
<td style="text-align:left;">
The contextual characteristic being recorded. If applicable, name should
include units of measurement, e.g. <code>CO2 concentration (ppm)</code>.
</td>
</tr>
<tr>
<td style="text-align:left;">
category
</td>
<td style="text-align:left;">
The category of context property, with options being <code>plot</code>,
<code>treatment</code>, <code>individual_context</code>,
<code>temporal</code> and <code>method</code>.
</td>
</tr>
<tr>
<td style="text-align:left;">
value
</td>
<td style="text-align:left;">
The measured value of a context property.
</td>
</tr>
<tr>
<td style="text-align:left;">
description
</td>
<td style="text-align:left;">
Description of a specific context property value.
</td>
</tr>
<tr>
<td style="text-align:left;">
link_id
</td>
<td style="text-align:left;">
Variable indicating which identifier column in the traits table contains
the specified <code>link_vals</code>.
</td>
</tr>
<tr>
<td style="text-align:left;">
link_vals
</td>
<td style="text-align:left;">
Unique integer identifiers that link between identifier columns in the
<code>traits</code> table and the contextual properties/values in the
<code>contexts</code> table.
</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="methods">methods<a class="anchor" aria-label="anchor" href="#methods"></a>
</h3>
<p><strong>Description:</strong> A table containing details on methods
with which data were collected, including time frame and source. Cross
referencing with the <code>traits</code> table is possible using
combinations of the variables <code>dataset_id</code>,
<code>trait_name</code>.</p>
<p><strong>Content:</strong></p>
<table class="table table table-striped table-hover table-condensed table-responsive" style="margin-left:30px; width: auto !important; ">
<thead><tr>
<th style="text-align:left;">
key
</th>
<th style="text-align:left;">
value
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
dataset_id
</td>
<td style="text-align:left;">
Primary identifier for each study contributed to AusTraits; most often
these are scientific papers, books, or online resources. By default this
should be the name of the first author and year of publication,
e.g. <code>Falster_2005</code>.
</td>
</tr>
<tr>
<td style="text-align:left;">
trait_name
</td>
<td style="text-align:left;">
Name of the trait sampled. Allowable values specified in the table
<code>definitions</code>.
</td>
</tr>
<tr>
<td style="text-align:left;">
methods
</td>
<td style="text-align:left;">
A textual description of the methods used to collect the trait data.
Whenever available, methods are taken near-verbatim from the referenced
source. Methods can include descriptions such as ‘measured on botanical
collections’, ‘data from the literature’, or a detailed description of
the field or lab methods used to collect the data.
</td>
</tr>
<tr>
<td style="text-align:left;">
description
</td>
<td style="text-align:left;">
A 1-2 sentence description of the purpose of the study.
</td>
</tr>
<tr>
<td style="text-align:left;">
sampling_strategy
</td>
<td style="text-align:left;">
A written description of how study locations were selected and how study
individuals were selected. When available, this information is lifted
verbatim from a published manuscript. For preserved specimens, this
field ideally indicates which records were ‘sampled’ to measure a
specific trait.
</td>
</tr>
<tr>
<td style="text-align:left;">
source_primary_key
</td>
<td style="text-align:left;">
Citation key for the primary source in <code>sources</code>. The key is
typically formatted as <code>Surname_year</code>.
</td>
</tr>
<tr>
<td style="text-align:left;">
source_primary_citation
</td>
<td style="text-align:left;">
Citation for the primary source. This detail is generated from the
primary source in the metadata.
</td>
</tr>
<tr>
<td style="text-align:left;">
source_secondary_key
</td>
<td style="text-align:left;">
Citation key for the secondary source in <code>sources</code>. The key
is typically formatted as <code>Surname_year</code>.
</td>
</tr>
<tr>
<td style="text-align:left;">
source_secondary_citation
</td>
<td style="text-align:left;">
Citations for the secondary source. This detail is generated from the
secondary source in the metadata.
</td>
</tr>
<tr>
<td style="text-align:left;">
source_original_dataset_key
</td>
<td style="text-align:left;">
Citation key for the original dataset_id in sources; for compilations.
The key is typically formatted as <code>Surname_year</code>.
</td>
</tr>
<tr>
<td style="text-align:left;">
source_original_dataset_citation
</td>
<td style="text-align:left;">
Citations for the original dataset_id in sources; for compilationse.
This detail is generated from the original source in the metadata.
</td>
</tr>
<tr>
<td style="text-align:left;">
data_collectors
</td>
<td style="text-align:left;">
The person (people) leading data collection for this study.
</td>
</tr>
<tr>
<td style="text-align:left;">
assistants
</td>
<td style="text-align:left;">
Names of additional people who played a more minor role in data
collection for the study.
</td>
</tr>
<tr>
<td style="text-align:left;">
austraits_curators
</td>
<td style="text-align:left;">
Names of AusTraits team member(s) who contacted the data collectors and
added the study to the AusTraits repository.
</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="exluded_data">exluded_data<a class="anchor" aria-label="anchor" href="#exluded_data"></a>
</h3>
<p><strong>Description:</strong> A table of data that did not pass
quality tests and so were excluded from the master dataset. The
structure is identical to that presented in the <code>traits</code>
table, only with an extra column called <code>error</code> indicating
why the record was excluded. Common reasons are
missing_unit_conversions, missing_value, and
unsupported_trait_value.</p>
<p><strong>Content:</strong></p>
<table class="table table table-striped table-hover table-condensed table-responsive" style="margin-left:30px; width: auto !important; ">
<thead><tr>
<th style="text-align:left;">
key
</th>
<th style="text-align:left;">
value
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
error
</td>
<td style="text-align:left;">
Indicating why the record was excluded. Common reasons are
missing_unit_conversions, missing_value, and unsupported_trait_value.
</td>
</tr>
<tr>
<td style="text-align:left;">
dataset_id
</td>
<td style="text-align:left;">
Primary identifier for each study contributed to AusTraits; most often
these are scientific papers, books, or online resources. By default this
should be the name of the first author and year of publication,
e.g. <code>Falster_2005</code>.
</td>
</tr>
<tr>
<td style="text-align:left;">
taxon_name
</td>
<td style="text-align:left;">
Scientific name of the taxon on which traits were sampled, without
authorship. When possible, this is the currently accepted (botanical) or
valid (zoological) scientific name, but might also be a higher taxonomic
level.
</td>
</tr>
<tr>
<td style="text-align:left;">
observation_id
</td>
<td style="text-align:left;">
A unique integral identifier for the observation, where an observation
is all measurements made on an individual at a single point in time. It
is important for joining traits coming from the same
<code>observation_id</code>. Within each dataset, observation_id’s are
unique combinations of <code>taxon_name</code>,
<code>population_id</code>, <code>individual_id</code>, and
<code>temporal_id</code>.
</td>
</tr>
<tr>
<td style="text-align:left;">
trait_name
</td>
<td style="text-align:left;">
Name of the trait sampled. Allowable values specified in the table
<code>definitions</code>.
</td>
</tr>
<tr>
<td style="text-align:left;">
value
</td>
<td style="text-align:left;">
The measured value of a trait.
</td>
</tr>
<tr>
<td style="text-align:left;">
unit
</td>
<td style="text-align:left;">
Units of the sampled trait value after aligning with AusTraits
standards.
</td>
</tr>
<tr>
<td style="text-align:left;">
entity_type
</td>
<td style="text-align:left;">
A categorical variable specifying the entity corresponding to the trait
values recorded.
</td>
</tr>
<tr>
<td style="text-align:left;">
value_type
</td>
<td style="text-align:left;">
A categorical variable describing the statistical nature of the trait
value recorded.
</td>
</tr>
<tr>
<td style="text-align:left;">
basis_of_value
</td>
<td style="text-align:left;">
A categorical variable describing how the trait value was obtained.
</td>
</tr>
<tr>
<td style="text-align:left;">
replicates
</td>
<td style="text-align:left;">
Number of replicate measurements that comprise a recorded trait
measurement. A numeric value (or range) is ideal and appropriate if the
value type is a <code>mean</code>, <code>median</code>, <code>min</code>
or <code>max</code>. For these value types, if replication is unknown
the entry should be <code>unknown</code>. If the value type is
<code>raw_value</code> the replicate value should be 1. If the trait is
categorical or the value indicates a measurement for an entire species
(or other taxon) replicate value should be <code>.na</code>.
</td>
</tr>
<tr>
<td style="text-align:left;">
basis_of_record
</td>
<td style="text-align:left;">
A categorical variable specifying from which kind of specimen traits
were recorded.
</td>
</tr>
<tr>
<td style="text-align:left;">
life_stage
</td>
<td style="text-align:left;">
A field to indicate the life stage or age class of the entity measured.
Standard values are <code>adult</code>, <code>sapling</code>,
<code>seedling</code> and <code>juvenile</code>.
</td>
</tr>
<tr>
<td style="text-align:left;">
population_id
</td>
<td style="text-align:left;">
A unique integer identifier for a population, where a population is
defined as individuals growing in the same location (location_id
/location_name) and plot (plot_id, a context category) and being
subjected to the same treatment (treatment_id, a context category).
</td>
</tr>
<tr>
<td style="text-align:left;">
individual_id
</td>
<td style="text-align:left;">
A unique integer identifier for an individual, with individuals numbered
sequentially within each dataset by taxon by population grouping. Most
often each row of data represents an individual, but in some datasets
trait data collected on a single individual is presented across multiple
rows of data, such as if the same trait is measured using different
methods or the same individual is measured repeatedly across time.
</td>
</tr>
<tr>
<td style="text-align:left;">
temporal_id
</td>
<td style="text-align:left;">
A unique integer identifier assigned where repeat observations are made
on the same individual (or population, or taxon) across time. The
identifier links to specific information in the context table.
</td>
</tr>
<tr>
<td style="text-align:left;">
source_id
</td>
<td style="text-align:left;">
For datasets that are compilations, an identifier for the original data
source.
</td>
</tr>
<tr>
<td style="text-align:left;">
location_id
</td>
<td style="text-align:left;">
A unique integer identifier for a location, with locations numbered
sequentially within a dataset. The identifier links to specific
information in the location table.
</td>
</tr>
<tr>
<td style="text-align:left;">
entity_context_id
</td>
<td style="text-align:left;">
A unique integer identifier indicating specific contextual properties of
an individual, possibly including the individual’s sex or caste (for
social insects).
</td>
</tr>
<tr>
<td style="text-align:left;">
plot_id
</td>
<td style="text-align:left;">
A unique integer identifier for a plot, where a plot is a distinct
collection of organisms within a single geographic location, such as
plants growing on different aspects or blocks in an experiment. The
identifier links to specific information in the context table.
</td>
</tr>
<tr>
<td style="text-align:left;">
treatment_id
</td>
<td style="text-align:left;">
A unique integer identifier for a treatment, where a treatment is any
experimental manipulation to an organism’s growing/living conditions.
The identifier links to specific information in the context table.
</td>
</tr>
<tr>
<td style="text-align:left;">
collection_date
</td>
<td style="text-align:left;">
Date sample was taken, in the format <code>yyyy-mm-dd</code>,
<code>yyyy-mm</code> or <code>yyyy</code>, depending on the resoluton
specified. Alternatively an overall range for the study can be
indicating, with the starting and ending sample date sepatated by a
<code>/</code>, as in 2010-10/2011-03
</td>
</tr>
<tr>
<td style="text-align:left;">
measurement_remarks
</td>
<td style="text-align:left;">
Brief comments or notes accompanying the trait measurement.
</td>
</tr>
<tr>
<td style="text-align:left;">
method_id
</td>
<td style="text-align:left;">
A unique integer identifier indicating a trait is measured multiple
times on the same entity, with different methods used for each entry.
This field is only used if a single trait is measured using multiple
methods within the same dataset. The identifier links to specific
information in the context table.
</td>
</tr>
<tr>
<td style="text-align:left;">
original_name
</td>
<td style="text-align:left;">
Name given to taxon in the original data supplied by the authors.
</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="taxa">taxa<a class="anchor" aria-label="anchor" href="#taxa"></a>
</h3>
<p><strong>Description:</strong> A table containing details on taxa that
are included in the table <a href="#traits"><code>traits</code></a>. We
have attempted to align species names with known taxonomic units in the
<a href="https://biodiversity.org.au/nsl/services/apc" class="external-link"><code>Australian Plant Census</code>
(APC)</a> and/or the <a href="https://biodiversity.org.au/nsl/services/APNI" class="external-link"><code>Australian Plant Names Index</code>
(APNI)</a>; the sourced information is released under a CC-BY3
license.</p>
<p>Version 0.9.0 of AusTraits contains records for 1245 different
taxa.</p>
<p><strong>Content:</strong></p>
<table class="table table table-striped table-hover table-condensed table-responsive" style="margin-left:30px; width: auto !important; ">
<thead><tr>
<th style="text-align:left;">
key
</th>
<th style="text-align:left;">
value
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
taxon_name
</td>
<td style="text-align:left;">
Scientific name of the taxon on which traits were sampled, without
authorship. When possible, this is the currently accepted (botanical) or
valid (zoological) scientific name, but might also be a higher taxonomic
level.
</td>
</tr>
<tr>
<td style="text-align:left;">
taxonomic_reference
</td>
<td style="text-align:left;">
Name of the taxonomy (tree) that contains this concept. ie. APC, AusMoss
etc.
</td>
</tr>
<tr>
<td style="text-align:left;">
taxon_rank
</td>
<td style="text-align:left;">
The taxonomic rank of the most specific name in the scientific name.
</td>
</tr>
<tr>
<td style="text-align:left;">
trinomial
</td>
<td style="text-align:left;">
The infraspecific taxon name match for an original name. This column is
assigned <code>na</code> for taxon name that are at a broader
taxonomic_resolution.
</td>
</tr>
<tr>
<td style="text-align:left;">
binomial
</td>
<td style="text-align:left;">
The species-level taxon name match for an original name. This column is
assigned <code>na</code> for taxon name that are at a broader
taxonomic_resolution.
</td>
</tr>
<tr>
<td style="text-align:left;">
genus
</td>
<td style="text-align:left;">
Genus of the taxon without authorship.
</td>
</tr>
<tr>
<td style="text-align:left;">
family
</td>
<td style="text-align:left;">
Family of the taxon.
</td>
</tr>
<tr>
<td style="text-align:left;">
taxon_distribution
</td>
<td style="text-align:left;">
Known distribution of the taxon, by Australian state.
</td>
</tr>
<tr>
<td style="text-align:left;">
establishment_means
</td>
<td style="text-align:left;">
Statement about whether an organism or organisms have been introduced to
a given place and time through the direct or indirect activity of modern
humans.
</td>
</tr>
<tr>
<td style="text-align:left;">
taxonomic_status
</td>
<td style="text-align:left;">
The status of the use of the scientificName as a label for the taxon in
regard to the ‘accepted (or valid) taxonomy’. The assigned taxonomic
status must be linked to a specific taxonomic reference that defines the
concept.
</td>
</tr>
<tr>
<td style="text-align:left;">
scientific_name
</td>
<td style="text-align:left;">
The full scientific name, with authorship and date information if known.
</td>
</tr>
<tr>
<td style="text-align:left;">
scientific_name_authorship
</td>
<td style="text-align:left;">
The authorship information for the scientific name formatted according
to the conventions of the applicable.
</td>
</tr>
<tr>
<td style="text-align:left;">
taxon_id
</td>
<td style="text-align:left;">
An identifier for the set of taxon information (data associated with the
taxon class). May be a global unique identifier or an identifier
specific to the data set. Must be resolvable within this dataset.
</td>
</tr>
<tr>
<td style="text-align:left;">
scientific_name_id
</td>
<td style="text-align:left;">
An identifier for the set of taxon information (data associated with the
taxon class). May be a global unique identifier or an identifier
specific to the data set. Must be resolvable within this dataset.
</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="taxonomic_updates">taxonomic_updates<a class="anchor" aria-label="anchor" href="#taxonomic_updates"></a>
</h3>
<p><strong>Description:</strong> A table of all taxonomic changes
implemented in the construction of AusTraits. Changes are determined by
comparing the originally submitted taxon name against the taxonomic
names listed in the taxonomic reference files, best placed in a
subfolder in the <code>config</code> folder . Cross referencing with the
<code>traits</code> table is possible using combinations of the
variables <code>dataset_id</code> and <code>taxon_name</code>.</p>
<p><strong>Content:</strong></p>
<table class="table table table-striped table-hover table-condensed table-responsive" style="margin-left:30px; width: auto !important; ">
<thead><tr>
<th style="text-align:left;">
key
</th>
<th style="text-align:left;">
value
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
dataset_id
</td>
<td style="text-align:left;">
Primary identifier for each study contributed to AusTraits; most often
these are scientific papers, books, or online resources. By default this
should be the name of the first author and year of publication,
e.g. <code>Falster_2005</code>.
</td>
</tr>
<tr>
<td style="text-align:left;">
original_name
</td>
<td style="text-align:left;">
Name given to taxon in the original data supplied by the authors.
</td>
</tr>
<tr>
<td style="text-align:left;">
cleaned_name
</td>
<td style="text-align:left;">
The taxon name without authorship after implementing automated syntax
standardisation and spelling changes as well as manually encoded syntax
alignments for this taxon in the metadata file for the corresponding
<code>dataset_id</code>. This name has not yet been matched to the
currently accepted (botanical) or valid (zoological) taxon name in cases
where there are taxonomic synonyms, isonyms, orthographic variants, etc.
</td>
</tr>
<tr>
<td style="text-align:left;">
taxonomic_resolution
</td>
<td style="text-align:left;">
The rank of the most specific taxon name (or scientific name) to which a
submitted orignal name resolves.
</td>
</tr>
<tr>
<td style="text-align:left;">
cleaned_scientific_name_id
</td>
<td style="text-align:left;">
An identifier for the cleaned name before it is updated to the currently
accepted name usage. This may be a global unique identifier or an
identifier specific to the data set. Must be resolvable within this
dataset.
</td>
</tr>
<tr>
<td style="text-align:left;">
cleaned_name_taxonomic_status
</td>
<td style="text-align:left;">
The status of the use of the <code>cleaned_name</code> as a label for a
taxon. Requires taxonomic opinion to define the scope of a taxon. Rules
of priority then are used to define the taxonomic status of the
nomenclature contained in that scope, combined with the experts opinion.
It must be linked to a specific taxonomic reference that defines the
concept.
</td>
</tr>
<tr>
<td style="text-align:left;">
cleaned_name_alternative_taxonomic_status
</td>
<td style="text-align:left;">
The taxonomic status of alternative taxonomic records with
<code>cleaned_name</code> as the accepted (botanical) or valid
(zoological) taxon name.
</td>
</tr>
<tr>
<td style="text-align:left;">
taxon_id
</td>
<td style="text-align:left;">
An identifier for the set of taxon information (data associated with the
taxon class). May be a global unique identifier or an identifier
specific to the data set. Must be resolvable within this dataset.
</td>
</tr>
<tr>
<td style="text-align:left;">
taxon_name
</td>
<td style="text-align:left;">
Scientific name of the taxon on which traits were sampled, without
authorship. When possible, this is the currently accepted (botanical) or
valid (zoological) scientific name, but might also be a higher taxonomic
level.
</td>
</tr>
</tbody>
</table>
<p>Both the original and the updated taxon names are included in the <a href="#traits"><code>traits</code></a> table.</p>
</div>
<div class="section level3">
<h3 id="definitions">definitions<a class="anchor" aria-label="anchor" href="#definitions"></a>
</h3>
<p><strong>Description:</strong> A copy of the definitions for all
tables and terms. Information included here was used to process data and
generate any documentation for the study.</p>
<p><strong>Details on trait definitions:</strong> The allowable trait
names and trait values are defined in the definitions file. Each trait
is labelled as either <code>numeric</code> or <code>categorical</code>.
An example of each type is as follows. For the full list, see the <a href="http://traitecoevo.github.io/austraits.build/articles/definitions.html">Trait
definitions vignette</a>.</p>
<p><strong>leaf_mass_per_area</strong></p>
<ul>
<li>number of records: 0</li>
<li>number of studies: 0</li>
</ul>
<p><strong>woodiness</strong></p>
<ul>
<li>number of records: 0</li>
<li>number of studies: 0</li>
</ul>
</div>
<div class="section level3">
<h3 id="contributors">contributors<a class="anchor" aria-label="anchor" href="#contributors"></a>
</h3>
<p><strong>Description:</strong> A table of people contributing to each
study.</p>
<p><strong>Content:</strong></p>
<table class="table table table-striped table-hover table-condensed table-responsive" style="margin-left:30px; width: auto !important; ">
<thead><tr>
<th style="text-align:left;">
key
</th>
<th style="text-align:left;">
value
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
dataset_id
</td>
<td style="text-align:left;">
Primary identifier for each study contributed to AusTraits; most often
these are scientific papers, books, or online resources. By default this
should be the name of the first author and year of publication,
e.g. <code>Falster_2005</code>.
</td>
</tr>
<tr>
<td style="text-align:left;">
last_name
</td>
<td style="text-align:left;">
Last name of the data collector.
</td>
</tr>
<tr>
<td style="text-align:left;">
given_name
</td>
<td style="text-align:left;">
Given names of the data collector.
</td>
</tr>
<tr>
<td style="text-align:left;">
ORCID
</td>
<td style="text-align:left;">
ORCID of the data collector.
</td>
</tr>
<tr>
<td style="text-align:left;">
affiliation
</td>
<td style="text-align:left;">
Last known institution or affiliation.
</td>
</tr>
<tr>
<td style="text-align:left;">
additional_role
</td>
<td style="text-align:left;">
Additional roles of data collector, mostly contact person.
</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="sources">sources<a class="anchor" aria-label="anchor" href="#sources"></a>
</h3>
<p>For each dataset in the compilation there is the option to list
primary and secondary citations. The primary citation is defined as,
<code>The secondary citation is defined as,</code></p>
<p>The element <code>sources</code> includes bibtex versions of all
sources which can be imported into your reference library:</p>
<pre><code><span><span class="fu">RefManageR</span><span class="fu">::</span><span class="fu"><a href="https://docs.ropensci.org/RefManageR/reference/WriteBib.html" class="external-link">WriteBib</a></span><span class="op">(</span><span class="va">austraits</span><span class="op">$</span><span class="va">sources</span>, <span class="st">"refs.bib"</span><span class="op">)</span> <span class="co"># write all sources to file</span></span>
<span><span class="fu">RefManageR</span><span class="fu">::</span><span class="fu"><a href="https://docs.ropensci.org/RefManageR/reference/WriteBib.html" class="external-link">WriteBib</a></span><span class="op">(</span><span class="va">austraits</span><span class="op">$</span><span class="va">sources</span><span class="op">[</span><span class="st">"Falster_2005_1"</span><span class="op">]</span>, <span class="st">"refs.bib"</span><span class="op">)</span> <span class="co"># write a single reference to a file</span></span></code></pre>
<p>Or individually viewed:</p>
<pre><code><span><span class="va">austraits</span><span class="op">$</span><span class="va">sources</span><span class="op">[</span><span class="st">"Falster_2005_1"</span><span class="op">]</span></span></code></pre>
<p>A formatted version of the sources also exists within the table <a href="#methods">methods</a>.</p>
</div>
<div class="section level3">
<h3 id="metadata">metadata<a class="anchor" aria-label="anchor" href="#metadata"></a>
</h3>
<p><strong>Description:</strong> Metadata associated with the dataset,
including title, creators, license, subject, funding sources.</p>
</div>
<div class="section level3">
<h3 id="build_info">build_info<a class="anchor" aria-label="anchor" href="#build_info"></a>
</h3>
<p><strong>Description:</strong> A description of the computing
environment used to create this version of the dataset, including
version number, git commit and R session_info.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Daniel Falster, Rachael Gallagher, Elizabeth Wenk, ARDC, ARC.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
