<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Building of the `AusTraits` data compilation • austraits.build</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Building of the `AusTraits` data compilation">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">austraits.build</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.9.1.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Contributing.html">Contributing to `AusTraits`</a>
    </li>
    <li>
      <a href="../articles/Docker.html">Docker container for reproducible compute environment</a>
    </li>
    <li>
      <a href="../articles/TipTricks.html">Tips &amp; Tricks for developers of `AusTraits`</a>
    </li>
    <li>
      <a href="../articles/Trait_definitions.html">Definitions for traits in `AusTraits`</a>
    </li>
    <li>
      <a href="../articles/Usage.html">Building of the `AusTraits` data compilation</a>
    </li>
    <li>
      <a href="../articles/test.html">test</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="Usage_files/kePrint-0.0.1/kePrint.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Building of the <code>AusTraits</code> data compilation</h1>
                        <h4 class="author">Daniel Falster, Rachael Gallagher, Sam Andrew, Dony Indiarto, James Lawson, Lizzy Wenk</h4>
            
            <h4 class="date">2020-03-17</h4>
      
      
      <div class="hidden name"><code>Usage.Rmd</code></div>

    </div>

    
    
<!-- hack to get indentation on 3rd level of floating TOC; see
https://stackoverflow.com/questions/46201753/rmarkdown-indentation-of-toc-items-in-html-output
 -->
<script>
$(document).ready(function() {
  $items = $('div#TOC li');
  $items.each(function(idx) {
    num_ul = $(this).parentsUntil('#TOC').length;
    $(this).css({'text-indent': num_ul * 10, 'padding-left': 0});
  });

});
</script><div id="overview" class="section level1">
<h1 class="hasAnchor">
<a href="#overview" class="anchor"></a>Overview</h1>
<p>This document describes the building of AusTraits - a compilation of plant traits for the Australian Flora. AusTraits harmonises data on 202 traits from 0 different sources, including field campaigns, published literature, taxonomic monographs, and individual species descriptions. Traits link to ecological strategy variation and vary in scope from physiological measures of performance (e.g. photosynthetic gas exchange, water-use efficiency) to morphological parameters (e.g. leaf size, seed size, maximum height). AusTraits contains curated and harmonised species- and genus-level observations coupled to, where available, contextual information on site properties.</p>
<p>This document provides information on the structure of AusTraits and corresponds to Version of the dataset.</p>
<!-- An overview of the actual data is provided in another document: see XXXX. -->
</div>
<div id="mission" class="section level1">
<h1 class="hasAnchor">
<a href="#mission" class="anchor"></a>Mission</h1>
<p>We envision AusTraits as an on-going collaborative community resource that:</p>
<ol style="list-style-type: decimal">
<li>Increases our collective understanding the Australian flora; and</li>
<li>Facilitates accumulating and sharing of trait data;</li>
<li>Builds a sense of community among contributors and users; and</li>
<li>Aspires to fully transparent and reproducible research of highest standard.</li>
</ol>
<p>To achieve these goals, we’re building a resource with the following objectives:</p>
<ul>
<li>The workflow for harmonising different datasets is made fully open and reproducible;</li>
<li>All data are shared under a standard open license permitting reuse;</li>
<li>Successive versions of the data remain available to ensure reproducibility of research;</li>
<li>Contributions towards AusTraits are recognised via invitations to co-author data papers, i.e. releases of the data resource, and subsequent citation;</li>
<li>Users of AusTraits are encouraged to recognise contributions (via citation) and also contribute towards the further improvement of the database.</li>
</ul>
<p>Please note that the AusTraits project is released with a <a href="CODE_OF_CONDUCT.html">Contributor Code of Conduct</a>. By contributing to this project, you agree to abide by its terms.</p>
<div id="license" class="section level2">
<h2 class="hasAnchor">
<a href="#license" class="anchor"></a>License</h2>
<p>Raw and harmonised data are made available under the Creative Commons CC-BY license XXXX</p>
</div>
<div id="access" class="section level2">
<h2 class="hasAnchor">
<a href="#access" class="anchor"></a>Access</h2>
<p>XXXXX</p>
<p>Wilkinson et al 2015 The FAIR Guiding Principles for scientific data management and stewardship <a href="http://doi.org/10.1038/sdata.2016.18" class="uri">http://doi.org/10.1038/sdata.2016.18</a></p>
</div>
<div id="citations" class="section level2">
<h2 class="hasAnchor">
<a href="#citations" class="anchor"></a>Citation</h2>
<p>XXXXX</p>
<p>Force 11 data citation: <a href="https://www.force11.org/datacitationprinciples" class="uri">https://www.force11.org/datacitationprinciples</a></p>
</div>
<div id="transparency" class="section level2">
<h2 class="hasAnchor">
<a href="#transparency" class="anchor"></a>Transparency</h2>
<p>Prior to the development of AusTraits, data on Australian plant traits existed as a series of largely disconnected datasets collected by individual laboratories or initiatives. Our goal is to harmonise these different sources.</p>
<p>vignettes/assets/Workflow.png</p>
<p>To facilitate the sharing of trait data under the FAIR principles (Findability, Accessibility, Interoperability, and Reusability), AusTraits uses a fully-reproducible workflow which exposes the decisions made in the processing of data into a harmonised and curated dataset (Figure 1). For instance, AusTraits makes no alterations to the primary datasets provided by co-authors, instead making use of R packages such as <code>make</code> and <code>remake</code> to configure and harmonise (standardise) data into a single curated resource. Several additional quality-assurance steps are undertaken, including detailed error-checking by dataset custodians.</p>
<p>Error checking is facilitated by the generation of bespoke reports for each dataset that can be compared to all observation in AusTraits and by the implementation of a series of tests in the <code>remake</code> pipeline that flag trait values that cannot be incorporated into AusTraits (e.g. constraining trait values to plausible minima and maxima; see Methods for full details). Whilst all care is taken to curate datasets in AusTraits this is not intended to absolve individual researchers from ensuring the appropriate use of the data provided.</p>
</div>
<div id="contributing" class="section level2">
<h2 class="hasAnchor">
<a href="#contributing" class="anchor"></a>Contributing</h2>
</div>
<div id="loading-data" class="section level2">
<h2 class="hasAnchor">
<a href="#loading-data" class="anchor"></a>Loading data</h2>
<p>There are several ways to access the AusTraits data, including</p>
<ul>
<li>downloading plain text files and loading these into R manually</li>
<li>using the <code>austraits</code> R package to fetch and load particular versions of the data (not yet available)</li>
<li>rebuilding the AusTraits data from scratch (see instructions above)</li>
<li>downloading the compressed <code>austraits.rds</code> binary and loading this directly into R.</li>
</ul>
<p>These instructions are written assuming you have followed the latter approach.</p>
<p>Let’s assume you have a variable <code>path</code> with the directory of your data. You can then load the AusTraits data package using the <code>readRDS</code> package</p>
<p>You should now have a object called <code>austraits</code> with the following elements:</p>
<p>[1] “traits” “sites” “methods” “excluded_data” [5] “taxonomy” “definitions” “contributors” “sources”<br>
[9] “build_info”</p>
<p>These names correspond to the sections described in full above under <a href="#elements_of_austraits">“elements of austraits”</a>.</p>
<p>The trait data are in the element called data</p>
<pre><code>## # A tibble: 484,158 x 10
##    dataset_id species_name site_name observation_id trait_name value unit 
##    &lt;chr&gt;      &lt;chr&gt;        &lt;chr&gt;     &lt;chr&gt;          &lt;chr&gt;      &lt;chr&gt; &lt;chr&gt;
##  1 Ahrens_20… Corymbia ca… BOO       Ahrens_2019_0… leaf_area  1288… mm2  
##  2 Ahrens_20… Corymbia ca… BOO       Ahrens_2019_0… leaf_delt… -27.… per …
##  3 Ahrens_20… Corymbia ca… BOO       Ahrens_2019_0… leaf_N_pe… 9     mg/g 
##  4 Ahrens_20… Corymbia ca… BOO       Ahrens_2019_0… modified_… 0.71… dime…
##  5 Ahrens_20… Corymbia ca… BOO       Ahrens_2019_0… specific_… 5.4   mm2/…
##  6 Ahrens_20… Corymbia ca… BOO       Ahrens_2019_0… water_ban… 1.06… dime…
##  7 Ahrens_20… Corymbia ca… BOO       Ahrens_2019_0… wood_dens… 0.60… mg/m…
##  8 Ahrens_20… Corymbia ca… BOO       Ahrens_2019_0… leaf_area  9647… mm2  
##  9 Ahrens_20… Corymbia ca… BOO       Ahrens_2019_0… leaf_delt… -27.… per …
## 10 Ahrens_20… Corymbia ca… BOO       Ahrens_2019_0… leaf_N_pe… 16    mg/g 
## # … with 484,148 more rows, and 3 more variables: value_type &lt;fct&gt;,
## #   replicates &lt;chr&gt;, original_name &lt;chr&gt;</code></pre>
<p>(Note that data appears as a <code>tibble</code> rather than a <code>data.frame</code>. Tibbles are an enhanced type dataframe, with some desirable features, such as not printing the whole dataframe to screen. (If you prefer, you can easily use a traditional R <code>data.frame</code> by converting it using <code>df &lt;- data.frame(austraits$traits)</code>. But from here on we’ll assume you’re sticking with a tibble.)</p>
</div>
</div>
<div id="structure" class="section level1">
<h1 class="hasAnchor">
<a href="#structure" class="anchor"></a>Structure of AusTraits</h1>
<p>(More here)</p>
<p>It is essential that users of AusTraits data are confident the data have the meaning they expect it to and were collected using methods they trust. As such, each dataset within AusTraits must include descriptions of the study, sites, and methods used as well as the data itself.</p>
<p>(More here)</p>
<p>The Extensible Observation Ontology (OBOE) is a formal ontology for capturing the semantics of scientific observation and measurement. The ontology supports researchers to add detailed semantic annotations to scientific data, thereby clarifying the inherent meaning of scientific observations.</p>
<blockquote>
<p>Mark Schildhauer, Matthew B. Jones, Shawn Bowers, Joshua Madin, Sergeui Krivov, Deana Pennington, Ferdinando Villa, Benjamin Leinfelder, Christopher Jones, and Margaret O’Brien. 2016. OBOE: the Extensible Observation Ontology, version 1.1. KNB Data Repository. <a href="http://doi.org/10.5063/F11C1TTM">doi:10.5063/F11C1TTM</a></p>
</blockquote>
<p>The original publication describing the OBOE concept is:</p>
<blockquote>
<p>Madin, J., S. Bowers, M. Schildhauer, S. Krivov, D. Pennington, and F. Villa. 2007. An ontology for describing and synthesizing ecological observation data. Ecological Informatics 2:279–296. <a href="http://doi.org/10.1016/j.ecoinf.2007.05.004">doi:10.1016/j.ecoinf.2007.05.004</a></p>
</blockquote>
<p>Source code on Github: <a href="https://github.com/NCEAS/oboe">NCEAS/oboe</a></p>
<p>(More here)</p>
<div id="elements" class="section level2">
<h2 class="hasAnchor">
<a href="#elements" class="anchor"></a>Elements of AusTraits</h2>
<p>The compiled AusTraits database has the following main components:</p>
<pre><code>austraits
├── traits
├── sites
├── methods
├── excluded_data
├── taxonomy
├── definitions
└── build_info</code></pre>
<p>These elements include all the data and contextual information submitted with each contributed dataset, but are not in the same format as data contributions. See <a href="#format">Format of raw data files for AusTraits</a> for how to format data to contribute.</p>
<p>Each component is the above table is defined as follows:</p>
<div id="traits" class="section level3">
<h3 class="hasAnchor">
<a href="#traits" class="anchor"></a>traits</h3>
<p><strong>Description:</strong> A table containing measurements of plant traits</p>
<p><strong>Content:</strong></p>
<table class="table table table-striped table-hover table-condensed table-responsive">
<thead><tr>
<th style="text-align:left;">
key
</th>
<th style="text-align:left;">
value
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
dataset_id
</td>
<td style="text-align:left;">
Primary identifier for each study contributed into AusTraits; most often these are scientific papers, books, or online resources. By default should be name of first author and year of publication, e.g. <code>Falster_2005</code>.
</td>
</tr>
<tr>
<td style="text-align:left;">
species_name
</td>
<td style="text-align:left;">
Name of species after aligning taxonomy with standard sources.
</td>
</tr>
<tr>
<td style="text-align:left;">
site_name
</td>
<td style="text-align:left;">
Name of site where species was sampled. Cross-references between similar columns in context and data.
</td>
</tr>
<tr>
<td style="text-align:left;">
observation_id
</td>
<td style="text-align:left;">
A unique identifier for the observation, useful for joining traits coming from the same observation_id. Observation ids are assigned automatically, based on <code>dataset_id</code> and row number in the file <code>data.csv</code>.
</td>
</tr>
<tr>
<td style="text-align:left;">
trait_name
</td>
<td style="text-align:left;">
Name of trait sampled. Allowable values specified in the table <code>traits</code>.
</td>
</tr>
<tr>
<td style="text-align:left;">
value
</td>
<td style="text-align:left;">
Value of trait sampled. In the master dataset these are all stored as characters, then converted to numeric later where appropriate.
</td>
</tr>
<tr>
<td style="text-align:left;">
unit
</td>
<td style="text-align:left;">
Units of the sampled trait value after aligning with AusTraits standards.
</td>
</tr>
<tr>
<td style="text-align:left;">
value_type
</td>
<td style="text-align:left;">
A categorical variable describing the type of trait value recorded
</td>
</tr>
<tr>
<td style="text-align:left;">
replicates
</td>
<td style="text-align:left;">
Number of replicate measurements that comprise the data points for the trait for each measurement. A numeric value (or range) is ideal and appropriate if the value type is a <code>mean</code>, <code>median</code>, <code>min</code> or <code>max</code>. For these value types, if replication is unknown the entry should be <code>unknown</code>. If the value type is <code>raw_value</code> the replicate value should be <code>1</code>. If the value type is <code>expert_mean</code>, <code>expert_min</code>, or <code>expert_max</code> the replicate value should be <code>.na</code>.
</td>
</tr>
<tr>
<td style="text-align:left;">
original_name
</td>
<td style="text-align:left;">
Taxonomic name given to species in the original data supplied by the authors
</td>
</tr>
</tbody>
</table>
</div>
<div id="sites" class="section level3">
<h3 class="hasAnchor">
<a href="#sites" class="anchor"></a>sites</h3>
<p><strong>Description:</strong> A table containing observations of site characteristics associated with information in <code>data</code>. Cross referencing between the two dataframes is possible using combinations of the variables <code>dataset_id</code>, <code>site_name</code>.</p>
<p><strong>Content:</strong></p>
<table class="table table table-striped table-hover table-condensed table-responsive">
<thead><tr>
<th style="text-align:left;">
key
</th>
<th style="text-align:left;">
value
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
dataset_id
</td>
<td style="text-align:left;">
Primary identifier for each study contributed into AusTraits; most often these are scientific papers, books, or online resources. By default should be name of first author and year of publication, e.g. <code>Falster_2005</code>.
</td>
</tr>
<tr>
<td style="text-align:left;">
site_name
</td>
<td style="text-align:left;">
Name of site where species was sampled. Cross-references between similar columns in context and data.
</td>
</tr>
<tr>
<td style="text-align:left;">
site_property
</td>
<td style="text-align:left;">
The site characteristic being recorded. Name should include units of measurement, e.g. <code>longitude (deg)</code>. Ideally we have at least these variables for each site - <code>longitude (deg)</code>, <code>latitude (deg)</code>, <code>description</code>.
</td>
</tr>
<tr>
<td style="text-align:left;">
value
</td>
<td style="text-align:left;">
Value of trait sampled. In the master dataset these are all stored as characters, then converted to numeric later where appropriate.
</td>
</tr>
</tbody>
</table>
</div>
<div id="methods" class="section level3">
<h3 class="hasAnchor">
<a href="#methods" class="anchor"></a>methods</h3>
<p><strong>Description:</strong> A table containing details on methods with which data were collected, including time frame and source</p>
<p><strong>Content:</strong></p>
<table class="table table table-striped table-hover table-condensed table-responsive">
<thead><tr>
<th style="text-align:left;">
key
</th>
<th style="text-align:left;">
value
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
dataset_id
</td>
<td style="text-align:left;">
Primary identifier for each study contributed into AusTraits; most often these are scientific papers, books, or online resources. By default should be name of first author and year of publication, e.g. <code>Falster_2005</code>.
</td>
</tr>
<tr>
<td style="text-align:left;">
trait_name
</td>
<td style="text-align:left;">
Name of trait sampled. Allowable values specified in the table <code>traits</code>.
</td>
</tr>
<tr>
<td style="text-align:left;">
methods
</td>
<td style="text-align:left;">
A textual description of the methods used to collect the trait data. Whenever available, methods are taken near-verbatim from referenced source. Methods can include descriptions such as ‘measured on herbarium specimens’,‘data from the literature’, or a detailed description of the field or lab methods used to collect the data.
</td>
</tr>
<tr>
<td style="text-align:left;">
year_collected_start
</td>
<td style="text-align:left;">
The year data collection commenced
</td>
</tr>
<tr>
<td style="text-align:left;">
year_collected_end
</td>
<td style="text-align:left;">
The year data collection was completed
</td>
</tr>
<tr>
<td style="text-align:left;">
description
</td>
<td style="text-align:left;">
A 1-2 sentence description of the purpose of the study.
</td>
</tr>
<tr>
<td style="text-align:left;">
collection_type
</td>
<td style="text-align:left;">
A field to indicate where the majority of plants on which traits were measured were collected - in the <code>field</code>, <code>lab</code>, <code>glasshouse</code>, <code>herbarium specimens</code>, or <code>literature</code>. The latter should only be used when the data were sourced from the literature and the collection type is unknown.
</td>
</tr>
<tr>
<td style="text-align:left;">
sample_age_class
</td>
<td style="text-align:left;">
A field to indicate if the study was completed on <code>adult</code> or <code>juvenile</code> plants
</td>
</tr>
<tr>
<td style="text-align:left;">
sampling_strategy
</td>
<td style="text-align:left;">
A written description of how study sites were selected and how study individuals were selected. When available, this information is lifted verbatim from a published manuscript. For herbarium studies, this field ideally indicates which herbaria were ‘sampled’ to measure a specific trait
</td>
</tr>
<tr>
<td style="text-align:left;">
source_primary_citation
</td>
<td style="text-align:left;">
Citation for primary source. This detail is generated from the primary source in the metadata
</td>
</tr>
<tr>
<td style="text-align:left;">
source_primary_key
</td>
<td style="text-align:left;">
Citation key for primary source in <code>sources</code>. The key is typically of format Surname_year
</td>
</tr>
<tr>
<td style="text-align:left;">
source_secondary_citation
</td>
<td style="text-align:left;">
Citations for an secondary source. This detail is generated from the secondary source in the metadata
</td>
</tr>
<tr>
<td style="text-align:left;">
source_secondary_key
</td>
<td style="text-align:left;">
Citation key for secondary source in <code>sources</code>. The key is typically of format Surname_year
</td>
</tr>
</tbody>
</table>
</div>
<div id="excluded_data" class="section level3">
<h3 class="hasAnchor">
<a href="#excluded_data" class="anchor"></a>excluded_data</h3>
<p><strong>Description:</strong> A table of data that did not pass quality test and so were excluded from the master dataset.</p>
<p><strong>Content:</strong> Structure is identical to that presented in the <code>traits</code> table, only with an extra column called <code>error</code> indicating why the record was excluded. Common reasons are missing_unit_conversions, missing_value, and unsupported_trait_value.</p>
</div>
<div id="taxonomy" class="section level3">
<h3 class="hasAnchor">
<a href="#taxonomy" class="anchor"></a>taxonomy</h3>
<p><strong>Description:</strong> A table containing details on species taxonomy associated with information in <code>data</code>. Cross referencing between the two dataframes is possible using the variable <code>species_name</code>. We have attempted to align species names with known taxonomic units, focussing primarily on the <a href="http://www.theplantlist.org/"><code>The Plant List</code> (TPL)</a> – a global working list of all known plant species. In addition we have tried to align these names with the <a href="https://biodiversity.org.au/nsl/services/apc"><code>Australian Plant Census</code> (APC)</a> and the <a href="https://biodiversity.org.au/nsl/services/APNI"><code>Australian Plant Names Index</code> (APNI)</a>.</p>
<p><strong>Content:</strong></p>
<table class="table table table-striped table-hover table-condensed table-responsive">
<thead><tr>
<th style="text-align:left;">
key
</th>
<th style="text-align:left;">
value
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
species_name
</td>
<td style="text-align:left;">
Name of species after aligning taxonomy with standard sources.
</td>
</tr>
<tr>
<td style="text-align:left;">
family
</td>
<td style="text-align:left;">
Family of aligned species_name
</td>
</tr>
<tr>
<td style="text-align:left;">
authority
</td>
<td style="text-align:left;">
Authority for aligned species_name
</td>
</tr>
<tr>
<td style="text-align:left;">
TPL_ID
</td>
<td style="text-align:left;">
ID of species in the TPL
</td>
</tr>
<tr>
<td style="text-align:left;">
status
</td>
<td style="text-align:left;">
Status of species in TPL
</td>
</tr>
<tr>
<td style="text-align:left;">
APC_name
</td>
<td style="text-align:left;">
Name of species in APC
</td>
</tr>
<tr>
<td style="text-align:left;">
APC_ID
</td>
<td style="text-align:left;">
ID of species in the APC
</td>
</tr>
<tr>
<td style="text-align:left;">
APNI_ID
</td>
<td style="text-align:left;">
ID of species in the APNI
</td>
</tr>
</tbody>
</table>
</div>
<div id="definitions" class="section level3">
<h3 class="hasAnchor">
<a href="#definitions" class="anchor"></a>definitions</h3>
<p><strong>Description:</strong> A copy of the definitions for all tables and terms. Information included here was used to process data and generate any documentation for the study.</p>
<p><strong>Content:</strong> A structured yaml file, represented as a list in R. See file <code>config/definitions.yaml</code> for more details.</p>
</div>
<div id="contributors" class="section level3">
<h3 class="hasAnchor">
<a href="#contributors" class="anchor"></a>contributors</h3>
<p><strong>Description:</strong> A list of people contributing to each study.</p>
<p><strong>Content:</strong></p>
<table class="table table table-striped table-hover table-condensed table-responsive">
<thead><tr>
<th style="text-align:left;">
key
</th>
<th style="text-align:left;">
value
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
dataset_id
</td>
<td style="text-align:left;">
Primary identifier for each study contributed into AusTraits; most often these are scientific papers, books, or online resources. By default should be name of first author and year of publication, e.g. <code>Falster_2005</code>.
</td>
</tr>
<tr>
<td style="text-align:left;">
name
</td>
<td style="text-align:left;">
Name of contributor
</td>
</tr>
<tr>
<td style="text-align:left;">
institution
</td>
<td style="text-align:left;">
Last known institution or affiliation
</td>
</tr>
<tr>
<td style="text-align:left;">
role
</td>
<td style="text-align:left;">
Their role in the study
</td>
</tr>
</tbody>
</table>
</div>
<div id="sources" class="section level3">
<h3 class="hasAnchor">
<a href="#sources" class="anchor"></a>sources</h3>
<p><strong>Description:</strong> Bibtex entries for all primary and secondary sources in the compilation</p>
</div>
<div id="build_info" class="section level3">
<h3 class="hasAnchor">
<a href="#build_info" class="anchor"></a>build_info</h3>
<p><strong>Description:</strong> Description of the computing environment used to create this version of the dataset, including version number, git commit and R session_info</p>
</div>
</div>
<div id="observation-ids" class="section level2">
<h2 class="hasAnchor">
<a href="#observation-ids" class="anchor"></a>Observation IDs</h2>
<p>Each trait measurement has an associated <code>observation_id</code>. Observation IDs bind together related measurements within a dataset. For example, if multiple traits were collected on the same individual, the <code>observation_id</code> allows us to gather these together. For floras, which report a species averages, the <code>observation_id</code> is determined via the species name. Importantly, the <code>observation_id</code> allows translation between long (e.g. with variables <code>trait_name</code> and <code>value</code>) and wide (e.g. with traits as columns) formats.</p>
<p>Generally, <code>observation_id</code> has the format <code>dataset_id_XX</code> where <code>XX</code> is a unique number within each dataset.</p>
<p>For datasets that arrive in wide format we assume each row has a unique <code>observation_id</code>.</p>
<p>For datasets that arrive in long format, the <code>observation_id</code> is assigned based on a specified grouping variable. This variable can be specified in the <code>metadata.yml</code> file under the section <code>variable_match</code>. If missing, <code>observation_id</code> is assigned based on <code>species_name</code>.</p>
</div>
<div id="value_types" class="section level2">
<h2 class="hasAnchor">
<a href="#value_types" class="anchor"></a>Values and Value types</h2>
<p>Each record in the table of <a href="#traits">trait data</a> has an associated <code>value</code> and <code>value_type</code>. Traits are either <code>numeric</code> or <code>categorical</code>.</p>
<p>For traits with numerical values, the recorded value has been converted into standardised units and we have check that the value can be converted into a number and lies within the allowable range.</p>
<p>For categorical variables, we only include records that are defined in the definitions (see <a href="#traits">trait definitions below</a>). Moreover, we use a format whereby</p>
<ul>
<li>we use <code>_</code> for multi-word terms, e.g. <code>semi_deciduous</code>
</li>
<li>use a space for situations where there are two possible values for that trait, e.g. <code>annual biennial</code> for something which is either annual or biennial</li>
</ul>
<p>Each trait measurement also an associated <code>value_type</code>, which gives <code>A categorical variable describing the type of trait value recorded</code>. Possible value_types are:</p>
<table class="table table table-striped table-hover table-condensed table-responsive">
<thead><tr>
<th style="text-align:left;">
key
</th>
<th style="text-align:left;">
value
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
raw_value
</td>
<td style="text-align:left;">
Value is a direct measurement
</td>
</tr>
<tr>
<td style="text-align:left;">
site_min
</td>
<td style="text-align:left;">
Value is the minimum of measurements on multiple individuals of the species at a single site
</td>
</tr>
<tr>
<td style="text-align:left;">
site_mean
</td>
<td style="text-align:left;">
Value is the mean or median of measurements on multiple individuals of the species at a single site
</td>
</tr>
<tr>
<td style="text-align:left;">
site_max
</td>
<td style="text-align:left;">
Value is the maximum of measurements on multiple individuals of the species at a single site
</td>
</tr>
<tr>
<td style="text-align:left;">
multisite_min
</td>
<td style="text-align:left;">
Value is the minimum of measurements on multiple individuals of the species across multiple sites
</td>
</tr>
<tr>
<td style="text-align:left;">
multisite_mean
</td>
<td style="text-align:left;">
Value is the mean or median of measurements on multiple individuals of the species across multiple sites
</td>
</tr>
<tr>
<td style="text-align:left;">
multisite_max
</td>
<td style="text-align:left;">
Value is the maximum of measurements on multiple individuals of the species across multiple sites
</td>
</tr>
<tr>
<td style="text-align:left;">
expert_min
</td>
<td style="text-align:left;">
Value is the minimum observed for a species across its range or in this particular dataset, as estimated by an expert based on their knowledge of the species. Data fitting this category include estimates from herbaria that represent a species’ entire range, and values for categorical variables obtained from a reference book, or identified by an expert.
</td>
</tr>
<tr>
<td style="text-align:left;">
expert_mean
</td>
<td style="text-align:left;">
Value is the average observed for a species across its range or in this particular dataset, as estimated by an expert based on their knowledge of the species. Data fitting this category include estimates from herbaria that represent a species’ entire range, and values for categorical variables obtained from a reference book, or identified by an expert.
</td>
</tr>
<tr>
<td style="text-align:left;">
expert_max
</td>
<td style="text-align:left;">
Value is the maximum observed for a species across its range or in this particular dataset, as estimated by an expert based on their knowledge of the species. Data fitting this category include estimates from herbaria that represent a species’ entire range, and values for categorical variables obtained from a reference book, or identified by an expert.
</td>
</tr>
<tr>
<td style="text-align:left;">
experiment_min
</td>
<td style="text-align:left;">
Value is the minimum of measurements from an experimental study either in the field or a glasshouse
</td>
</tr>
<tr>
<td style="text-align:left;">
experiment_mean
</td>
<td style="text-align:left;">
Value is the mean or median of measurements from an experimental study either in the field or a glasshouse
</td>
</tr>
<tr>
<td style="text-align:left;">
experiment_max
</td>
<td style="text-align:left;">
Value is the maximum of measurements from an experimental study either in the field or a glasshouse
</td>
</tr>
<tr>
<td style="text-align:left;">
individual_mean
</td>
<td style="text-align:left;">
Value is a mean of replicate measurements on an individual (usually for experimental ecophysiology studies)
</td>
</tr>
<tr>
<td style="text-align:left;">
unknown
</td>
<td style="text-align:left;">
Value type is not currently known
</td>
</tr>
</tbody>
</table>
<p>AusTraits does not include intra-individual observations. WHen multiple measurements per individual are submitted to AusTraits, we take the mean of the values and record the value_type as <code>individual_mean</code>.</p>
</div>
<div id="taxonomy" class="section level2">
<h2 class="hasAnchor">
<a href="#taxonomy" class="anchor"></a>Species taxonomy</h2>
<p>Within AusTraits there are records for 22191 different species. A full list of all known species is available in the table <code>taxonomy</code> (see details above).</p>
<p>We have attempted to align species names with known taxonomic units, focussing primarily on the <a href="http://www.theplantlist.org/"><code>The Plant List</code> (TPL)</a> – a global working list of all known plant species. In addition we have tried to align these names with the <a href="https://biodiversity.org.au/nsl/services/apc"><code>Australian Plant Census</code> (APC)</a> and the <a href="https://biodiversity.org.au/nsl/services/APNI"><code>Australian Plant Names Index</code> (APNI)</a>. The <code>APNI_ID</code> can also be used to access relevant records for the species in the <a href="https://bie.ala.org.au/"><code>Atlas of Living Australia</code></a> (ALA).</p>
<p>Links to species records in these systems can be accessed via the relevant IDs as in these examples.</p>
<ul>
<li>The Plant List: <a href="http://www.theplantlist.org/tpl1.1/record/kew-450649">http://www.theplantlist.org/tpl1.1/record/kew-450649</a> where <code>kew-450649</code> is the <code>TPL_ID</code>
</li>
<li>Australian Plant Census: <a href="https://biodiversity.org.au/nsl/services/node/apc/2908862">https://biodiversity.org.au/nsl/services/node/apc/2908862</a> where <code>2908862</code> is the <code>APC_ID</code>
</li>
<li>Australian Plant Names Index: <a href="http://id.biodiversity.org.au/node/apni/2899106">http://id.biodiversity.org.au/node/apni/2899106</a> where <code>2899106</code> is the <code>APNI_ID</code>
</li>
<li>Atlas of Living Australia: <a href="https://bie.ala.org.au/species/http://id.biodiversity.org.au/node/apni/2899106">https://bie.ala.org.au/species/http://id.biodiversity.org.au/node/apni/2899106</a> where <code>2899106</code> is the <code>APNI_ID</code>.</li>
</ul>
</div>
<div id="sources-1" class="section level2">
<h2 class="hasAnchor">
<a href="#sources-1" class="anchor"></a>Sources</h2>
<p>For each dataset in the compilation there is the option to list a primary and secondary citation. The primary citation The original study in which data were collected while the secondary citation is A subsequent study where data were compiled or re-analysed and then made available. These references are included in two places:</p>
<ol style="list-style-type: decimal">
<li>Within the table <a href="#methods">methods</a>, where we provide a formatted version of each.</li>
<li>In the element <a href="#sources">sources</a>, where we provided bibtex versions of all sources which can be imported into your reference library. The keys for these references are listed within the <span id="methods">methods</span>.</li>
</ol>
<p>As noted above under <a href="#citations">citations</a>, anyone using the compilation is expected to cite primary sources.</p>
</div>
<div id="trait_defs" class="section level2">
<h2 class="hasAnchor">
<a href="#trait_defs" class="anchor"></a>Trait definitions</h2>
<p>Below is the standard definition for each trait in AusTraits (drawn from the the file <code>config/definitions.yml</code>). As described above in the section on <a href="#value_types">Value types</a>, traits are labelled as either <code>numeric</code> or <code>categorical</code>.</p>
<p><strong>accessory_cost_mass</strong></p>
<ul>
<li>label: Mass of seed accessory costs</li>
<li>description: Mass of seed accessory costs, the proportion of a fruit that does not develop into a seed</li>
<li>number of records: 47</li>
<li>number of studies: 1</li>
<li>type: numeric</li>
<li>units: mg</li>
<li>allowable range: 0.01 - 10000 mg</li>
</ul>
<p><strong>accessory_cost_fraction</strong></p>
<ul>
<li>label: Fraction of total reproductive investment to non-seed tissues</li>
<li>description: The fraction of total reproductive investment required to mature a seed that is invested in non-seed tissues</li>
<li>number of records: 47</li>
<li>number of studies: 1</li>
<li>type: numeric</li>
<li>units: mg/mg</li>
<li>allowable range: 0.01 - 1 mg/mg</li>
</ul>
<p><strong>bark_delta13C</strong></p>
<ul>
<li>label: Bark carbon (C) isotope signature (delta 13C)</li>
<li>description: Bark carbon stable isotope signature</li>
<li>number of records: 174</li>
<li>number of studies: 1</li>
<li>type: numeric</li>
<li>units: per mille</li>
<li>allowable range: -50 - 0 per mille</li>
</ul>
<p><strong>bark_delta15N</strong></p>
<ul>
<li>label: Bark nitrogen (N) isotope signature (delta 15N)</li>
<li>description: Bark nitrogen stable isotope signature</li>
<li>number of records: 174</li>
<li>number of studies: 1</li>
<li>type: numeric</li>
<li>units: per mille</li>
<li>allowable range: -25 - 75 per mille</li>
</ul>
<p><strong>bark_mass_area</strong></p>
<ul>
<li>label: Bark mass per unit surface area of stem</li>
<li>description: Bark mass per unit surface area of stem</li>
<li>number of records: 27</li>
<li>number of studies: 1</li>
<li>type: numeric</li>
<li>units: g/cm2</li>
<li>allowable range: 0.01 - 1 g/cm2</li>
</ul>
<p><strong>bark_C_per_dry_mass</strong></p>
<ul>
<li>label: Bark carbon (C) content per unit bark dry mass</li>
<li>description: Bark carbon (C) content per unit bark dry mass</li>
<li>number of records: 174</li>
<li>number of studies: 1</li>
<li>type: numeric</li>
<li>units: mg/g</li>
<li>allowable range: 50 - 750 mg/g</li>
</ul>
<p><strong>bark_N_per_dry_mass</strong></p>
<ul>
<li>label: Bark nitrogen (N) content per unit bark dry mass</li>
<li>description: Bark nitrogen (N) content per unit bark dry mass</li>
<li>number of records: 174</li>
<li>number of studies: 1</li>
<li>type: numeric</li>
<li>units: mg/g</li>
<li>allowable range: 0.1 - 100 mg/g</li>
</ul>
<p><strong>bark_thickness</strong></p>
<ul>
<li>label: Bark thickness</li>
<li>description: Thickness of the bark of the stem</li>
<li>number of records: 1373</li>
<li>number of studies: 6</li>
<li>type: numeric</li>
<li>units: mm</li>
<li>allowable range: 0.01 - 50 mm</li>
</ul>
<p><strong>branch_mass_fraction</strong></p>
<ul>
<li>label: Fraction of plant dry mass comprised of branch material</li>
<li>description: Fraction of plant dry mass comprised of branch material</li>
<li>number of records: 45</li>
<li>number of studies: 1</li>
<li>type: numeric</li>
<li>units: mg/mg</li>
<li>allowable range: 0 - 1 mg/mg</li>
</ul>
<p><strong>sprout_depth</strong></p>
<ul>
<li>label: Depth below ground (negative number) or height above ground (positive number) from which buds emerge following a disturbance (i.e. fire)</li>
<li>description: Depth of resprouting shoots</li>
<li>number of records: 0</li>
<li>number of studies: 0</li>
<li>type: numeric</li>
<li>units: mm</li>
<li>allowable range: -100 - 100 mm</li>
</ul>
<p><strong>carotenoid_per_area</strong></p>
<ul>
<li>label: Leaf carotenoid content per unit leaf area</li>
<li>description: Leaf carotenoid content per unit leaf area</li>
<li>number of records: 0</li>
<li>number of studies: 0</li>
<li>type: numeric</li>
<li>units: umol/m2</li>
<li>allowable range: 10 - 1000 umol/m2</li>
</ul>
<p><strong>carotenoid_per_dry_mass</strong></p>
<ul>
<li>label: Leaf carotenoid content per unit leaf dry mass</li>
<li>description: Leaf carotenoid content per unit leaf dry mass</li>
<li>number of records: 0</li>
<li>number of studies: 0</li>
<li>type: numeric</li>
<li>units: mmol/kg</li>
<li>allowable range: 0.1 - 10 mmol/kg</li>
</ul>
<p><strong>cell_epidermis_Ca_per_fresh_mass</strong></p>
<ul>
<li>label: Ca content of epidermal cells</li>
<li>description: Ca content of epidermal cells</li>
<li>number of records: 20</li>
<li>number of studies: 1</li>
<li>type: numeric</li>
<li>units: umol/g</li>
<li>allowable range: 1 - 500 umol/g</li>
</ul>
<p><strong>cell_hypodermis_Ca_per_fresh_mass</strong></p>
<ul>
<li>label: Ca content of hypodermis cells</li>
<li>description: Ca content of hypodermis cells</li>
<li>number of records: 9</li>
<li>number of studies: 1</li>
<li>type: numeric</li>
<li>units: umol/g</li>
<li>allowable range: 1 - 500 umol/g</li>
</ul>
<p><strong>cell_internal_parenchyma_Ca_per_fresh_mass</strong></p>
<ul>
<li>label: Ca content of internal parenchyma cells</li>
<li>description: Ca content of internal parenchyma cells</li>
<li>number of records: 12</li>
<li>number of studies: 1</li>
<li>type: numeric</li>
<li>units: umol/g</li>
<li>allowable range: 1 - 500 umol/g</li>
</ul>
<p><strong>cell_palisade_mesophyll_Ca_per_fresh_mass</strong></p>
<ul>
<li>label: Ca content of palisade mesophyll cells</li>
<li>description: Ca content of palisade mesophyll cells</li>
<li>number of records: 20</li>
<li>number of studies: 1</li>
<li>type: numeric</li>
<li>units: umol/g</li>
<li>allowable range: 1 - 100 umol/g</li>
</ul>
<p><strong>cell_sclerenchyma_Ca_per_fresh_mass</strong></p>
<ul>
<li>label: Ca content of sclerenchyma cells</li>
<li>description: Ca content of sclerenchyma cells</li>
<li>number of records: 17</li>
<li>number of studies: 1</li>
<li>type: numeric</li>
<li>units: umol/g</li>
<li>allowable range: 1 - 100 umol/g</li>
</ul>
<p><strong>cell_spongy_mesophyll_Ca_per_fresh_mass</strong></p>
<ul>
<li>label: Ca content of spongy mesophyll cells</li>
<li>description: Ca content of spongy mesophyll cells</li>
<li>number of records: 9</li>
<li>number of studies: 1</li>
<li>type: numeric</li>
<li>units: umol/g</li>
<li>allowable range: 1 - 1000 umol/g</li>
</ul>
<p><strong>cell_epidermis_P_per_fresh_mass</strong></p>
<ul>
<li>label: P content of epidermal cells</li>
<li>description: P content of epidermal cells</li>
<li>number of records: 20</li>
<li>number of studies: 1</li>
<li>type: numeric</li>
<li>units: umol/g</li>
<li>allowable range: 0 - 10 umol/g</li>
</ul>
<p><strong>cell_hypodermis_P_per_fresh_mass</strong></p>
<ul>
<li>label: P content of hypodermis cells</li>
<li>description: P content of hypodermis cells</li>
<li>number of records: 9</li>
<li>number of studies: 1</li>
<li>type: numeric</li>
<li>units: umol/g</li>
<li>allowable range: 0 - 10 umol/g</li>
</ul>
<p><strong>cell_internal_parenchyma_P_per_fresh_mass</strong></p>
<ul>
<li>label: P content of internal parenchyma cells</li>
<li>description: P content of internal parenchyma cells</li>
<li>number of records: 12</li>
<li>number of studies: 1</li>
<li>type: numeric</li>
<li>units: umol/g</li>
<li>allowable range: 0 - 10 umol/g</li>
</ul>
<p><strong>cell_palisade_mesophyll_P_per_fresh_mass</strong></p>
<ul>
<li>label: P content of palisade mesophyll cells</li>
<li>description: P content of palisade mesophyll cells</li>
<li>number of records: 20</li>
<li>number of studies: 1</li>
<li>type: numeric</li>
<li>units: umol/g</li>
<li>allowable range: 0 - 30 umol/g</li>
</ul>
<p><strong>cell_sclerenchyma_P_per_fresh_mass</strong></p>
<ul>
<li>label: P content of sclerenchyma cells</li>
<li>description: P content of sclerenchyma cells</li>
<li>number of records: 17</li>
<li>number of studies: 1</li>
<li>type: numeric</li>
<li>units: umol/g</li>
<li>allowable range: 0 - 10 umol/g</li>
</ul>
<p><strong>cell_spongy_mesophyll_P_per_fresh_mass</strong></p>
<ul>
<li>label: P content of spongy mesophyll cells</li>
<li>description: P content of spongy mesophyll cells</li>
<li>number of records: 9</li>
<li>number of studies: 1</li>
<li>type: numeric</li>
<li>units: umol/g</li>
<li>allowable range: 0 - 30 umol/g</li>
</ul>
<p><strong>chlorophyll_per_area</strong></p>
<ul>
<li>label: Sum of chlorophyll A and B per leaf area</li>
<li>description: Sum of chlorophyll A and B per leaf area</li>
<li>number of records: 0</li>
<li>number of studies: 0</li>
<li>type: numeric</li>
<li>units: umol/m2</li>
<li>allowable range: 50 - 2000 umol/m2</li>
</ul>
<p><strong>chlorophyll_A_per_area</strong></p>
<ul>
<li>label: Chlorophyll A content relative to leaf area</li>
<li>description: Chlorophyll A content relative to leaf area</li>
<li>number of records: 0</li>
<li>number of studies: 0</li>
<li>type: numeric</li>
<li>units: umol/m2</li>
<li>allowable range: 50 - 2000 umol/m2</li>
</ul>
<p><strong>chlorophyll_B_per_area</strong></p>
<ul>
<li>label: Sum of Chlorophyll B content relative to leaf area</li>
<li>description: Chlorophyll B content relative to leaf area</li>
<li>number of records: 0</li>
<li>number of studies: 0</li>
<li>type: numeric</li>
<li>units: umol/m2</li>
<li>allowable range: 10 - 1000 umol/m2</li>
</ul>
<p><strong>chlorophyll_per_dry_mass</strong></p>
<ul>
<li>label: Leaf chlorophyll content per unit leaf dry mass</li>
<li>description: Leaf chlorophyll content per unit leaf dry mass</li>
<li>number of records: 0</li>
<li>number of studies: 0</li>
<li>type: numeric</li>
<li>units: mmol/kg</li>
<li>allowable range: 0.1 - 10 mmol/kg</li>
</ul>
<p><strong>chlorophyll_A_per_dry_mass</strong></p>
<ul>
<li>label: Leaf chlorophyll A content per unit leaf dry mass</li>
<li>description: Leaf chlorophyll A content per unit leaf dry mass</li>
<li>number of records: 0</li>
<li>number of studies: 0</li>
<li>type: numeric</li>
<li>units: mmol/kg</li>
<li>allowable range: 0.1 - 10 mmol/kg</li>
</ul>
<p><strong>chlorophyll_B_per_dry_mass</strong></p>
<ul>
<li>label: Leaf chlorophyll B content per unit leaf dry mass</li>
<li>description: Leaf chlorophyll B content per unit leaf dry mass</li>
<li>number of records: 0</li>
<li>number of studies: 0</li>
<li>type: numeric</li>
<li>units: mmol/kg</li>
<li>allowable range: 0.01 - 10 mmol/kg</li>
</ul>
<p><strong>chlorophyll_A_B_ratio</strong></p>
<ul>
<li>label: Ratio of chlorophyll A to chlorophyll B</li>
<li>description: Ratio of chlorophyll A to chlorophyll B</li>
<li>number of records: 0</li>
<li>number of studies: 0</li>
<li>type: numeric</li>
<li>units: umol/umol</li>
<li>allowable range: 1 - 10 umol/umol</li>
</ul>
<p><strong>ca</strong></p>
<ul>
<li>label: ambient CO2 concentration</li>
<li>description: ambient CO2 concentration (external CO2 concentration)</li>
<li>number of records: 713</li>
<li>number of studies: 2</li>
<li>type: numeric</li>
<li>units: umol/mol</li>
<li>allowable range: 50 - 1000 umol/mol</li>
</ul>
<p><strong>cc</strong></p>
<ul>
<li>label: CO2 concentration inside chloroplasts</li>
<li>description: CO2 concentration inside chloroplasts</li>
<li>number of records: 0</li>
<li>number of studies: 0</li>
<li>type: numeric</li>
<li>units: umol/umol</li>
<li>allowable range: 50 - 1000 umol/umol</li>
</ul>
<p><strong>ci</strong></p>
<ul>
<li>label: internal CO2 concentration under ambient conditions</li>
<li>description: CO2 concentration in interstitial spaces under ambient conditions</li>
<li>number of records: 0</li>
<li>number of studies: 0</li>
<li>type: numeric</li>
<li>units: umol/mol</li>
<li>allowable range: 50 - 1000 umol/mol</li>
</ul>
<p><strong>ci_at_Amax</strong></p>
<ul>
<li>label: internal CO2 concentration during Amax measurement</li>
<li>description: CO2 concentration in interstitial spaces during Amax measurement</li>
<li>number of records: 0</li>
<li>number of studies: 0</li>
<li>type: numeric</li>
<li>units: umol/mol</li>
<li>allowable range: 50 - 1000 umol/mol</li>
</ul>
<p><strong>ci_at_Asat</strong></p>
<ul>
<li>label: internal CO2 concentration during Asat measurement</li>
<li>description: CO2 concentration in interstitial spaces during Asat measurement</li>
<li>number of records: 737</li>
<li>number of studies: 3</li>
<li>type: numeric</li>
<li>units: umol/mol</li>
<li>allowable range: 50 - 1000 umol/mol</li>
</ul>
<p><strong>ci_over_ca</strong></p>
<ul>
<li>label: Ratio of internal to external CO2 concentrations</li>
<li>description: Ratio of internal to external CO2 concentrations</li>
<li>number of records: 0</li>
<li>number of studies: 0</li>
<li>type: numeric</li>
<li>units: umol/umol</li>
<li>allowable range: 0 - 1 umol/umol</li>
</ul>
<p><strong>cotyledon_position</strong></p>
<ul>
<li>label: Cotyledon position at germination</li>
<li>description: Binary variable distinguishing between seedlings where the cotyledon remains within the seed coat versus emerges from the seed coat at germination.</li>
<li>number of records: 1738</li>
<li>number of studies: 1</li>
<li>type: categorical</li>
<li>allowable values:
<ul>
<li>
<em>cryptocotylar</em>: A type of seed germination in which the cotyledons remain within the seed coat at germination.</li>
<li>
<em>phanerocotylar</em>: A type of seed germination in which the cotyledons emerge from the seed coat.</li>
</ul>
</li>
</ul>
<p><strong>cotyledon_type</strong></p>
<ul>
<li>label: Cotyledon type</li>
<li>description: Binary variable distinguishing between glabrous versus hairy cotyledons</li>
<li>number of records: 586</li>
<li>number of studies: 1</li>
<li>type: categorical</li>
<li>allowable values:
<ul>
<li>
<em>glabrous</em>: Cotyledon lacks hairs</li>
<li>
<em>hairy</em>: Cotyledon has hairs</li>
</ul>
</li>
</ul>
<p><strong>upper_cuticle_thickness</strong></p>
<ul>
<li>label: Upper cuticle thickness</li>
<li>description: Thickness of the upper cuticle</li>
<li>number of records: 163</li>
<li>number of studies: 4</li>
<li>type: numeric</li>
<li>units: um</li>
<li>allowable range: 0.1 - 100 um</li>
</ul>
<p><strong>lower_cuticle_thickness</strong></p>
<ul>
<li>label: Lower cuticle thickness</li>
<li>description: Thickness of the lower cuticle</li>
<li>number of records: 142</li>
<li>number of studies: 3</li>
<li>type: numeric</li>
<li>units: um</li>
<li>allowable range: 0.1 - 100 um</li>
</ul>
<p><strong>diaspore_mass</strong></p>
<ul>
<li>label: Mass of entire diaspore</li>
<li>description: Mass of seed including dispersal appendages</li>
<li>number of records: 314</li>
<li>number of studies: 2</li>
<li>type: numeric</li>
<li>units: mg</li>
<li>allowable range: 0.001 - 10000 mg</li>
</ul>
<p><strong>dispersal_appendage</strong></p>
<ul>
<li>label: Appendage of propagule which facilitates dispersal</li>
<li>description: Type of dispersal appendage present</li>
<li>number of records: 1675</li>
<li>number of studies: 7</li>
<li>type: categorical</li>
<li>allowable values:
<ul>
<li>
<em>aril</em>: Fleshy outgrowth of a seed, that often attracts animals like birds or ants</li>
<li>
<em>awns</em>: A slender, bristle-like projection, e.g. from the back or tip of the glumes and lemmas in some grasses and on the fruit of some other species.</li>
<li>
<em>awn_bristle</em>: REPLACE WITH ‘awns bristles’</li>
<li>
<em>barbs</em>: A rear-facing point, as in a fish hook that aid in seed and fruit dispersal. Includes “hooks”, a term used in Ranunculus</li>
<li>
<em>beak</em>: Name of dispersal appendage in the genus Carex</li>
<li>
<em>bladdery_wings</em>: Maybe referring to the wings present on the pollen grains of some conifers to aid in their dispersal by wind</li>
<li>
<em>bristles</em>: Bristle-like projections</li>
<li>
<em>bract</em>: REPLACE WITH ‘bracts’</li>
<li>
<em>bracts</em>: When the bracts below the inflorescence are the dispersal appendage. For the family Poaceae “bract” refers to the glumes, the modified membranous bracts surrounding the spikelet of a grass.</li>
<li>
<em>caruncle</em>: A type of eliasome, term especially used in the family Euphorbiaceae; an outgrowth or appendage at or near the hilum of certain seeds that aids in dispersal. Includes “strophiole”</li>
<li>
<em>curved_awn</em>: A curved, slender awn</li>
<li>
<em>dry_dehiscent</em>: Dry fruit that dehisces in some manner</li>
<li>
<em>drupe</em>: Fruit with membranous exocarp, fleshy mesocarp and hard endocarp</li>
<li>
<em>elaiosome</em>: Fleshy (often fatty) appendage on seeds that attracts ants</li>
<li>
<em>enclosing_wing</em>: Wing-line seed extentions that enclose the seed</li>
<li>
<em>exarillate</em>: A seed that lacks an aril</li>
<li>
<em>feathery_style</em>: When a feathery style aids in seed dispersal</li>
<li>
<em>fleshy_dehiscent_capsule</em>: A fleshy fruit that splits open and releases seeds at maturity; fruit type</li>
<li>
<em>fleshy_fruit</em>: A fruit where part or all of the pericarp (fruit wall) is fleshy at maturity; fruit type</li>
<li>
<em>fleshy_wings_capsule</em>: A fruit whose capsule or wings are fleshy</li>
<li>
<em>floral_parts</em>: When a plant’s floral parts, the petals and/or sepals, are persistent and aid in seed and fruit dispersal</li>
<li>
<em>funicle</em>: The stalk that joins the seed to the pod frame; aids in dispersal when persistent</li>
<li>
<em>glumes</em>: The modified membranous bracts surrounding the spikelet of a grass; also see ‘bract’</li>
<li>
<em>hairs</em>: Modified hairs aid in seed dispersal</li>
<li>
<em>indehiscent</em>: Fruit that does not open at maturity in a pre-defined way, but instead relies on predation or decomposition to release the seeds;fruit type</li>
<li>
<em>inflated_parts</em>: When some part of the seed, fruit, or associated tissues is inflated, aiding in seed or fruit dispersal</li>
<li>
<em>none</em>: When a fruit and associated tissues lack any dispersal appendages; see also exarillate</li>
<li>
<em>paddles</em>: unknown meaning; currently used only for “Brachyscome dentata”</li>
<li>
<em>pappus</em>: The pappus is the modified calyx, the part of an individual floret, that surrounds the base of the corolla tube in flower heads of the plant family Asteraceae</li>
<li>
<em>placental_endocarp</em>: Exact dispersal mechanism unclear; currently used only for “Crowea saligna” (maybe others now)</li>
<li>
<em>pseudo-wing</em>: UNKNOWN</li>
<li>
<em>receptacle</em>: A winged achene, a type of fruit in which a flattened wing of fibrous, papery tissue develops from the ovary wall</li>
<li>
<em>samara</em>: Fleshy seed coat; term especially used for cycads; dispersal appendage?</li>
<li>
<em>sarcotesta</em>: When scales assist in seed or fruit dispersal</li>
<li>
<em>scales</em>: Seed containing a sac of air</li>
<li>
<em>seed_airsac</em>: A seed wing that arises from one side; dispersal appendage</li>
<li>
<em>seed_unilaterally_winged</em>: Seed wing absent</li>
<li>
<em>seed_wing_obsolete</em>: Modified short hairs aid in seed dispersal</li>
<li>
<em>short_hairs</em>: Spines that aid in seed and fruit dispersal</li>
<li>
<em>spines</em>: Outgrowth of the hilum region which restricts water movement into and out of some seeds, especially legumes</li>
<li>
<em>strophiole</em>: REPLACE WITH ‘strophiole’</li>
<li>
<em>strophiole_on_seed</em>: Cases where a species’ style, stigma or carpel are persistent and aid in seed or fruit dispersal</li>
<li>
<em>style</em>: A fruit where some tissue layers have become flattened and papery to function as wings; includes reference to a winged nut and winged achene (samara); fruit type</li>
<li>
<em>winged_fruit</em>: Referring to wing-like seed extensions that aid in wind dispersal</li>
<li>
<em>wings</em>: UNSURE; assume it refers to conifers, where woody bracts protect the seeds</li>
<li>
<em>woody_bract_scales</em>: Fleshy outgrowth of a seed, that often attracts animals like birds or ants</li>
</ul>
</li>
</ul>
<p><strong>dispersal_syndrome</strong></p>
<ul>
<li>label: Dispersal syndrome</li>
<li>description: Type of dispersal syndrome displayed by species, although the list includes many dispersal appendages and fruit types. Many definitions come from Kew Botanic Gardens website.</li>
<li>number of records: 2451</li>
<li>number of studies: 18</li>
<li>type: categorical</li>
<li>allowable values:
<ul>
<li>
<em>adhesion</em>: Dispersal syndrome where fruit is transported by attaching itself to something, usually an animal; true dispersal syndrome; will overlap with exozoochory; also called epizoochory</li>
<li>
<em>anemochory</em>: Dispersal by wind; true dispersal syndrome</li>
<li>
<em>animal_vector</em>: Seeds are dispersed by an animal; true dispersal syndrome; should be replaced with zoochory</li>
<li>
<em>aril</em>: Fleshy outgrowth of a seed, that often attracts animals like birds or ants; dispersal appendage</li>
<li>
<em>atelochory</em>: Dispersal is prevented; also called antitelochory</li>
<li>
<em>ballistic</em>: Seeds are launched away from the plant by explosion as soon as the seed capsule opens; true dispersal syndrome</li>
<li>
<em>chamaechory</em>: diaspore is rolled along ground surface by wind</li>
<li>
<em>dyszoochory</em>: diaspore is eaten intentionally by an animal</li>
<li>
<em>elaiosome</em>: Fleshy (often fatty) appendage on seeds that attracts ants; dispersal appendage</li>
<li>
<em>endozoochory</em>: Diaspores that dispersed through ingestion by animals, either intentionally or unintentionally and are transported before dropping off; includes the term ingestion; true dispersal syndrome</li>
<li>
<em>exozoochory</em>: When seeds are dispersed on the surface of an animal; will overlap significantly with the terms adhesion and animal_vector</li>
<li>
<em>fleshy_fruit</em>: A fruit where part or all of the pericarp (fruit wall) is fleshy at maturity; fruit type</li>
<li>
<em>fleshy_wings_capsule</em>: A fruit whose capsule or wings are fleshy</li>
<li>
<em>hydrochory</em>: Dispersal on the surface of water; see also the term water; true dispersal syndrome</li>
<li>
<em>mobile</em>: UNKNOWN</li>
<li>
<em>myrmecochory</em>: Dispersules with elaiosomes (specialised nutritious appendages) that make them attractive for capture, transport and use by ants or related insects; true dispersal syndrome</li>
<li>
<em>nautohydrochory</em>: water; e.g. floating/submerged in fresh/saltwater currents</li>
<li>
<em>ombrohydrochory</em>: water; diaspore is propelled by action of rain on plant structure/wetting by rain or dew</li>
<li>
<em>parent_plant_or_diaspore</em>: methods originating from parent plant or diaspore</li>
<li>
<em>synzoochory</em>: animal; diaspore is carried intentionally</li>
<li>
<em>unassisted</em>: Seeds are dispersed without assistance; true dispersal syndrome; same as barochory; seeds drop to the ground close to or beneath the parent plant</li>
<li>
<em>undefined</em>: UNKNOWN</li>
<li>
<em>vertebrate</em>: Seeds are dispersed by a vertebrate species; true dispersal syndrome</li>
<li>
<em>water</em>: Seeds dispersal depends in some way on water; should use hydrochory instead</li>
<li>
<em>wind</em>: Seeds are dispersed by wind; true dispersal syndrome; should use anemochory instead</li>
<li>
<em>xerochasywater</em>: wetting by rain or dew; also called hygrochasy</li>
</ul>
</li>
</ul>
<p><strong>dispersers</strong></p>
<ul>
<li>label: Types of animals dispersing fruit</li>
<li>description: Types of animals dispersing fruit</li>
<li>number of records: 0</li>
<li>number of studies: 0</li>
<li>type: categorical</li>
<li>allowable values:
<ul>
<li>
<em>aborigines</em>: aborigines</li>
<li>
<em>ants</em>: amt</li>
<li>
<em>birds</em>: birds</li>
<li>
<em>flying_foxes</em>: flying foxes</li>
<li>
<em>mammals</em>: mammals</li>
</ul>
</li>
</ul>
<p><strong>dormancy_type</strong></p>
<ul>
<li>label: Dormancy type</li>
<li>description: Classification for seed dormancy</li>
<li>number of records: 13901</li>
<li>number of studies: 1</li>
<li>type: categorical</li>
<li>allowable values:
<ul>
<li>
<em>morphophysiological_dormancy</em>: Seeds exhibit morphophysiological dormancy</li>
<li>
<em>non_dormant</em>: Seeds are non-dormant</li>
<li>
<em>physical_dormancy</em>: Seeds exhibit physical dormancy</li>
<li>
<em>physiological_dormancy</em>: Seeds exhibit physiological dormancy</li>
</ul>
</li>
</ul>
<p><strong>embryo_colour</strong></p>
<ul>
<li>label: Embryo colour</li>
<li>description: Binary variable distinguishing between embryos that are green versus colourless</li>
<li>number of records: 296</li>
<li>number of studies: 1</li>
<li>type: categorical</li>
<li>allowable values:
<ul>
<li>
<em>colourless</em>: Colourless embryo</li>
<li>
<em>green</em>: Green embryo</li>
</ul>
</li>
</ul>
<p><strong>upper_epidermis_thickness</strong></p>
<ul>
<li>label: Upper epidermis thickness</li>
<li>description: Thickness of the upper epidermis</li>
<li>number of records: 149</li>
<li>number of studies: 3</li>
<li>type: numeric</li>
<li>units: um</li>
<li>allowable range: 1 - 100 um</li>
</ul>
<p><strong>lower_epidermis_thickness</strong></p>
<ul>
<li>label: Lower epidermis thickness</li>
<li>description: Thickness of the lower epidermis</li>
<li>number of records: 151</li>
<li>number of studies: 3</li>
<li>type: numeric</li>
<li>units: um</li>
<li>allowable range: 1 - 100 um</li>
</ul>
<p><strong>epithelial_cell_density</strong></p>
<ul>
<li>label: Epithelial cell density on the leaf surface</li>
<li>description: Epithelial cell density on the leaf surface</li>
<li>number of records: 0</li>
<li>number of studies: 0</li>
<li>type: numeric</li>
<li>units: count/mm2</li>
<li>allowable range: 100 - 10000 count/mm2</li>
</ul>
<p><strong>epithelial_cell_layer_thickness</strong></p>
<ul>
<li>label: Epithelial cell layer thickness on the leaf surface</li>
<li>description: Epithelial cell layer thickness on the leaf surface</li>
<li>number of records: 0</li>
<li>number of studies: 0</li>
<li>type: numeric</li>
<li>units: um</li>
<li>allowable range: 1 - 100 um</li>
</ul>
<p><strong>fire_response</strong></p>
<ul>
<li>label: Resprouts or is killed by fire</li>
<li>description: Distinguishes between plants that are killed by fire and resprout following fire</li>
<li>number of records: 3554</li>
<li>number of studies: 8</li>
<li>type: categorical</li>
<li>allowable values:
<ul>
<li>
<em>fire_killed</em>: Plants killed by hot fires</li>
<li>
<em>weak_resprouting</em>: Plant shows weak resprouting following fire</li>
<li>
<em>intermediate_resprouting</em>: Plant shows intermediate resprouting following fire</li>
<li>
<em>strong_resprouting</em>: Plant shows strong resprouting following fire</li>
<li>
<em>resprouts</em>: Plants resprout from underground storage organ following fire. (For studies that don’t differentiate between respouting strength)</li>
</ul>
</li>
</ul>
<p><strong>fire_response_detailed</strong></p>
<ul>
<li>label: Resprouts or is killed by fire</li>
<li>description: Detailed information distinguishing between plants that are killed by fire and resprout following fire</li>
<li>number of records: 46</li>
<li>number of studies: 1</li>
<li>type: categorical</li>
<li>allowable values:
<ul>
<li>
<em>fire_killed</em>: Plants killed by hot fires</li>
<li>
<em>weak_resprouting</em>: Plant shows weak resprouting following fire</li>
<li>
<em>intermediate_resprouting</em>: Plant shows intermediate resprouting following fire</li>
<li>
<em>strong_resprouting</em>: Plant shows strong resprouting following fire</li>
</ul>
</li>
</ul>
<p><strong>fire_cued_seeding</strong></p>
<ul>
<li>label: Fire-cued seeding</li>
<li>description: Distinguishes between plants that do and do not have fire-cued seeding</li>
<li>number of records: 0</li>
<li>number of studies: 0</li>
<li>type: categorical</li>
<li>allowable values:
<ul>
<li>
<em>fire_cued_seeding</em>: Plants that germinate robustly following fire</li>
<li>
<em>no_fire_cued_seeding</em>: Plants that do not show increased seeding following fire</li>
</ul>
</li>
</ul>
<p><strong>flower_colour</strong></p>
<ul>
<li>label: Flower colour</li>
<li>description: Flower colour, with six possible outcomes</li>
<li>number of records: 8633</li>
<li>number of studies: 1</li>
<li>type: categorical</li>
<li>allowable values:
<ul>
<li>
<em>blue_purple</em>: blue or purple</li>
<li>
<em>green</em>: green flower</li>
<li>
<em>pink</em>: pink flower</li>
<li>
<em>red_brown</em>: red or brown flower</li>
<li>
<em>white_cream</em>: white or cream flower</li>
<li>
<em>yellow_orange</em>: yellow orange flower</li>
</ul>
</li>
</ul>
<p><strong>flower_count_maximum</strong></p>
<ul>
<li>label: Maximum flower number</li>
<li>description: Maximum flower number produced</li>
<li>number of records: 0</li>
<li>number of studies: 0</li>
<li>type: numeric</li>
<li>units: count</li>
<li>allowable range: 0 - 1000000 count</li>
</ul>
<p><strong>flowering_time</strong></p>
<ul>
<li>label: Range of flowering period</li>
<li>description: Months during which species is flowering; keyed as a sequences of 12 0s (not flowering) and 1s (flowering) starting with January</li>
<li>number of records: 15454</li>
<li>number of studies: 26</li>
<li>type: character</li>
<li>allowable values:
<ul>
<li>**:</li>
</ul>
</li>
</ul>
<p><strong>fluorescence_Jmax_per_mass</strong></p>
<ul>
<li>label: Capacity for photosynthetic electron transport, measured through chlorophyll fluorescence, on a per mass basis</li>
<li>description: Capacity for photosynthetic electron transport, measured through chlorophyll fluorescence, on a per mass basis</li>
<li>number of records: 0</li>
<li>number of studies: 0</li>
<li>type: numeric</li>
<li>units: umol/g/s</li>
<li>allowable range: 0 - 3 umol/g/s</li>
</ul>
<p><strong>fluorescence_Vcmax_per_mass</strong></p>
<ul>
<li>label: Maximum carboxylase activity of ribulose 1,5-bisphosphate carboxylase/oxygenase (Rubisco), measured through chlorophyll fluorescence, on a per mass basis</li>
<li>description: Maximum carboxylase activity of ribulose 1,5-bisphosphate carboxylase/oxygenase (Rubisco), measured through chlorophyll fluorescence, on a per mass basis</li>
<li>number of records: 0</li>
<li>number of studies: 0</li>
<li>type: numeric</li>
<li>units: umol/g</li>
<li>allowable range: 0 - 1 umol/g</li>
</ul>
<p><strong>fluorescence_Jmax_over_Vcmax</strong></p>
<ul>
<li>label: Ratio of photosynthetic electron transport capacity to maximum Rubisco activity, measured through chlorophyll fluorescence</li>
<li>description: Ratio of photosynthetic electron transport capacity to maximum Rubisco activity, measured through chlorophyll fluorescence</li>
<li>number of records: 0</li>
<li>number of studies: 0</li>
<li>type: numeric</li>
<li>units: 1/s</li>
<li>allowable range: 0 - 5 1/s</li>
</ul>
<p><strong>fruit_mass</strong></p>
<ul>
<li>label: Fruit mass</li>
<li>description: The weight of a fruit, including the seed</li>
<li>number of records: 0</li>
<li>number of studies: 0</li>
<li>type: numeric</li>
<li>units: mg</li>
<li>allowable range: 1 - 10000 mg</li>
</ul>
<p><strong>fruit_width</strong></p>
<ul>
<li>label: Fruit diameter</li>
<li>description: The longest width dimension of a fruit; orthogonal to the length</li>
<li>number of records: 531</li>
<li>number of studies: 5</li>
<li>type: numeric</li>
<li>units: mm</li>
<li>allowable range: 0.1 - 100 mm</li>
</ul>
<p><strong>fruit_length</strong></p>
<ul>
<li>label: Fruit length</li>
<li>description: Longest fruit dimension or if clearly recognizable the length from its base to its apex</li>
<li>number of records: 656</li>
<li>number of studies: 5</li>
<li>type: numeric</li>
<li>units: mm</li>
<li>allowable range: 0.1 - 1000 mm</li>
</ul>
<p><strong>fruit_breadth</strong></p>
<ul>
<li>label: Fruit breadth</li>
<li>description: The shorter width dimension of a fruit; orthogonal to the length</li>
<li>number of records: 0</li>
<li>number of studies: 0</li>
<li>type: numeric</li>
<li>units: mm</li>
<li>allowable range: 0.1 - 100 mm</li>
</ul>
<p><strong>fruit_type</strong></p>
<ul>
<li>label: Fruit type</li>
<li>description: Fruit types</li>
<li>number of records: 4484</li>
<li>number of studies: 2</li>
<li>type: categorical</li>
<li>allowable values:
<ul>
<li>
<em>achene</em>: A small, dry, indehiscent one-seeded fruit with a thin wall</li>
<li>
<em>anthocarp_nut</em>: A false fruit consisting of the true fruit and the base of the floral whorls; instance where the true fruit is a nut</li>
<li>
<em>anthocarp_viscid</em>: UNCERTAIN what the viscid refers to (See anthocarp above)</li>
<li>
<em>berry</em>: Fleshy, indehiscent fruit in which the seeds are generally more than 1 and are not encased in a stone</li>
<li>
<em>berry_inf</em>: UNCERTAIN how this differs from general definition for berry</li>
<li>
<em>capsule</em>: Dry fruit from compound pistil, nearly always dehiscent</li>
<li>
<em>capsule_explosive</em>: Dry fruit from compound pistil, that releases its seeds in an explosive manner</li>
<li>
<em>capsule_aril</em>: Dry fruit from compound pistil, with a fleshy outgrowth (aril) to attract pollinators</li>
<li>
<em>capsule_wing</em>: Dry fruit from compound pistil, with a wing for dispersal by wind</li>
<li>
<em>caryopsis</em>: a dry one-seeded fruit in which the ovary wall is united with the seed coat, typical of grasses and cereals</li>
<li>
<em>cypsela</em>: An achene developed from an inferior bicarpellary ovary fused with the calyx tube (in Asteraceae)</li>
<li>
<em>dehiscent_pod</em>: Fleshy or pulpy, indehiscent, superficially berry-like fruit in which 1 seed is encased in a stone or more than 1 seed is encased in an equal number of free or variously fused stones</li>
<li>
<em>drupe</em>: Fibrous or dry drupe; indehiscent, superficially berry-like fruit in which 1 seed is encased in a stone or more than 1 seed is encased in an equal number of free or variously fused stones</li>
<li>
<em>drupe_fibrous</em>: Dry fruit from a simple pistil, dehiscent on generally only one side, along a single suture</li>
<li>
<em>fig</em>: Fruits from a legume; a dry or somewhat fleshy, 1- to many-seeded fruit from a simple pistil, typically dehiscent longitudinally along two sutures and splitting into halves that remain joined at the base</li>
<li>
<em>follicle</em>: Fruits from a legume; a legume (see above) that is indehiscent</li>
<li>
<em>indehiscent_pod</em>: Mostly dry, sometimes fleshy or pulpy, usually indehiscent fruit in which a single seed is encased in a hard shell</li>
<li>
<em>legume</em>: A nut that is enclosed in a winglike bract</li>
<li>
<em>legume_indehiscent</em>: Small, dry nut or nut-like fruit, usually several of which are produced by a single flower; especially an achene</li>
<li>
<em>mericarp</em>: A legume or superficially similar fruit</li>
<li>
<em>multiple_fruit</em>: A winged achene, a type of fruit in which a flattened wing of fibrous, papery tissue develops from the ovary wall</li>
<li>
<em>nut</em>: A dry fruit which splits into individual carpels</li>
<li>
<em>nut_winged</em>: A fleshy hollow receptacle that develops into a multiple fruit, as in the fig</li>
<li>
<em>nutlet</em>: A multiple fruit consisting of several united fruits, originating from several originally free carpels, usually fleshy</li>
<li>
<em>pod</em>: Small, bladder-like, thin-walled indehiscent fruit</li>
<li>
<em>samara</em>: Small, bladder-like, thin-walled indehiscent fruit, that is partially fleshy</li>
<li>
<em>schizocarp</em>: Small, bladder-like, thin-walled indehiscent fruit, that is spiny</li>
<li>
<em>syconium</em>: A small, dry, indehiscent one-seeded fruit with a thin wall</li>
<li>
<em>syncarp</em>: A false fruit consisting of the true fruit and the base of the floral whorls; instance where the true fruit is a nut</li>
<li>
<em>utricle</em>: UNCERTAIN what the viscid refers to (See anthocarp above)</li>
<li>
<em>utricle_fleshy</em>: Fleshy, indehiscent fruit in which the seeds are generally more than 1 and are not encased in a stone</li>
<li>
<em>utricle_spiny</em>: UNCERTAIN how this differs from general definition for berry</li>
</ul>
</li>
</ul>
<p><strong>fruit_type_botany</strong></p>
<ul>
<li>label: Fruit type (fleshy / dry)</li>
<li>description: Binary variable, dividing fruits into ‘dry’ versus ‘fleshy’ based on their appearance</li>
<li>number of records: 4044</li>
<li>number of studies: 1</li>
<li>type: categorical</li>
<li>allowable values:
<ul>
<li>
<em>dry</em>: fruit whose appearance is dry</li>
<li>
<em>fleshy</em>: fruit whose appearance is fleshy</li>
</ul>
</li>
</ul>
<p><strong>fruit_type_function</strong></p>
<ul>
<li>label: Fruit type (fleshy / dry)</li>
<li>description: Binary variable, dividing fruits into ‘dry’ versus ‘fleshy’ based on their function</li>
<li>number of records: 4126</li>
<li>number of studies: 1</li>
<li>type: categorical</li>
<li>allowable values:
<ul>
<li>
<em>dry</em>: fruit that are functionally dry</li>
<li>
<em>fleshy</em>: fruit that are functionally fleshy</li>
</ul>
</li>
</ul>
<p><strong>fruiting_time</strong></p>
<ul>
<li>label: Plant reproductive phenology timing (fruiting)</li>
<li>description: Months during which species is fruiting; keyed as a sequences of 12 0s (not flowering) and 1s (flowering) starting with January</li>
<li>number of records: 1144</li>
<li>number of studies: 4</li>
<li>type: character</li>
<li>allowable values:
<ul>
<li>**:</li>
</ul>
</li>
</ul>
<p><strong>fv_over_fm</strong></p>
<ul>
<li>label: Fv/Fm</li>
<li>description: Chlorophyll fluorescence measurement that indicates whether plant stress affects photo-system II in a dark adapted state</li>
<li>number of records: 20</li>
<li>number of studies: 1</li>
<li>type: numeric</li>
<li>units: x/x</li>
<li>allowable range: 0.2 - 1 x/x</li>
</ul>
<p><strong>germination</strong></p>
<ul>
<li>label: Germination (proportion)</li>
<li>description: Proportion of seeds that germinate</li>
<li>number of records: 0</li>
<li>number of studies: 0</li>
<li>type: numeric</li>
<li>units: n/n</li>
<li>allowable range: 0 - 1 n/n</li>
</ul>
<p><strong>germination_treatment</strong></p>
<ul>
<li>label: Germination treatment</li>
<li>description: Seed treatment required for germination</li>
<li>number of records: 3530</li>
<li>number of studies: 2</li>
<li>type: categorical</li>
<li>allowable values:
<ul>
<li>
<em>apply_chemical</em>: apply generic chemical treatment</li>
<li>
<em>crack_seed</em>: same as remove seed_coat</li>
<li>
<em>excise_pericarp</em>: heating the seed</li>
<li>
<em>heat</em>: soaked</li>
<li>
<em>imbibed</em>: soaked in a mild bleach solution</li>
<li>
<em>imbibed_bleach</em>: soaked in a mild solution of calcium hypochlorite (is this always the same as standard bleach?)</li>
<li>
<em>imbibed_calcium_hypochlorite</em>: soaked in water that has boiled and then cooled slightly</li>
<li>
<em>imbibed_cooled_boiling</em>: soaked in distilled water</li>
<li>
<em>imbibed_distilled</em>: soaked in hot, but not boiling water</li>
<li>
<em>imbibed_hot_water</em>: soaked in near-boiling water</li>
<li>
<em>imbibed_near_boiling</em>: soaked and seed casing removed</li>
<li>
<em>imbibed_remove_casing</em>: soaked to leach salts</li>
<li>
<em>imbibed_remove_salt</em>: soaked in seed primer</li>
<li>
<em>imbibed_seed_primer</em>: soaked, then seed is nicked</li>
<li>
<em>imbibed_then_nicked</em>: make incision into pericarp</li>
<li>
<em>incise_pericarp</em>: remove bracts surrounding the seed</li>
<li>
<em>remove_bracts</em>: remove the mericarp plug</li>
<li>
<em>remove_mericarp_plug</em>: same as excise pericarp</li>
<li>
<em>remove_seed_casing</em>: rinse the seed (generic)</li>
<li>
<em>remove_seed_coat</em>: Remove or partially remove a covering that is not the seed coat</li>
<li>
<em>rinse</em>: Remove or partially remove a covering that is not the seed coat. Covering sufficiently tough a vice is required to break it.</li>
<li>
<em>scarify_file_seed_coat</em>: Chip, nick, or otherwise partially remove the seed coat</li>
<li>
<em>scarify_fruit</em>: Sand the seed coat</li>
<li>
<em>scarify_fruit_vice</em>: Simulate smoke conditions for a period using the disk method.</li>
<li>
<em>scarify_nick_seed_coat</em>: Soak seed in smoked water</li>
<li>
<em>scarify_pierce_seed_coat</em>: Seed sterilised in calcium hydroxide (or similar)</li>
<li>
<em>scarify_remove_aril</em>: Seed sterilised in 10% doemestos (or similar)</li>
<li>
<em>scarify_remove_seed_coat</em>: Cold treatment</li>
<li>
<em>scarify_sand_seed_coat</em>: apply generic chemical treatment</li>
<li>
<em>scarify_seed_coat_vice</em>: same as remove seed_coat</li>
<li>
<em>simulate_smoke</em>: heating the seed</li>
<li>
<em>smoked_water</em>: soaked</li>
<li>
<em>sterilise_calcium_hydroxide</em>: soaked in a mild bleach solution</li>
<li>
<em>sterilise_domestos</em>: soaked in a mild solution of calcium hypochlorite (is this always the same as standard bleach?)</li>
<li>
<em>stratification</em>: soaked in water that has boiled and then cooled slightly</li>
</ul>
</li>
</ul>
<p><strong>genome_size</strong></p>
<ul>
<li>label: Plant genome size</li>
<li>description: Mass of the plant’s genome</li>
<li>number of records: 92</li>
<li>number of studies: 1</li>
<li>type: numeric</li>
<li>units: pg</li>
<li>allowable range: 1 - 10 pg</li>
</ul>
<p><strong>glaucous</strong></p>
<ul>
<li>label: Glaucous</li>
<li>description: Variable indicating if a plant’s leaves are glaucous or not</li>
<li>number of records: 5</li>
<li>number of studies: 1</li>
<li>type: categorical</li>
<li>allowable values:
<ul>
<li>
<em>0</em>: not glaucous</li>
<li>
<em>1</em>: glaucous</li>
</ul>
</li>
</ul>
<p><strong>growth_habit</strong></p>
<ul>
<li>label: Growth habit (data actually describes plant vegetative reproduction capacity)</li>
<li>description: Variable that defines a combination of growth habit and plant vegetative reproductive potential</li>
<li>number of records: 245</li>
<li>number of studies: 2</li>
<li>type: categorical</li>
<li>allowable values:
<ul>
<li>
<em>prostrate</em>: Plants that lie flat on the ground</li>
<li>
<em>rhizomatous</em>: Plants having modified underground stems</li>
<li>
<em>stoloniferous</em>: Plants having horizontal branches from the base of the plant that produce new plants from buds at its tip or nodes</li>
<li>
<em>tufted</em>: A dense clump attached at the base or growing close together; can refer to grasses, shrubs, or trees</li>
</ul>
</li>
</ul>
<p><strong>huber_value</strong></p>
<ul>
<li>label: Huber value</li>
<li>description: Sapwood area to leaf area ratio</li>
<li>number of records: 541</li>
<li>number of studies: 6</li>
<li>type: numeric</li>
<li>units: mm2_sapwood/mm2_leaf</li>
<li>allowable range: 1e-06 - 0.1 mm2_sapwood/mm2_leaf</li>
</ul>
<p><strong>hypocotyl_type</strong></p>
<ul>
<li>label: Hypocotyl type</li>
<li>description: Binary variable distinguishing between glabrous versus hairy hypocotyls (the embryonic axis to which the cotyeledons are attached).</li>
<li>number of records: 567</li>
<li>number of studies: 1</li>
<li>type: categorical</li>
<li>allowable values:
<ul>
<li>
<em>glabrous</em>: Hypocotyl lacks hairs</li>
<li>
<em>hairy</em>: Hypocotyl has hairs</li>
</ul>
</li>
</ul>
<p><strong>upper_hypodermis_thickness</strong></p>
<ul>
<li>label: Upper hypodermis thickness</li>
<li>description: Thickness of the upper hypodermis</li>
<li>number of records: 8</li>
<li>number of studies: 2</li>
<li>type: numeric</li>
<li>units: um</li>
<li>allowable range: 1 - 100 um</li>
</ul>
<p><strong>lower_hypodermis_thickness</strong></p>
<ul>
<li>label: Lower hypodermis thickness</li>
<li>description: Thickness of the lower hypodermis</li>
<li>number of records: 4</li>
<li>number of studies: 2</li>
<li>type: numeric</li>
<li>units: um</li>
<li>allowable range: 1 - 100 um</li>
</ul>
<p><strong>leaf_Al_per_dry_mass</strong></p>
<ul>
<li>label: Leaf aluminium (Al) content per unit leaf dry mass</li>
<li>description: Leaf aluminium (Al) content per unit leaf dry mass</li>
<li>number of records: 418</li>
<li>number of studies: 5</li>
<li>type: numeric</li>
<li>units: mg/kg</li>
<li>allowable range: 0.01 - 10000 mg/kg</li>
</ul>
<p><strong>leaf_angle</strong></p>
<ul>
<li>label: Leaf angle, relative to horizontal</li>
<li>description: Leaf angle, relative to horizontal</li>
<li>number of records: 0</li>
<li>number of studies: 0</li>
<li>type: numeric</li>
<li>units: degrees</li>
<li>allowable range: -90 - 90 degrees</li>
</ul>
<p><strong>leaf_area</strong></p>
<ul>
<li>label: Leaf area</li>
<li>description: Area of the leaf surface</li>
<li>number of records: 13895</li>
<li>number of studies: 47</li>
<li>type: numeric</li>
<li>units: mm2</li>
<li>allowable range: 0.1 - 1e+07 mm2</li>
</ul>
<p><strong>leaf_area_ratio</strong></p>
<ul>
<li>label: Leaf area per unit plant dry mass</li>
<li>description: Ratio of leaf area to total plant dry mass</li>
<li>number of records: 109</li>
<li>number of studies: 4</li>
<li>type: numeric</li>
<li>units: mm2/mg</li>
<li>allowable range: 1 - 100 mm2/mg</li>
</ul>
<p><strong>leaf_absorption</strong></p>
<ul>
<li>label: Proportion of incoming light that is absorbed by the leaf</li>
<li>description: Proportion of incoming light between 400-700 nm that is absorbed by the leaf</li>
<li>number of records: 0</li>
<li>number of studies: 0</li>
<li>type: numeric</li>
<li>units: umol/umol</li>
<li>allowable range: 0.01 - 1 umol/umol</li>
</ul>
<p><strong>leaf_reflectance</strong></p>
<ul>
<li>label: Proportion of incoming light that is reflected by the leaf</li>
<li>description: Proportion of incoming light between 400-700 nm that is reflected by the leaf</li>
<li>number of records: 0</li>
<li>number of studies: 0</li>
<li>type: numeric</li>
<li>units: umol/umol</li>
<li>allowable range: 0.001 - 1 umol/umol</li>
</ul>
<p><strong>leaf_transmission</strong></p>
<ul>
<li>label: Proportion of incoming light that is transmitted through the leaf</li>
<li>description: Proportion of incoming light between 400-700 nm that is transmitted through the leaf</li>
<li>number of records: 0</li>
<li>number of studies: 0</li>
<li>type: numeric</li>
<li>units: umol/umol</li>
<li>allowable range: 0.001 - 1 umol/umol</li>
</ul>
<p><strong>leaf_PRI</strong></p>
<ul>
<li>label: Photochemical Reflectance Index</li>
<li>description: The photochemical reflectance index measures plant responses to stress, by indicating changes in carotenoid pigments in live foliage.</li>
<li>number of records: 0</li>
<li>number of studies: 0</li>
<li>type: numeric</li>
<li>units: umol/umol</li>
<li>allowable range: -1 - 1 umol/umol</li>
</ul>
<p><strong>leaf_arrangement</strong></p>
<ul>
<li>label: Arrangement of leaves</li>
<li>description: Describes leaf arrangement on the stem</li>
<li>number of records: 5983</li>
<li>number of studies: 1</li>
<li>type: categorical</li>
<li>allowable values:
<ul>
<li>
<em>alternate</em>: A single leaf is attached at each stem node</li>
<li>
<em>opposite</em>: Leaves attach to the stem in pairs</li>
<li>
<em>reduced_to_scales</em>: Plant lacks true leaves. Leaves are instead small scales</li>
<li>
<em>whorled_tufted_crowded</em>: Three or more leaves are attached at each node</li>
</ul>
</li>
</ul>
<p><strong>leaf_B_per_dry_mass</strong></p>
<ul>
<li>label: Leaf boron (B) content per unit leaf dry mass</li>
<li>description: Leaf boron (B) content per unit leaf dry mass</li>
<li>number of records: 490</li>
<li>number of studies: 4</li>
<li>type: numeric</li>
<li>units: mg/kg</li>
<li>allowable range: 0.1 - 1000 mg/kg</li>
</ul>
<p><strong>leaf_C_per_dry_mass</strong></p>
<ul>
<li>label: Leaf carbon (C) content per unit leaf dry mass</li>
<li>description: Leaf carbon (C) content per unit leaf dry mass</li>
<li>number of records: 1126</li>
<li>number of studies: 7</li>
<li>type: numeric</li>
<li>units: mg/g</li>
<li>allowable range: 50 - 750 mg/g</li>
</ul>
<p><strong>leaf_Ca_per_dry_mass</strong></p>
<ul>
<li>label: Leaf calcium (Ca) content per unit leaf dry mass</li>
<li>description: Leaf calcium (Ca) content per unit leaf dry mass</li>
<li>number of records: 816</li>
<li>number of studies: 8</li>
<li>type: numeric</li>
<li>units: mg/g</li>
<li>allowable range: 0.01 - 1000 mg/g</li>
</ul>
<p><strong>leaf_cell_wall_fraction</strong></p>
<ul>
<li>label: Leaf cell wall fraction (fraction of cell wall material recovered from total leaf biomass)</li>
<li>description: Fraction of total leaf biomass that is cell wall material</li>
<li>number of records: 55</li>
<li>number of studies: 2</li>
<li>type: numeric</li>
<li>units: dimensionless</li>
<li>allowable range: 0.1 - 1 dimensionless</li>
</ul>
<p><strong>leaf_cell_wall_N</strong></p>
<ul>
<li>label: Leaf cell wall nitrogen concentration</li>
<li>description: Proportion of leaf cell wall material that is nitrogen</li>
<li>number of records: 29</li>
<li>number of studies: 1</li>
<li>type: numeric</li>
<li>units: mmol/g</li>
<li>allowable range: 0.1 - 10 mmol/g</li>
</ul>
<p><strong>leaf_cell_wall_N_fraction</strong></p>
<ul>
<li>label: Leaf cell wall nitrogen fraction</li>
<li>description: Proportion of all N in leaves that is found in the leaf cell walls</li>
<li>number of records: 29</li>
<li>number of studies: 1</li>
<li>type: numeric</li>
<li>units: dimensionless</li>
<li>allowable range: 0.01 - 1 dimensionless</li>
</ul>
<p><strong>leaf_CN_ratio</strong></p>
<ul>
<li>label: Leaf carbon/nitrogen (C/N) ratio</li>
<li>description: Leaf carbon/nitrogen (C/N) ratio</li>
<li>number of records: 310</li>
<li>number of studies: 2</li>
<li>type: numeric</li>
<li>units: g/g</li>
<li>allowable range: 10 - 1000 g/g</li>
</ul>
<p><strong>leaf_compoundness</strong></p>
<ul>
<li>label: Leaf compoundness</li>
<li>description: Indicates whether or not a leaf is compound; different ‘simple’ terminology used by different studies</li>
<li>number of records: 18388</li>
<li>number of studies: 17</li>
<li>type: categorical</li>
<li>allowable values:
<ul>
<li>
<em>compound</em>: A leaf that is divided into multiple leaflets</li>
<li>
<em>leafless</em>: A stem that lacks leaves</li>
<li>
<em>lobed</em>: A leaf that is lobed but not compound</li>
<li>
<em>palmate</em>: A leaf that is palmately compound</li>
<li>
<em>phyllode</em>: Stem segment that functions like a leaf</li>
<li>
<em>simple</em>: A leaf with a single undivided blade</li>
<li>
<em>trifoliate</em>: A leaf that is divided into three leaflets; a special category of compound</li>
<li>
<em>unifoliate</em>: A leaf that consists of a single leaflet; a special category of compound</li>
</ul>
</li>
</ul>
<p><strong>leaf_Cu_per_dry_mass</strong></p>
<ul>
<li>label: Leaf copper (Cu) content per unit leaf dry mass</li>
<li>description: Leaf copper (Cu) content per unit leaf dry mass</li>
<li>number of records: 768</li>
<li>number of studies: 6</li>
<li>type: numeric</li>
<li>units: mg/kg</li>
<li>allowable range: 0.01 - 150 mg/kg</li>
</ul>
<p><strong>leaf_dark_respiration_per_area</strong></p>
<ul>
<li>label: Leaf respiration rate per unit leaf area</li>
<li>description: Leaf respiration rate per unit leaf area</li>
<li>number of records: 216</li>
<li>number of studies: 3</li>
<li>type: numeric</li>
<li>units: umolCO2/m2/s</li>
<li>allowable range: 1e-05 - 10 umolCO2/m2/s</li>
</ul>
<p><strong>leaf_dark_respiration_per_dry_mass</strong></p>
<ul>
<li>label: Leaf respiration rate per unit leaf dry mass</li>
<li>description: Leaf respiration rate per unit leaf dry mass</li>
<li>number of records: 307</li>
<li>number of studies: 4</li>
<li>type: numeric</li>
<li>units: umolCO2/g/s</li>
<li>allowable range: 0.001 - 100 umolCO2/g/s</li>
</ul>
<p><strong>leaf_light_respiration_per_area</strong></p>
<ul>
<li>label: Leaf respiration rate per unit leaf area under light</li>
<li>description: Leaf respiration rate per unit leaf area under light</li>
<li>number of records: 42</li>
<li>number of studies: 1</li>
<li>type: numeric</li>
<li>units: umolCO2/m2/s</li>
<li>allowable range: 1e-05 - 10000 umolCO2/m2/s</li>
</ul>
<p><strong>leaf_delta13C</strong></p>
<ul>
<li>label: Leaf carbon (C) isotope signature (delta 13C)</li>
<li>description: Leaf carbon stable isotope signature</li>
<li>number of records: 1570</li>
<li>number of studies: 12</li>
<li>type: numeric</li>
<li>units: per mille</li>
<li>allowable range: -50 - 0 per mille</li>
</ul>
<p><strong>leaf_delta15N</strong></p>
<ul>
<li>label: Leaf nitrogen (N) isotope signature (delta 15N)</li>
<li>description: Leaf nitrogen stable isotope signature</li>
<li>number of records: 725</li>
<li>number of studies: 6</li>
<li>type: numeric</li>
<li>units: per mille</li>
<li>allowable range: -25 - 75 per mille</li>
</ul>
<p><strong>leaf_density</strong></p>
<ul>
<li>label: Leaf tissue density</li>
<li>description: Leaf tissue density</li>
<li>number of records: 0</li>
<li>number of studies: 0</li>
<li>type: numeric</li>
<li>units: mg/mm3</li>
<li>allowable range: 0.1 - 10 mg/mm3</li>
</ul>
<p><strong>leaf_dry_mass</strong></p>
<ul>
<li>label: Leaf dry mass</li>
<li>description: Leaf dry mass</li>
<li>number of records: 2448</li>
<li>number of studies: 11</li>
<li>type: numeric</li>
<li>units: mg</li>
<li>allowable range: 0.1 - 15000 mg</li>
</ul>
<p><strong>leaf_dry_matter_content</strong></p>
<ul>
<li>label: Leaf dry mass per unit leaf fresh mass (Leaf dry matter content, LDMC)</li>
<li>description: Leaf dry mass per unit leaf fresh mass. (See also leaf_water_content_per_mass, the ratio of water content to leaf dry mass, recorded by some studies.)</li>
<li>number of records: 1129</li>
<li>number of studies: 11</li>
<li>type: numeric</li>
<li>units: g/g</li>
<li>allowable range: 0.001 - 20 g/g</li>
</ul>
<p><strong>leaf_elastic_modulus</strong></p>
<ul>
<li>label: Leaf elastic modulus</li>
<li>description: The ratio of the change in cell turgor relative to the change in cell volume</li>
<li>number of records: 0</li>
<li>number of studies: 0</li>
<li>type: numeric</li>
<li>units: mmol/m2/s</li>
<li>allowable range: 0.1 - 100 mmol/m2/s</li>
</ul>
<p><strong>leaf_Fe_per_dry_mass</strong></p>
<ul>
<li>label: Leaf iron (Fe) content per unit leaf dry mass</li>
<li>description: Leaf iron (Fe) content per unit leaf dry mass</li>
<li>number of records: 768</li>
<li>number of studies: 6</li>
<li>type: numeric</li>
<li>units: mg/kg</li>
<li>allowable range: 1 - 10000 mg/kg</li>
</ul>
<p><strong>leaf_fracture_force</strong></p>
<ul>
<li>label: Leaf force of fracture</li>
<li>description: Measures of leaf mechanical resistant; how much force is required to shear, punch or rip a leaf</li>
<li>number of records: 156</li>
<li>number of studies: 3</li>
<li>type: numeric</li>
<li>units: N</li>
<li>allowable range: 0.01 - 10 N</li>
</ul>
<p><strong>leaf_fresh_mass</strong></p>
<ul>
<li>label: Leaf fresh mass</li>
<li>description: Leaf fresh mass</li>
<li>number of records: 136</li>
<li>number of studies: 2</li>
<li>type: numeric</li>
<li>units: mg</li>
<li>allowable range: 0.1 - 25000 mg</li>
</ul>
<p><strong>leaf_fresh_mass_per_area</strong></p>
<ul>
<li>label: Leaf fresh mass per leaf area</li>
<li>description: Leaf fresh mass per leaf area</li>
<li>number of records: 108</li>
<li>number of studies: 1</li>
<li>type: numeric</li>
<li>units: g/m2</li>
<li>allowable range: 200 - 2000 g/m2</li>
</ul>
<p><strong>leaf_hairs_adult</strong></p>
<ul>
<li>label: Presence of hairs on adult leaves</li>
<li>description: Binary variable describing whether or not adult leaves have hairs</li>
<li>number of records: 137</li>
<li>number of studies: 2</li>
<li>type: categorical</li>
<li>allowable values:
<ul>
<li>
<em>0</em>: leaves on adult plants do not have hairs</li>
<li>
<em>1</em>: leaves on adult plants do have hairs</li>
</ul>
</li>
</ul>
<p><strong>leaf_hairs_juvenile</strong></p>
<ul>
<li>label: Presence of hairs on juvenile leaves</li>
<li>description: Binary variable describing whether or not juvenile leaves have hairs</li>
<li>number of records: 75</li>
<li>number of studies: 1</li>
<li>type: categorical</li>
<li>allowable values:
<ul>
<li>
<em>0</em>: leaves on juvenile plants do not have hairs</li>
<li>
<em>1</em>: leaves on juvenile plants do have hairs</li>
</ul>
</li>
</ul>
<p><strong>leaf_hydraulic_conductivity</strong></p>
<ul>
<li>label: Leaf hydraulic conductivity</li>
<li>description: Measure of how efficiently water is transported through the leaf, determined as the ratio of water flow rate through the leaf to the difference in water potential across the leaf, standardised to leaf area.</li>
<li>number of records: 19</li>
<li>number of studies: 1</li>
<li>type: numeric</li>
<li>units: mmol/m2/s/MPa</li>
<li>allowable range: 0.1 - 100 mmol/m2/s/MPa</li>
</ul>
<p><strong>leaf_hydraulic_vulnerability</strong></p>
<ul>
<li>label: Leaf hydraulic vulnerability</li>
<li>description: The leaf water potential value at which leaf hydraulic conductance have declined by 50% from the mean maximum rate</li>
<li>number of records: 20</li>
<li>number of studies: 1</li>
<li>type: numeric</li>
<li>units: neg_MPa</li>
<li>allowable range: 0.001 - 40 neg_MPa</li>
</ul>
<p><strong>leaf_K_per_area</strong></p>
<ul>
<li>label: Leaf potassium (K) content per unit leaf area</li>
<li>description: Leaf potassium (K) content per unit leaf area</li>
<li>number of records: 18</li>
<li>number of studies: 1</li>
<li>type: numeric</li>
<li>units: g/m2</li>
<li>allowable range: 0.1 - 10 g/m2</li>
</ul>
<p><strong>leaf_K_per_dry_mass</strong></p>
<ul>
<li>label: Leaf potassium (K) content per unit leaf dry mass</li>
<li>description: Leaf potassium (K) content per unit leaf dry mass</li>
<li>number of records: 840</li>
<li>number of studies: 9</li>
<li>type: numeric</li>
<li>units: mg/g</li>
<li>allowable range: 0.1 - 100 mg/g</li>
</ul>
<p><strong>leaf_length</strong></p>
<ul>
<li>label: Leaf length</li>
<li>description: Length of the leaf, including petiole and rachis in compound leaves</li>
<li>number of records: 37579</li>
<li>number of studies: 33</li>
<li>type: numeric</li>
<li>units: mm</li>
<li>allowable range: 0.1 - 1e+05 mm</li>
</ul>
<p><strong>leaf_lifespan</strong></p>
<ul>
<li>label: Leaf lifespan (longevity)</li>
<li>description: Leaf lifespan (longevity)</li>
<li>number of records: 270</li>
<li>number of studies: 4</li>
<li>type: numeric</li>
<li>units: month</li>
<li>allowable range: 1 - 1000 month</li>
</ul>
<p><strong>leaf_lignin_per_dry_mass</strong></p>
<ul>
<li>label: Leaf lignin per unit leaf dry mass</li>
<li>description: Leaf lignin per unit leaf dry mass</li>
<li>number of records: 0</li>
<li>number of studies: 0</li>
<li>type: numeric</li>
<li>units: mg/mg</li>
<li>allowable range: 0.001 - 1 mg/mg</li>
</ul>
<p><strong>leaf_margin</strong></p>
<ul>
<li>label: Leaf margin</li>
<li>description: Description of leaf margin are lobed versus entire. Trait values could be expanded to include toothed as a separate category</li>
<li>number of records: 10192</li>
<li>number of studies: 5</li>
<li>type: categorical</li>
<li>allowable values:
<ul>
<li>
<em>entire</em>: Leaf margin without lobes or toothed</li>
<li>
<em>lobed</em>: Leaf margin that is either lobed or toothed</li>
</ul>
</li>
</ul>
<p><strong>leaf_mass_fraction</strong></p>
<ul>
<li>label: Leaf mass fraction</li>
<li>description: Ratio of leaf dry mass to total plant dry mass</li>
<li>number of records: 162</li>
<li>number of studies: 2</li>
<li>type: numeric</li>
<li>units: g/g</li>
<li>allowable range: 0.1 - 1 g/g</li>
</ul>
<p><strong>leaf_mass_to_stem_mass</strong></p>
<ul>
<li>label: Ratio of leaf dry mass to stem dry mass</li>
<li>description: Ratio of leaf dry mass to stem dry mass</li>
<li>number of records: 144</li>
<li>number of studies: 1</li>
<li>type: numeric</li>
<li>units: g/g</li>
<li>allowable range: 0.1 - 10 g/g</li>
</ul>
<p><strong>leaf_mesophyll_conductance_per_area</strong></p>
<ul>
<li>label: Mesophyll conductance per unit leaf area</li>
<li>description: Rate of CO2 movement between chloroplasts and sub-stomatal cavities (intracellular space), per unit leaf area</li>
<li>number of records: 0</li>
<li>number of studies: 0</li>
<li>type: numeric</li>
<li>units: mmolCO2/g/s</li>
<li>allowable range: 0.01 - 1 mmolCO2/g/s</li>
</ul>
<p><strong>leaf_mesophyll_conductance_per_mass</strong></p>
<ul>
<li>label: Mesophyll conductance per unit leaf mass</li>
<li>description: Rate of CO2 movement between chloroplasts and sub-stomatal cavities (intracellular space), per unit leaf mass</li>
<li>number of records: 0</li>
<li>number of studies: 0</li>
<li>type: numeric</li>
<li>units: molCO2/m2/s</li>
<li>allowable range: 0.01 - 10 molCO2/m2/s</li>
</ul>
<p><strong>leaf_Mg_per_dry_mass</strong></p>
<ul>
<li>label: Leaf magnesium (Mg) content per unit leaf dry mass</li>
<li>description: Leaf magnesium (Mg) content per unit leaf dry mass</li>
<li>number of records: 786</li>
<li>number of studies: 7</li>
<li>type: numeric</li>
<li>units: mg/g</li>
<li>allowable range: 0.01 - 100 mg/g</li>
</ul>
<p><strong>leaf_Mn_per_dry_mass</strong></p>
<ul>
<li>label: Leaf manganese (Mn) content per unit leaf dry mass</li>
<li>description: Leaf manganese (Mn) content per unit leaf dry mass</li>
<li>number of records: 768</li>
<li>number of studies: 6</li>
<li>type: numeric</li>
<li>units: mg/kg</li>
<li>allowable range: 1 - 10000 mg/kg</li>
</ul>
<p><strong>leaf_Mo_per_dry_mass</strong></p>
<ul>
<li>label: Leaf molybdenum (Mo) content per unit leaf dry mass</li>
<li>description: Leaf molybdenum (Mo) content per unit leaf dry mass</li>
<li>number of records: 302</li>
<li>number of studies: 2</li>
<li>type: numeric</li>
<li>units: mg/kg</li>
<li>allowable range: 0 - 100 mg/kg</li>
</ul>
<p><strong>leaf_N_per_area</strong></p>
<ul>
<li>label: Leaf nitrogen (N) content per unit leaf area</li>
<li>description: Leaf nitrogen (N) content per unit leaf area</li>
<li>number of records: 1062</li>
<li>number of studies: 10</li>
<li>type: numeric</li>
<li>units: g/m2</li>
<li>allowable range: 0.2 - 20 g/m2</li>
</ul>
<p><strong>leaf_N_per_dry_mass</strong></p>
<ul>
<li>label: Leaf nitrogen (N) content per unit leaf dry mass</li>
<li>description: Leaf nitrogen (N) content per unit leaf dry mass</li>
<li>number of records: 4294</li>
<li>number of studies: 33</li>
<li>type: numeric</li>
<li>units: mg/g</li>
<li>allowable range: 0.1 - 100 mg/g</li>
</ul>
<p><strong>leaf_Na_per_dry_mass</strong></p>
<ul>
<li>label: Leaf sodium (Na) content per unit leaf dry mass</li>
<li>description: Leaf sodium (Na) content per unit leaf dry mass</li>
<li>number of records: 593</li>
<li>number of studies: 6</li>
<li>type: numeric</li>
<li>units: mg/kg</li>
<li>allowable range: 1 - 20000 mg/kg</li>
</ul>
<p><strong>leaf_Ni_per_dry_mass</strong></p>
<ul>
<li>label: Leaf nickel (Ni) content per unit leaf dry mass</li>
<li>description: Leaf nickel (Ni) content per unit leaf dry mass</li>
<li>number of records: 0</li>
<li>number of studies: 0</li>
<li>type: numeric</li>
<li>units: mg/kg</li>
<li>allowable range: 10 - 1e+05 mg/kg</li>
</ul>
<p><strong>leaf_P_per_area</strong></p>
<ul>
<li>label: Leaf phosphorus (P) content per unit leaf area</li>
<li>description: Leaf phosphorus (P) content per unit leaf area</li>
<li>number of records: 595</li>
<li>number of studies: 7</li>
<li>type: numeric</li>
<li>units: g/m2</li>
<li>allowable range: 0.005 - 1 g/m2</li>
</ul>
<p><strong>leaf_P_per_dry_mass</strong></p>
<ul>
<li>label: Leaf phosphorus (P) content per unit leaf dry mass</li>
<li>description: Leaf phosphorus (P) content per unit leaf dry mass</li>
<li>number of records: 2145</li>
<li>number of studies: 21</li>
<li>type: numeric</li>
<li>units: mg/g</li>
<li>allowable range: 0.03 - 12 mg/g</li>
</ul>
<p><strong>leaf_phenology</strong></p>
<ul>
<li>label: Leaf phenology type (evergreen, deciduous, semi-deciduous)</li>
<li>description: Variable indicating whether a plant has deciduous versus evergreen leaves; different types of deciduousness included as trait values</li>
<li>number of records: 8677</li>
<li>number of studies: 19</li>
<li>type: categorical</li>
<li>allowable values:
<ul>
<li>
<em>deciduous</em>: Plant where all leaves are shed yearly, either due to drought or cold</li>
<li>
<em>evergreen</em>: Plant which retains its leaves year-round</li>
<li>
<em>facultative_drought_deciduous</em>: Plant that sometimes sheds its years in response to drought</li>
<li>
<em>semi_deciduous</em>: Plant that sheds its years each year for just a very brief period of time</li>
</ul>
</li>
</ul>
<p><strong>leaf_photosynthetic_nitrogen_use_efficiency</strong></p>
<ul>
<li>label: Leaf photosynthesis rate per unit leaf nitrogen (N) content (photosynthetic nitrogen use efficiency, PNUE)</li>
<li>description: Ratio of photosynthesis (CO2 assimilation rate) to leaf nitrogen content</li>
<li>number of records: 162</li>
<li>number of studies: 2</li>
<li>type: numeric</li>
<li>units: umolCO2/molN/s</li>
<li>allowable range: 0.1 - 800 umolCO2/molN/s</li>
</ul>
<p><strong>leaf_photosynthetic_nitrogen_use_efficiency_maximum</strong></p>
<ul>
<li>label: Leaf photosynthesis rate per unit leaf nitrogen (N) content (photosynthetic nitrogen use efficiency, PNUE) at saturating light and CO2 conditions</li>
<li>description: Ratio of photosynthesis (CO2 assimilation rate) to leaf nitrogen content at saturating light and CO2 conditions</li>
<li>number of records: 104</li>
<li>number of studies: 2</li>
<li>type: numeric</li>
<li>units: umolCO2/molN/s</li>
<li>allowable range: 0.1 - 800 umolCO2/molN/s</li>
</ul>
<p><strong>leaf_photosynthetic_nitrogen_use_efficiency_saturated</strong></p>
<ul>
<li>label: Leaf photosynthesis rate per unit leaf nitrogen (N) content (photosynthetic nitrogen use efficiency, PNUE) at saturating light conditions but ambient CO2 conditions</li>
<li>description: Ratio of photosynthesis (CO2 assimilation rate) to leaf nitrogen content at saturating light conditions but ambient CO2 conditions</li>
<li>number of records: 128</li>
<li>number of studies: 2</li>
<li>type: numeric</li>
<li>units: umolCO2/molN/s</li>
<li>allowable range: 0.1 - 800 umolCO2/molN/s</li>
</ul>
<p><strong>leaf_photosynthetic_phosphorus_use_efficiency</strong></p>
<ul>
<li>label: Photosynthetic phosphorus use efficiency</li>
<li>description: Ratio of photosynthetic rate to leaf phosphorus concentration</li>
<li>number of records: 0</li>
<li>number of studies: 0</li>
<li>type: numeric</li>
<li>units: umolCO2/molP/s</li>
<li>allowable range: 500 - 50000 umolCO2/molP/s</li>
</ul>
<p><strong>leaf_photosynthetic_phosphorus_use_efficiency_maximum</strong></p>
<ul>
<li>label: Leaf photosynthesis rate per unit leaf phosphorus (P) content (photosynthetic phosphorus use efficiency, PPUE) at saturating light and CO2 conditions</li>
<li>description: Ratio of photosynthesis (CO2 assimilation rate) to leaf phosphorus content at saturating light and CO2 conditions</li>
<li>number of records: 73</li>
<li>number of studies: 1</li>
<li>type: numeric</li>
<li>units: umolCO2/molP/s</li>
<li>allowable range: 500 - 50000 umolCO2/molP/s</li>
</ul>
<p><strong>leaf_photosynthetic_phosphorus_use_efficiency_saturated</strong></p>
<ul>
<li>label: Leaf photosynthesis rate per unit leaf phosphorus (P) content (photosynthetic phosphorus use efficiency, PPUE) at saturating light conditions but ambient CO2 conditions</li>
<li>description: Ratio of photosynthesis (CO2 assimilation rate) to leaf phosphorus content at saturating light conditions but ambient CO2 conditions</li>
<li>number of records: 73</li>
<li>number of studies: 1</li>
<li>type: numeric</li>
<li>units: umolCO2/molP/s</li>
<li>allowable range: 500 - 50000 umolCO2/molP/s</li>
</ul>
<p><strong>leaf_photosynthetic_water_use_efficiency</strong></p>
<ul>
<li>label: Leaf photosynthesis rate per unit leaf transpiration (water use efficiency, WUE)</li>
<li>description: WUE; Ratio of photosynthesis (CO2 assimilation rate) to leaf transpiration (water loss)</li>
<li>number of records: 91</li>
<li>number of studies: 2</li>
<li>type: numeric</li>
<li>units: umolCO2/mmolH2O</li>
<li>allowable range: 0.01 - 1 umolCO2/mmolH2O</li>
</ul>
<p><strong>leaf_S_per_dry_mass</strong></p>
<ul>
<li>label: Leaf sulphur (S) content per unit leaf dry mass</li>
<li>description: Leaf sulphur (S) content per unit leaf dry mass</li>
<li>number of records: 768</li>
<li>number of studies: 6</li>
<li>type: numeric</li>
<li>units: mg/kg</li>
<li>allowable range: 100 - 50000 mg/kg</li>
</ul>
<p><strong>leaf_saturated_water_content_per_mass</strong></p>
<ul>
<li>label: Water content per unit mass of saturated leaf</li>
<li>description: Ratio of water in a saturated leaf (maximal water holding capacity at full turgidity) to leaf dry mass</li>
<li>number of records: 389</li>
<li>number of studies: 4</li>
<li>type: numeric</li>
<li>units: g/g</li>
<li>allowable range: 0.1 - 10 g/g</li>
</ul>
<p><strong>stem_saturated_water_content_per_mass</strong></p>
<ul>
<li>label: Water content per unit mass of saturated stem</li>
<li>description: Ratio of water in a saturated stem (maximal water holding capacity at full turgidity) to stem dry mass</li>
<li>number of records: 64</li>
<li>number of studies: 1</li>
<li>type: numeric</li>
<li>units: g/g</li>
<li>allowable range: 0.1 - 10 g/g</li>
</ul>
<p><strong>bark_saturated_water_content_per_mass</strong></p>
<ul>
<li>label: Water content per unit mass of saturated bark</li>
<li>description: Ratio of water in a saturated bark (maximal water holding capacity at full turgidity) to bark dry mass</li>
<li>number of records: 64</li>
<li>number of studies: 1</li>
<li>type: numeric</li>
<li>units: g/g</li>
<li>allowable range: 0.1 - 10 g/g</li>
</ul>
<p><strong>leaf_shape</strong></p>
<ul>
<li>label: Leaf shape</li>
<li>description: Leaf shape</li>
<li>number of records: 3354</li>
<li>number of studies: 14</li>
<li>type: categorical</li>
<li>allowable values:
<ul>
<li>
<em>acicular</em>: Needle-shaped</li>
<li>
<em>article</em>: Leaf comprised of jointed segments, as in family Casuarinaceae</li>
<li>
<em>branchlets_articles</em>: Branches with article leaves, as in family Casuarinaceae</li>
<li>
<em>broadly_lanceolate</em>: Like lanceolate, but broader</li>
<li>
<em>broadly_obovate</em>: Like obovate, but broader</li>
<li>
<em>broadly_ovate</em>: Like ovate, but broader</li>
<li>
<em>circular</em>: Circular</li>
<li>
<em>cladode</em>: A flattened stem or internode that resembles and functions as a leaf</li>
<li>
<em>clavate</em>: Slender, broadening towards the tip; club-shaped, with the thickened part at the tip</li>
<li>
<em>cordate</em>: Heart-shaped</li>
<li>
<em>cuneate</em>: Triangular, wedge-shaped, stem attaches to point</li>
<li>
<em>deltoid</em>: Triangular, stem attaches to side</li>
<li>
<em>dimidate</em>: A leaf in which only one side is developed</li>
<li>
<em>elliptical</em>: Oval shaped, without a distinct point</li>
<li>
<em>elliptical-oblong</em>: Oval shaped, without a distinct point, but more elongated</li>
<li>
<em>falcate</em>: Sickle-shaped</li>
<li>
<em>filiform</em>: Thread-like or filament-shaped</li>
<li>
<em>fusiform</em>: Spindle-shaped and tapering at both ends</li>
<li>
<em>half_terete</em>: Semi-circular in cross-section</li>
<li>
<em>hastate</em>: Spear-shaped - pointed, with barbs, shaped like a spear point, with flaring pointed lobes at the base</li>
<li>
<em>lanceolate</em>: Long, wider in the middle, shaped like a lance tip</li>
<li>
<em>leafless</em>: No leaf present</li>
<li>
<em>linear</em>: Long and very narrow like a blade of grass</li>
<li>
<em>linear_elliptical</em>: A particularly long, narrow elliptical leaf</li>
<li>
<em>linear_oblanceolate</em>: A particularly long, narrow oblanceolate leaf</li>
<li>
<em>linear_obovate</em>: A particularly long, narrow obovate leaf</li>
<li>
<em>lobed</em>: Being divided by clefts, may be pinnately lobed or palmately lobed</li>
<li>
<em>lorate</em>: Having the form of a strap</li>
<li>
<em>narrowly_cuneate</em>: like cuneate, but narrower</li>
<li>
<em>narrowly_elliptical</em>: like elliptical, but narrower</li>
<li>
<em>narrowly_oblong</em>: like oblong, but narrower</li>
<li>
<em>narrowly_obovate</em>: like obovate, but narrower</li>
<li>
<em>narrowly_obtrullate</em>: like obtrullate, but narrower</li>
<li>
<em>narrowly_ovate</em>: like ovate, but narrower</li>
<li>
<em>narrowly_subulate</em>: like subulate, but narrower</li>
<li>
<em>narrowly_rhomboidal</em>: like rhomboidal, but narrower</li>
<li>
<em>narrowly_triangular</em>: like triangular, but narrower</li>
<li>
<em>needle</em>: Needle-shaped</li>
<li>
<em>obcordate</em>: Heart-shaped, stem attaches at the tapering end</li>
<li>
<em>oblanceolate</em>: Much longer than wide and with the widest portion near the tip, reversed lanceolate</li>
<li>
<em>oblate</em>: An elongate leaf that is wider in the middle than at the two ends</li>
<li>
<em>oblong</em>: Having an elongated form with slightly parallel sides, roughly rectangular</li>
<li>
<em>obtriangular</em>: A triangular leaf where the stem attaches to the tapering end</li>
<li>
<em>obtrullate</em>: In the form of a reversed trowel (with longer sides meeting at the base)</li>
<li>
<em>obovate</em>: Teardrop-shaped, stem attaches to the tapering end; reversed ovate</li>
<li>
<em>obovoid</em>: A 3-dimensional shape; teardrop-shaped, stem attaches to the tapering end</li>
<li>
<em>orbicular</em>: Circular</li>
<li>
<em>ovate</em>: Oval, egg-shaped, with a tapering point and the widest portion near the petiole</li>
<li>
<em>palm</em>: UNCERTAIN if this refers to a leaf from a palm (Araceeae) or a palmately lobed leaf</li>
<li>
<em>palmately_lobed</em>: Palm-shaped, i.e., with lobes or leaflets stemming from the leaf base</li>
<li>
<em>peltate</em>: A round leaf where the petiole attaches near the center</li>
<li>
<em>phyllode</em>: Stem segment that functions like a leaf</li>
<li>
<em>reniform</em>: Shaped like a kidney - an oval with an inward curve on one side</li>
<li>
<em>rhomboidal</em>: Diamond-shaped</li>
<li>
<em>scale</em>: Scale leaf</li>
<li>
<em>spathulate</em>: Spoon-shaped; having a broad flat end which tapers to the base</li>
<li>
<em>spine</em>: Leaf reduced to a spine</li>
<li>
<em>straight</em>: Leaf with very straight sides</li>
<li>
<em>subulate</em>: Slender and tapering to a point; awl-shaped.</li>
<li>
<em>terete</em>: Circular in cross-section; more or less cylindrical without grooves or ridges</li>
<li>
<em>terete_compressed</em>: Like terete, but compressed</li>
<li>
<em>triangular</em>: Triangular leaf where stem attaches to a side; similar to deltoid</li>
<li>
<em>trigonous_terete</em>: Terete, but triangular in cross section</li>
<li>
<em>widely_obovate</em>: like obovate, but wider</li>
<li>
<em>widely_obtrullate</em>: In the form of a reversed trowel (with longer sides meeting at the base), but wider</li>
</ul>
</li>
</ul>
<p><strong>leaf_specific_conductivity</strong></p>
<ul>
<li>label: Leaf specific hydraulic conductivity (Kl)</li>
<li>description: Kl; the ratio of leaf hydraulic conductivity to the leaf area distil to the segment</li>
<li>number of records: 68</li>
<li>number of studies: 2</li>
<li>type: numeric</li>
<li>units: 10^4 x kg/m/s/MPa</li>
<li>allowable range: 0.1 - 100 10^4 x kg/m/s/MPa</li>
</ul>
<p><strong>leaf_spines_adult</strong></p>
<ul>
<li>label: Presence of spines on adult leaves</li>
<li>description: Binary variable describing whether or not adult leaves have spines</li>
<li>number of records: 86</li>
<li>number of studies: 1</li>
<li>type: categorical</li>
<li>allowable values:
<ul>
<li>
<em>0</em>: leaves on adult plants do not have spines</li>
<li>
<em>1</em>: leaves on adult plants do have spines</li>
</ul>
</li>
</ul>
<p><strong>leaf_thickness</strong></p>
<ul>
<li>label: Leaf thickness</li>
<li>description: Thickness of the leaf lamina</li>
<li>number of records: 1169</li>
<li>number of studies: 14</li>
<li>type: numeric</li>
<li>units: mm</li>
<li>allowable range: 0.01 - 51 mm</li>
</ul>
<p><strong>leaf_toughness</strong></p>
<ul>
<li>label: Leaf toughness</li>
<li>description: The work (energy) required to fracture a leaf by tearing, shearing or punching</li>
<li>number of records: 156</li>
<li>number of studies: 3</li>
<li>type: numeric</li>
<li>units: N/m</li>
<li>allowable range: 100 - 10000 N/m</li>
</ul>
<p><strong>leaf_transpiration</strong></p>
<ul>
<li>label: Leaf transpiration under ambient conditions</li>
<li>description: Rate of water loss from leaf under ambient conditions</li>
<li>number of records: 0</li>
<li>number of studies: 0</li>
<li>type: numeric</li>
<li>units: mmol/m2/s</li>
<li>allowable range: 1e-04 - 10000 mmol/m2/s</li>
</ul>
<p><strong>leaf_transpiration_at_Amax</strong></p>
<ul>
<li>label: Leaf transpiration during Amax measurement</li>
<li>description: Rate of water loss from leaf during Amax measurement</li>
<li>number of records: 5</li>
<li>number of studies: 1</li>
<li>type: numeric</li>
<li>units: mmol/m2/s</li>
<li>allowable range: 1e-04 - 10000 mmol/m2/s</li>
</ul>
<p><strong>leaf_transpiration_at_Asat</strong></p>
<ul>
<li>label: Leaf transpiration during Asat measurement</li>
<li>description: Rate of water loss from leaf during Asat measurement</li>
<li>number of records: 190</li>
<li>number of studies: 2</li>
<li>type: numeric</li>
<li>units: mmol/m2/s</li>
<li>allowable range: 1e-04 - 10000 mmol/m2/s</li>
</ul>
<p><strong>leaf_turgor_loss_point</strong></p>
<ul>
<li>label: Leaf turgor loss point</li>
<li>description: Water potential at which a leaf loses turgor</li>
<li>number of records: 166</li>
<li>number of studies: 3</li>
<li>type: numeric</li>
<li>units: neg_MPa</li>
<li>allowable range: 0.01 - 100 neg_MPa</li>
</ul>
<p><strong>leaf_type</strong></p>
<ul>
<li>label: Leaf type</li>
<li>description: Broad definitions of leaf type</li>
<li>number of records: 601</li>
<li>number of studies: 8</li>
<li>type: categorical</li>
<li>allowable values:
<ul>
<li>
<em>article</em>: Leaf comprised of jointed segments</li>
<li>
<em>broadleaf</em>: Flat leaf lamina</li>
<li>
<em>leafless</em>: Lacks leaves</li>
<li>
<em>needle</em>: Needle or awl-shaped leaf</li>
<li>
<em>phyllode</em>: Winged leaf stalk which functions as a leaf</li>
<li>
<em>scale_leaf</em>: A small flat leaf resembling a scale</li>
</ul>
</li>
</ul>
<p><strong>leaf_water_content_per_area</strong></p>
<ul>
<li>label: Leaf water content per unit area</li>
<li>description: The ratio of the mass of water in a leaf relative to its surface area</li>
<li>number of records: 47</li>
<li>number of studies: 2</li>
<li>type: numeric</li>
<li>units: g/m2</li>
<li>allowable range: 10 - 1000 g/m2</li>
</ul>
<p><strong>leaf_water_content_per_mass</strong></p>
<ul>
<li>label: Leaf water content per unit mass</li>
<li>description: The ratio of the mass of water in a leaf relative to its dry mass. (See also leaf_dry_matter_content, the ratio of a leaf’s dry mass to fresh mass, that is recorded by a greater number of studies.)</li>
<li>number of records: 356</li>
<li>number of studies: 3</li>
<li>type: numeric</li>
<li>units: g/g</li>
<li>allowable range: 0.01 - 10 g/g</li>
</ul>
<p><strong>leaf_width</strong></p>
<ul>
<li>label: Leaf width</li>
<li>description: The longest width axis of a leaf; orthogonal to its length</li>
<li>number of records: 36659</li>
<li>number of studies: 35</li>
<li>type: numeric</li>
<li>units: mm</li>
<li>allowable range: 0.01 - 1e+05 mm</li>
</ul>
<p><strong>leaf_Zn_per_dry_mass</strong></p>
<ul>
<li>label: Leaf zinc (Zn) content per unit leaf dry mass</li>
<li>description: Leaf zinc (Zn) content per unit leaf dry mass</li>
<li>number of records: 768</li>
<li>number of studies: 6</li>
<li>type: numeric</li>
<li>units: mg/kg</li>
<li>allowable range: 0.1 - 1000 mg/kg</li>
</ul>
<p><strong>life_form</strong></p>
<ul>
<li>label: Life form</li>
<li>description: Raunkiaer classification; Categorical classification of plants according to shoot-apex or bud protection</li>
<li>number of records: 431</li>
<li>number of studies: 3</li>
<li>type: categorical</li>
<li>allowable values:
<ul>
<li>
<em>aerophytes</em>: Plants that obtain moisture and nutrients from the air and rain</li>
<li>
<em>chamaephyte</em>: Plants with buds on persistent shoots near the soil surface</li>
<li>
<em>cryptophytes</em>: plants have subterranean or under water resting buds</li>
<li>
<em>epiphyte</em>: Plant that grows on the surface of another plant</li>
<li>
<em>geophyte</em>: Plants with bulbs that rest in dry soil as a rhizome, bulb or corm</li>
<li>
<em>helophytes</em>: Plants with bulbs that rest in marshy or wet soil</li>
<li>
<em>hemicryptophyte</em>: Plants with buds at or near the soil surface</li>
<li>
<em>hydrophytes</em>: Plants with bulbs that rest submerged under water</li>
<li>
<em>phanerophyte</em>: Plants with dominant buds on branches which project freely into the air; normally woody perennials</li>
<li>
<em>therophyte</em>: Annual plants that complete their lives rapidly in favourable conditions and survive the unfavourable cold or dry season in the form of seeds</li>
</ul>
</li>
</ul>
<p><strong>life_history</strong></p>
<ul>
<li>label: Life history</li>
<li>description: Categorical description of plant’s life history</li>
<li>number of records: 23646</li>
<li>number of studies: 38</li>
<li>type: categorical</li>
<li>allowable values:
<ul>
<li>
<em>annual</em>: Plant that lives for only a single year</li>
<li>
<em>biennial</em>: Plant with a two-year lifespan</li>
<li>
<em>ephemeral</em>: Very short-lived plant; the lifespan is usually only a few months</li>
<li>
<em>indefinite</em>: Very long-lived plant, whose exact lifespan is unknown. Category also includes plants that reproduce vegetatively, such that their ‘genetic’ lifespan is indefinite</li>
<li>
<em>long_lived</em>: Very long-lived plant, whose exact lifespan is unknown</li>
<li>
<em>perennial</em>: Plant that lives for 3 or more growing seasons</li>
<li>
<em>short_lived_perennial</em>: A short-lived perennial</li>
<li>
<em>short_lived</em>: Unclear term that only suggests a plant is not long-lived - could be ephemeral, annual or biennial</li>
</ul>
</li>
</ul>
<p><strong>lifespan</strong></p>
<ul>
<li>label: Life span (years)</li>
<li>description: Broad categories of plant life span, in years</li>
<li>number of records: 230</li>
<li>number of studies: 1</li>
<li>type: categorical</li>
<li>allowable values:
<ul>
<li>
<em>&lt;1</em>: Plant that lives less than 1 year</li>
<li>
<em>&lt;5</em>: Plant that lives less than 5 years</li>
<li>
<em>&lt;10</em>: Plant that lives less than 10 years</li>
<li>
<em>&lt;20</em>: Plant that lives less than 20 years</li>
<li>
<em>&lt;50</em>: Plant that lives less than 50 years</li>
<li>
<em>&lt;100</em>: Plant that lives less than 100 years</li>
<li>
<em>&gt;100</em>: Plant that lives more than 100 years</li>
<li>
<em>?</em>: Plant with unknown lifespan</li>
</ul>
</li>
</ul>
<p><strong>mating_system</strong></p>
<ul>
<li>label: Plant mating system</li>
<li>description: Plant mating system</li>
<li>number of records: 0</li>
<li>number of studies: 0</li>
<li>type: categorical</li>
<li>allowable values:
<ul>
<li>
<em>andromonoecious</em>: Plants with hermaphrodite flowers and staminate flowers (producing just male gametes) on the same plant but no pistillate (female) flowers.</li>
<li>
<em>dioecious</em>: Plants that produce male and female flowers on separate plants</li>
<li>
<em>gynomonoecious</em>: Plants with hermaphrodite flowers and pistillate flowers (producing just female gametes) on the same plant but no staminate (male) flowers.</li>
<li>
<em>hermaphrodite</em>: Flowers can produce both male and female gametes</li>
<li>
<em>monoecious</em>: Plants that have both male and female flowers on the same plant, but an individual flower produces only male or female gametes.</li>
</ul>
</li>
</ul>
<p><strong>modified_NDVI</strong></p>
<ul>
<li>label: modified normalized difference vegetation index (modified NDVI)</li>
<li>description: modified normalized difference vegetation index (modified NDVI), based on Landsat data</li>
<li>number of records: 453</li>
<li>number of studies: 1</li>
<li>type: numeric</li>
<li>units: dimensionless</li>
<li>allowable range: 0.1 - 1 dimensionless</li>
</ul>
<p><strong>modulus_of_elasticity</strong></p>
<ul>
<li>label: Bulk modulus of elasticity (Young’s modulus)</li>
<li>description: A measure of xylem’s resistance to being deformed elastically (i.e., non-permanently) when a stress is applied to it</li>
<li>number of records: 191</li>
<li>number of studies: 2</li>
<li>type: numeric</li>
<li>units: Pa</li>
<li>allowable range: 1 - 100000 Pa</li>
</ul>
<p><strong>modulus_of_rupture</strong></p>
<ul>
<li>label: Bulk modulus of rupture</li>
<li>description: A measure of the force required to rupture xylem vessel</li>
<li>number of records: 125</li>
<li>number of studies: 1</li>
<li>type: numeric</li>
<li>units: Pa</li>
<li>allowable range: 1 - 100000 Pa</li>
</ul>
<p><strong>nitrogen_fixing</strong></p>
<ul>
<li>label: Plant nitrogen fixation capacity</li>
<li>description: Binary variable describing whether or not a plant hosts a nitrogen-fixing bacteria</li>
<li>number of records: 2164</li>
<li>number of studies: 16</li>
<li>type: categorical</li>
<li>allowable values:
<ul>
<li>
<em>0</em>: plant does not exhibit nitrogen-fixation</li>
<li>
<em>1</em>: plant does exhibit nitrogen-fixation</li>
</ul>
</li>
</ul>
<p><strong>N_to_P_ratio</strong></p>
<ul>
<li>label: Ratio of N to P dry mass</li>
<li>description: Ratio of N to P dry mass</li>
<li>number of records: 79</li>
<li>number of studies: 1</li>
<li>type: numeric</li>
<li>units: dimensionless</li>
<li>allowable range: 1 - 100 dimensionless</li>
</ul>
<p><strong>osmotic_potential</strong></p>
<ul>
<li>label: Osmotic potential</li>
<li>description: Potential for water to move across a semi-permeable membrane based on solute concentration</li>
<li>number of records: 80</li>
<li>number of studies: 1</li>
<li>type: numeric</li>
<li>units: MPa</li>
<li>allowable range: -10 - 0 MPa</li>
</ul>
<p><strong>palisade_cell_length</strong></p>
<ul>
<li>label: Length of individual palisade cells</li>
<li>description: Length of individual palisade cells</li>
<li>number of records: 0</li>
<li>number of studies: 0</li>
<li>type: numeric</li>
<li>units: um</li>
<li>allowable range: 10 - 1000 um</li>
</ul>
<p><strong>lower_palisade_cell_thickness</strong></p>
<ul>
<li>label: Lower palisade cell thickness</li>
<li>description: Thickness (length) of lower palisade cells</li>
<li>number of records: 62</li>
<li>number of studies: 2</li>
<li>type: numeric</li>
<li>units: um</li>
<li>allowable range: 10 - 1000 um</li>
</ul>
<p><strong>upper_palisade_cell_thickness</strong></p>
<ul>
<li>label: Upper palisade cell thickness</li>
<li>description: Thickness (length) of upper palisade cells</li>
<li>number of records: 95</li>
<li>number of studies: 2</li>
<li>type: numeric</li>
<li>units: um</li>
<li>allowable range: 10 - 1000 um</li>
</ul>
<p><strong>palisade_cell_width</strong></p>
<ul>
<li>label: Width of individual palisade cells</li>
<li>description: Width of individual palisade cells</li>
<li>number of records: 0</li>
<li>number of studies: 0</li>
<li>type: numeric</li>
<li>units: um</li>
<li>allowable range: 1 - 50 um</li>
</ul>
<p><strong>palisade_layer_number</strong></p>
<ul>
<li>label: Number of layers of palisade cells</li>
<li>description: Number of layers of palisade cells</li>
<li>number of records: 0</li>
<li>number of studies: 0</li>
<li>type: numeric</li>
<li>units: count</li>
<li>allowable range: 1 - 10 count</li>
</ul>
<p><strong>photosynthetic_pathway</strong></p>
<ul>
<li>label: Photosynthetic pathway</li>
<li>description: Type of photosynthetic pathway displayed by plants</li>
<li>number of records: 4580</li>
<li>number of studies: 12</li>
<li>type: categorical</li>
<li>allowable values:
<ul>
<li>
<em>c3</em>: Plant using the ‘standard’ photosynthetic pathway, where a 3-carbon compound is produced after the first stage in the photosynthetic pathway</li>
<li>
<em>c4</em>: Plants in which the photosynthetic light-dependent reactions and the Calvin cycle are physically separated to reduce photo respiration</li>
<li>
<em>c3-c4</em>: Species that shift between c3 and c4 photosynthesis</li>
<li>
<em>cam</em>: Plants which display crassulacean acid metabolism</li>
<li>
<em>c3-cam</em>: Species that shift between c3 and cam photosynthesis</li>
<li>
<em>c4-cam</em>: Species that shift between c4 and cam photosynthesis</li>
</ul>
</li>
</ul>
<p><strong>photosynthetic_rate_per_area</strong></p>
<ul>
<li>label: Leaf photosynthesis rate per unit leaf area</li>
<li>description: Rate at which a plant consumes carbon dioxide through photosynthesis, normalised to leaf area</li>
<li>number of records: 18</li>
<li>number of studies: 1</li>
<li>type: numeric</li>
<li>units: umolCO2/m2/s</li>
<li>allowable range: 0.25 - 50 umolCO2/m2/s</li>
</ul>
<p><strong>photosynthetic_rate_per_area_maximum</strong></p>
<ul>
<li>label: Leaf photosynthesis rate per unit leaf area at saturating light and CO2 conditions</li>
<li>description: Rate at which a plant consumes carbon dioxide through photosynthesis at saturating light and CO2 conditions, normalised to leaf area</li>
<li>number of records: 99</li>
<li>number of studies: 1</li>
<li>type: numeric</li>
<li>units: umolCO2/m2/s</li>
<li>allowable range: 0.25 - 85 umolCO2/m2/s</li>
</ul>
<p><strong>photosynthetic_rate_per_area_saturated</strong></p>
<ul>
<li>label: Leaf photosynthesis rate per unit leaf area at saturating light conditions</li>
<li>description: Rate at which a plant consumes carbon dioxide through photosynthesis at saturating light conditions but ambient CO2 conditions, normalised to leaf area</li>
<li>number of records: 1559</li>
<li>number of studies: 14</li>
<li>type: numeric</li>
<li>units: umolCO2/m2/s</li>
<li>allowable range: 0.25 - 85 umolCO2/m2/s</li>
</ul>
<p><strong>photosynthetic_rate_per_dry_mass</strong></p>
<ul>
<li>label: Leaf photosynthesis rate per unit leaf dry mass</li>
<li>description: Maximum rate at which a plant consumes carbon dioxide through photosynthesis, normalised to leaf dry mass</li>
<li>number of records: 0</li>
<li>number of studies: 0</li>
<li>type: numeric</li>
<li>units: umolCO2/g/s</li>
<li>allowable range: 0.01 - 1 umolCO2/g/s</li>
</ul>
<p><strong>photosynthetic_rate_per_dry_mass_maximum</strong></p>
<ul>
<li>label: Leaf photosynthesis rate per unit leaf dry mass at saturating light and CO2 conditions</li>
<li>description: Maximum rate at which a plant consumes carbon dioxide through photosynthesis at saturating light and CO2 conditions, normalised to leaf dry mass</li>
<li>number of records: 99</li>
<li>number of studies: 1</li>
<li>type: numeric</li>
<li>units: umolCO2/g/s</li>
<li>allowable range: 0.01 - 1 umolCO2/g/s</li>
</ul>
<p><strong>photosynthetic_rate_per_dry_mass_saturated</strong></p>
<ul>
<li>label: Leaf photosynthesis rate per unit leaf dry mass at saturating light conditions</li>
<li>description: Maximum rate at which a plant consumes carbon dioxide through photosynthesis at saturating light conditions but ambient CO2 conditions, normalised to leaf dry mass</li>
<li>number of records: 1338</li>
<li>number of studies: 9</li>
<li>type: numeric</li>
<li>units: umolCO2/g/s</li>
<li>allowable range: 0.01 - 1 umolCO2/g/s</li>
</ul>
<p><strong>plant_growth_form</strong></p>
<ul>
<li>label: Plant Growth Form</li>
<li>description: Different growth forms displayed by plants, including both standard plant growth form descriptors (tree, shrub, etc.) and specific plant characteristics (i.e. patristic)</li>
<li>number of records: 35549</li>
<li>number of studies: 55</li>
<li>type: categorical</li>
<li>allowable values:
<ul>
<li>
<em>aquatic</em>: Grows primarily in water</li>
<li>
<em>aquatic_herb</em>: Herb that grows primarily in water</li>
<li>
<em>climber</em>: Plant that climbs up another plant’s trunk, branches, rather than being able to support itself</li>
<li>
<em>climber_tree</em>: Tree that climbs up another plant</li>
<li>
<em>climber_herb</em>: Herbaceous plant that climbs up another plant</li>
<li>
<em>climber_shrub</em>: Shrubby plant that climbs up another plant</li>
<li>
<em>creeper</em>: Creeper that roots along stems</li>
<li>
<em>cycad</em>: Cycad</li>
<li>
<em>epiphyte</em>: Plant that grows on top of another plant</li>
<li>
<em>fern</em>: A group of vascular plants that reproduce via spores</li>
<li>
<em>geophyte</em>: A perennial plant with an underground food storage organ, such as a bulb, tuber, corm, or rhizome</li>
<li>
<em>graminoid</em>: Herbaceous plants with a grass-like morphology</li>
<li>
<em>grass-tree</em>: Term specific refers to members of the family Xanthorrhoaceae, monocots with a tree-like growth-form</li>
<li>
<em>hemi-epiphyte</em>: Plant that partially grows on top of another plant</li>
<li>
<em>herb</em>: A seed-bearing plant which does not have a woody stem and dies down to the ground after flowering</li>
<li>
<em>palm</em>: Flowering plants in the family Arecaceae, most with large, compound evergreen leaves referred to as fronds</li>
<li>
<em>parasite</em>: A plant that derives some or all of its nutritional requirements from another living plant</li>
<li>
<em>prostrate</em>: Plant that lies flat against the ground</li>
<li>
<em>subshrub</em>: A dwarf shrub, especially one that is only woody at the base</li>
<li>
<em>succulent</em>: A plant where some parts are thick and fleshy and usually involved in water storage</li>
<li>
<em>shrub</em>: A woody plant that has multiple stems arising at or near ground level</li>
<li>
<em>treelet</em>: A woody plant that is intermediate between a shrub and a tree</li>
<li>
<em>tree</em>: A tall, woody, perennial plant, usually with a single main trunk</li>
</ul>
</li>
</ul>
<p><strong>plant_height</strong></p>
<ul>
<li>label: Plant height</li>
<li>description: Vegetative plant height</li>
<li>number of records: 50262</li>
<li>number of studies: 53</li>
<li>type: numeric</li>
<li>units: m</li>
<li>allowable range: 0.001 - 130 m</li>
</ul>
<p><strong>plant_width</strong></p>
<ul>
<li>label: Plant width</li>
<li>description: The width of the plant canopy</li>
<li>number of records: 2234</li>
<li>number of studies: 3</li>
<li>type: numeric</li>
<li>units: m</li>
<li>allowable range: 0.001 - 20 m</li>
</ul>
<p><strong>plant_breadth</strong></p>
<ul>
<li>label: Plant breadth</li>
<li>description: The breadth of the plant canopy (smaller of two dimensions)</li>
<li>number of records: 0</li>
<li>number of studies: 0</li>
<li>type: numeric</li>
<li>units: m</li>
<li>allowable range: 0.001 - 20 m</li>
</ul>
<p><strong>pollination_syndrome</strong></p>
<ul>
<li>label: Pollination syndrome</li>
<li>description: Pollination syndrome</li>
<li>number of records: 0</li>
<li>number of studies: 0</li>
<li>type: categorical</li>
<li>allowable values:
<ul>
<li>
<em>animal</em>: Generic cath-all for any form of animal pollination</li>
<li>
<em>wind</em>: wind-pollinated</li>
</ul>
</li>
</ul>
<p><strong>regen_strategy</strong></p>
<ul>
<li>label: Regeneration strategy</li>
<li>description: Different regeneration strategies displayed by plants. Trait values include both generic terms and quite specific ones. See Pausus, Lamont et al. 2018, doi.org/10.1111/nph.14982 for trait values used and detailed desciptions of recolonization ability and level of fire protection provided by each regeneration strategy. This trait includes terminology for storage organs and regeneration strategies following fire. The trait “fire_response” is a binary trait distinguishing between fire-killed and regenerating species.</li>
<li>number of records: 2780</li>
<li>number of studies: 6</li>
<li>type: categorical</li>
<li>allowable values:
<ul>
<li>
<em>aboveground</em>: Plants that resprout from an aboveground organ, but the type of organ isn’t specified.</li>
<li>
<em>aboveground_caudex</em>: Plant regenerates from above-ground apical meristems. Also termed a caudex. Term refers to palm-like plants with aerial buds that resprout, including grass trees, palms, tree ferns, and pandanus.</li>
<li>
<em>basal</em>: Plant regenerates from the base of the stem (versus underground). Generally synonymous with “lignotuber”, especially among Australian plants.</li>
<li>
<em>belowground_caudex</em>: Plant regenerates from belowe-ground caudex.</li>
<li>
<em>bud-bearing_root</em>: Lateral root that produces vertical suckers</li>
<li>
<em>bulb</em>: Below ground storage organ derived from stem (and sometimes leaf) tissue.</li>
<li>
<em>corm</em>: Below ground storage organ derived from stem tissue.</li>
<li>
<em>epicormic</em>: Plants that resprout from epicormic buds; these are buds beneath the bark on stems that survive fire</li>
<li>
<em>fleshy_underground_organ</em>: Plants that use fleshy, generally underground organs (swellings) to regenerate; examples include corms, tubers, bulbs, and caudex</li>
<li>
<em>lignotuber</em>: Swollen woody structures are located at the transition between the stem base and root crown of woody shrubs, mallees and small trees, and are formed from stem tissues. Buds are located over the entire structure. No colonisation ability.</li>
<li>
<em>fire_killed</em>: Plant that is fire-killed and regenerates only by seed. (An obligate re-seeder.)</li>
<li>
<em>resprouts</em>: Plant reprouts following fire, but the mechanism is not specified.</li>
<li>
<em>rhizome</em>: Plant has an underground stem; this stem may be woody or non-woody. Rhizomes are both a mechanism for vegetative spread and a strategy to survive fire.</li>
<li>
<em>root_crown</em>: Plant that produces a few vertical shoots from the root crown.</li>
<li>
<em>root_tuber</em>: Below ground storage organ derived from root tissue.</li>
<li>
<em>seed</em>: Plant that regenerate only or primarily by seed. These are mostly fire-killed species, but the not always</li>
<li>
<em>stem_tuber</em>: Below ground storage organ derived from stem tissue.</li>
<li>
<em>woody_rootstock</em>: Plant that regenerate by resprouting from wood rootstock.</li>
<li>
<em>woody_rootstock_or_lignotuber</em>: Plant that regenerates by resprouting from either a woody rootstock or lignotuber.</li>
<li>
<em>no_storage_organ</em>: No storage organ present</li>
<li>
<em>storage_organ</em>: Storage organ present, but the type is unknown</li>
<li>
<em>underground</em>: Plants that resprout from an underground organ, but the type of organ isn’t specified.</li>
<li>
<em>unknown</em>: Assume this trait value means it is unknown if a storage organ is present or a plant can regenerate, not that the type is unknown</li>
</ul>
</li>
</ul>
<p><strong>resorption_leaf_N</strong></p>
<ul>
<li>label: Nitrogen resorption from leaves</li>
<li>description: Nitrogen resorption from leaves</li>
<li>number of records: 86</li>
<li>number of studies: 1</li>
<li>type: numeric</li>
<li>units: dimensionless</li>
<li>allowable range: -1.1 - 1 dimensionless</li>
</ul>
<p><strong>resorption_leaf_P</strong></p>
<ul>
<li>label: Phosphorus resorption from leaves</li>
<li>description: Phosphorus resorption from leaves</li>
<li>number of records: 90</li>
<li>number of studies: 1</li>
<li>type: numeric</li>
<li>units: dimensionless</li>
<li>allowable range: -1.1 - 1 dimensionless</li>
</ul>
<p><strong>root_diameter</strong></p>
<ul>
<li>label: Root diameter</li>
<li>description: Root diameter</li>
<li>number of records: 0</li>
<li>number of studies: 0</li>
<li>type: numeric</li>
<li>units: mm</li>
<li>allowable range: 0.01 - 1000 mm</li>
</ul>
<p><strong>root_dry_matter_content</strong></p>
<ul>
<li>label: Root dry mass per unit root fresh mass (Root dry matter content)</li>
<li>description: Root dry mass per unit root fresh mass</li>
<li>number of records: 0</li>
<li>number of studies: 0</li>
<li>type: numeric</li>
<li>units: mg/g</li>
<li>allowable range: 10 - 1000 mg/g</li>
</ul>
<p><strong>root_mass_fraction</strong></p>
<ul>
<li>label: Fraction of plant dry mass comprised of root material</li>
<li>description: Fraction of plant dry mass comprised of root material</li>
<li>number of records: 0</li>
<li>number of studies: 0</li>
<li>type: numeric</li>
<li>units: mg/mg</li>
<li>allowable range: 0.01 - 1 mg/mg</li>
</ul>
<p><strong>root_morphology</strong></p>
<ul>
<li>label: Categorical root descriptions, sense Cannon 1949</li>
<li>description: Categorical root descriptions sensu Cannon 1949, A Tentative Classification of Root Systems, Ecology, doi.org/10.2307/1932458</li>
<li>number of records: 30</li>
<li>number of studies: 1</li>
<li>type: categorical</li>
<li>allowable values:
<ul>
<li>
<em>type_i</em>: The primary root is fibrous or fleshy and is relatively long; mesophytic</li>
<li>
<em>type_ii</em>: The primary root is relatively short and is fibrous. Roots are well branched in a dry habitat but branch little in one that is wet.</li>
<li>
<em>type_iii</em>: The primary root is usually long and fibrous. The first order lateral roots are well branched and are usually short; they are rather evenly distributed along the primary root; xerophytic</li>
<li>
<em>type_iv</em>: The primary root is usually long and slender like that of Type II. The first order lateral roots are well branched. Those closest to the surface of the ground, or not far below, are relatively long; xerophytic and mesophytic</li>
<li>
<em>type_v</em>: The primary root is divided-forked or branched-and fibrous, and with few first order lateral roots, or, in age, apparently with none; xerophytic</li>
<li>
<em>type_vi</em>: The primary root is usually as in Type V. The first order lateral roots are mostly on the upper portion of the primary root and are thus relatively superficial; xerophytic or in well aerated sandy soil where the rainfall is not light.</li>
<li>
<em>type_vii</em>: Adventitious roots that are in one group and arise on a short horizontal, or vertical, axis of the shoot. They are fibrous or fleshly; mesophytic and hydrophytic</li>
<li>
<em>type_viii</em>: Adventitious roots that are in one group and are fibrous and fleshy, or fibrous and thick. The fibrous adventitious roots are absorbing-anchoring roots and are usually with laterals to the second order; meophytic</li>
<li>
<em>type_ix</em>: Adventitious roots that arise on the aerial stem and branches of some herbaceous species, shrubs, and trees; on creepers, vines, and runners; and on subterranean rhizomes and stolons</li>
</ul>
</li>
</ul>
<p><strong>root_shoot_ratio</strong></p>
<ul>
<li>label: Root to shoot ratio</li>
<li>description: Ratio of root mass to shoot mass</li>
<li>number of records: 0</li>
<li>number of studies: 0</li>
<li>type: numeric</li>
<li>units: mg/mg</li>
<li>allowable range: 0.01 - 1 mg/mg</li>
</ul>
<p><strong>root_structure</strong></p>
<ul>
<li>label: Root structure</li>
<li>description: Specific specialized types of root structures and root symbioses. <a href="https://www.mycorrhizas.info/" class="uri">https://www.mycorrhizas.info/</a> provides detailed information for types of mycorrhizal associations.</li>
<li>number of records: 1004</li>
<li>number of studies: 6</li>
<li>type: categorical</li>
<li>allowable values:
<ul>
<li>
<em>arbuscular_mycorrhizal</em>: Symbiosis in which the mycorrhizal fungi does penetrate the cortex cells of the plant roots and its hyphae form arbuscules and vesicles inside the plant root; endomycorrhizal, AM, VAM</li>
<li>
<em>carnivorous</em>: non-mycorrhizal carnivorous root</li>
<li>
<em>cluster_roots</em>: cluster roots in non-proteoid species</li>
<li>
<em>coiling_vam</em>: Associations that spread predominantly by intracellular hyphal coils within roots; a sub-category of arbuscular mycorrhizal association</li>
<li>
<em>dauciform_root</em>: A type of cluster root found in the family Cyperaceae</li>
<li>
<em>ectomycorrhizal</em>: Symbiosis in which the mycorrhizal fungi does not penetrate the cortex cells of the plant roots</li>
<li>
<em>ericoid_mycorrhizal</em>: A form of arbuscular mycorrhizal relationship only found on plants in the order Ericales. The fungi’s hyphae penetrate the plant root but do not form arbuscules</li>
<li>
<em>fine_roots</em>: Particularly fine roots for nutrient uptake</li>
<li>
<em>hemiparasitic_root</em>: Partially parasitic root</li>
<li>
<em>long_root_hairs</em>: Specialised long root hairs</li>
<li>
<em>orchid_mycorrhizal</em>: A mycorrhizal relationship specific to orchid species, and for most orchids, essential for seedlings to establish</li>
<li>
<em>non_mycorrhizal</em>: Plants lacking a mycorrhizal symbiont</li>
<li>
<em>parasitic_root</em>: Parasitic root</li>
<li>
<em>proteoid_root</em>: Also known as cluster roots, are plant roots that form clusters of closely spaced short lateral rootlets and aid in nutrient upake in nutrient-poor soils; common in members of the family Proteaceae, but present in other families as well</li>
<li>
<em>root_hairs</em>: Specialised root hairs</li>
<li>
<em>sand-binding</em>: Persistent sheaths of sand that form around the roots of species in the families Restionaceae, Cyperaceae, Haemodoraceae, and Lomandraceae in western Australia. Assumed to function in nutrient uptake.</li>
<li>
<em>subepidermal_mycorrhizal</em>: Mycorrhizae where hyphae grow in a cavity under epidermal cells. Only in the genus Thysanotus ( family Laxmaniaceae).</li>
</ul>
</li>
</ul>
<p><strong>root_wood_density</strong></p>
<ul>
<li>label: Root wood density</li>
<li>description: For root wood, the ratio of mass to volume, yielding density</li>
<li>number of records: 173</li>
<li>number of studies: 2</li>
<li>type: numeric</li>
<li>units: mg/mm3</li>
<li>allowable range: 0.1 - 10 mg/mm3</li>
</ul>
<p><strong>sapwood_specific_conductivity</strong></p>
<ul>
<li>label: Sapwood specific conductivity (Ks)</li>
<li>description: Ks; Describes the flow rate of water (kg/s) along a stem for a given drop in pressure (1/MPa), normalised to the length of the segment (1/m). Calculated as leaf hydraulic conductivity divided by the sapwood cross-sectional area where the measurement is taken.</li>
<li>number of records: 171</li>
<li>number of studies: 3</li>
<li>type: numeric</li>
<li>units: kg/m/s/MPa</li>
<li>allowable range: 0.05 - 50 kg/m/s/MPa</li>
</ul>
<p><strong>seed_breadth</strong></p>
<ul>
<li>label: Seed breadth</li>
<li>description: The shorter width axis of a seed; orthogonal to its length</li>
<li>number of records: 1281</li>
<li>number of studies: 13</li>
<li>type: numeric</li>
<li>units: mm</li>
<li>allowable range: 0.01 - 100 mm</li>
</ul>
<p><strong>seed_count</strong></p>
<ul>
<li>label: Seed count</li>
<li>description: Number of seeds produced by a plant</li>
<li>number of records: 0</li>
<li>number of studies: 0</li>
<li>type: numeric</li>
<li>units: n</li>
<li>allowable range: 0 - 1e+05 n</li>
</ul>
<p><strong>seed_length</strong></p>
<ul>
<li>label: Seed length</li>
<li>description: Longest seed dimension</li>
<li>number of records: 13458</li>
<li>number of studies: 28</li>
<li>type: numeric</li>
<li>units: mm</li>
<li>allowable range: 0.01 - 1000 mm</li>
</ul>
<p><strong>seed_mass</strong></p>
<ul>
<li>label: Seed dry mass</li>
<li>description: Seed carbon dry mass</li>
<li>number of records: 41847</li>
<li>number of studies: 29</li>
<li>type: numeric</li>
<li>units: mg</li>
<li>allowable range: 1e-04 - 1e+05 mg</li>
</ul>
<p><strong>seed_mass_reserve</strong></p>
<ul>
<li>label: Seed dry mass reserve</li>
<li>description: Energy reserves stored in seeds that are mobilized at the time of germination; on a carbon dry mass basis</li>
<li>number of records: 104</li>
<li>number of studies: 2</li>
<li>type: numeric</li>
<li>units: mg</li>
<li>allowable range: 1e-04 - 1000 mg</li>
</ul>
<p><strong>seed_oil_content</strong></p>
<ul>
<li>label: Seed oil content</li>
<li>description: Seed oil content as a fraction of total seed weight, usually on a dry weight basis</li>
<li>number of records: 268</li>
<li>number of studies: 1</li>
<li>type: numeric</li>
<li>units: g/g</li>
<li>allowable range: 0.001 - 1 g/g</li>
</ul>
<p><strong>seed_P_concentration</strong></p>
<ul>
<li>label: Seed phosphorus concentration</li>
<li>description: Seed phosphorus concentration on a weight basis</li>
<li>number of records: 77</li>
<li>number of studies: 2</li>
<li>type: numeric</li>
<li>units: mg/g</li>
<li>allowable range: 1 - 100 mg/g</li>
</ul>
<p><strong>seed_protein_content</strong></p>
<ul>
<li>label: Seed protein content</li>
<li>description: Seed protein content as a fraction of total seed weight</li>
<li>number of records: 146</li>
<li>number of studies: 1</li>
<li>type: numeric</li>
<li>units: g/g</li>
<li>allowable range: 1e-05 - 1 g/g</li>
</ul>
<p><strong>seed_shape</strong></p>
<ul>
<li>label: Seed shape</li>
<li>description: Possible seed shapes. Note that some terms currently used refer to 2-dimensional shapes, not 3-dimensional shapes.</li>
<li>number of records: 2966</li>
<li>number of studies: 8</li>
<li>type: categorical</li>
<li>allowable values:
<ul>
<li>
<em>angular</em>: Doesn’t refer to a specific shape, but simply that instead of being smoothly rounded, it is angular</li>
<li>
<em>broadly_ellipsoid</em>: Like ellipsoid, but broader</li>
<li>
<em>broadly_obovoid</em>: A 3-dimensional shape; like obovoid, but particularly broad in the middle</li>
<li>
<em>circular</em>: Both width dimensions are similar; rounded; REPLACE WITH ‘globoid’ - maybe</li>
<li>
<em>comma_shaped</em>: Seed shaped liked a comma; see also ‘falcate’</li>
<li>
<em>compressed</em>: Seed flatter</li>
<li>
<em>compressed_cylinder</em>: Like cylinder, but a flatter 3-dimensional shape</li>
<li>
<em>compressed_ellipsoid</em>: Like ellipsoid, but a flatter 3-dimensional shape</li>
<li>
<em>compressed_globoid</em>: Like globoid, but a flatter 3-dimensional shape</li>
<li>
<em>compressed_lachrimiform</em>: Like lachrimiform, but a flatter 3-dimensional shape</li>
<li>
<em>compressed_obovoid</em>: Like obovoid, but a flatter 3-dimensional shape</li>
<li>
<em>compressed_ovoid</em>: Like ovoid, but a flatter 3-dimensional shape</li>
<li>
<em>compressed_sectoroid</em>: Like sectoroid, but a flatter 3-dimensional shape</li>
<li>
<em>conical</em>: Shaped like a cone; the opposite of turbinate</li>
<li>
<em>cordiform</em>: Heart-shaped</li>
<li>
<em>cylindrical</em>: A 3-dimensional shape; shaped like a cylinder; see also ‘terete’ which refers to particular thin cylindrical shape; should maybe be cylinder?</li>
<li>
<em>discoid</em>: Resembling a disc, with two convex faces; similar to plano-convex and lenticular, flatter</li>
<li>
<em>discoid_two_lobed</em>: Resembling a disc, but with two lobes</li>
<li>
<em>ellipsoid</em>: A 3-dimensional shape; elliptic in outline and with a length:breadth ratio between 3:2 and 2:1</li>
<li>
<em>elongated_lachrimiform</em>: Like lachrimiform, but extended along the longest axis</li>
<li>
<em>elongated_ovoid</em>: Like ovoid, but extended along the longest axis</li>
<li>
<em>elongated_pyramidal</em>: Like pyramidal, but extended along the longest axis</li>
<li>
<em>elongated_reniform</em>: Like reniform, but extended along the longest axis</li>
<li>
<em>elongated_sectoroid</em>: Like sectoroid, but extended along the longest axis</li>
<li>
<em>falcate</em>: Sickle-shaped; see also ‘comma-shaped’</li>
<li>
<em>flat</em>: Suggests a quite flattened seed, but no information about shape</li>
<li>
<em>fusiform</em>: Spindle-shaped, cigar-shaped, circular in cross-section and tapering at both ends</li>
<li>
<em>globoid</em>: Globe-shaped; includes ‘globular’ and ‘globose’</li>
<li>
<em>inflated_pyramidal</em>: Like pyramidal, but inflated</li>
<li>
<em>inflated_trigonal_pyramidal</em>: Like trigonal_pyramidal, but inflated</li>
<li>
<em>irregular</em>: Irregular shape</li>
<li>
<em>irregular_lachrimiform</em>: Like lachrimiform, but somewhat irregular</li>
<li>
<em>irregular_ovoid</em>: Like ovoid, but somewhat irregular</li>
<li>
<em>lachrimiform</em>: A 3-dimensional shape; tear-shaped</li>
<li>
<em>lanceolate</em>: A 2-dimensional shape; much longer than wide, with the widest point below the middle; in reference to a seed, assume a rounded 2-D shape; REPLACE WITH OVOID</li>
<li>
<em>lenticular</em>: A 3-dimensional shape; shaped like a biconvex lens. Similar to plano-convex??</li>
<li>
<em>muricate</em>: has numerous short hard outgrowths</li>
<li>
<em>obclavate</em>: A 2 or 3-dimensional shape; slender, broadening towards the base; club-shaped, with the thickened part at the base</li>
<li>
<em>obloid</em>: A 3-dimensional shape, where the length is a few times greater than the width, with sides almost parallel and ends rounded.</li>
<li>
<em>obovoid</em>: A 3-dimensional shape; similar to ovoid, but attached at the narrower end and with a length:breadth ratio between 3:2 and 2:1</li>
<li>
<em>orbicular</em>: Fairly flat, but round in 2-dimensional outline; see also compressed_globoid</li>
<li>
<em>ovoid</em>: Egg-shaped (i.e., widest below the middle)</li>
<li>
<em>pointed_ovoid</em>: Like ovoid, but with one pointier end</li>
<li>
<em>pointed_samara</em>: A type of fruit in which a flattened wing of fibrous, papery tissue develops from the ovary wall; variant that is more pointed; not truly a shape</li>
<li>
<em>pyramidal</em>: Pyramid; cross-section unknown</li>
<li>
<em>pyriform</em>: A 3-dimensional shape; resembling a pear, attached at the broader end</li>
<li>
<em>rectangular</em>: REPLACE WITH ‘rectangular_prism’</li>
<li>
<em>rectangular_prism</em>: Prism that is rectangular in cross=section</li>
<li>
<em>reniform</em>: Having a circular or roughly circular shape with a notch, like a kidney</li>
<li>
<em>rhomboid</em>: Diamond shaped in outline with the broadest axis in the middle and with a length:breadth ratio between 3:2 and 2:1</li>
<li>
<em>round</em>: Seed with round outline; seems to be a subcategory of samara</li>
<li>
<em>samara</em>: A type of fruit in which a flattened wing of fibrous, papery tissue develops from the ovary wall; not truly a shape</li>
<li>
<em>sectoroid</em>: Shaped like an orange segment</li>
<li>
<em>semigloboid</em>: Semi-circular in cross-section</li>
<li>
<em>sinusoidal</em>: S-shaped; includes ‘sigmoidal’</li>
<li>
<em>spathulate</em>: Spoon-shaped</li>
<li>
<em>spherical</em>: A 3-dimensional shape; round with all dimensions approximately equal</li>
<li>
<em>spheroid</em>: A 3-dimensional shape; nearly, but imperfectly spherical</li>
<li>
<em>square_prism</em>: Prism that is square in cross=section</li>
<li>
<em>square_pyramidal</em>: Pyramid that is square in cross-section</li>
<li>
<em>subcylindrical</em>: Nearly or almost cylindrical in shape</li>
<li>
<em>subgloboid</em>: Nearly or almost globoid or spheroid in shape</li>
<li>
<em>suborbicular</em>: Nearly or almost orbicular in shape</li>
<li>
<em>subtriangular</em>: Nearly or almost ‘trigonal_prism’ in shape</li>
<li>
<em>terete</em>: Circular in cross-section; more or less cylindrical without grooves or ridges</li>
<li>
<em>triangular</em>: subset of samara possibly; 3-d structure should be trigonal_prism or trigonal_pyramidal</li>
<li>
<em>trigonal_prism</em>: Prism that is triangular in cross-section</li>
<li>
<em>trigonal_pyramidal</em>: Pyramid that is triangular in cross-section</li>
<li>
<em>turbinate</em>: A 3-dimensional shape; top shaped; inversely conical</li>
<li>
<em>wedge_shaped</em>: A 2 or 3-dimensional shape; tapers to a point at one end; includes the 2-D classificatiion “cuneate” and possibly plants that are more conical in outline</li>
</ul>
</li>
</ul>
<p><strong>seed_storage_location</strong></p>
<ul>
<li>label: Location where seeds are stored at maturity</li>
<li>description: Location where seeds are stored at maturity</li>
<li>number of records: 586</li>
<li>number of studies: 1</li>
<li>type: categorical</li>
<li>allowable values:
<ul>
<li>
<em>canopy</em>: Seeds stored in the canopy, usually still within a woody cone</li>
<li>
<em>soil</em>: Seeds stored in the soil</li>
</ul>
</li>
</ul>
<p><strong>seed_texture</strong></p>
<ul>
<li>label: Seed texture</li>
<li>description: The texture of a seed</li>
<li>number of records: 960</li>
<li>number of studies: 1</li>
<li>type: categorical</li>
<li>allowable values:
<ul>
<li>
<em>acutely_reticulate</em>: seed surface with a thickened linear ridge</li>
<li>
<em>bumpy</em>: seed surface deeply wrinkled</li>
<li>
<em>chartaceous</em>: seed surface finely wrinkled</li>
<li>
<em>ciliate</em>: seed surface with a fine-scaled pattern</li>
<li>
<em>costate</em>: seed surface hairy</li>
<li>
<em>deeply_reticulate</em>: seed surface covered with short blunt protruberances</li>
<li>
<em>deeply_rugose</em>: seed surface pitted</li>
<li>
<em>dimpled</em>: seed surface rough</li>
<li>
<em>finely_reticulate</em>: wrinkled</li>
<li>
<em>finely_rugose</em>: rough (synonym seed)</li>
<li>
<em>finely_textured</em>: seed surface with a shallow linear ridge</li>
<li>
<em>fluted</em>: seed surface with shallow pits</li>
<li>
<em>hairs</em>: seed surface resembles a shallow net or network</li>
<li>
<em>lineolate</em>: seed surface with shallow wrinkles</li>
<li>
<em>micronate-tubercular</em>: seed surface slightly wrinkled</li>
<li>
<em>multi-faceted</em>: seed surface smooth</li>
<li>
<em>papillate</em>: seed surface covered with hairs with radiating papillae</li>
<li>
<em>papillose</em>: seed surface marked with more or less parallel longitudinal ridges</li>
<li>
<em>pitted</em>: seed surface covered with densely matted woolly hairs</li>
<li>
<em>plumose</em>: seed surface covered with small rounded warts, protuberances or nodules</li>
<li>
<em>raised_edge</em>: seed surface wavy</li>
<li>
<em>reticulate</em>: seed surface covered with short, dense, silky, upright hairs; soft and velvety</li>
<li>
<em>ribbed</em>: seed surface very rough</li>
<li>
<em>rough</em>: seed surface with a thickened linear ridge</li>
<li>
<em>rugose</em>: seed surface deeply wrinkled</li>
<li>
<em>scabrous</em>: seed surface finely wrinkled</li>
<li>
<em>scalar</em>: seed surface with a fine-scaled pattern</li>
<li>
<em>scaly</em>: seed surface hairy</li>
<li>
<em>shallowly_costate</em>: seed surface covered with short blunt protruberances</li>
<li>
<em>shallowly_pitted</em>: seed surface pitted</li>
<li>
<em>shallowly_reticulate</em>: seed surface rough</li>
<li>
<em>shallowly_rugose</em>: wrinkled</li>
<li>
<em>slightly_rugose</em>: rough (synonym seed)</li>
<li>
<em>smooth</em>: seed surface with a shallow linear ridge</li>
<li>
<em>stellate</em>: seed surface with shallow pits</li>
<li>
<em>striate</em>: seed surface resembles a shallow net or network</li>
<li>
<em>tomentose</em>: seed surface with shallow wrinkles</li>
<li>
<em>tuberculate</em>: seed surface slightly wrinkled</li>
<li>
<em>undulate</em>: seed surface smooth</li>
<li>
<em>velutinous</em>: seed surface covered with hairs with radiating papillae</li>
<li>
<em>very_rough</em>: seed surface marked with more or less parallel longitudinal ridges</li>
</ul>
</li>
</ul>
<p><strong>seed_viability</strong></p>
<ul>
<li>label: Seed viability</li>
<li>description: Proportion of seeds that are viable</li>
<li>number of records: 65</li>
<li>number of studies: 1</li>
<li>type: numeric</li>
<li>units: n/n</li>
<li>allowable range: 0 - 1 n/n</li>
</ul>
<p><strong>seed_volume</strong></p>
<ul>
<li>label: Seed volume</li>
<li>description: Volume of a seed</li>
<li>number of records: 519</li>
<li>number of studies: 1</li>
<li>type: numeric</li>
<li>units: mm3</li>
<li>allowable range: 0.001 - 1e+05 mm3</li>
</ul>
<p><strong>seed_width</strong></p>
<ul>
<li>label: Seed width</li>
<li>description: The longest width dimension of a seed; orthogonal to the length</li>
<li>number of records: 8435</li>
<li>number of studies: 25</li>
<li>type: numeric</li>
<li>units: mm</li>
<li>allowable range: 0.01 - 100 mm</li>
</ul>
<p><strong>seedling_first_leaf</strong></p>
<ul>
<li>label: First leaf style</li>
<li>description: Binary variable distinguishing between seedlings where the first leaf is scale-like (cataphyll) versus leaf-like.</li>
<li>number of records: 940</li>
<li>number of studies: 1</li>
<li>type: categorical</li>
<li>allowable values:
<ul>
<li>
<em>scale-like</em>: When the first leaf at germination is scale-like, also referred to as a cataphyll</li>
<li>
<em>leaf-like</em>: When the first leaf at germination is leaf-like</li>
</ul>
</li>
</ul>
<p><strong>seedling_first_node</strong></p>
<ul>
<li>label: Leaf count at first node</li>
<li>description: Binary variable distinguishing between seedlings where the leaves at the first node are single versus paired.</li>
<li>number of records: 841</li>
<li>number of studies: 1</li>
<li>type: categorical</li>
<li>allowable values:
<ul>
<li>
<em>single</em>: When there is a single leaf at the first node</li>
<li>
<em>paired</em>: When there is a pair of leaves at the first node</li>
</ul>
</li>
</ul>
<p><strong>senesced_leaf_Ca_per_dry_mass</strong></p>
<ul>
<li>label: Senesced leaf calcium (Ca) content per unit leaf dry mass</li>
<li>description: Senesced leaf calcium (Ca) content per unit leaf dry mass</li>
<li>number of records: 254</li>
<li>number of studies: 1</li>
<li>type: numeric</li>
<li>units: mg/g</li>
<li>allowable range: 0.01 - 1000 mg/g</li>
</ul>
<p><strong>senesced_leaf_Cu_per_dry_mass</strong></p>
<ul>
<li>label: Senesced leaf copper (Cu) content per unit leaf dry mass</li>
<li>description: Senesced leaf copper (Cu) content per unit leaf dry mass</li>
<li>number of records: 254</li>
<li>number of studies: 1</li>
<li>type: numeric</li>
<li>units: mg/kg</li>
<li>allowable range: 0.01 - 100 mg/kg</li>
</ul>
<p><strong>senesced_leaf_Fe_per_dry_mass</strong></p>
<ul>
<li>label: Senesced leaf iron (Fe) content per unit leaf dry mass</li>
<li>description: Senesced leaf iron (Fe) content per unit leaf dry mass</li>
<li>number of records: 254</li>
<li>number of studies: 1</li>
<li>type: numeric</li>
<li>units: mg/kg</li>
<li>allowable range: 1 - 10000 mg/kg</li>
</ul>
<p><strong>senesced_leaf_K_per_dry_mass</strong></p>
<ul>
<li>label: Senesced leaf potassium (K) content per unit leaf dry mass</li>
<li>description: Senesced leaf potassium (K) content per unit leaf dry mass</li>
<li>number of records: 340</li>
<li>number of studies: 2</li>
<li>type: numeric</li>
<li>units: mg/g</li>
<li>allowable range: 0.1 - 100 mg/g</li>
</ul>
<p><strong>senesced_leaf_Mg_per_dry_mass</strong></p>
<ul>
<li>label: Senesced leaf magnesium (Mg) content per unit leaf dry mass</li>
<li>description: Senesced leaf magnesium (Mg) content per unit leaf dry mass</li>
<li>number of records: 254</li>
<li>number of studies: 1</li>
<li>type: numeric</li>
<li>units: mg/g</li>
<li>allowable range: 0.01 - 100 mg/g</li>
</ul>
<p><strong>senesced_leaf_Mn_per_dry_mass</strong></p>
<ul>
<li>label: Senesced leaf manganese (Mn) content per unit leaf dry mass</li>
<li>description: Senesced leaf manganese (Mn) content per unit leaf dry mass</li>
<li>number of records: 254</li>
<li>number of studies: 1</li>
<li>type: numeric</li>
<li>units: mg/kg</li>
<li>allowable range: 1 - 10000 mg/kg</li>
</ul>
<p><strong>senesced_leaf_Mo_per_dry_mass</strong></p>
<ul>
<li>label: Senesced leaf molybdenum (Mo) content per unit leaf dry mass</li>
<li>description: Senesced leaf molybdenum (Mo) content per unit leaf dry mass</li>
<li>number of records: 176</li>
<li>number of studies: 1</li>
<li>type: numeric</li>
<li>units: mg/kg</li>
<li>allowable range: 0 - 30 mg/kg</li>
</ul>
<p><strong>senesced_leaf_N_per_dry_mass</strong></p>
<ul>
<li>label: Senesced leaf nitrogen (N) content per unit leaf dry mass</li>
<li>description: Senesced leaf nitrogen (N) content per unit leaf dry mass</li>
<li>number of records: 425</li>
<li>number of studies: 3</li>
<li>type: numeric</li>
<li>units: mg/g</li>
<li>allowable range: 0.1 - 100 mg/g</li>
</ul>
<p><strong>senesced_leaf_Na_per_dry_mass</strong></p>
<ul>
<li>label: Senesced leaf sodium (Na) content per unit leaf dry mass</li>
<li>description: Senesced leaf sodium (Na) content per unit leaf dry mass</li>
<li>number of records: 0</li>
<li>number of studies: 0</li>
<li>type: numeric</li>
<li>units: mg/kg</li>
<li>allowable range: 1 - 20000 mg/kg</li>
</ul>
<p><strong>senesced_leaf_Ni_per_dry_mass</strong></p>
<ul>
<li>label: Senesced leaf nickel (Ni) content per unit leaf dry mass</li>
<li>description: Senesced leaf nickel (Ni) content per unit leaf dry mass</li>
<li>number of records: 0</li>
<li>number of studies: 0</li>
<li>type: numeric</li>
<li>units: mg/kg</li>
<li>allowable range: 10 - 1e+05 mg/kg</li>
</ul>
<p><strong>senesced_leaf_P_per_dry_mass</strong></p>
<ul>
<li>label: Senesced leaf phosphorus (P) content per unit leaf dry mass</li>
<li>description: Senesced leaf phosphorus (P) content per unit leaf dry mass</li>
<li>number of records: 430</li>
<li>number of studies: 3</li>
<li>type: numeric</li>
<li>units: mg/g</li>
<li>allowable range: 0.01 - 12 mg/g</li>
</ul>
<p><strong>senesced_leaf_S_per_dry_mass</strong></p>
<ul>
<li>label: Senesced leaf sulphur (S) content per unit leaf dry mass</li>
<li>description: Senesced leaf sulphur (S) content per unit leaf dry mass</li>
<li>number of records: 254</li>
<li>number of studies: 1</li>
<li>type: numeric</li>
<li>units: mg/kg</li>
<li>allowable range: 100 - 50000 mg/kg</li>
</ul>
<p><strong>senesced_leaf_Zn_per_dry_mass</strong></p>
<ul>
<li>label: Senesced leaf zinc (Zn) content per unit leaf dry mass</li>
<li>description: Senesced leaf zinc (Zn) content per unit leaf dry mass</li>
<li>number of records: 254</li>
<li>number of studies: 1</li>
<li>type: numeric</li>
<li>units: mg/kg</li>
<li>allowable range: 0 - 100 mg/kg</li>
</ul>
<p><strong>serotiny</strong></p>
<ul>
<li>label: Serotiny</li>
<li>description: When a fruit or cone only releases its seeds following an environmental trigger, often fire</li>
<li>number of records: 379</li>
<li>number of studies: 4</li>
<li>type: categorical</li>
<li>allowable values:
<ul>
<li>
<em>not_serotinous</em>: Plant does NOT display serotiny</li>
<li>
<em>serotinous</em>: Plant does display serotiny</li>
<li>
<em>serotiny_low</em>: Plant displays a low level of serotiny</li>
<li>
<em>serotiny_moderate</em>: Plant displays a moderate level of serotiny</li>
</ul>
</li>
</ul>
<p><strong>soil_salinity_tolerance</strong></p>
<ul>
<li>label: Salt tolerance</li>
<li>description: Maximum salinity tolerated by a species, reported as the conductivity of the soil</li>
<li>number of records: 264</li>
<li>number of studies: 1</li>
<li>type: numeric</li>
<li>units: dS/m</li>
<li>allowable range: 0.1 - 100 dS/m</li>
</ul>
<p><strong>soluable_protein_per_area</strong></p>
<ul>
<li>label: Soluble protein per leaf area</li>
<li>description: Mass of soluble protein per leaf area</li>
<li>number of records: 0</li>
<li>number of studies: 0</li>
<li>type: numeric</li>
<li>units: g/m2</li>
<li>allowable range: 0.1 - 20 g/m2</li>
</ul>
<p><strong>insoluable_protein_per_area</strong></p>
<ul>
<li>label: Insoluble protein per leaf area</li>
<li>description: Mass of insoluble protein per leaf area</li>
<li>number of records: 0</li>
<li>number of studies: 0</li>
<li>type: numeric</li>
<li>units: g/m2</li>
<li>allowable range: 0.1 - 20 g/m2</li>
</ul>
<p><strong>soluable_sugars_per_area</strong></p>
<ul>
<li>label: Soluble sugars per leaf area</li>
<li>description: Mass of soluble sugars per leaf area</li>
<li>number of records: 0</li>
<li>number of studies: 0</li>
<li>type: numeric</li>
<li>units: g/m2</li>
<li>allowable range: 1 - 100 g/m2</li>
</ul>
<p><strong>specific_leaf_area</strong></p>
<ul>
<li>label: Leaf area per unit leaf dry mass (specific leaf area, SLA)</li>
<li>description: Leaf area per unit leaf dry mass; SLA</li>
<li>number of records: 23664</li>
<li>number of studies: 69</li>
<li>type: numeric</li>
<li>units: mm2/mg</li>
<li>allowable range: 0.4 - 500 mm2/mg</li>
</ul>
<p><strong>specific_root_area</strong></p>
<ul>
<li>label: Root area per unit root dry mass (specific root area)</li>
<li>description: Root area per unit root dry mass</li>
<li>number of records: 0</li>
<li>number of studies: 0</li>
<li>type: numeric</li>
<li>units: mm2/mg</li>
<li>allowable range: 0.5 - 500 mm2/mg</li>
</ul>
<p><strong>specific_root_length</strong></p>
<ul>
<li>label: Root length per unit root dry mass (specific root length, SRL)</li>
<li>description: Root length per unit root dry mass; SRL</li>
<li>number of records: 200</li>
<li>number of studies: 2</li>
<li>type: numeric</li>
<li>units: m/g</li>
<li>allowable range: 0.5 - 500 m/g</li>
</ul>
<p><strong>spongy_mesophyll_thickness</strong></p>
<ul>
<li>label: Spongy mesophyll cell thickness</li>
<li>description: Thickness of the spongy mesophyll cells</li>
<li>number of records: 75</li>
<li>number of studies: 2</li>
<li>type: numeric</li>
<li>units: um</li>
<li>allowable range: 10 - 1000 um</li>
</ul>
<p><strong>starch_per_area</strong></p>
<ul>
<li>label: Starch per leaf area</li>
<li>description: Mass of starch per leaf area</li>
<li>number of records: 0</li>
<li>number of studies: 0</li>
<li>type: numeric</li>
<li>units: g/m2</li>
<li>allowable range: 0.1 - 50 g/m2</li>
</ul>
<p><strong>stem_C_per_dry_mass</strong></p>
<ul>
<li>label: Stem carbon (C) content per unit stem dry mass</li>
<li>description: Stem carbon (C) content per unit stem dry mass</li>
<li>number of records: 82</li>
<li>number of studies: 1</li>
<li>type: numeric</li>
<li>units: mg/g</li>
<li>allowable range: 100 - 1000 mg/g</li>
</ul>
<p><strong>stem_count_categorical</strong></p>
<ul>
<li>label: Stem Count, categorical</li>
<li>description: Number of stems present, expressed in groups, where categories were 1=1; 2-3=2; 4-10=3; 11-30=4; and &gt;30=5. Used by Peter Vesk.</li>
<li>number of records: 0</li>
<li>number of studies: 0</li>
<li>type: numeric</li>
<li>units: count</li>
<li>allowable range: 0 - 5 count</li>
</ul>
<p><strong>stem_cross_section_area</strong></p>
<ul>
<li>label: Stem cross-sectional area</li>
<li>description: Cross-sectional area of the stem</li>
<li>number of records: 0</li>
<li>number of studies: 0</li>
<li>type: numeric</li>
<li>units: mm2</li>
<li>allowable range: 1 - 1000 mm2</li>
</ul>
<p><strong>stem_cross_section_area_no_bark</strong></p>
<ul>
<li>label: Stem cross-sectional area without bark</li>
<li>description: Cross-sectional area of the stem once bark is removed</li>
<li>number of records: 0</li>
<li>number of studies: 0</li>
<li>type: numeric</li>
<li>units: mm2</li>
<li>allowable range: 1 - 1000 mm2</li>
</ul>
<p><strong>stem_dry_matter_content</strong></p>
<ul>
<li>label: Stem dry mass per unit stem fresh mass (Stem dry matter content)</li>
<li>description: Stem dry mass per unit stem fresh mass</li>
<li>number of records: 230</li>
<li>number of studies: 1</li>
<li>type: numeric</li>
<li>units: mg/g</li>
<li>allowable range: 10 - 1000 mg/g</li>
</ul>
<p><strong>stem_hydraulic_conductivity</strong></p>
<ul>
<li>label: Hydraulic conductivity (Kh)</li>
<li>description: Kh; Measure of how efficiently water is transported through the leaf, determined as the ratio of water flow rate through the leaf to the difference in water potential across the leaf, standardised to leaf area.</li>
<li>number of records: 30</li>
<li>number of studies: 1</li>
<li>type: numeric</li>
<li>units: kg*m/s/MPa</li>
<li>allowable range: 1e-08 - 1e-04 kg*m/s/MPa</li>
</ul>
<p><strong>stem_N_per_dry_mass</strong></p>
<ul>
<li>label: Stem carbon (N) content per unit stem dry mass</li>
<li>description: Stem carbon (N) content per unit stem dry mass</li>
<li>number of records: 82</li>
<li>number of studies: 1</li>
<li>type: numeric</li>
<li>units: mg/g</li>
<li>allowable range: 0.1 - 1000 mg/g</li>
</ul>
<p><strong>basal_diameter</strong></p>
<ul>
<li>label: Stem diameter</li>
<li>description: Diameter at the base of the plant, usually “DBH” except in short plants</li>
<li>number of records: 901</li>
<li>number of studies: 1</li>
<li>type: numeric</li>
<li>units: mm</li>
<li>allowable range: 0.1 - 1000 mm</li>
</ul>
<p><strong>stomatal_conductance_per_area</strong></p>
<ul>
<li>label: Stomatal conductance per unit leaf area under ambient conditions</li>
<li>description: Rate of water loss through stomata under ambient conditions, per unit leaf area</li>
<li>number of records: 18</li>
<li>number of studies: 1</li>
<li>type: numeric</li>
<li>units: mmolH2O/m2/s</li>
<li>allowable range: -0.1 - 2000 mmolH2O/m2/s</li>
</ul>
<p><strong>stomatal_conductance_per_area_at_Amax</strong></p>
<ul>
<li>label: Stomatal conductance per unit leaf area under saturated light and CO2 conditions</li>
<li>description: Rate of water loss through stomata, per unit leaf area under saturated light and CO2 conditions</li>
<li>number of records: 5</li>
<li>number of studies: 1</li>
<li>type: numeric</li>
<li>units: mmolH2O/m2/s</li>
<li>allowable range: -0.1 - 2000 mmolH2O/m2/s</li>
</ul>
<p><strong>stomatal_conductance_per_area_at_Asat</strong></p>
<ul>
<li>label: Stomatal conductance per unit leaf area under saturated light conditions</li>
<li>description: Rate of water loss through stomata, per unit leaf area under saturated light conditions</li>
<li>number of records: 763</li>
<li>number of studies: 5</li>
<li>type: numeric</li>
<li>units: mmolH2O/m2/s</li>
<li>allowable range: -0.1 - 2000 mmolH2O/m2/s</li>
</ul>
<p><strong>stomatal_density</strong></p>
<ul>
<li>label: Stomatal density on the leaf surface</li>
<li>description: Stomatal density on the leaf surface</li>
<li>number of records: 0</li>
<li>number of studies: 0</li>
<li>type: numeric</li>
<li>units: count/mm2</li>
<li>allowable range: 1 - 1000 count/mm2</li>
</ul>
<p><strong>support_fraction</strong></p>
<ul>
<li>label: Proportion of shoot dry mass that is stems (versus leaves)</li>
<li>description: Proportion of shoot dry mass that is stems (versus leaves)</li>
<li>number of records: 0</li>
<li>number of studies: 0</li>
<li>type: numeric</li>
<li>units: g/g</li>
<li>allowable range: 0 - 1 g/g</li>
</ul>
<p><strong>tap_root</strong></p>
<ul>
<li>label: Tap root</li>
<li>description: Binary variable describing whether or not a plant has a tap root present</li>
<li>number of records: 75</li>
<li>number of studies: 1</li>
<li>type: categorical</li>
<li>allowable values:
<ul>
<li>
<em>0</em>: tap root not present</li>
<li>
<em>1</em>: tap root present</li>
</ul>
</li>
</ul>
<p><strong>thickest_root_diameter</strong></p>
<ul>
<li>label: Thickest root diameter</li>
<li>description: Diameter of the thickest root</li>
<li>number of records: 75</li>
<li>number of studies: 1</li>
<li>type: numeric</li>
<li>units: mm</li>
<li>allowable range: 0.01 - 1000 mm</li>
</ul>
<p><strong>twig_area</strong></p>
<ul>
<li>label: Terminal twig cross-sectional area</li>
<li>description: The cross-sectional area of the terminal twig</li>
<li>number of records: 58</li>
<li>number of studies: 1</li>
<li>type: numeric</li>
<li>units: mm2</li>
<li>allowable range: 0.05 - 50 mm2</li>
</ul>
<p><strong>twig_length</strong></p>
<ul>
<li>label: Terminal twig length</li>
<li>description: The length of the terminal twig</li>
<li>number of records: 33</li>
<li>number of studies: 1</li>
<li>type: numeric</li>
<li>units: mm</li>
<li>allowable range: 1 - 1000 mm</li>
</ul>
<p><strong>vapor_pressure_deficit</strong></p>
<ul>
<li>label: Vapor pressure deficit</li>
<li>description: Vapor pressure deficit across the leaf surface</li>
<li>number of records: 0</li>
<li>number of studies: 0</li>
<li>type: numeric</li>
<li>units: kPa</li>
<li>allowable range: 0 - 5 kPa</li>
</ul>
<p><strong>vegetative_regeneration</strong></p>
<ul>
<li>label: Ability to regenerate and spread through the growth and division of vegetative material.</li>
<li>description: Ability to regenerate and spread through the growth and division of vegetative material.</li>
<li>number of records: 759</li>
<li>number of studies: 3</li>
<li>type: categorical</li>
<li>allowable values:
<ul>
<li>
<em>not_vegetative</em>: Plant that regenerate and spread only by seed</li>
<li>
<em>vegetative</em>: Generic term including all plants that can regenerate and spread through the growth and division of vegetative material, not just from seed; examples include runners, rhizomes, and bulbs</li>
</ul>
</li>
</ul>
<p><strong>water_band_index</strong></p>
<ul>
<li>label: Water band index</li>
<li>description: Water band index, the ratio of the reflectance at 970 nm / 900 nm, recorded from the spectro-radiometer.</li>
<li>number of records: 453</li>
<li>number of studies: 1</li>
<li>type: numeric</li>
<li>units: dimensionless</li>
<li>allowable range: 1 - 1.1 dimensionless</li>
</ul>
<p><strong>water_potential_midday</strong></p>
<ul>
<li>label: Midday water potential</li>
<li>description: A plant’s water potential during the heat of the day</li>
<li>number of records: 300</li>
<li>number of studies: 5</li>
<li>type: numeric</li>
<li>units: MPa</li>
<li>allowable range: -10 - 0 MPa</li>
</ul>
<p><strong>water_potential_predawn</strong></p>
<ul>
<li>label: Pre-dawn water potential</li>
<li>description: A plant’s water potential just before sunrise</li>
<li>number of records: 200</li>
<li>number of studies: 3</li>
<li>type: numeric</li>
<li>units: MPa</li>
<li>allowable range: -6 - 0 MPa</li>
</ul>
<p><strong>water_tolerance</strong></p>
<ul>
<li>label: Water (and salt) tolerance strategies</li>
<li>description: Water-tolerance categories</li>
<li>number of records: 564</li>
<li>number of studies: 1</li>
<li>type: categorical</li>
<li>allowable values:
<ul>
<li>
<em>glycophyte</em>: Plants that cannot survive even one tenth the salt concentration found in seawater</li>
<li>
<em>halophyte</em>: Plants that complete their life cycle in a salty environment; many survive in seawater or even higher concentrations of salt</li>
<li>
<em>hydrohalophyte</em>: Tidal swamp or ‘mangrove’ and coastal lagoon elements and temperate-zone salt-marsh species are classed together.</li>
<li>
<em>phreatophyte</em>: Deep-rooted plants—usually trees—obtaining water from a deep underground source that may or may not be saline.</li>
<li>
<em>psammophile</em>: Sand-loving plants commonly found in littoral strand or inland sandfields</li>
<li>
<em>weedy</em>: This inexact term is meant to indicate a high degree of adaptability to sites, and proven tendency toward aggressive colonization of either dry or wet sites, or rarely, both. In general, weeds invade and colonize highly disturbed sites or areas.</li>
<li>
<em>xerohalophyte</em>: Plants adapted to inland salt desert and saline habitats</li>
<li>
<em>xerophyte</em>: Drought-tolerant and drought-adapted plants</li>
</ul>
</li>
</ul>
<p><strong>water_use_efficiency</strong></p>
<ul>
<li>label: Water use efficiency</li>
<li>description: Rate of carbon dioxide uptake relative to water loss, per unit leaf area</li>
<li>number of records: 746</li>
<li>number of studies: 3</li>
<li>type: numeric</li>
<li>units: mmolCO2/molH2O/m2/s</li>
<li>allowable range: 1 - 10 mmolCO2/molH2O/m2/s</li>
</ul>
<p><strong>wood_delta13C</strong></p>
<ul>
<li>label: Wood carbon (C) isotope signature (delta 13C)</li>
<li>description: Wood carbon stable isotope signature</li>
<li>number of records: 169</li>
<li>number of studies: 1</li>
<li>type: numeric</li>
<li>units: per mille</li>
<li>allowable range: -50 - 0 per mille</li>
</ul>
<p><strong>wood_delta15N</strong></p>
<ul>
<li>label: Wood nitrogen (N) isotope signature (delta 15N)</li>
<li>description: Wood nitrogen stable isotope signature</li>
<li>number of records: 169</li>
<li>number of studies: 1</li>
<li>type: numeric</li>
<li>units: per mille</li>
<li>allowable range: -25 - 75 per mille</li>
</ul>
<p><strong>wood_density</strong></p>
<ul>
<li>label: Stem dry mass per unit fresh stem volume</li>
<li>description: Stem dry mass per unit stem fresh volume (stem specific density or SSD or wood density)</li>
<li>number of records: 6325</li>
<li>number of studies: 41</li>
<li>type: numeric</li>
<li>units: mg/mm3</li>
<li>allowable range: 0.1 - 1.3 mg/mm3</li>
</ul>
<p><strong>wood_C_per_dry_mass</strong></p>
<ul>
<li>label: Wood carbon (C) content per unit wood dry mass</li>
<li>description: Wood carbon (C) content per unit wood dry mass</li>
<li>number of records: 169</li>
<li>number of studies: 1</li>
<li>type: numeric</li>
<li>units: mg/g</li>
<li>allowable range: 50 - 750 mg/g</li>
</ul>
<p><strong>wood_N_per_dry_mass</strong></p>
<ul>
<li>label: Wood nitrogen (N) content per unit wood dry mass</li>
<li>description: Wood nitrogen (N) content per unit wood dry mass</li>
<li>number of records: 169</li>
<li>number of studies: 1</li>
<li>type: numeric</li>
<li>units: mg/g</li>
<li>allowable range: 0.1 - 100 mg/g</li>
</ul>
<p><strong>woodiness</strong></p>
<ul>
<li>label: Woodiness</li>
<li>description: A plant’s degree of lignification in stems</li>
<li>number of records: 6582</li>
<li>number of studies: 6</li>
<li>type: categorical</li>
<li>allowable values:
<ul>
<li>
<em>herbaceous</em>: Plant with non-lignified stems</li>
<li>
<em>semi_woody</em>: Plant with partially lignified stems</li>
<li>
<em>woody</em>: Plant that produces secondary xylem, have lignin</li>
</ul>
</li>
</ul>
</div>
</div>
<div id="building-and-contributing-to-austraits" class="section level1">
<h1 class="hasAnchor">
<a href="#building-and-contributing-to-austraits" class="anchor"></a>Building and contributing to AusTraits</h1>
<p>Anyone wishing to use the AusTraits data will need to begin by building AusTraits to create a current compiled dataset. Below, under <a href="#approach">Approach</a> we describe conceptually how individual datasets are merged into the complete AusTraits compilation and under <a href="#building">Rebuilding from scratch</a> is the code needed to actually build AusTraits.</p>
<div id="approach" class="section level2">
<h2 class="hasAnchor">
<a href="#approach" class="anchor"></a>Approach</h2>
<p>Data in AusTraits is stored by</p>
<p>XXXXX Daniel, can you write this?</p>
<p>(More here)</p>
</div>
<div id="structure2" class="section level2">
<h2 class="hasAnchor">
<a href="#structure2" class="anchor"></a>Structure of raw data files for AusTraits</h2>
<p>(More here)</p>
<p>XXXXX</p>
<p>(needs updating)</p>
<pre><code>austraits
├── austraits.Rproj
├── config
├── data
├── R
├── README.md
├── remake.yml
├── tests
└── vignettes</code></pre>
<p>The folder <code>data</code> contains a long list of folders, one for each study and each containing two files:</p>
<pre><code>data
├── Angevin_2010
│   ├── data.csv
│   └── metadata.yml
├── Barlow_1981
│   ├── data.csv
│   └── metadata.yml
├── Bean_1997
│   ├── data.csv
│   └── metadata.yml
├── Blackman_2014
│   ├── data.csv
│   └── metadata.yml
├── ....
</code></pre>
<p>where <code>Angevin_2010</code>, <code>Barlow_1981</code>, <code>Bean_1997</code>, <code>Westoby_2014</code> are each a unique <code>dataset_id</code> in the final dataset.</p>
</div>
<div id="building" class="section level2">
<h2 class="hasAnchor">
<a href="#building" class="anchor"></a>Rebuilding from scratch</h2>
<p>XXXXX clone austraits from github (Daniel include details)</p>
<p>XXXXX Once you have cloned AusTraits onto your computer, run the following code to ensure you have a compiled version of AusTraits and have sourced any relevant functions</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" title="1">austraits &lt;-<span class="st"> </span>remake<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/remake/man/make.html">make</a></span>(<span class="st">"austraits"</span>)</a>
<a class="sourceLine" id="cb5-2" title="2"><span class="kw"><a href="https://rdrr.io/r/base/source.html">source</a></span>(<span class="st">"R/support.R"</span>)</a>
<a class="sourceLine" id="cb5-3" title="3"><span class="kw"><a href="https://rdrr.io/r/base/source.html">source</a></span>(<span class="st">"R/setup.R"</span>)</a>
<a class="sourceLine" id="cb5-4" title="4"><span class="kw"><a href="https://rdrr.io/r/base/source.html">source</a></span>(<span class="st">"R/steps.R"</span>)</a>
<a class="sourceLine" id="cb5-5" title="5"><span class="kw"><a href="https://rdrr.io/r/base/source.html">source</a></span>(<span class="st">"R/austraits.R"</span>)</a>
<a class="sourceLine" id="cb5-6" title="6"><span class="kw"><a href="https://rdrr.io/r/base/source.html">source</a></span>(<span class="st">"R/support.R"</span>)</a>
<a class="sourceLine" id="cb5-7" title="7"><span class="kw"><a href="https://rdrr.io/r/base/source.html">source</a></span>(<span class="st">"R/report_utils.R"</span>)</a>
<a class="sourceLine" id="cb5-8" title="8"><span class="kw"><a href="https://rdrr.io/r/base/source.html">source</a></span>(<span class="st">"R/report_notetaker.R"</span>)</a></code></pre></div>
<p>Within the compiled matrix <code>austraits</code> are the elements listed in <a href="#elements">Elements of AusTraits</a>. The actual trait data can be found in the dataframe <code>austraits$traits</code>.</p>
</div>
</div>
<div id="format" class="section level1">
<h1 class="hasAnchor">
<a href="#format" class="anchor"></a>Format of raw data files for AusTraits</h1>
<p>This section describes how to prepare / modify the raw data from individual studies for inclusion in the AusTraits compilation.</p>
<p>Data from each study is organised into a separate folder, with two files:<br> - <code>data.csv</code>: a table containing the actual trait data in comma-separated values format, ideally with each trait in a single column and each observation on a single row (i.e. wide format).<br> - <code>metadata.yml</code>: contains further information about the study, maps trait names and units onto standard types, and lists any substitutions applied to the data in processing (see below).</p>
<p>Below are the Instructions on how to format files to contribute a new study to AusTraits. It is important that all steps are followed so that our automated workflow proceeds without problems.</p>
<ol style="list-style-type: decimal">
<li>XXXXX create a github branch?</li>
<li>Create a new folder within the folder <code>data</code> with a name corresponding to the paper or study, e.g. <code>Gallagher_2014</code> (do not include <em>et al</em> or similar).</li>
<li>Prepare the files <code>data.csv</code> and <code>metadata.yml</code> and place them within the folder, as per instructions <a href="#create_data">below</a>.</li>
<li>Add the new study into the build framework and rebuild AusTraits, as described under <a href="#adding">Adding data to AusTraits</a> below.</li>
<li>Run tests on the contributed data and correct the <code>data.csv</code> and <code>metadata.yml</code>files as necessary. See the section <a href="#running_tests">Tests</a> below for more details.</li>
<li>Generate and proofread a report on the data. In particular, check that numeric trait values fall within a logical range relative to other studies and that individual trait observations are not unnecessarily excluded because their trait values are unsupported.</li>
<li>Return to step 3 if changes are made to the <code>data.csv</code> or <code>metadata.yml</code> files.</li>
<li>Push to GitHub.</li>
</ol>
<div id="create_data" class="section level3">
<h3 class="hasAnchor">
<a href="#create_data" class="anchor"></a>Data.csv</h3>
<p>The file <code>data.csv</code> file can be in either long or wide format.</p>
<p>Required columns include the species name, the trait name (column in long format, header in wide format), units (column in long format, part of header in wide format), site (if applicable), and trait values.</p>
<p>If multiple trait measurements were made on the same individual or are the mean of a species’ measurements from the same site, they should be kept linked. If the data is in wide format, each row should represent measurements made on a single individual or a single species-by-site mean, with different trait values as consecutive columns. If the data is in long format, an observation ID (or other identifier) must be assigned to identify which measurements are linked to a unique individual or site.</p>
<p>Keep the data file in rawest form possible (i.e. as few changes as possible) but it must be a single csv file. Additional custom R code may be required to make the file exactly compatible with the AusTraits format, but these changes should be executed as AusTraits is compiled and should be in the <code>metadata.yml</code> file under <code>config/custom_R_code</code>. Any files used to create the submitted <code>data.csv</code> file (e.g. Excel …) should be supplied so they can be archived in a subfolder within the study folder named <code>raw</code>.</p>
</div>
<div id="metadata-yml" class="section level3">
<h3 class="hasAnchor">
<a href="#metadata-yml" class="anchor"></a>Metadata.yml</h3>
<p>The metadata is compiled in a <code>.yml</code> file, a structured data file where information is presented in a hierarchical format (see <a href="#yaml">Appendix for details</a>). There are 9 values at the top hierarchical level: source, people, dataset, sites, config, traits, substitutions, taxonomic_updates, questions. These are each described below.</p>
<p>As a start, you may want to checkout some examples from <a href="https://github.com/traitecoevo/austraits/tree/master/data">existing studies in AusTraits</a>, e.g. <a href="https://github.com/traitecoevo/austraits/blob/master/data/Angevin_2010/metadata.yml">Angevin_2010</a> or <a href="https://github.com/traitecoevo/austraits/blob/master/data/Wright_2004/metadata.yml">Wright_2004</a>.</p>
<div id="source" class="section level4">
<h4 class="hasAnchor">
<a href="#source" class="anchor"></a>Source</h4>
<p>This section provides citation details for the original source(s) for the data, whether it is a published journal article, book, website, or thesis. In general we aim to reference the primary source. References are written in structured yml format, under the category <code>source</code> and then sub-groupings <code>primary</code> and <code>secondary</code>. General guidelines for describing a source</p>
<ul>
<li>maximum of one primary and secondary source allowed</li>
<li>elements are names as in <a href="https://en.wikipedia.org/wiki/BibTeX">bibtex format</a>
</li>
<li>keys should be named in the format <code>Surname_year</code> and should be identical to the name given to the dataset folder.</li>
<li>a secondary source may be needed if the main collector is not an author on the paper where data was released, or data were otherwise released via a subsequent study.</li>
<li>if your data is from an unpublished study, only include the elements that are applicable</li>
</ul>
<p>Following are some examples for different types of source.</p>
<p>A journal article:</p>
<pre><code>source:
  primary:
    key: Falster_2005
    bibtype: Article
    author: Daniel S. Falster, Mark Westoby
    year: 2005 title: Alternative height strategies among 45 dicot rain forest species from tropical Queensland, Australia
    journal: Journal of Ecology volume: 93 pages: 521--535
    publisher: Wiley-Blackwell
    doi: 10.1111/j.0022-0477.2005.00992.x
  secondary: .na</code></pre>
<p>A book:</p>
<pre><code>source:
  primary:
    key: Cooper_2004
    bibtype: Book
    author: Wendy Cooper, William T. Cooper
    year: 2004
    title: Fruits of the Australian tropical rainforest
    publisher: Nokomis Editions
    isbn: '9780958174213'
  secondary: .na</code></pre>
<p>An online resource:</p>
<pre><code>source:
  primary: key: WAH_1998
    bibtype: Misc
    author: Western Australian Herbarium
    year: 1998
    title: FloraBase--the Western Australian Flora
    publisher: Department of Parks and Wildlife
    url: https://florabase.dpaw.wa.gov.au/
  secondary: .na</code></pre>
<p>An unpublished resource:</p>
<pre><code>source:
  primary:
    key: Duncan_1998
    bibtype: Unpublished
    author: David H. Duncan
    year: 1998
    title: Leaf anatomy of Australian plant species
    note: Collected while at Macquarie University
  secondary: .na</code></pre>
<p>Note that in these first examples <code>secondary</code> is set as <code>.na</code>. If a secondary source is included it may look like:</p>
<pre><code>  primary:
    key: Chave_2009
    bibtype: Article
    author: Jerome Chave, David Coomes, Steven Jansen, Simon L. Lewis, Nathan G. Swenson, Amy E. Zanne
    year: 2009 title: Towards a worldwide wood economics spectrum
    journal: Ecology Letters
    volume: 12
    pages: 351--366
    publisher: Wiley-Blackwell
    doi: 10.1111/j.1461-0248.2009.01285.x
  secondary:
    key: Zanne_2009
    bibtype: Misc
    author: Amy E. Zanne, G. Lopez-Gonzalez, David A. Coomes, Jugo Ilic, Steven Jansen, Simon L. Lewis, Regis B. Miller, Nathan G. Swenson, Michael C. Wiemann, Jerome   Chave
    year: 2009
    title: 'Data from: Towards a worldwide wood economics spectrum'
    volume: .na
    pages: .na
    publisher: Dryad Digital Repository
    doi: 10.5061/dryad.234</code></pre>
</div>
<div id="people" class="section level4">
<h4 class="hasAnchor">
<a href="#people" class="anchor"></a>People</h4>
<p>This section provides a list of the key contributors to the study, their respective institutions and roles in the study. Roles are defined as follows:</p>
<table class="table table table-striped table-hover table-condensed table-responsive">
<thead><tr>
<th style="text-align:left;">
key
</th>
<th style="text-align:left;">
value
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
collector
</td>
<td style="text-align:left;">
The person (people) leading data collection (up to 2 permitted)
</td>
</tr>
<tr>
<td style="text-align:left;">
contributor
</td>
<td style="text-align:left;">
Person responsible for entering data into AusTraits
</td>
</tr>
<tr>
<td style="text-align:left;">
lab_leader
</td>
<td style="text-align:left;">
Leader of lab group at time of collection
</td>
</tr>
<tr>
<td style="text-align:left;">
assistant
</td>
<td style="text-align:left;">
Anyone else who assisted in collection of the data
</td>
</tr>
<tr>
<td style="text-align:left;">
contact
</td>
<td style="text-align:left;">
The person to contact with questions about the data set
</td>
</tr>
</tbody>
</table>
<p>An example is as follows:</p>
<pre><code>people:
- name: Daniel Falster
  institution: Macquarie University
  role: collector, contact, contributor
- name: Mark Westoby
  institution: Macquarie University
  role: lab_leader
</code></pre>
<p>Note that only the AusTraits custodians have the contributors e-mail addresses on file. This information will not be directly available to AusTraits users or new contributors via Github.</p>
</div>
<div id="dataset" class="section level4">
<h4 class="hasAnchor">
<a href="#dataset" class="anchor"></a>Dataset</h4>
<p>This section details about the study where the data were collected</p>
<p>The following elements are included under the element <code>dataset</code>:</p>
<ul>
<li>
<strong>year_collected_start</strong>: The year data collection commenced</li>
<li>
<strong>year_collected_end</strong>: The year data collection was completed</li>
<li>
<strong>description</strong>: A 1-2 sentence description of the purpose of the study.</li>
<li>
<strong>collection_type</strong>: A field to indicate where the majority of plants on which traits were measured were collected - in the <code>field</code>, <code>lab</code>, <code>glasshouse</code>, <code>herbarium specimens</code>, or <code>literature</code>. The latter should only be used when the data were sourced from the literature and the collection type is unknown.</li>
<li>
<strong>sample_age_class</strong>: A field to indicate if the study was completed on <code>adult</code> or <code>juvenile</code> plants</li>
<li>
<strong>sampling_strategy</strong>: A written description of how study sites were selected and how study individuals were selected. When available, this information is lifted verbatim from a published manuscript. For herbarium studies, this field ideally indicates which herbaria were ‘sampled’ to measure a specific trait</li>
<li>
<strong>original_file</strong>: The name of the file initially submitted to AusTraits</li>
<li>
<strong>notes</strong>: Generic notes about the study and processing of data</li>
</ul>
<p>An example is as follows:</p>
<pre><code>  year_collected_start: 2004
  year_collected_end: 2004
  description: Trait values for species with faster versus slower height growth following disturbance for Myall Lakes species.
  collection_type: field
  sample_age_class: adult
  sampling_strategy: Fire is a recurrent disturbance in the park (interval – 0–30 years; Fox and Fox 1986). A mosaic of fire histories has facilitated previous use of space-for-time substitutions in studies of small mammal succession (Fox and McKay 1981). Here we employ the same methodology to reconstruct species height-growth trajectories (Enright and Goldblum 1999). Sites were identified at a range of times since fire with the use of NSW national parks GIS fire history records and personal observations of Karen Ross (Ross et al. 2002). Patches of vegetation 1, 2, 4, 8, 10, 12, 15, 27 and 28 years since fire were identified. Where possible several patches within a given age class were surveyed to determine species presence or absence. Nineteen species recorded in a majority of patches were selected for further study. This included eight resprouting species and 11 obligate seeders (full list in Appendix 1).
  original_file: Falster &amp; Westoby 2005 Oikos appendix.doc
  notes: none</code></pre>
</div>
<div id="config" class="section level4">
<h4 class="hasAnchor">
<a href="#config" class="anchor"></a>Config</h4>
<p>This section information on the format of the submitted data file</p>
<p>Values are as follows:</p>
<ul>
<li>
<strong>data_is_long_format</strong>: Indicates if the data spreadsheet has a vertical (long) or horizontal (wide) configuration with <code>yes</code> or <code>no</code> terminology</li>
<li>
<strong>variable_match</strong>: Identifies which information is in each column in the file <code>data.csv</code>, excluding the actual trait data. One element within <code>variable_match</code> must be <code>species_name</code>. Datasets with <code>data_is_long_format</code> set to <code>yes</code> must also identify which column includes data on <code>value</code> and <code>trait_name</code>. Other allowed values include <code>site_name</code> and <code>observation_id</code>, if these columns are present in the <code>data.csv</code> file.</li>
<li>
<strong>custom_R_code</strong>: A field where additional R code can be included. This allows for custom manipulation of the data in the submitted spreadsheet into a different format for easy integration with AusTraits. <code>.na</code> indicates no custom R code was used.</li>
</ul>
<p>An example is</p>
<pre><code>config:
  data_is_long_format: yes
  variable_match:
    species_name: Taxon
    value: trait value
    trait_name: trait
  custom_R_code: mutate(data, `trait value` = ifelse(trait == 'flowering time', convert_month_range_vec_to_binary(`trait value`), `trait value`))</code></pre>
<p>A common use of the <code>custom_R_code</code> is to automate the conversion of a verbal description of flowering or fruiting periods into the supported trait values, as occurs in this example. It might also be used if values for a single trait are expressed across multiple columns and need to be merged. See the <code>Catford_2014</code> as an example of this. Additional examples on adding <code>custom_R_code</code> are provided <a href="#custom_R_code">in the appendix</a>.</p>
</div>
<div id="traits-1" class="section level4">
<h4 class="hasAnchor">
<a href="#traits-1" class="anchor"></a>Traits</h4>
<p>This section A translation table mapping traits and units in the original study onto corresponding variables in AusTraits. Also specified here are methods used to collect the data.</p>
<p>For each trait submitted to AusTraits, there is the following information:</p>
<ul>
<li>
<strong>var_in</strong>: Names of trait in the original data submitted</li>
<li>
<strong>unit_in</strong>: Units of trait in the original data submitted</li>
<li>
<strong>trait_name</strong>: Name of trait sampled. Allowable values specified in the table <code>traits</code>.</li>
<li>
<strong>value_type</strong>: A categorical variable describing the type of trait value recorded</li>
<li>
<strong>replicates</strong>: Number of replicate measurements that comprise the data points for the trait for each measurement. A numeric value (or range) is ideal and appropriate if the value type is a <code>mean</code>, <code>median</code>, <code>min</code> or <code>max</code>. For these value types, if replication is unknown the entry should be <code>unknown</code>. If the value type is <code>raw_value</code> the replicate value should be <code>1</code>. If the value type is <code>expert_mean</code>, <code>expert_min</code>, or <code>expert_max</code> the replicate value should be <code>.na</code>.</li>
<li>
<strong>methods</strong>: A textual description of the methods used to collect the trait data. Whenever available, methods are taken near-verbatim from referenced source. Methods can include descriptions such as ‘measured on herbarium specimens’,‘data from the literature’, or a detailed description of the field or lab methods used to collect the data.</li>
</ul>
<p>Values under <code>trait_name</code> must be allowable values, as described under <a href="#trait_defs">the section trait definitions</a>. Similarly, values under <code>value_type</code> must be allowable values, as described under <a href="#value_types">the section Value types</a>.</p>
<p>An example is as follows:</p>
<pre><code>traits:
- var_in: LMA (mg mm-2)
  unit_in: mg/mm2
  trait_name: specific_leaf_area
  value_type: site_mean
  replicates: 3
  methods: LMA was calculated as the leaf dry mass (oven-dried for 48 hours at 65 °C) divided by leaf size. It was measured on the first five fully expanded leaves at the tip of each individual.
- var_in: leaf size (mm2)
  unit_in: mm2
  trait_name: leaf_area
  value_type: site_mean
  replicates: 3
  methods: Leaf size was calculated as the one-sided leaf area (flat bed scanner). It was measured on the first five fully expanded leaves at the tip of each individual.</code></pre>
</div>
<div id="substitutions" class="section level4">
<h4 class="hasAnchor">
<a href="#substitutions" class="anchor"></a>Substitutions</h4>
<p>This section provides a list of any “find and replace” substitutions needed to get the data to get it into the right format</p>
<p>Substitutions are required whenever the exact word(s) used to describe a categorical trait value in AusTraits are different from the vocabulary used by the author in the <code>data.csv</code> file. It is preferable that authors make changes using <code>substitutions</code> rather than changing the <code>data.csv</code> file. See the section <a href="#trait_defs">trait definitions</a> for a list of supported values for each trait.</p>
<p>Each substitution is documented using the following elements:</p>
<ul>
<li>
<strong>trait_name</strong>: Name of variable to search</li>
<li>
<strong>find</strong>: Value to find</li>
<li>
<strong>replace</strong>: Replacement value</li>
</ul>
<p>An example is as follows:</p>
<pre><code>substitutions:
- trait_name: life_history
  find: p
  replace: perennial
- trait_name: plant_growth_form
  find: s
  replace: shrub
- ...</code></pre>
</div>
<div id="taxonomic-updates" class="section level4">
<h4 class="hasAnchor">
<a href="#taxonomic-updates" class="anchor"></a>Taxonomic updates</h4>
<p>This section provides a list of taxonomic name changes needed to align with current taxonomy Taxonomic updates are required to align species names with an accepted species list (for details, see details on <a href="#taxonomy">Taxonony</a>.</p>
<p>Each substitution is documented using the following elements:</p>
<ul>
<li>
<strong>find</strong>: Taxonomic name given to species in the original data supplied by the authors</li>
<li>
<strong>replace</strong>: Name of species after aligning taxonomy with standard sources.</li>
<li>
<strong>reason</strong>: Records why the change was implemented, e.g. <code>typos</code>, <code>taxonomic synonyms</code>, and <code>standardising spellings</code>
</li>
</ul>
</div>
</div>
<div id="taxonomic-updates-1" class="section level3">
<h3 class="hasAnchor">
<a href="#taxonomic-updates-1" class="anchor"></a>Taxonomic updates</h3>
<p>An example is as follows:</p>
<pre><code>taxonomic_updates:
- find: Carissa lanceolata
  replace: Carissa spinarum
  reason: Synonym reported by TaxonStand (2018-09-19)
- find: Melaleuca pallida
  replace: Callistemon pallidus
  reason: Synonym reported by TaxonStand (2018-09-19)</code></pre>
<div id="questions" class="section level4">
<h4 class="hasAnchor">
<a href="#questions" class="anchor"></a>Questions</h4>
<p>This section provides a place to record any queries we have about the dataset (recorded as a named array), including notes on any additional traits that may have been collected in the study but have not been incorporated into austraits.</p>
<p>An example is as follows:</p>
<pre><code>questions:
  questions for author: Triglochin procera has very different seed masses in the main traits spreadsheet and the field seeds worksheet. Which is correct? There are a number of species with values in the field leaves worksheet that are absent in the main traits worksheet - we have included this data into AusTraits; please advise if this was inappropriate.
  austraits: need to map aquatic_terrestrial onto an actual trait once one is created.</code></pre>
</div>
</div>
<div id="building2" class="section level2">
<h2 class="hasAnchor">
<a href="#building2" class="anchor"></a>Building &amp; rebuilding AusTraits</h2>
<p>To build AusTraits from scratch:..,</p>
<p>(More here)</p>
<p>You can then rebuild AusTraits, including your dataset.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb18-1" title="1">austraits &lt;-<span class="st"> </span>remake<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/remake/man/make.html">make</a></span>(<span class="st">"austraits"</span>)</a></code></pre></div>
</div>
<div id="adding" class="section level2">
<h2 class="hasAnchor">
<a href="#adding" class="anchor"></a>Adding data to AusTraits</h2>
<p>This section describes how to prepare / modify the raw data from individual studies for inclusion in the AusTraits compilation.</p>
<p>Below are the Instructions on how to format files to contribute a new study to AusTraits. It is important that all steps are followed so that our automated workflow proceeds without problems.</p>
<ol style="list-style-type: decimal">
<li>XXXXX create a github branch?</li>
<li>Create a new folder within the folder <code>data</code> with a name corresponding to the paper or study, e.g. <code>Gallagher_2014</code> (do not include <em>et al</em> or similar).</li>
<li>Prepare the files <code>data.csv</code> and <code>metadata.yml</code> and place them within the folder, as per instructions <a href="#create_data">below</a>.</li>
<li>Add the new study into the build framework and rebuild AusTraits, as described under <a href="#adding">Adding data to AusTraits</a> below.</li>
<li>Run tests on the contributed data and correct the <code>data.csv</code> and <code>metadata.yml</code>files as necessary. See the section <a href="#running_tests">Tests</a> below for more details.</li>
<li>Generate and proofread a report on the data. In particular, check that numeric trait values fall within a logical range relative to other studies and that individual trait observations are not unnecessarily excluded because their trait values are unsupported.</li>
<li>Return to step 3 if changes are made to the <code>data.csv</code> or <code>metadata.yml</code> files.</li>
<li>Push to GitHub.</li>
</ol>
<p>It may help to download one of the <a href="https://github.com/traitecoevo/austraits/tree/master/data">existing datasets</a> and use it as a template for your own files and a guide on required content. You should look at the files in the <a href="https://github.com/traitecoevo/austraits/tree/master/config">config folder</a>, in particular the <code>definitions</code> files for the list of traits we cover and the supported trait values for each trait. Or read through information on the supported traits and trait values <a href="#trait_defs">here</a></p>
<p>Once you have prepared your <code>data.csv</code> and <code>metadata.yml</code> files within a folder in the <code>data</code> directory, you can incorporate the new data into AusTraits by running:</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb19-1" title="1"><span class="kw">austraits_rebuild_remake_setup</span>()</a></code></pre></div>
<p>(This step updates the file <code>remake.yml</code> with appropriate rules for the new dataset; similarly if you remove datasets, do the same. At this stage, <a href="https://github.com/richfitz/remake">remake</a> offers no looping constructs (on purpose) so for now we generate the remake file using <a href="https://github.com/edwindj/whisker">whisker</a>.)</p>
<p>You can then rebuild AusTraits, including your dataset.</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb20-1" title="1">austraits &lt;-<span class="st"> </span>remake<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/remake/man/make.html">make</a></span>(<span class="st">"austraits"</span>)</a></code></pre></div>
<div id="constructing-the-metadata-yml-file" class="section level3">
<h3 class="hasAnchor">
<a href="#constructing-the-metadata-yml-file" class="anchor"></a>Constructing the <code>metadata.yml</code> file</h3>
<p>One way to construct the <code>metadata.yml</code> file is to use one of the existing files and modify yours to follow the same format. As a start, checkout some examples from <a href="https://github.com/traitecoevo/austraits/tree/master/data">existing studies in AusTraits</a>, e.g. <a href="https://github.com/traitecoevo/austraits/blob/master/data/Angevin_2010/metadata.yml">Angevin_2010</a> or <a href="https://github.com/traitecoevo/austraits/blob/master/data/Wright_2004/metadata.yml">Wright_2004</a>.</p>
<p>Note, when editing the <code>metadata.yml</code>, edits should be made in a proper text editor (Microsoft word tends to stuff up the formatting). For example, Rstudio works.</p>
<p>To assist you in constructing the <code>metadata.yml</code> file, we have developed functions to help fill in the different sections of the file. If you wish to include additional elements, you can afterwards edit the file further.</p>
<p>To use the functions, make sure you first run the following, to make the functions available</p>
</div>
<div id="creating-a-template" class="section level3">
<h3 class="hasAnchor">
<a href="#creating-a-template" class="anchor"></a>Creating a template</h3>
<p>The first function creates a basic template for your the <code>metadata.yml</code> file for your study. Assuming you have already created a file <code>data.csv</code> in the folder <code>data/your_dataset_id</code>, run</p>
<p>The function will ask a series of questions and then create a relatively empty file <code>data/your_dataset_id/metadata.yml</code>. The key questions are:</p>
<ul>
<li>Is the data long vs wide? A wide dataset has each variable(i.e. trait ) as a column. A long dataset has each variable as a row and column as a species. <em>Select column for ‘species_name’ </em>Select column for ‘trait_name’</li>
</ul>
</div>
<div id="source-1" class="section level3">
<h3 class="hasAnchor">
<a href="#source-1" class="anchor"></a>Source</h3>
<p>Three functions are available to help entering citation details for the source data.</p>
<p>The function <code>metadata_create_template</code> creates a template for the primary source with default fields for a journal article, which you can then edit manually.</p>
<p>Alternatively, if you have a <code>doi</code> for your study, use the function:</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb21-1" title="1"><span class="kw">metadata_add_source_doi</span>(dataset_id, doi)</a></code></pre></div>
<p>and the different elements within source will automatically be generated. By default, details are added as a primary source. To override this, specify the type</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb22-1" title="1"><span class="kw">metadata_add_source_doi</span>(dataset_id, doi, <span class="dt">type=</span><span class="st">"secondary"</span>)</a></code></pre></div>
<p>Alternatively, if you have reference details saved in a bibtex file called <code>myref.bib</code> you can use the function</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb23-1" title="1"><span class="kw">metadata_add_source_doi</span>(dataset_id, <span class="dt">file =</span> <span class="st">"myref.bib"</span>)</a></code></pre></div>
<p>(These options require the package <a href="https://github.com/ropensci/rcrossref">rcorssref</a> and <a href="https://github.com/ropensci/RefManageR/">RefManageR</a> to be installed.)</p>
</div>
<div id="people-1" class="section level3">
<h3 class="hasAnchor">
<a href="#people-1" class="anchor"></a>People</h3>
<p>The function <code>metadata_create_template</code> creates a template for entering details about people, which you can then edit manually.</p>
</div>
<div id="testing-custom-r-code" class="section level3">
<h3 class="hasAnchor">
<a href="#testing-custom-r-code" class="anchor"></a>Testing custom R code</h3>
<p>Occasionally all the changes we want to make to dataset may not fit into the prescribed workflow used in AusTraits. For example, we assume each trait has a single unit. But there are a few datasets where data on different rows have different units. So we want to make to make some custom modifications to this particular dataset before the common pipeline of operations gets applied. To make this possible, the workflow allows for some custom R code to be run as a first step in the processing pipeline. That pipeline (in the function <a href="https://github.com/traitecoevo/austraits/blob/master/R/steps.R"><code>load_study</code></a>) looks like this:</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb24-1" title="1">  data &lt;-<span class="st"> </span></a>
<a class="sourceLine" id="cb24-2" title="2"><span class="st">    </span><span class="kw">read_csv</span>(filename_data_raw, <span class="dt">col_types =</span> <span class="kw">cols</span>()) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb24-3" title="3"><span class="st">    </span><span class="kw">custom_manipulation</span>(metadata[[<span class="st">"config"</span>]][[<span class="st">"custom_R_code"</span>]])() <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb24-4" title="4"><span class="st">    </span><span class="kw">parse_data</span>(dataset_id, metadata) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb24-5" title="5"><span class="st">    </span>...</a></code></pre></div>
<p>Note the second line. After loading the csv file, we can apply some Custom R code small manipulations to the dataframe before processing it. Custom R code is valid R code, but written inside the <code>metadata.yml</code> file. While developing this, you’ll want to test your code. This can be achieved by running the function</p>
<p>which returns a data frame, showing how the datasets looks after being manipulated.mes</p>
</div>
<div id="traits-2" class="section level3">
<h3 class="hasAnchor">
<a href="#traits-2" class="anchor"></a>Traits</h3>
<p>Add traits</p>
<!-- TODO: You will be asked to indicate the columns you wish to keep as distinct traits -->
</div>
<div id="sites-1" class="section level3">
<h3 class="hasAnchor">
<a href="#sites-1" class="anchor"></a>Sites</h3>
<p>Add sites details</p>
<p>This function assumes you have site details stored in wide format, in R:</p>
<!-- TODO: need to indicate that the following code needs to be run with the assignment site_data for the previous code to work. -->
<pre><code>## # A tibble: 2 x 6
##   site_name description `elevation (m)` `latitude (deg)` `longitude (deg…
##   &lt;chr&gt;     &lt;chr&gt;                 &lt;dbl&gt;            &lt;dbl&gt;            &lt;dbl&gt;
## 1 Atherton  Tropical r…             800            -17.1             146.
## 2 Cape Tri… Complex me…              25            -16.1             145.
## # … with 1 more variable: `rainfall (mm)` &lt;dbl&gt;</code></pre>
<p>If your data is in a file, you’ll need to read it in first.</p>
</div>
</div>
<div id="substitutions-1" class="section level2">
<h2 class="hasAnchor">
<a href="#substitutions-1" class="anchor"></a>Substitutions</h2>
<p>Substitutions can be added by running:</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb26-1" title="1"><span class="kw">metadata_add_substitution</span>(dataset_id, trait_name, find, replace)</a></code></pre></div>
<p>where <code>find</code> is the trait value used in the data.csv file and <code>replace</code> is the trait value supported by AusTraits.</p>
</div>
<div id="taxonomic-updates-2" class="section level2">
<h2 class="hasAnchor">
<a href="#taxonomic-updates-2" class="anchor"></a>Taxonomic Updates</h2>
<p>We’ve implemented code to semi-automate the checking of names using the R package <a href="https://cran.r-project.org/web/packages/Taxonstand/index.html">Taxonstand</a> (for more documentation see <a href="https://www.rdocumentation.org/packages/Taxonstand/versions/2.1/topics/TPL">here</a>). To generate a suggested name change for a specific study run:</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb27-1" title="1"><span class="kw">metadata_check_taxa</span>(<span class="st">"Westoby_2014"</span>)</a></code></pre></div>
<p>If TaxonStand finds taxonomic changes to make it will add the relevant lines of code directly to the metadata.yml file.</p>
<p>TaxonStand has been configured in the above function to only permit relatively certain changes (e.g. with a minor change to spelling or known synonym).</p>
<p>There are additional arguments you can add for the function <code>metadata_check_taxa</code> including: - <code>update</code> where the default is TRUE, meaning changes found will be added to the <code>metadata.yml</code> file - <code>typos</code> where the default is FALSE, meaning typos will not be corrected - <code>diffchar</code> which indicates the number of characters that can be different for a typo-match. Here the default is two.</p>
<p>Therefore, if you want the function <code>metadata_check_taxa</code> to correct 1 and 2 character typos, run the function as follows:</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb28-1" title="1"><span class="kw">metadata_check_taxa</span>(<span class="st">"Westoby_2014"</span>, <span class="dt">typos=</span><span class="ot">TRUE</span>)</a></code></pre></div>
<p>If TaxonStand fails to find a suitable alignment, and you have identified one yourself, you can add it to the metadata by running</p>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb29-1" title="1"><span class="kw">metadata_add_taxnomic_change</span>(study, find, replace, reason)</a></code></pre></div>
</div>
<div id="running_tests" class="section level2">
<h2 class="hasAnchor">
<a href="#running_tests" class="anchor"></a>Tests</h2>
<p>You can also run some automated tests to ensure the dataset meets required setup. The tests run through a collection of pre-specified checks on the files for each study. The output alerts you to possible issues needing to be fixed, by comparing the data in the files with expected structure and allowed values, as specified in the definitions.</p>
<p>To run the tests, the variable <code>dataset_ids</code> must be defined in the global namespace, containing a vector of ids to check. For example</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb30-1" title="1"><span class="co"># load relevant functions</span></a>
<a class="sourceLine" id="cb30-2" title="2"><span class="kw"><a href="https://rdrr.io/r/base/source.html">source</a></span>(<span class="st">"R/setup.R"</span>)</a>
<a class="sourceLine" id="cb30-3" title="3"></a>
<a class="sourceLine" id="cb30-4" title="4"><span class="co"># Tests run test on one study</span></a>
<a class="sourceLine" id="cb30-5" title="5">dataset_ids &lt;-<span class="st"> "Bragg_2002"</span></a>
<a class="sourceLine" id="cb30-6" title="6"><span class="kw">austraits_run_tests</span>()</a>
<a class="sourceLine" id="cb30-7" title="7"></a>
<a class="sourceLine" id="cb30-8" title="8"><span class="co"># Tests run test on all studies</span></a>
<a class="sourceLine" id="cb30-9" title="9">dataset_ids &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/list.files.html">dir</a></span>(<span class="st">"data"</span>)</a>
<a class="sourceLine" id="cb30-10" title="10"><span class="kw">austraits_run_tests</span>()</a></code></pre></div>
</div>
<div id="reports-quality-checks" class="section level2">
<h2 class="hasAnchor">
<a href="#reports-quality-checks" class="anchor"></a>Reports / quality checks</h2>
<p>To enable better quality checks we have code to generate a report on the data in each study.</p>
<p>(Reports are written in <a href="https://rstudio.github.io/rmarkdown/">Rmarkdown</a> and generated via the <a href="https://cran.r-project.org/web/packages/knitr/">knitr</a> package. The template is stored in the folder <code>vignettes</code> called <code>report_study.html</code>).</p>
<p>To generate a report for a particular study:</p>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb31-1" title="1">austraits &lt;-<span class="st"> </span>remake<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/remake/man/make.html">make</a></span>(<span class="st">"austraits"</span>)</a>
<a class="sourceLine" id="cb31-2" title="2"><span class="kw"><a href="https://rdrr.io/r/base/source.html">source</a></span>(<span class="st">"R/report_utils.R"</span>)</a>
<a class="sourceLine" id="cb31-3" title="3"><span class="kw">build_study_report</span>(<span class="st">"Wright_2002"</span>)</a></code></pre></div>
<p><strong>Guidelines for writing report code</strong></p>
<ul>
<li>use <a href="https://rmarkdown.rstudio.com/lesson-3.html">knitr chunk options</a> to customise when code is shown and how output is displayed.</li>
<li>use <a href="http://htmlpreview.github.io/?https://github.com/nicercode/2018_BEES_regression/blob/master/tidyverse.html">tidyverse style and format</a>
</li>
<li>use <a href="https://cran.r-project.org/web/packages/kableExtra/vignettes/awesome_table_in_html.html">kableExtra for styling tables</a>
</li>
</ul>
<p><strong>Maps:</strong> We use the package <a href="https://cran.r-project.org/web/packages/leaflet/index.html">leaflet</a> to generate interactive maps via the JavaScript ‘Leaflet’ framework and based on the <a href="https://www.openstreetmap.org/">Open street map</a>.</p>
</div>
<div id="pushing-to-github" class="section level2">
<h2 class="hasAnchor">
<a href="#pushing-to-github" class="anchor"></a>Pushing to GitHub</h2>
<p>By far our preferred way of contributing is for you to fork the database in github, add your dataset then send us a <a href="https://help.github.com/articles/using-pull-requests/">pull request</a>. If this is not possible, you could email the relevant files (see above) to Rachael Gallagher.</p>
</div>
</div>
<div id="usage-examples" class="section level1">
<h1 class="hasAnchor">
<a href="#usage-examples" class="anchor"></a>Usage examples</h1>
<p>The following section provides examples for users of <code>R</code> on manipulating the AusTraits dataset. Users of other platforms can follow similar steps using commands specific to their platform.</p>
<p>We like using functions from packages like <a href="https://dplyr.tidyverse.org/">dplyr</a>, <a href="https://readr.tidyverse.org/">readr</a>, <a href="https://tidyr.tidyverse.org/">tidyr</a> and <a href="https://magrittr.tidyverse.org/"><code>magrittr</code></a> in the <a href="https://www.tidyverse.org/">tidyverse</a> to manipulate data. These packages can be loaded individually, or together by running:</p>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb32-1" title="1"><span class="co"># install.packages("tidyverse")</span></a>
<a class="sourceLine" id="cb32-2" title="2"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(tidyverse)</a></code></pre></div>
<p>The examples below assume you have the tidyverse installed and loaded. This is not essential, but without these packages the examples below won’t work.</p>
<p>For anyone needing to learn about the tidyverse, you will in particular want to read up on <a href="https://dplyr.tidyverse.org/">dplyr</a> and <a href="https://magrittr.tidyverse.org/">the pipe operator <code>%&gt;%</code> from <code>magrittr</code></a>, which pipes output from one command into the next command.</p>
<p>In addition, we have included some helper functions for common manipulations of the data. These are available in the file <code>austraits.R</code> (TODO: where will users access this). To make these functions available, source the file prior to use:</p>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb33-1" title="1"><span class="kw"><a href="https://rdrr.io/r/base/source.html">source</a></span>(<span class="st">"R/austraits.R"</span>)</a></code></pre></div>
<div id="taking-subsets" class="section level2">
<h2 class="hasAnchor">
<a href="#taking-subsets" class="anchor"></a>Taking subsets</h2>
<p>Using the <code>dplyr</code>, it’s easy to subset by rows using <code>filter</code>. This approach can be used to filter the data to particular trait or study:</p>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb34-1" title="1">austraits<span class="op">$</span>traits <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb34-2" title="2"><span class="st">  </span><span class="kw"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span>(trait_name <span class="op">==</span><span class="st"> "leaf_area"</span>)</a></code></pre></div>
<pre><code>## # A tibble: 13,895 x 10
##    dataset_id species_name site_name observation_id trait_name value unit 
##    &lt;chr&gt;      &lt;chr&gt;        &lt;chr&gt;     &lt;chr&gt;          &lt;chr&gt;      &lt;chr&gt; &lt;chr&gt;
##  1 Ahrens_20… Corymbia ca… BOO       Ahrens_2019_0… leaf_area  1288… mm2  
##  2 Ahrens_20… Corymbia ca… BOO       Ahrens_2019_0… leaf_area  9647… mm2  
##  3 Ahrens_20… Corymbia ca… BOO       Ahrens_2019_0… leaf_area  1250… mm2  
##  4 Ahrens_20… Corymbia ca… BOO       Ahrens_2019_0… leaf_area  1519… mm2  
##  5 Ahrens_20… Corymbia ca… BOO       Ahrens_2019_0… leaf_area  1094… mm2  
##  6 Ahrens_20… Corymbia ca… BOO       Ahrens_2019_0… leaf_area  2129… mm2  
##  7 Ahrens_20… Corymbia ca… BOO       Ahrens_2019_0… leaf_area  1499… mm2  
##  8 Ahrens_20… Corymbia ca… BOO       Ahrens_2019_0… leaf_area  19767 mm2  
##  9 Ahrens_20… Corymbia ca… BOO       Ahrens_2019_0… leaf_area  2060… mm2  
## 10 Ahrens_20… Corymbia ca… BOO       Ahrens_2019_0… leaf_area  1338… mm2  
## # … with 13,885 more rows, and 3 more variables: value_type &lt;fct&gt;,
## #   replicates &lt;chr&gt;, original_name &lt;chr&gt;</code></pre>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb36-1" title="1">austraits<span class="op">$</span>traits <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb36-2" title="2"><span class="st">  </span><span class="kw"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span>(dataset_id <span class="op">==</span><span class="st"> "Westoby_2014"</span>)</a></code></pre></div>
<pre><code>## # A tibble: 3,888 x 10
##    dataset_id species_name site_name observation_id trait_name value unit 
##    &lt;chr&gt;      &lt;chr&gt;        &lt;chr&gt;     &lt;chr&gt;          &lt;chr&gt;      &lt;chr&gt; &lt;chr&gt;
##  1 Westoby_2… Acacia deal… Bothwell  Westoby_2014_… bark_thic… 0.21… mm   
##  2 Westoby_2… Acacia deal… Bothwell  Westoby_2014_… huber_val… 8.22… mm2_…
##  3 Westoby_2… Acacia deal… Bothwell  Westoby_2014_… leaf_area  6575  mm2  
##  4 Westoby_2… Acacia deal… Bothwell  Westoby_2014_… leaf_B_pe… 21.6… mg/kg
##  5 Westoby_2… Acacia deal… Bothwell  Westoby_2014_… leaf_C_pe… 523.2 mg/g 
##  6 Westoby_2… Acacia deal… Bothwell  Westoby_2014_… leaf_Ca_p… 3.31… mg/g 
##  7 Westoby_2… Acacia deal… Bothwell  Westoby_2014_… leaf_Cu_p… 13.1… mg/kg
##  8 Westoby_2… Acacia deal… Bothwell  Westoby_2014_… leaf_dark… 1.96  umol…
##  9 Westoby_2… Acacia deal… Bothwell  Westoby_2014_… leaf_dark… 0.02… umol…
## 10 Westoby_2… Acacia deal… Bothwell  Westoby_2014_… leaf_delt… -28.… per …
## # … with 3,878 more rows, and 3 more variables: value_type &lt;fct&gt;,
## #   replicates &lt;chr&gt;, original_name &lt;chr&gt;</code></pre>
<p><code>filter</code> can also be used to select observations for a specific string of text, such as a single genus:</p>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb38-1" title="1">austraits<span class="op">$</span>traits <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb38-2" title="2"><span class="st">  </span><span class="kw"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/grep.html">grepl</a></span>(<span class="st">"Daviesia"</span>, species_name))</a></code></pre></div>
<pre><code>## # A tibble: 3,804 x 10
##    dataset_id species_name site_name observation_id trait_name value unit 
##    &lt;chr&gt;      &lt;chr&gt;        &lt;chr&gt;     &lt;chr&gt;          &lt;chr&gt;      &lt;chr&gt; &lt;chr&gt;
##  1 ANBG_2019  Daviesia ab… site_at_… ANBG_2019_1362 seed_mass  2.569 mg   
##  2 ANBG_2019  Daviesia bu… site_at_… ANBG_2019_1363 dispersal… aril  &lt;NA&gt; 
##  3 ANBG_2019  Daviesia bu… site_at_… ANBG_2019_1363 seed_leng… 2.89  mm   
##  4 ANBG_2019  Daviesia bu… site_at_… ANBG_2019_1363 seed_shape reni… &lt;NA&gt; 
##  5 ANBG_2019  Daviesia bu… site_at_… ANBG_2019_1363 seed_text… smoo… &lt;NA&gt; 
##  6 ANBG_2019  Daviesia bu… site_at_… ANBG_2019_1363 seed_width 1.8   mm   
##  7 ANBG_2019  Daviesia co… site_at_… ANBG_2019_1364 seed_mass  7.77… mg   
##  8 ANBG_2019  Daviesia ep… site_at_… ANBG_2019_1365 seed_mass  69.6… mg   
##  9 ANBG_2019  Daviesia ge… site_at_… ANBG_2019_1366 seed_mass  11.6… mg   
## 10 ANBG_2019  Daviesia ge… site_at_… ANBG_2019_1367 dispersal… aril  &lt;NA&gt; 
## # … with 3,794 more rows, and 3 more variables: value_type &lt;fct&gt;,
## #   replicates &lt;chr&gt;, original_name &lt;chr&gt;</code></pre>
<p>We can select particular columns from the whole or filtered dataset using <code>select</code>:</p>
<div class="sourceCode" id="cb40"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb40-1" title="1">austraits<span class="op">$</span>traits <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb40-2" title="2"><span class="st">  </span><span class="kw">select</span>(species_name, observation_id, trait_name, value)</a></code></pre></div>
<pre><code>## # A tibble: 484,158 x 4
##    species_name        observation_id  trait_name          value      
##    &lt;chr&gt;               &lt;chr&gt;           &lt;chr&gt;               &lt;chr&gt;      
##  1 Corymbia calophylla Ahrens_2019_001 leaf_area           12883.7    
##  2 Corymbia calophylla Ahrens_2019_001 leaf_delta13C       -27.98     
##  3 Corymbia calophylla Ahrens_2019_001 leaf_N_per_dry_mass 9          
##  4 Corymbia calophylla Ahrens_2019_001 modified_NDVI       0.711535497
##  5 Corymbia calophylla Ahrens_2019_001 specific_leaf_area  5.4        
##  6 Corymbia calophylla Ahrens_2019_001 water_band_index    1.060028751
##  7 Corymbia calophylla Ahrens_2019_001 wood_density        0.600120627
##  8 Corymbia calophylla Ahrens_2019_002 leaf_area           9647.7     
##  9 Corymbia calophylla Ahrens_2019_002 leaf_delta13C       -27.33     
## 10 Corymbia calophylla Ahrens_2019_002 leaf_N_per_dry_mass 16         
## # … with 484,148 more rows</code></pre>
</div>
<div id="extracting-a-particular-trait-or-study" class="section level2">
<h2 class="hasAnchor">
<a href="#extracting-a-particular-trait-or-study" class="anchor"></a>Extracting a particular trait or study</h2>
<p>The file <code>austraits.R</code> also include pre-built functions for extracting a particular trait or study. The difference with the examples above, is that these functions also subset the other tables and elements of the data, not just the main data table. So we need to provide the entire <code>austraits</code> object and a <code>dataset_id</code>:</p>
<div class="sourceCode" id="cb42"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb42-1" title="1">traits_B14 &lt;-<span class="st"> </span><span class="kw">extract_dataset</span>(austraits, <span class="st">"Westoby_2014"</span>)</a></code></pre></div>
<p>As wiith the the main <code>austraits</code> object, this new dataset has multiple elements:</p>
<div class="sourceCode" id="cb43"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb43-1" title="1"><span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(traits_B14)</a></code></pre></div>
<pre><code>## [1] "traits"        "sites"         "methods"       "excluded_data"
## [5] "taxonomy"      "definitions"</code></pre>
<p>but the length of the datasets is smaller:</p>
<div class="sourceCode" id="cb45"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb45-1" title="1">count_rows &lt;-<span class="st"> </span><span class="cf">function</span>(obj) {<span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(obj<span class="op">$</span>traits), <span class="kw"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(obj<span class="op">$</span>sites), <span class="kw"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(obj<span class="op">$</span>taxonomy))}</a>
<a class="sourceLine" id="cb45-2" title="2"><span class="co"># main data</span></a>
<a class="sourceLine" id="cb45-3" title="3">austraits <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">count_rows</span>()</a></code></pre></div>
<pre><code>## [1] 484158  16136  22617</code></pre>
<div class="sourceCode" id="cb47"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb47-1" title="1"><span class="co"># subset</span></a>
<a class="sourceLine" id="cb47-2" title="2">traits_B14 <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">count_rows</span>()</a></code></pre></div>
<pre><code>## [1] 3888  188  127</code></pre>
<p>A similar function enables sub-setting to a particular trait, including the various other tables:</p>
<div class="sourceCode" id="cb49"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb49-1" title="1">traits_LS &lt;-<span class="st"> </span><span class="kw">extract_trait</span>(austraits, <span class="st">"leaf_width"</span>)</a></code></pre></div>
<pre><code>## Warning in `[.bibentry`(austraits$sources, keys): subscript out of bounds</code></pre>
<p>As with the the main <code>austraits</code> object, this new dataset has multiple elements:</p>
<div class="sourceCode" id="cb51"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb51-1" title="1"><span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(traits_LS)</a></code></pre></div>
<pre><code>## [1] "traits"        "sites"         "methods"       "excluded_data"
## [5] "taxonomy"      "definitions"   "contributors"  "sources"      
## [9] "build_info"</code></pre>
<p>but the length of the datasets is smaller:</p>
<div class="sourceCode" id="cb53"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb53-1" title="1"><span class="co"># main data</span></a>
<a class="sourceLine" id="cb53-2" title="2">austraits <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">count_rows</span>()</a></code></pre></div>
<pre><code>## [1] 484158  16136  22617</code></pre>
<div class="sourceCode" id="cb55"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb55-1" title="1"><span class="co"># subset</span></a>
<a class="sourceLine" id="cb55-2" title="2">traits_LS <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">count_rows</span>()</a></code></pre></div>
<pre><code>## [1] 36659   503 13684</code></pre>
</div>
<div id="merging-with-taxonomic-data" class="section level2">
<h2 class="hasAnchor">
<a href="#merging-with-taxonomic-data" class="anchor"></a>Merging with taxonomic data</h2>
<p>Limited taxonomic information is provided in the main table <code>data</code>, only <code>original_name</code> (the original name provided by the authors) and <code>species_name</code> (the aligned taxonomic name; see <a href="#taxonomy">species taxonomy</a> for details). But further information is available in the table <code>taxonomy</code>:</p>
<div class="sourceCode" id="cb57"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb57-1" title="1">austraits<span class="op">$</span>taxonomy</a></code></pre></div>
<pre><code>## # A tibble: 22,617 x 9
##    species_name  genus  family  authority  TPL_ID status APC_name APC_ID APNI_ID
##    &lt;chr&gt;         &lt;chr&gt;  &lt;chr&gt;   &lt;chr&gt;      &lt;chr&gt;  &lt;chr&gt;  &lt;chr&gt;    &lt;chr&gt;  &lt;chr&gt;  
##  1 Abelmoschus   Abelm… Malvac… &lt;NA&gt;       &lt;NA&gt;   &lt;NA&gt;   &lt;NA&gt;     &lt;NA&gt;   &lt;NA&gt;   
##  2 Abelmoschus … Abelm… Malvac… (L.) Wigh… kew-2… Accep… Abelmos… 28979… 55929  
##  3 Abelmoschus … Abelm… Malvac… (L.) Medi… kew-2… Accep… Abelmos… 29010… 55937  
##  4 Abelmoschus … Abelm… Malvac… Medik.     kew-2… Accep… Abelmos… 29005… 55953  
##  5 Abroma augus… Abroma Malvac… (L.) L.f.  kew-2… Accep… Abroma … 29048… 199881 
##  6 Abrophyllum … Abrop… Rousse… (F.Muell.… kew-2… Accep… Abrophy… 28936… 56085  
##  7 Abrotanella … Abrot… Compos… (Hook.f.)… gcc-9… Accep… unknown  unkno… unknown
##  8 Abrotanella … Abrot… Compos… (F.Muell.… gcc-3… Accep… Abrotan… 29005… 56120  
##  9 Abrotanella … Abrot… Compos… &lt;NA&gt;       &lt;NA&gt;   &lt;NA&gt;   &lt;NA&gt;     &lt;NA&gt;   &lt;NA&gt;   
## 10 Abrus precat… Abrus  Legumi… L.         ild-2… Accep… Abrus p… 29193… 56149  
## # … with 22,607 more rows</code></pre>
<p>You can merge this information in with the main dataset using the <code>species_name</code> as a an identifier linking both tables:</p>
<div class="sourceCode" id="cb59"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb59-1" title="1">traits2 &lt;-<span class="st"> </span>austraits<span class="op">$</span>traits <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb59-2" title="2"><span class="st">          </span><span class="kw">left_join</span>(<span class="dt">by=</span><span class="st">"species_name"</span>, austraits<span class="op">$</span>taxonomy)</a>
<a class="sourceLine" id="cb59-3" title="3"></a>
<a class="sourceLine" id="cb59-4" title="4"><span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(traits2)</a></code></pre></div>
<pre><code>##  [1] "dataset_id"     "species_name"   "site_name"      "observation_id"
##  [5] "trait_name"     "value"          "unit"           "value_type"    
##  [9] "replicates"     "original_name"  "genus"          "family"        
## [13] "authority"      "TPL_ID"         "status"         "APC_name"      
## [17] "APC_ID"         "APNI_ID"</code></pre>
<p>Similarly we could merge in only family information by using <code>select</code> to reduce the second table before merging:</p>
<div class="sourceCode" id="cb61"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb61-1" title="1">traits2 &lt;-<span class="st"> </span>austraits<span class="op">$</span>traits <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb61-2" title="2"><span class="st">          </span><span class="kw">left_join</span>(<span class="dt">by=</span><span class="st">"species_name"</span>, <span class="kw">select</span>(austraits<span class="op">$</span>taxonomy, species_name, family))</a>
<a class="sourceLine" id="cb61-3" title="3"></a>
<a class="sourceLine" id="cb61-4" title="4"><span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(traits2)</a></code></pre></div>
<pre><code>##  [1] "dataset_id"     "species_name"   "site_name"      "observation_id"
##  [5] "trait_name"     "value"          "unit"           "value_type"    
##  [9] "replicates"     "original_name"  "family"</code></pre>
</div>
<div id="merging-with-location-data" class="section level2">
<h2 class="hasAnchor">
<a href="#merging-with-location-data" class="anchor"></a>Merging with location data</h2>
<p>As with taxonomic information, limited information about location is provided in the main table <code>data</code>. But further information is available in the table <code>context</code>:</p>
<div class="sourceCode" id="cb63"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb63-1" title="1">austraits<span class="op">$</span>sites</a></code></pre></div>
<pre><code>## # A tibble: 16,136 x 4
##    dataset_id  site_name site_property   value     
##    &lt;chr&gt;       &lt;chr&gt;     &lt;chr&gt;           &lt;chr&gt;     
##  1 Ahrens_2019 BOO       latitude (deg)  -34.638944
##  2 Ahrens_2019 BOO       longitude (deg) 116.123806
##  3 Ahrens_2019 BOO       max temp (C)    25.6      
##  4 Ahrens_2019 BOO       MAP (mm)        1159      
##  5 Ahrens_2019 BOO       aridity index   1.05      
##  6 Ahrens_2019 BRA       latitude (deg)  -33.916389
##  7 Ahrens_2019 BRA       longitude (deg) 115.083278
##  8 Ahrens_2019 BRA       max temp (C)    26.1      
##  9 Ahrens_2019 BRA       MAP (mm)        1072      
## 10 Ahrens_2019 BRA       aridity index   0.96      
## # … with 16,126 more rows</code></pre>
<p>As with the main table of trait data, this information is in long format. Depending on the study, a variety of contextual information may have been provided, including</p>
<div class="sourceCode" id="cb65"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb65-1" title="1">austraits<span class="op">$</span>sites<span class="op">$</span>site_property <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/table.html">table</a></span>() <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/sort.html">sort</a></span>(<span class="dt">decreasing=</span><span class="ot">TRUE</span>) <span class="op">%&gt;%</span><span class="st"> </span>.[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>]</a></code></pre></div>
<pre><code>## .
##  latitude (deg) longitude (deg)       collector        locality     description 
##            3502            3502            2389            2389             347 
##         geology   elevation (m)            date           stage             age 
##             166             117             113             113             110</code></pre>
<p>In this example we’ll just focus on location details. Before merging in with the main data table let’s convert this data from long to wide format:</p>
<div class="sourceCode" id="cb67"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb67-1" title="1">sites &lt;-<span class="st"> </span>austraits<span class="op">$</span>sites <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb67-2" title="2"><span class="st">  </span><span class="kw"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span>(site_property <span class="op">%in%</span><span class="st">  </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"longitude (deg)"</span>,<span class="st">"latitude (deg)"</span>)) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb67-3" title="3"><span class="st">  </span><span class="kw">spread</span>(site_property, value)</a>
<a class="sourceLine" id="cb67-4" title="4">sites</a></code></pre></div>
<pre><code>## # A tibble: 3,502 x 4
##    dataset_id  site_name `latitude (deg)` `longitude (deg)`
##    &lt;chr&gt;       &lt;chr&gt;     &lt;chr&gt;            &lt;chr&gt;            
##  1 Ahrens_2019 BOO       -34.638944       116.123806       
##  2 Ahrens_2019 BRA       -33.916389       115.083278       
##  3 Ahrens_2019 CAR       -34.419639       115.821306       
##  4 Ahrens_2019 CHID      -31.8682         116.222973       
##  5 Ahrens_2019 CRI       -34.6015         118.742665       
##  6 Ahrens_2019 HRI       -30.3114         115.201584       
##  7 Ahrens_2019 KIN       -34.081167       116.330444       
##  8 Ahrens_2019 LUP       -32.520667       116.499          
##  9 Ahrens_2019 MOG       -31.0986         116.0508938      
## 10 Ahrens_2019 PEE       -32.684667       115.742667       
## # … with 3,492 more rows</code></pre>
<p>We can now merge in with the main data table using combinations of <code>dataset_id</code> and <code>site_name</code> as identifiers from both tables:</p>
<div class="sourceCode" id="cb69"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb69-1" title="1">traits_sites &lt;-<span class="st"> </span><span class="kw">left_join</span>(<span class="dt">by=</span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"dataset_id"</span>, <span class="st">"site_name"</span>), austraits<span class="op">$</span>traits, sites)</a>
<a class="sourceLine" id="cb69-2" title="2"></a>
<a class="sourceLine" id="cb69-3" title="3"><span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(traits_sites)</a></code></pre></div>
<pre><code>##  [1] "dataset_id"      "species_name"    "site_name"       "observation_id" 
##  [5] "trait_name"      "value"           "unit"            "value_type"     
##  [9] "replicates"      "original_name"   "latitude (deg)"  "longitude (deg)"</code></pre>
</div>
<div id="removing-suspected-duplicates" class="section level2">
<h2 class="hasAnchor">
<a href="#removing-suspected-duplicates" class="anchor"></a>Removing suspected duplicates</h2>
<p>AusTraits compiles data from many different sources and not all of these will be unique. Data from one study may be repeated in another. We have provided a workflow for removing suspected duplicate data via the function <code>remove_suspected_duplicates</code>. The function takes as an argument the entire <code>austraits</code> object and returns a modified object in which suspected duplicates have been moved from the main <code>traits</code> table to the <code>excluded_data</code> table, noting the <code>observartion_id</code> of the matching data in the error column. To use this function you mast pass in the entire <code>austraits</code> object:</p>
<div class="sourceCode" id="cb71"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb71-1" title="1">austraits_deduped &lt;-<span class="st"> </span><span class="kw">remove_suspected_duplicates</span>(austraits)</a></code></pre></div>
<p>Suspected duplicates are identified based on identical matches between the variables <code>species_name</code>, <code>trait_name</code>, and <code>value</code>. Priority is given to older sources, as indicated by the year in the <code>dataset_id</code>.</p>
<p>Checking the number of rows in original dataset and modified dataset we can see that 94141 records have been removed from <code>traits</code>:</p>
<div class="sourceCode" id="cb72"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb72-1" title="1"><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dt">traits =</span> <span class="kw"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(austraits<span class="op">$</span>traits), </a>
<a class="sourceLine" id="cb72-2" title="2">  <span class="dt">excluded =</span> <span class="kw"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(austraits<span class="op">$</span>excluded_data), </a>
<a class="sourceLine" id="cb72-3" title="3">  <span class="dt">total =</span> <span class="kw"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(austraits<span class="op">$</span>traits)<span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(austraits<span class="op">$</span>excluded_data))</a></code></pre></div>
<pre><code>##   traits excluded    total 
##   484158   242414   726572</code></pre>
<div class="sourceCode" id="cb74"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb74-1" title="1"><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dt">traits =</span> <span class="kw"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(austraits_deduped<span class="op">$</span>traits), </a>
<a class="sourceLine" id="cb74-2" title="2">    <span class="dt">excluded =</span> <span class="kw"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(austraits_deduped<span class="op">$</span>excluded_data), </a>
<a class="sourceLine" id="cb74-3" title="3">  <span class="dt">total =</span> <span class="kw"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(austraits_deduped<span class="op">$</span>traits)<span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(austraits_deduped<span class="op">$</span>excluded_data))</a></code></pre></div>
<pre><code>##   traits excluded    total 
##   390017   336555   726572</code></pre>
</div>
<div id="converting-from-long-to-wide-format" class="section level2">
<h2 class="hasAnchor">
<a href="#converting-from-long-to-wide-format" class="anchor"></a>Converting from long to wide format</h2>
<p>As discussed above, the main data table in <code>austraits</code> is distributed in long format. A long format has each measurement on a different row, and just a single column for all the value. By contrast, a dataset in wide format has data for each different trait in it’s own column. The reason we use long format is mainly for efficiency. We have data from 202 different traits, so in wide format we’d need at least this many columns. With many species, the file size gets very large very quickly. Other benefits of long format are that we can include columns for <code>units</code>, <code>value_type</code>, and <code>replicates</code>, which we could not in wide format (without created 3 more columns for each traits).</p>
<p>For many analyses, however, you’ll want to use wide format. We have therefore provided the ability to convert between the two formats. In wide format, each trait appears in it’s own column and measurements from the same observation (ideally individual, but in floras may be a species) appear on the same row.</p>
<p>As the entire austraits dataset is quite large, we recommend first sub-setting the data before spreading from long to wide format. You may also need to reduce the number of measurements in any one observation (see next section). For this first example we’ll use the dataset <code>Westoby_2014</code>.</p>
<p>First extract the trait data from the <code>Westoby_2014</code> study:</p>
<div class="sourceCode" id="cb76"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb76-1" title="1">data &lt;-<span class="st"> </span>austraits<span class="op">$</span>traits <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb76-2" title="2"><span class="st">  </span><span class="kw"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span>(dataset_id <span class="op">==</span><span class="st"> "Westoby_2014"</span>)</a>
<a class="sourceLine" id="cb76-3" title="3">data</a></code></pre></div>
<pre><code>## # A tibble: 3,888 x 10
##    dataset_id species_name site_name observation_id trait_name value unit 
##    &lt;chr&gt;      &lt;chr&gt;        &lt;chr&gt;     &lt;chr&gt;          &lt;chr&gt;      &lt;chr&gt; &lt;chr&gt;
##  1 Westoby_2… Acacia deal… Bothwell  Westoby_2014_… bark_thic… 0.21… mm   
##  2 Westoby_2… Acacia deal… Bothwell  Westoby_2014_… huber_val… 8.22… mm2_…
##  3 Westoby_2… Acacia deal… Bothwell  Westoby_2014_… leaf_area  6575  mm2  
##  4 Westoby_2… Acacia deal… Bothwell  Westoby_2014_… leaf_B_pe… 21.6… mg/kg
##  5 Westoby_2… Acacia deal… Bothwell  Westoby_2014_… leaf_C_pe… 523.2 mg/g 
##  6 Westoby_2… Acacia deal… Bothwell  Westoby_2014_… leaf_Ca_p… 3.31… mg/g 
##  7 Westoby_2… Acacia deal… Bothwell  Westoby_2014_… leaf_Cu_p… 13.1… mg/kg
##  8 Westoby_2… Acacia deal… Bothwell  Westoby_2014_… leaf_dark… 1.96  umol…
##  9 Westoby_2… Acacia deal… Bothwell  Westoby_2014_… leaf_dark… 0.02… umol…
## 10 Westoby_2… Acacia deal… Bothwell  Westoby_2014_… leaf_delt… -28.… per …
## # … with 3,878 more rows, and 3 more variables: value_type &lt;fct&gt;,
## #   replicates &lt;chr&gt;, original_name &lt;chr&gt;</code></pre>
<p>Now use the inbuilt function <code>spread_trait_data</code> to convert from long to wide:</p>
<div class="sourceCode" id="cb78"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb78-1" title="1">traits_spread &lt;-<span class="st"> </span><span class="kw">spread_trait_data</span>(data)</a></code></pre></div>
<p>This function spreads not only the data in the column <code>value</code>, but also other columns. The returned object is a list with a number of wide tables:</p>
<div class="sourceCode" id="cb79"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb79-1" title="1"><span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(traits_spread)</a></code></pre></div>
<pre><code>## [1] "value"      "unit"       "value_type" "replicates"</code></pre>
<p>Looking at each of these we see a table in wide format (i.e. with each trait being a column), as well as identifying columns at the start:</p>
<div class="sourceCode" id="cb81"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb81-1" title="1">traits_spread<span class="op">$</span>value</a></code></pre></div>
<pre><code>## # A tibble: 140 x 38
##    dataset_id species_name site_name observation_id original_name bark_thickness
##    &lt;chr&gt;      &lt;chr&gt;        &lt;chr&gt;     &lt;chr&gt;          &lt;chr&gt;         &lt;chr&gt;         
##  1 Westoby_2… Acacia aneu… Fowlers … Westoby_2014_… Acacia aneura &lt;NA&gt;          
##  2 Westoby_2… Acacia coll… Round Hi… Westoby_2014_… Acacia colle… 0.16476       
##  3 Westoby_2… Acacia cras… Cardwell  Westoby_2014_… Acacia crass… 0.23161       
##  4 Westoby_2… Acacia deal… Bothwell  Westoby_2014_… Acacia dealb… 0.21691       
##  5 Westoby_2… Acacia deal… Longley   Westoby_2014_… Acacia dealb… 0.18926       
##  6 Westoby_2… Acacia flav… Cardwell  Westoby_2014_… Acacia flave… 0.20286       
##  7 Westoby_2… Acacia flav… Princess… Westoby_2014_… Acacia flave… 0.2924        
##  8 Westoby_2… Acacia havi… Round Hi… Westoby_2014_… Acacia havil… 0.22115       
##  9 Westoby_2… Acacia holo… Claravil… Westoby_2014_… Acacia holos… 0.21568       
## 10 Westoby_2… Acacia lept… Princess… Westoby_2014_… Acacia lepto… 0.21386       
## # … with 130 more rows, and 32 more variables: huber_value &lt;chr&gt;,
## #   leaf_area &lt;chr&gt;, leaf_B_per_dry_mass &lt;chr&gt;, leaf_C_per_dry_mass &lt;chr&gt;,
## #   leaf_Ca_per_dry_mass &lt;chr&gt;, leaf_Cu_per_dry_mass &lt;chr&gt;,
## #   leaf_dark_respiration_per_area &lt;chr&gt;,
## #   leaf_dark_respiration_per_dry_mass &lt;chr&gt;, leaf_delta13C &lt;chr&gt;,
## #   leaf_Fe_per_dry_mass &lt;chr&gt;, leaf_K_per_dry_mass &lt;chr&gt;,
## #   leaf_Mg_per_dry_mass &lt;chr&gt;, leaf_Mn_per_dry_mass &lt;chr&gt;,
## #   leaf_Mo_per_dry_mass &lt;chr&gt;, leaf_N_per_dry_mass &lt;chr&gt;,
## #   leaf_Na_per_dry_mass &lt;chr&gt;, leaf_P_per_dry_mass &lt;chr&gt;,
## #   leaf_S_per_dry_mass &lt;chr&gt;, leaf_specific_conductivity &lt;chr&gt;,
## #   leaf_transpiration_at_Asat &lt;chr&gt;, leaf_Zn_per_dry_mass &lt;chr&gt;,
## #   modulus_of_elasticity &lt;chr&gt;, modulus_of_rupture &lt;chr&gt;,
## #   photosynthetic_rate_per_area_saturated &lt;chr&gt;,
## #   photosynthetic_rate_per_dry_mass_saturated &lt;chr&gt;, plant_height &lt;chr&gt;,
## #   sapwood_specific_conductivity &lt;chr&gt;, specific_leaf_area &lt;chr&gt;,
## #   water_potential_midday &lt;chr&gt;, water_potential_predawn &lt;chr&gt;,
## #   water_use_efficiency &lt;chr&gt;, wood_density &lt;chr&gt;</code></pre>
<p>The other tables have an identical format</p>
<div class="sourceCode" id="cb83"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb83-1" title="1">traits_spread<span class="op">$</span>unit</a></code></pre></div>
<pre><code>## # A tibble: 140 x 38
##    dataset_id species_name site_name observation_id original_name bark_thickness
##    &lt;chr&gt;      &lt;chr&gt;        &lt;chr&gt;     &lt;chr&gt;          &lt;chr&gt;         &lt;chr&gt;         
##  1 Westoby_2… Acacia aneu… Fowlers … Westoby_2014_… Acacia aneura &lt;NA&gt;          
##  2 Westoby_2… Acacia coll… Round Hi… Westoby_2014_… Acacia colle… mm            
##  3 Westoby_2… Acacia cras… Cardwell  Westoby_2014_… Acacia crass… mm            
##  4 Westoby_2… Acacia deal… Bothwell  Westoby_2014_… Acacia dealb… mm            
##  5 Westoby_2… Acacia deal… Longley   Westoby_2014_… Acacia dealb… mm            
##  6 Westoby_2… Acacia flav… Cardwell  Westoby_2014_… Acacia flave… mm            
##  7 Westoby_2… Acacia flav… Princess… Westoby_2014_… Acacia flave… mm            
##  8 Westoby_2… Acacia havi… Round Hi… Westoby_2014_… Acacia havil… mm            
##  9 Westoby_2… Acacia holo… Claravil… Westoby_2014_… Acacia holos… mm            
## 10 Westoby_2… Acacia lept… Princess… Westoby_2014_… Acacia lepto… mm            
## # … with 130 more rows, and 32 more variables: huber_value &lt;chr&gt;,
## #   leaf_area &lt;chr&gt;, leaf_B_per_dry_mass &lt;chr&gt;, leaf_C_per_dry_mass &lt;chr&gt;,
## #   leaf_Ca_per_dry_mass &lt;chr&gt;, leaf_Cu_per_dry_mass &lt;chr&gt;,
## #   leaf_dark_respiration_per_area &lt;chr&gt;,
## #   leaf_dark_respiration_per_dry_mass &lt;chr&gt;, leaf_delta13C &lt;chr&gt;,
## #   leaf_Fe_per_dry_mass &lt;chr&gt;, leaf_K_per_dry_mass &lt;chr&gt;,
## #   leaf_Mg_per_dry_mass &lt;chr&gt;, leaf_Mn_per_dry_mass &lt;chr&gt;,
## #   leaf_Mo_per_dry_mass &lt;chr&gt;, leaf_N_per_dry_mass &lt;chr&gt;,
## #   leaf_Na_per_dry_mass &lt;chr&gt;, leaf_P_per_dry_mass &lt;chr&gt;,
## #   leaf_S_per_dry_mass &lt;chr&gt;, leaf_specific_conductivity &lt;chr&gt;,
## #   leaf_transpiration_at_Asat &lt;chr&gt;, leaf_Zn_per_dry_mass &lt;chr&gt;,
## #   modulus_of_elasticity &lt;chr&gt;, modulus_of_rupture &lt;chr&gt;,
## #   photosynthetic_rate_per_area_saturated &lt;chr&gt;,
## #   photosynthetic_rate_per_dry_mass_saturated &lt;chr&gt;, plant_height &lt;chr&gt;,
## #   sapwood_specific_conductivity &lt;chr&gt;, specific_leaf_area &lt;chr&gt;,
## #   water_potential_midday &lt;chr&gt;, water_potential_predawn &lt;chr&gt;,
## #   water_use_efficiency &lt;chr&gt;, wood_density &lt;chr&gt;</code></pre>
<p>…etc</p>
<p>The new dataset can then be used to plot traits against each other, e.g.</p>
<div class="sourceCode" id="cb85"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb85-1" title="1">traits_spread<span class="op">$</span>value <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb85-2" title="2"><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">specific_leaf_area =</span> <span class="kw"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span>(specific_leaf_area), </a>
<a class="sourceLine" id="cb85-3" title="3">         <span class="dt">photosynthetic_rate_per_area =</span> <span class="kw"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span>(photosynthetic_rate_per_dry_mass_saturated)) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb85-4" title="4"><span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x=</span>specific_leaf_area, <span class="dt">y =</span> photosynthetic_rate_per_dry_mass_saturated, <span class="dt">colour =</span>site_name)) <span class="op">+</span></a>
<a class="sourceLine" id="cb85-5" title="5"><span class="st">  </span><span class="kw">geom_point</span>()</a></code></pre></div>
<p><img src="../docs/figures/documentation_unnamed-chunk-56-1.png" width="700"></p>
<p>There is also a function to convert back from wide to long format:</p>
<div class="sourceCode" id="cb86"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb86-1" title="1">traits2 &lt;-<span class="st"> </span><span class="kw">gather_trait_data</span>(traits_spread, austraits<span class="op">$</span>definitions)</a></code></pre></div>
<pre><code>## Note: Using an external vector in selections is ambiguous.
## ℹ Use `all_of(id_variables)` instead of `id_variables` to silence this message.
## ℹ See &lt;https://tidyselect.r-lib.org/reference/faq-external-vector.html&gt;.
## This message is displayed once per session.</code></pre>
<p>This manipulation should recover the original dataset.</p>
<p>Original:</p>
<div class="sourceCode" id="cb88"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb88-1" title="1">traits2</a></code></pre></div>
<pre><code>## # A tibble: 3,888 x 10
##    dataset_id species_name site_name observation_id trait_name value unit 
##    &lt;chr&gt;      &lt;chr&gt;        &lt;chr&gt;     &lt;chr&gt;          &lt;chr&gt;      &lt;chr&gt; &lt;chr&gt;
##  1 Westoby_2… Acacia deal… Bothwell  Westoby_2014_… bark_thic… 0.21… mm   
##  2 Westoby_2… Acacia deal… Bothwell  Westoby_2014_… huber_val… 8.22… mm2_…
##  3 Westoby_2… Acacia deal… Bothwell  Westoby_2014_… leaf_area  6575  mm2  
##  4 Westoby_2… Acacia deal… Bothwell  Westoby_2014_… leaf_B_pe… 21.6… mg/kg
##  5 Westoby_2… Acacia deal… Bothwell  Westoby_2014_… leaf_C_pe… 523.2 mg/g 
##  6 Westoby_2… Acacia deal… Bothwell  Westoby_2014_… leaf_Ca_p… 3.31… mg/g 
##  7 Westoby_2… Acacia deal… Bothwell  Westoby_2014_… leaf_Cu_p… 13.1… mg/kg
##  8 Westoby_2… Acacia deal… Bothwell  Westoby_2014_… leaf_dark… 1.96  umol…
##  9 Westoby_2… Acacia deal… Bothwell  Westoby_2014_… leaf_dark… 0.02… umol…
## 10 Westoby_2… Acacia deal… Bothwell  Westoby_2014_… leaf_delt… -28.… per …
## # … with 3,878 more rows, and 3 more variables: value_type &lt;chr&gt;,
## #   replicates &lt;chr&gt;, original_name &lt;chr&gt;</code></pre>
<p>After conversion</p>
<div class="sourceCode" id="cb90"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb90-1" title="1">data</a></code></pre></div>
<pre><code>## # A tibble: 3,888 x 10
##    dataset_id species_name site_name observation_id trait_name value unit 
##    &lt;chr&gt;      &lt;chr&gt;        &lt;chr&gt;     &lt;chr&gt;          &lt;chr&gt;      &lt;chr&gt; &lt;chr&gt;
##  1 Westoby_2… Acacia deal… Bothwell  Westoby_2014_… bark_thic… 0.21… mm   
##  2 Westoby_2… Acacia deal… Bothwell  Westoby_2014_… huber_val… 8.22… mm2_…
##  3 Westoby_2… Acacia deal… Bothwell  Westoby_2014_… leaf_area  6575  mm2  
##  4 Westoby_2… Acacia deal… Bothwell  Westoby_2014_… leaf_B_pe… 21.6… mg/kg
##  5 Westoby_2… Acacia deal… Bothwell  Westoby_2014_… leaf_C_pe… 523.2 mg/g 
##  6 Westoby_2… Acacia deal… Bothwell  Westoby_2014_… leaf_Ca_p… 3.31… mg/g 
##  7 Westoby_2… Acacia deal… Bothwell  Westoby_2014_… leaf_Cu_p… 13.1… mg/kg
##  8 Westoby_2… Acacia deal… Bothwell  Westoby_2014_… leaf_dark… 1.96  umol…
##  9 Westoby_2… Acacia deal… Bothwell  Westoby_2014_… leaf_dark… 0.02… umol…
## 10 Westoby_2… Acacia deal… Bothwell  Westoby_2014_… leaf_delt… -28.… per …
## # … with 3,878 more rows, and 3 more variables: value_type &lt;fct&gt;,
## #   replicates &lt;chr&gt;, original_name &lt;chr&gt;</code></pre>
<p>Just to be sure we can run a check:</p>
<div class="sourceCode" id="cb92"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb92-1" title="1"><span class="kw"><a href="https://rdrr.io/r/base/all.equal.html">all.equal</a></span>(data, traits2)</a></code></pre></div>
<pre><code>## [1] "Incompatible type for column `value_type`: x factor, y character"</code></pre>
<p>Note that the function <code>gather_trait_data</code> also needs the element <code>austraits$definitions</code> to run successfully.</p>
<div id="datasets-with-multiple-measurements-per-observation" class="section level3">
<h3 class="hasAnchor">
<a href="#datasets-with-multiple-measurements-per-observation" class="anchor"></a>Datasets with multiple measurements per observation</h3>
<p>The above example works easily because there is only a single measurement of each trait per <code>observation_id</code>. But this is not always the case. Many datasets include multiple records per <code>observation_id</code>, for example studies reporting multiple <code>value_types</code> per <code>observation_id</code>. An example is the <code>AusGrass_2014</code> datasets:</p>
<div class="sourceCode" id="cb94"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb94-1" title="1">traits &lt;-<span class="st"> </span>austraits<span class="op">$</span>traits <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb94-2" title="2"><span class="st">  </span><span class="kw"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span>(dataset_id <span class="op">==</span><span class="st"> "GrassBase_2014"</span>)</a>
<a class="sourceLine" id="cb94-3" title="3">traits</a></code></pre></div>
<pre><code>## # A tibble: 6,678 x 10
##    dataset_id species_name site_name observation_id trait_name value unit 
##    &lt;chr&gt;      &lt;chr&gt;        &lt;chr&gt;     &lt;chr&gt;          &lt;chr&gt;      &lt;chr&gt; &lt;chr&gt;
##  1 GrassBase… Acrachne ra… &lt;NA&gt;      GrassBase_201… leaf_leng… 130   mm   
##  2 GrassBase… Acrachne ra… &lt;NA&gt;      GrassBase_201… leaf_leng… 750   mm   
##  3 GrassBase… Acrachne ra… &lt;NA&gt;      GrassBase_201… leaf_width 3     mm   
##  4 GrassBase… Acrachne ra… &lt;NA&gt;      GrassBase_201… leaf_width 12    mm   
##  5 GrassBase… Acrachne ra… &lt;NA&gt;      GrassBase_201… plant_hei… 1     m    
##  6 GrassBase… Acrachne ra… &lt;NA&gt;      GrassBase_201… seed_leng… 1.1   mm   
##  7 GrassBase… Acrachne ra… &lt;NA&gt;      GrassBase_201… seed_leng… 1.1   mm   
##  8 GrassBase… Agrostis au… &lt;NA&gt;      GrassBase_201… leaf_leng… 10    mm   
##  9 GrassBase… Agrostis au… &lt;NA&gt;      GrassBase_201… leaf_leng… 50    mm   
## 10 GrassBase… Agrostis au… &lt;NA&gt;      GrassBase_201… leaf_width 0.4   mm   
## # … with 6,668 more rows, and 3 more variables: value_type &lt;fct&gt;,
## #   replicates &lt;chr&gt;, original_name &lt;chr&gt;</code></pre>
<p>Note the dataset includes multiple value_types: <code>expert_min</code> and <code>expert_max</code>. If you try running <code>spread_trait_data</code> on this data it will fail:</p>
<div class="sourceCode" id="cb96"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb96-1" title="1">traits_spread &lt;-<span class="st"> </span><span class="kw">spread_trait_data</span>(traits)</a></code></pre></div>
<pre><code>## Error: Each row of output must be identified by a unique combination of keys. Keys are shared for 6452 rows .....</code></pre>
<p>So before converting from long to wide we use an additional function called <code>bind_trait_values</code> to collapse all the different measurements for a given <code>trait_name</code> and <code>observation_id</code> into a single cell:</p>
<div class="sourceCode" id="cb98"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb98-1" title="1">traits_bind &lt;-<span class="st"> </span><span class="kw">bind_trait_values</span>(traits)</a></code></pre></div>
<p>This function binds data from multiple rows together, separating with the string `<code>--</code>. This occurs for columns <code>value</code>, <code>value_type</code>, and <code>replicates</code>. Looking at the table we can see the bound values:</p>
<div class="sourceCode" id="cb99"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb99-1" title="1">traits_bind <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb99-2" title="2"><span class="st">  </span><span class="kw"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/grep.html">grepl</a></span>(<span class="st">"--"</span>, value_type)) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb99-3" title="3"><span class="st">  </span><span class="kw">select</span>(<span class="op">-</span>dataset_id, <span class="op">-</span>site_name, <span class="op">-</span>original_name)</a></code></pre></div>
<pre><code>## # A tibble: 3,226 x 7
##    species_name   observation_id  trait_name  value unit  value_type  replicates
##    &lt;chr&gt;          &lt;chr&gt;           &lt;chr&gt;       &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;       &lt;chr&gt;     
##  1 Acrachne race… GrassBase_2014… leaf_length 130-… mm    expert_min… NA--NA    
##  2 Acrachne race… GrassBase_2014… leaf_width  3--12 mm    expert_min… NA--NA    
##  3 Acrachne race… GrassBase_2014… seed_length 1.1-… mm    expert_min… NA--NA    
##  4 Agrostis aust… GrassBase_2014… leaf_length 10--… mm    expert_min… NA--NA    
##  5 Agrostis aust… GrassBase_2014… leaf_width  0.4-… mm    expert_min… NA--NA    
##  6 Agrostis aust… GrassBase_2014… seed_length 2--2  mm    expert_min… NA--NA    
##  7 Agrostis bett… GrassBase_2014… leaf_width  1.5-… mm    expert_min… NA--NA    
##  8 Agrostis bett… GrassBase_2014… seed_length 2--2  mm    expert_min… NA--NA    
##  9 Agrostis diem… GrassBase_2014… leaf_length 120-… mm    expert_min… NA--NA    
## 10 Agrostis diem… GrassBase_2014… leaf_width  2--2  mm    expert_min… NA--NA    
## # … with 3,216 more rows</code></pre>
<p>Using this new dataset we can now go from long to wide:</p>
<div class="sourceCode" id="cb101"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb101-1" title="1">traits_spread &lt;-<span class="st"> </span><span class="kw">spread_trait_data</span>(traits_bind)</a></code></pre></div>
<p>The bound data values are carried over to the new dataset:</p>
<div class="sourceCode" id="cb102"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb102-1" title="1">traits_spread<span class="op">$</span>value</a></code></pre></div>
<pre><code>## # A tibble: 1,054 x 9
##    dataset_id species_name site_name observation_id original_name leaf_length
##    &lt;chr&gt;      &lt;chr&gt;        &lt;chr&gt;     &lt;chr&gt;          &lt;chr&gt;         &lt;chr&gt;      
##  1 GrassBase… Acrachne ra… &lt;NA&gt;      GrassBase_201… Acrachne rac… 130--750   
##  2 GrassBase… Agropyron b… &lt;NA&gt;      GrassBase_201… Australopyru… 50--160    
##  3 GrassBase… Agrostis ad… &lt;NA&gt;      GrassBase_201… Lachnagrosti… 80--250    
##  4 GrassBase… Agrostis au… &lt;NA&gt;      GrassBase_201… Agrostis aus… 10--50     
##  5 GrassBase… Agrostis be… &lt;NA&gt;      GrassBase_201… Agrostis bet… &lt;NA&gt;       
##  6 GrassBase… Agrostis bi… &lt;NA&gt;      GrassBase_201… Lachnagrosti… 50--150    
##  7 GrassBase… Agrostis di… &lt;NA&gt;      GrassBase_201… Agrostis die… 120--120   
##  8 GrassBase… Agrostis dr… &lt;NA&gt;      GrassBase_201… Lachnagrosti… 30--150    
##  9 GrassBase… Agrostis jo… &lt;NA&gt;      GrassBase_201… Agrostis joy… &lt;NA&gt;       
## 10 GrassBase… Agrostis mu… &lt;NA&gt;      GrassBase_201… Agrostis mue… 30--100    
## # … with 1,044 more rows, and 3 more variables: leaf_width &lt;chr&gt;,
## #   plant_height &lt;chr&gt;, seed_length &lt;chr&gt;</code></pre>
<p>As previously, we can also convert back again:</p>
<div class="sourceCode" id="cb104"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb104-1" title="1">traits_bind2 &lt;-<span class="st"> </span><span class="kw">gather_trait_data</span>(traits_spread, austraits<span class="op">$</span>definitions)</a></code></pre></div>
<p>Thsi gets us back to the intermediate dataset</p>
<div class="sourceCode" id="cb105"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb105-1" title="1"><span class="kw"><a href="https://rdrr.io/r/base/all.equal.html">all.equal</a></span>(traits_bind, traits_bind2)</a></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<p>And then finally we can separate out the values that have been bound together:</p>
<div class="sourceCode" id="cb107"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb107-1" title="1">traits2 &lt;-<span class="st"> </span><span class="kw">separate_trait_values</span>(traits_bind2, austraits<span class="op">$</span>definitions)</a></code></pre></div>
<p>We should now have a dataset identical to our original. Just to be sure we can run a check:</p>
<div class="sourceCode" id="cb108"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb108-1" title="1"><span class="kw"><a href="https://rdrr.io/r/base/all.equal.html">all.equal</a></span>(traits, traits2)</a></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<p>Using pipes we can put all these steps together to short the full range of steps:</p>
<div class="sourceCode" id="cb110"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb110-1" title="1">traits <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb110-2" title="2"><span class="st"> </span><span class="kw">bind_trait_values</span>() <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb110-3" title="3"><span class="st"> </span><span class="kw">spread_trait_data</span>() <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb110-4" title="4"><span class="st"> </span><span class="kw">gather_trait_data</span>() <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb110-5" title="5"><span class="st"> </span><span class="kw">separate_trait_values</span>(austraits<span class="op">$</span>definitions) -&gt;<span class="st"> </span>traits2</a>
<a class="sourceLine" id="cb110-6" title="6"></a>
<a class="sourceLine" id="cb110-7" title="7"><span class="kw"><a href="https://rdrr.io/r/base/all.equal.html">all.equal</a></span>(traits, traits2)</a></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
</div>
</div>
<div id="categorical-vs-numerical-traits" class="section level2">
<h2 class="hasAnchor">
<a href="#categorical-vs-numerical-traits" class="anchor"></a>Categorical vs numerical traits</h2>
<p>List of categorical traits:</p>
<div class="sourceCode" id="cb112"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb112-1" title="1">traits_all &lt;-<span class="st"> </span>austraits<span class="op">$</span>traits<span class="op">$</span>trait_name <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/unique.html">unique</a></span>()  <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/sort.html">sort</a></span>()</a>
<a class="sourceLine" id="cb112-2" title="2"></a>
<a class="sourceLine" id="cb112-3" title="3">traits_all[<span class="kw">trait_is_categorical</span>(traits_all, austraits<span class="op">$</span>definitions)]</a></code></pre></div>
<pre><code>##  [1] "cotyledon_position"      "cotyledon_type"         
##  [3] "dispersal_appendage"     "dispersal_syndrome"     
##  [5] "dormancy_type"           "embryo_colour"          
##  [7] "fire_response"           "fire_response_detailed" 
##  [9] "flower_colour"           "flowering_time"         
## [11] "fruit_type"              "fruit_type_botany"      
## [13] "fruit_type_function"     "fruiting_time"          
## [15] "germination_treatment"   "glaucous"               
## [17] "growth_habit"            "hypocotyl_type"         
## [19] "leaf_arrangement"        "leaf_compoundness"      
## [21] "leaf_hairs_adult"        "leaf_hairs_juvenile"    
## [23] "leaf_margin"             "leaf_phenology"         
## [25] "leaf_shape"              "leaf_spines_adult"      
## [27] "leaf_type"               "life_form"              
## [29] "life_history"            "lifespan"               
## [31] "nitrogen_fixing"         "photosynthetic_pathway" 
## [33] "plant_growth_form"       "regen_strategy"         
## [35] "root_morphology"         "root_structure"         
## [37] "seed_shape"              "seed_storage_location"  
## [39] "seed_texture"            "seedling_first_leaf"    
## [41] "seedling_first_node"     "serotiny"               
## [43] "tap_root"                "vegetative_regeneration"
## [45] "water_tolerance"         "woodiness"</code></pre>
<p>list of numerical traits:</p>
<div class="sourceCode" id="cb114"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb114-1" title="1">traits_all[<span class="kw">trait_is_numeric</span>(traits_all, austraits<span class="op">$</span>definitions)]</a></code></pre></div>
<pre><code>##   [1] "accessory_cost_fraction"                                
##   [2] "accessory_cost_mass"                                    
##   [3] "bark_C_per_dry_mass"                                    
##   [4] "bark_delta13C"                                          
##   [5] "bark_delta15N"                                          
##   [6] "bark_mass_area"                                         
##   [7] "bark_N_per_dry_mass"                                    
##   [8] "bark_saturated_water_content_per_mass"                  
##   [9] "bark_thickness"                                         
##  [10] "basal_diameter"                                         
##  [11] "branch_mass_fraction"                                   
##  [12] "ca"                                                     
##  [13] "cell_epidermis_Ca_per_fresh_mass"                       
##  [14] "cell_epidermis_P_per_fresh_mass"                        
##  [15] "cell_hypodermis_Ca_per_fresh_mass"                      
##  [16] "cell_hypodermis_P_per_fresh_mass"                       
##  [17] "cell_internal_parenchyma_Ca_per_fresh_mass"             
##  [18] "cell_internal_parenchyma_P_per_fresh_mass"              
##  [19] "cell_palisade_mesophyll_Ca_per_fresh_mass"              
##  [20] "cell_palisade_mesophyll_P_per_fresh_mass"               
##  [21] "cell_sclerenchyma_Ca_per_fresh_mass"                    
##  [22] "cell_sclerenchyma_P_per_fresh_mass"                     
##  [23] "cell_spongy_mesophyll_Ca_per_fresh_mass"                
##  [24] "cell_spongy_mesophyll_P_per_fresh_mass"                 
##  [25] "ci_at_Asat"                                             
##  [26] "diaspore_mass"                                          
##  [27] "fruit_length"                                           
##  [28] "fruit_width"                                            
##  [29] "fv_over_fm"                                             
##  [30] "genome_size"                                            
##  [31] "huber_value"                                            
##  [32] "leaf_Al_per_dry_mass"                                   
##  [33] "leaf_area"                                              
##  [34] "leaf_area_ratio"                                        
##  [35] "leaf_B_per_dry_mass"                                    
##  [36] "leaf_C_per_dry_mass"                                    
##  [37] "leaf_Ca_per_dry_mass"                                   
##  [38] "leaf_cell_wall_fraction"                                
##  [39] "leaf_cell_wall_N"                                       
##  [40] "leaf_cell_wall_N_fraction"                              
##  [41] "leaf_CN_ratio"                                          
##  [42] "leaf_Cu_per_dry_mass"                                   
##  [43] "leaf_dark_respiration_per_area"                         
##  [44] "leaf_dark_respiration_per_dry_mass"                     
##  [45] "leaf_delta13C"                                          
##  [46] "leaf_delta15N"                                          
##  [47] "leaf_dry_mass"                                          
##  [48] "leaf_dry_matter_content"                                
##  [49] "leaf_Fe_per_dry_mass"                                   
##  [50] "leaf_fracture_force"                                    
##  [51] "leaf_fresh_mass"                                        
##  [52] "leaf_fresh_mass_per_area"                               
##  [53] "leaf_hydraulic_conductivity"                            
##  [54] "leaf_hydraulic_vulnerability"                           
##  [55] "leaf_K_per_area"                                        
##  [56] "leaf_K_per_dry_mass"                                    
##  [57] "leaf_length"                                            
##  [58] "leaf_lifespan"                                          
##  [59] "leaf_light_respiration_per_area"                        
##  [60] "leaf_mass_fraction"                                     
##  [61] "leaf_mass_to_stem_mass"                                 
##  [62] "leaf_Mg_per_dry_mass"                                   
##  [63] "leaf_Mn_per_dry_mass"                                   
##  [64] "leaf_Mo_per_dry_mass"                                   
##  [65] "leaf_N_per_area"                                        
##  [66] "leaf_N_per_dry_mass"                                    
##  [67] "leaf_Na_per_dry_mass"                                   
##  [68] "leaf_P_per_area"                                        
##  [69] "leaf_P_per_dry_mass"                                    
##  [70] "leaf_photosynthetic_nitrogen_use_efficiency"            
##  [71] "leaf_photosynthetic_nitrogen_use_efficiency_maximum"    
##  [72] "leaf_photosynthetic_nitrogen_use_efficiency_saturated"  
##  [73] "leaf_photosynthetic_phosphorus_use_efficiency_maximum"  
##  [74] "leaf_photosynthetic_phosphorus_use_efficiency_saturated"
##  [75] "leaf_photosynthetic_water_use_efficiency"               
##  [76] "leaf_S_per_dry_mass"                                    
##  [77] "leaf_saturated_water_content_per_mass"                  
##  [78] "leaf_specific_conductivity"                             
##  [79] "leaf_thickness"                                         
##  [80] "leaf_toughness"                                         
##  [81] "leaf_transpiration_at_Amax"                             
##  [82] "leaf_transpiration_at_Asat"                             
##  [83] "leaf_turgor_loss_point"                                 
##  [84] "leaf_water_content_per_area"                            
##  [85] "leaf_water_content_per_mass"                            
##  [86] "leaf_width"                                             
##  [87] "leaf_Zn_per_dry_mass"                                   
##  [88] "lower_cuticle_thickness"                                
##  [89] "lower_epidermis_thickness"                              
##  [90] "lower_hypodermis_thickness"                             
##  [91] "lower_palisade_cell_thickness"                          
##  [92] "modified_NDVI"                                          
##  [93] "modulus_of_elasticity"                                  
##  [94] "modulus_of_rupture"                                     
##  [95] "N_to_P_ratio"                                           
##  [96] "osmotic_potential"                                      
##  [97] "photosynthetic_rate_per_area"                           
##  [98] "photosynthetic_rate_per_area_maximum"                   
##  [99] "photosynthetic_rate_per_area_saturated"                 
## [100] "photosynthetic_rate_per_dry_mass_maximum"               
## [101] "photosynthetic_rate_per_dry_mass_saturated"             
## [102] "plant_height"                                           
## [103] "plant_width"                                            
## [104] "resorption_leaf_N"                                      
## [105] "resorption_leaf_P"                                      
## [106] "root_wood_density"                                      
## [107] "sapwood_specific_conductivity"                          
## [108] "seed_breadth"                                           
## [109] "seed_length"                                            
## [110] "seed_mass"                                              
## [111] "seed_mass_reserve"                                      
## [112] "seed_oil_content"                                       
## [113] "seed_P_concentration"                                   
## [114] "seed_protein_content"                                   
## [115] "seed_viability"                                         
## [116] "seed_volume"                                            
## [117] "seed_width"                                             
## [118] "senesced_leaf_Ca_per_dry_mass"                          
## [119] "senesced_leaf_Cu_per_dry_mass"                          
## [120] "senesced_leaf_Fe_per_dry_mass"                          
## [121] "senesced_leaf_K_per_dry_mass"                           
## [122] "senesced_leaf_Mg_per_dry_mass"                          
## [123] "senesced_leaf_Mn_per_dry_mass"                          
## [124] "senesced_leaf_Mo_per_dry_mass"                          
## [125] "senesced_leaf_N_per_dry_mass"                           
## [126] "senesced_leaf_P_per_dry_mass"                           
## [127] "senesced_leaf_S_per_dry_mass"                           
## [128] "senesced_leaf_Zn_per_dry_mass"                          
## [129] "soil_salinity_tolerance"                                
## [130] "specific_leaf_area"                                     
## [131] "specific_root_length"                                   
## [132] "spongy_mesophyll_thickness"                             
## [133] "stem_C_per_dry_mass"                                    
## [134] "stem_dry_matter_content"                                
## [135] "stem_hydraulic_conductivity"                            
## [136] "stem_N_per_dry_mass"                                    
## [137] "stem_saturated_water_content_per_mass"                  
## [138] "stomatal_conductance_per_area"                          
## [139] "stomatal_conductance_per_area_at_Amax"                  
## [140] "stomatal_conductance_per_area_at_Asat"                  
## [141] "thickest_root_diameter"                                 
## [142] "twig_area"                                              
## [143] "twig_length"                                            
## [144] "upper_cuticle_thickness"                                
## [145] "upper_epidermis_thickness"                              
## [146] "upper_hypodermis_thickness"                             
## [147] "upper_palisade_cell_thickness"                          
## [148] "water_band_index"                                       
## [149] "water_potential_midday"                                 
## [150] "water_potential_predawn"                                
## [151] "water_use_efficiency"                                   
## [152] "wood_C_per_dry_mass"                                    
## [153] "wood_delta13C"                                          
## [154] "wood_delta15N"                                          
## [155] "wood_density"                                           
## [156] "wood_N_per_dry_mass"</code></pre>
</div>
</div>
<div id="appendices" class="section level1">
<h1 class="hasAnchor">
<a href="#appendices" class="anchor"></a>Appendices</h1>
<div id="acknowledgements" class="section level2">
<h2 class="hasAnchor">
<a href="#acknowledgements" class="anchor"></a>Acknowledgements</h2>
</div>
<div id="primary-sources" class="section level2">
<h2 class="hasAnchor">
<a href="#primary-sources" class="anchor"></a>Primary sources</h2>
<p>AusTraits is built from the following primary and secondary sources:</p>
<ul>
<li>C. W. Ahrens, M. E. Andrew, R. A. Mazanec, K. X. Ruthrof, A. Challis, G. Hardy, M. Byrne, D. T. Tissue, and P. D. Rymer. “Plant functional traits differ in adaptability and are predicted to be differentially affected by climate change”. <em>Ecology and Evolution</em> (2019). doi: <a href="https://doi.org/10.1002%2Fece3.5890">10.1002/ece3.5890</a>.</li>
<li>Australian National Botanic Garden. <em>The National Seed Bank</em>. 2018. url: <a href="http://www.anbg.gov.au/gardens/living/seedbank/">http://www.anbg.gov.au/gardens/living/seedbank/</a>.</li>
<li>T. Angevin. “Unpublished data: Trait data collected via Honours project at La Trobe University”. 2010.</li>
<li>B. A. Barlow, H. T. Clifford, A. S. George, and A. K. A. McCusker. <em>Flora of Australia (online)</em>. 1981. url: <a href="http://www.environment.gov.au/biodiversity/abrs/online-resources/flora/main/">http://www.environment.gov.au/biodiversity/abrs/online-resources/flora/main/</a>.</li>
<li>A. Bean. “A revision of Baeckea (Myrtaceae) in eastern Australia, Malesia and south-east Asia”. <em>Telopea</em> 7 (1997), pp. 245-268. doi: <a href="https://doi.org/10.1.1.520.6315%26rep%3Drep1%26type%3Dpdf">10.1.1.520.6315&amp;rep=rep1&amp;type=pdf</a>.</li>
<li>C. J. Blackman, T. J. Brodribb, and G. J. Jordan. “Leaf hydraulic vulnerability is related to conduit dimensions and drought resistance across a diverse range of woody angiosperms”. <em>New Phytologist</em> 188.4 (2010), pp. 1113-1123. doi: <a href="https://doi.org/10.1111%2Fj.1469-8137.2010.03439.x">10.1111/j.1469-8137.2010.03439.x</a>.</li>
<li>C. J. Blackman, S. M. Gleason, Y. Chang, A. M. Cook, C. Laws, and M. Westoby. “Leaf hydraulic vulnerability to drought is linked to site water availability across a broad range of species and climates”. <em>Annals of Botany</em> 114 (2014), pp. 435-440. doi: <a href="https://doi.org/10.1093%2Faob%2Fmcu131">10.1093/aob/mcu131</a>.</li>
<li>E. Bolza. <em>Properties and uses of 175 timber species from Papua New Guinea and West Irian</em>. Victoria (Australia) CSIRO, Div. of Building Research, 1975.</li>
<li>J. G. Bragg and M. Westoby. “Leaf size and foraging for light in a sclerophyll woodland”. <em>Functional Ecology</em> 16 (2002), pp. 633-639. doi: <a href="https://doi.org/10.1046%2Fj.1365-2435.2002.00661.x">10.1046/j.1365-2435.2002.00661.x</a>.</li>
<li>Brisbane Rainforest Action &amp; Information Network. <em>Trait measurements for Australian rainforest species</em>. 2016. url: <a href="http://www.brisrain.org.au/">http://www.brisrain.org.au/</a>.</li>
<li>A. L. Briggs and J. W. Morgan. “Seed characteristics and soil surface patch type interact to affect germination of semi-arid woodland species”. <em>Plant Ecol</em> 212 (2010), pp. 91-103. doi: <a href="https://doi.org/10.1007%2Fs11258-010-9806-x">10.1007/s11258-010-9806-x</a>.</li>
<li>J. Brock and A. Dunlop. <em>Native plants of northern Australia</em>. Frenchs Forest, N.S.W.: Reed New Holland, 1993. ISBN: 9781877069246.</li>
<li>G. E. Burrows. “Comparative anatomy of the photosynthetic organs of 39 xeromorphic species from subhumid New South Wales, Australia”. <em>International Journal of Plant Sciences</em> 162 (2001), pp. 411-430. doi: <a href="https://doi.org/10.1086%2F319579">10.1086/319579</a>.</li>
<li>D. W. Butler, S. M. Gleason, I. Davidson, Y. Onoda, and M. Westoby. “Safety and streamlining of woody shoots in wind: an empirical study across 39 species in tropical Australia”. <em>New Phytologist</em> 193.1 (2011), pp. 137-149. doi: <a href="https://doi.org/10.1111%2Fj.1469-8137.2011.03887.x">10.1111/j.1469-8137.2011.03887.x</a>.</li>
<li>CAB International. <em>Forestry Compendium</em>. 2009. url: <a href="http://www.cabi.org/fc/">http://www.cabi.org/fc/</a>.</li>
<li>J. A. Catford, B. J. Downes, C. J. Gippel, and P. A. Vesk. “Flow regulation reduces native plant cover and facilitates exotic invasion in riparian wetlands”. <em>Journal of Applied Ecology</em> 48.2 (2011), pp. 432-442. doi: <a href="https://doi.org/10.1111%2Fj.1365-2664.2010.01945.x">10.1111/j.1365-2664.2010.01945.x</a>.</li>
<li>J. A. Catford, W. K. Morris, P. A. Vesk, C. J. Gippel, and B. J. Downes. “Species and environmental characteristics point to flow regulation and drought as drivers of riparian plant invasion”. <em>Diversity and Distributions</em> 20 (2014), pp. 1084-1096. doi: <a href="https://doi.org/10.1111%2Fddi.12225">10.1111/ddi.12225</a>.</li>
<li>G. T. Chandler, M. D. Crisp, L. W. Cayzer, and R. J. Bayer. “Monograph of Gastrolobium (Fabaceae: Mirbelieae)”. <em>Australian Systematic Botany</em> 15 (2002), p. 619. doi: <a href="https://doi.org/10.1071%2Fsb01010">10.1071/sb01010</a>.</li>
<li>J. Chave, D. Coomes, S. Jansen, S. L. Lewis, N. G. Swenson, and A. E. Zanne. “Towards a worldwide wood economics spectrum”. <em>Ecology Letters</em> 12 (2009), pp. 351-366. doi: <a href="https://doi.org/10.1111%2Fj.1461-0248.2009.01285.x">10.1111/j.1461-0248.2009.01285.x</a>.</li>
<li>Chen, S.-C., Cornwell, W., Z. H. &amp;. Moles, and A. T. “Plants show more flesh in the tropics: variation in fruit type along latitudinal and climatic gradients.”. <em>Ecography</em> 40 (2017), pp. 531-538. doi: <a href="https://doi.org/abs%2F10.1111%2Fecog.02010">abs/10.1111/ecog.02010</a>.</li>
<li>R. J. Chinnock. <em>Eremophila and allied genera: A monograph of the plant family Myoporaceae</em>. Dural, NSW: Rosenberg, 2007. ISBN: 9781877058165.</li>
<li>M. Chudnoff. <em>Tropical timbers of the world</em>. US Department of Agriculture, Forest Service, 1984, p. 472. ISBN: 3935638825.</li>
<li>The French agricultural research &amp; international cooperation organization (CIRAD). <em>Wood density data</em>. 2009. url: <a href="http://www.cirad.fr/">http://www.cirad.fr/</a>.</li>
<li>P. J. Clarke, M. J. Lawes, B. P. Murphy, J. Russell-Smith, C. E. Nano, R. Bradstock, N. J. Enright, J. B. Fontaine, C. R. Gosper, I. Radford, J. J. Midgley, and R. M. Gunton. “A synthesis of postfire recovery traits of woody plants in Australian ecosystems”. <em>Science of The Total Environment</em> 534 (2015), pp. 31-42. doi: <a href="https://doi.org/10.1016%2Fj.scitotenv.2015.04.002">10.1016/j.scitotenv.2015.04.002</a>.</li>
<li>W. Cooper and W. T. Cooper. <em>Fruits of the Australian tropical rainforest</em>. Nokomis Editions, 2004. ISBN: 9780958174213.</li>
<li>T. G. Cornelissen, F. A. O. Silveira, and M. J. K. O. et al. “Global herbivory: how much plant tissue do insects eat?”. 2019.</li>
<li>Centre for Plant Biodiversity Research. <em>EUCLID 2.0: Eucalypts of Australia</em>. 2002.</li>
<li>L. Craven. “A taxonomic Revision of Calytrix Labill. (Myrtaceae)”. <em>Brunonia</em> 10 (1987), pp. 1-138. doi: <a href="https://doi.org/10.1071%2Fbru9870001">10.1071/bru9870001</a>.</li>
<li>L. A. Craven, B. J. Lepschi, and K. J. Cowley. “Melaleuca (Myrtaceae) of Western Australia: Five new species, three new combinations, one new name and a new state record.”. <em>Nuytsia</em> 20 (2010), pp. 27-36.</li>
<li>M. D. Crisp, L. Cayzer, G. T. Chandler, and L. G. Cook. “A monograph of Daviesia (Mirbelieae, Faboideae, Fabaceae)”. <em>Phytotaxa</em> 300 (2017), pp. 1-308. doi: <a href="https://doi.org/10.11646%2Fphytotaxa.300.1.1">10.11646/phytotaxa.300.1.1</a>.</li>
<li>E. Cross. “Unpublished data: Woodlands of western Victoria trait database”. 2011.</li>
<li>S. A. Cunningham, B. Summerhayes, and M. Westoby. “Evolutionary divergences in leaf structure and chemistry, comparing rainfall and soil nutrient gradients”. <em>Ecological Monographs</em> 69 (1999), pp. 569-588. doi: <a href="https://doi.org/10.1890%2F0012-9615%281999%29069%5B0569%3AEDILSA%5D2.0.CO%3B2">10.1890/0012-9615(1999)069[0569:EDILSA]2.0.CO;2</a>.</li>
<li>M. D. Denton, E. J. Veneklaas, F. M. Freimoser, and H. Lambers. “Banksia species (Proteaceae) from severely phosphorus-impoverished soils exhibit extreme efficiency in the use and re-mobilization of phosphorus”. <em>Plant, Cell and Environment</em> 30 (2007), pp. 1557-1565. doi: <a href="https://doi.org/10.1111%2Fj.1365-3040.2007.01733.x">10.1111/j.1365-3040.2007.01733.x</a>.</li>
<li>H. E. Desch and J. M. Dinwoodie. <em>Timber structure, properties, conversion and use</em>. Palgrave Macmillan, 1996. ISBN: 156022861X.</li>
<li>M. Dimitri and J. Biloni. <em>Libro del árbol. Vol 1. Esencias forestales indígenas de la Argentina; and Vol 2. Esencias Forestales indígenas de la Argentina de aplicación industrial</em>. Editorial Celulosa Argentina S. A., 1973.</li>
<li>D. H. Duncan. “Unpublished manuscript: Structural basis of variation in leaf mass per area in three plant clades”. 1998.</li>
<li>J. M. Dwyer and D. C. Laughlin. “Constraints on trait combinations explain climatic drivers of biodiversity: The importance of trait covariance in community assembly”. <em>Ecology Letters</em> 20.7 (2017), pp. 872-882. doi: <a href="https://doi.org/10.1111%2Fele.12781">10.1111/ele.12781</a>.</li>
<li>J. M. Dwyer and R. Mason. “Plant community responses to thinning in densely regenerating Acacia harpophylla forest”. <em>Restoration Ecology</em> 26.1 (2017), pp. 97-105. doi: <a href="https://doi.org/doi.org%2F10.1111%2Frec.12536">doi.org/10.1111/rec.12536</a>.</li>
<li>D. Eamus and H. Prichard. “A cost-benefit analysis of leaves of four Australian savanna species”. <em>Tree Physiology</em> 18 (1998), pp. 537-545. doi: <a href="https://doi.org/10.1093%2Ftreephys%2F18.8-9.537">10.1093/treephys/18.8-9.537</a>.</li>
<li>D. Eamus, B. Myers, G. Duff, and D. Williams. “Seasonal changes in photosynthesis of eight savanna tree species”. <em>Tree Physiology</em> 19 (1999), pp. 665-671. doi: <a href="https://doi.org/10.1093%2Ftreephys%2F19.10.665">10.1093/treephys/19.10.665</a>.</li>
<li>D. E. B. Myers and G. Duff. “A cost-benefit analysis of leaves of eight Australian savanna tree species of differing life-span”. <em>Photosynthetica</em> 36 (1999), pp. 575-586. doi: <a href="https://doi.org/10.1023%2FA%3A1007048222329">10.1023/A:1007048222329</a>.</li>
<li>C. Edwards, J. Read, and G. D. Sanson. “Characterising sclerophylly: some mechanical properties of leaves from heath and forest”. <em>Oecologia</em> 123.2 (2000), pp. 158-167. doi: <a href="https://doi.org/10.1007%2Fs004420051001">10.1007/s004420051001</a>.</li>
<li>C. Edwards, G. D. Sanson, N. Aranwela, and J. Read. “Relationships between sclerophylly, leaf biomechanical properties and leaf anatomy in some Australian heath and forest species”. <em>Plant Biosystems - An International Journal Dealing with all Aspects of Plant Biology</em> 134 (2000), pp. 261-277. doi: <a href="https://doi.org/10.1080%2F11263500012331350445">10.1080/11263500012331350445</a>.</li>
<li>M. Esperon-Rodriguez, S. A. Power, M. G.Tjoelker, A. Challis, R. M. Marchin, and P. D. Rymer. “Functional adaptations of urban trees along a climatic gradient.”. 2019.</li>
<li>M. Esperon-Rodriguez, S. A. Power, M. G.Tjoelker, and P. D. Rymer. <em>Which Plant Where Living Lab</em>. 2019.</li>
<li>D. S. Falster and M. Westoby. “Alternative height strategies among 45 dicot rain forest species from tropical Queensland, Australia”. <em>Journal of Ecology</em> 93 (2005), pp. 521-535. doi: <a href="https://doi.org/10.1111%2Fj.0022-0477.2005.00992.x">10.1111/j.0022-0477.2005.00992.x</a>.</li>
<li>D. S. Falster and M. Westoby. “Tradeoffs between height growth rate, stem persistence and maximum height among plant species in a post-fire succession”. <em>Oikos</em> 111 (2005), pp. 57-66. doi: <a href="https://doi.org/10.1111%2Fj.0030-1299.2005.13383.x">10.1111/j.0030-1299.2005.13383.x</a>.</li>
<li>J. H. Flynn and C. D. Holder. <em>A guide to useful woods of the world</em>. 2nd. Forest Products Society, 2001. ISBN: 1892529157.</li>
<li>C. R. Fonseca, J. M. Overton, B. Collins, and M. Westoby. “Shifts in trait-combinations along rainfall and phosphorus gradients”. <em>Journal of Ecology</em> 88 (2000), pp. 964-977. doi: <a href="https://doi.org/10.1046%2Fj.1365-2745.2000.00506.x">10.1046/j.1365-2745.2000.00506.x</a>.</li>
<li>P. G. McDonald, C. R. Fonseca, J. M. Overton, and M. Westoby. “Leaf-size divergence along rainfall and soil-nutrient gradients: is the method of size reduction common among clades?”. <em>Functional Ecology</em> 17.1 (2003), pp. 50-57. doi: <a href="https://doi.org/10.1046%2Fj.1365-2435.2003.00698.x">10.1046/j.1365-2435.2003.00698.x</a>.</li>
<li>P. I. Forster. “A taxonomic revision of Alyxia (Apocynaceae) in Australia”. <em>Australian Systematic Botany</em> 5 (1992), pp. 547-580. doi: <a href="https://doi.org/10.1071%2Fsb9920547">10.1071/sb9920547</a>.</li>
<li>P. I. Forster. “New names and combinations in Marsdenia (Asclepiadaceae: Marsdenieae) from Asia and Malesia (excluding Papusia)”. <em>Australian Systematic Botany</em> 8 (1995), pp. 691-701. doi: <a href="https://doi.org/10.1071%2Fsb9950691">10.1071/sb9950691</a>.</li>
<li>J. L. Funk, R. J. Standish, W. D. Stock, and F. Valladares. “Plant functional traits of dominant native and invasive species in mediterranean-climate ecosystems”. <em>Ecology</em> 97.1 (2016), pp. 75-83. doi: <a href="https://doi.org/10.1890%2F15-0974.1">10.1890/15-0974.1</a>.</li>
<li>R. V. Gallagher, M. R. Leishman, J. T. Miller, C. Hui, D. M. Richardson, J. Suda, and P. Trávníček. “Invasiveness in introduced Australian acacias: The role of species traits and genome size”. <em>Diversity and Distributions</em> 17 (2011), pp. 884-897. doi: <a href="https://doi.org/10.1111%2Fj.1472-4642.2011.00805.x">10.1111/j.1472-4642.2011.00805.x</a>.</li>
<li>R. Gardiner, L. Shoo, and J. M. Dwyer. “Look to seedling heights, rather than functional traits, to explain survival during extreme heat stress in the early stages of subtropical rainforest restoration”. <em>Journal of Applied Ecology</em> 56.12 (2019), pp. 2687-2697. doi: <a href="https://doi.org/doi.org%2F10.1111%2F1365-2664.13505">doi.org/10.1111/1365-2664.13505</a>.</li>
<li>S. M. Gleason, C. J. Blackman, A. M. Cook, C. A. Laws, and M. Westoby. “Whole-plant capacitance, embolism resistance and slow transpiration rates all contribute to longer desiccation times in woody angiosperms from arid and wet habitats”. <em>Tree Physiology</em> 34.3 (2014), pp. 275-284. doi: <a href="https://doi.org/10.1093%2Ftreephys%2Ftpu001">10.1093/treephys/tpu001</a>.</li>
<li>E. Goble-Garratt, D. Bell, and W. Loneragan. “Floristic and leaf structure patterns along a shallow elevational gradient”. <em>Australian Journal of Botany</em> 29 (1981), p. 329. doi: <a href="https://doi.org/10.1071%2Fbt9810329">10.1071/bt9810329</a>.</li>
<li>B. Goldsmith and D. Carter. <em>The indigenous timbers of Zimbabwe</em>. 9. 1981, p. 406. ISBN: 0869232940.</li>
<li>W. Clayton, M. Vorontsova, and K. H. &amp;. H. Williamson. <em>GrassBase - The online world grass flora</em>. 2006. url: <a href="http://www.kew.org/data/grasses-db.html">http://www.kew.org/data/grasses-db.html</a>.</li>
<li>P. K. Groom and B. B. Lamont. “Reproductive ecology of non-sprouting and re-sprouting Hakea species (Proteaceae) in southwestern Australia.”. In: <em>Gondwanan Heritage</em>. Ed. by J. C. S.D. Hopper M. Harvey and A. George. Surrey Beatty, Chipping Norton, 1996, pp. 239-248. doi: <a href="https://doi.org/unknown">unknown</a>.</li>
<li>P. K. Groom and B. B. Lamont. “Phosphorus accumulation in Proteaceae seeds: a synthesis”. <em>Plant and Soil</em> 334.1-2 (2009), pp. 61-72. doi: <a href="https://doi.org/10.1007%2Fs11104-009-0135-6">10.1007/s11104-009-0135-6</a>.</li>
<li>C. L. Gross. “The reproductive ecology of Canavalia rosea (Fabaceae) on Anak Krakatau, Indonesia”. <em>Australian Journal of Botany</em> 41 (1993), p. 591. doi: <a href="https://doi.org/10.1071%2Fbt9930591">10.1071/bt9930591</a>.</li>
<li>P. J. Grubb and D. J. Metcalfe. “Adaptation and inertia in the Australian tropical lowland rain-forest flora: Contradictory trends in intergeneric and intrageneric comparisons of seed size in relation to light demand”. <em>Functional Ecology</em> 10 (1996), p. 512. doi: <a href="https://doi.org/10.2307%2F2389944">10.2307/2389944</a>.</li>
<li>P. J. Grubb, R. V. Jackson, I. M. Barberis, J. N. Bee, D. A. Coomes, N. J. Dominy, M. A. S. D. L. Fuente, P. W. Lucas, D. J. Metcalfe, J. Svenning, I. M. Turner, and O. Vargas. “Monocot leaves are eaten less than dicot leaves in tropical lowland rain forests: Correlations with toughness and leaf presentation”. <em>Annals of Botany</em> 101 (2008), pp. 1379-1389. doi: <a href="https://doi.org/10.1093%2Faob%2Fmcn047">10.1093/aob/mcn047</a>.</li>
<li>C. G. Pereira, P. L. Clode, R. S. Oliveira, and H. Lambers. “Eudicots from severely phosphorus-impoverished environments preferentially allocate phosphorus to their mesophyll”. <em>New Phytologist</em> 218.3 (2018), pp. 959-973. doi: <a href="https://doi.org/10.1111%2Fnph.15043">10.1111/nph.15043</a>.</li>
<li>C. G. Pereira, P. E. Hayes, O. S. O’Sullivan, L. K. Weerasinghe, P. L. Clode, O. K. Atkin, and H. Lambers. “Trait convergence in photosynthetic nutrient-use efficiency along a 2-million year dune chronosequence in a global biodiversity hotspot”. <em>Journal of Ecology</em> 107.4 (2019). Ed. by S. Schwinning, pp. 2006-2023. doi: <a href="https://doi.org/10.1111%2F1365-2745.13158">10.1111/1365-2745.13158</a>.</li>
<li>M. T. Harrison, E. J. Edwards, G. D. Farquhar, A. B. Nicotra, and J. R. Evans. “Nitrogen in cell walls of sclerophyllous leaves accounts for little of the variation in photosynthetic nitrogen-use efficiency”. <em>Plant, Cell and Environment</em> 32 (2009), pp. 259-270. doi: <a href="https://doi.org/10.1111%2Fj.1365-3040.2008.01918.x">10.1111/j.1365-3040.2008.01918.x</a>.</li>
<li>F. Hassiotou, J. R. Evans, M. Ludwig, and E. J. Veneklaas. “Stomatal crypts may facilitate diffusion of CO2 to adaxial mesophyll cells in thick sclerophylls”. <em>Plant, Cell and Environment</em> 32 (2009), pp. 1596-1611. doi: <a href="https://doi.org/10.1111%2Fj.1365-3040.2009.02024.x">10.1111/j.1365-3040.2009.02024.x</a>.</li>
<li>P. Hayes, B. L. Turner, H. Lambers, and E. Laliberte. “Foliar nutrient concentrations and resorption efficiency in plants of contrasting nutrient-acquisition strategies along a 2-million-year dune chronosequence”. <em>Journal of Ecology</em> 102.2 (2013). Ed. by P. Bellingham, pp. 396-410. doi: <a href="https://doi.org/10.1111%2F1365-2745.12196">10.1111/1365-2745.12196</a>.</li>
<li>P. E. Hayes, P. L. Clode, R. S. Oliveira, and H. Lambers. “Proteaceae from phosphorus-impoverished habitats preferentially allocate phosphorus to photosynthetic cells: An adaptation improving phosphorus-use efficiency”. <em>Plant, Cell and Environment</em> 41.3 (2018), pp. 605-619. doi: <a href="https://doi.org/10.1111%2Fpce.13124">10.1111/pce.13124</a>.</li>
<li>M. L. Henery and M. Westoby. “Seed mass and seed nutrient content as predictors of seed output variation between species”. <em>Oikos</em> 92 (2001), pp. 479-490. doi: <a href="https://doi.org/10.1034%2Fj.1600-0706.2001.920309.x">10.1034/j.1600-0706.2001.920309.x</a>.</li>
<li>L. Hong, R. Lemmens, S. Prawirohatmodjo, I. Soerianegara, M. Sosef, and W. W. (Editors). <em>Plant resources of south east Asia: Timber trees</em>. Springer-Verlag Berlin and Heidelberg GmbH &amp; Co. KG, 1999, p. World biodiversity Database CD rom series. ISBN: 9783540147732.</li>
<li>L. Hughes and B. Rice. “Unpublished data: Traits data for 643 species”. 1992.</li>
<li>K. Hughes. “Unpublished data: Wood and stem density data for Australian plant species”. 2005.</li>
<li>B. P. M. Hyland, T. Whiffin, D. Christophel, B. Gray, and R. W. Elick. <em>Australian tropical rain forest plants trees, shrubs and vines</em>. CSIRO Publishing, 2003. ISBN: 0643068724.</li>
<li>World Agroforestry Centre (ICRAF). <em>The wood density database</em>. 2009. url: <a href="http://www.worldagroforestry.org/output/wood-density-database">http://www.worldagroforestry.org/output/wood-density-database</a>.</li>
<li>J. Ilic, D. Boland, M. McDonald, D. G, and P. Blakemore. <em>Woody density phase 1 - State of Knowledge. National Carbon Accounting System. Technical Report 18</em>. Tech. rep. Australian Greenhouse Office, Canberra, Australia., 2000.</li>
<li>M. Islam, D. W. Turner, and M. A. Adams. “Phosphorus availability and the growth, mineral composition and nutritive value of ephemeral forbs and associated perennials from the Pilbara, Western Australia”. <em>Australian Journal of Experimental Agriculture</em> 39 (1999), pp. 149-159. doi: <a href="https://doi.org/10.1071%2FEA98133">10.1071/EA98133</a>.</li>
<li>M. Islam and M. Adams. “Mineral content and nutritive value of native grasses and the response to added phosphorus in a Pilbara rangeland”. <em>Tropical Grasslands</em> 33 (1999), pp. 193-200.</li>
<li>D. Jin and D. Hochuli. “Urbanisation increases arthropod abundance and herbivory on a native weed”. 2019.</li>
<li>G. Jordan. “Unpublished data: Leaf traits for Tasmanian species”. 2007.</li>
<li>E. Jurado. “Diaspore weight, dispersal, growth form and perenniality of central Australian plants”. <em>Journal of Ecology</em> 79 (1993), pp. 811-828. url: <a href="https://www.jstor.org/stable/2260669">https://www.jstor.org/stable/2260669</a>.</li>
<li>G. Keighery. “Taxonomy of the Calytrix ecalycata complex (Myrtaceae)”. <em>Nuytsia</em> 15 (2004), pp. 261-268. url: <a href="https://florabase.dpaw.wa.gov.au/nuytsia/article/409">https://florabase.dpaw.wa.gov.au/nuytsia/article/409</a>.</li>
<li>Royal Botanic Gardens Kew. <em>Seed Information Database (SID) and Seed Bank Database</em>. 2019. url: <a href="http://data.kew.org/sid/">http://data.kew.org/sid/</a>.</li>
<li>Royal Botanic Gardens Kew. <em>Seed Information Database (SID) and Seed Bank Database</em>. 2019. url: <a href="http://data.kew.org/sid/">http://data.kew.org/sid/</a>.</li>
<li>Royal Botanic Gardens Kew. <em>Oil content data from Seed Information Database (SID) and Seed Bank Database</em>. 2019. url: <a href="http://data.kew.org/sid/">http://data.kew.org/sid/</a>.</li>
<li>Royal Botanic Gardens Kew. <em>Seed dispersal data from the Seed Information Database (SID) and Seed Bank Database</em>. 2019. url: <a href="http://data.kew.org/sid/">http://data.kew.org/sid/</a>.</li>
<li>Royal Botanic Gardens Kew. <em>Germination data from the Seed Information Database (SID) and Seed Bank Database</em>. 2019. url: <a href="http://data.kew.org/sid/">http://data.kew.org/sid/</a>.</li>
<li>K. J. E. Knox and P. J. Clarke. “Fire severity and nutrient availability do not constrain resprouting in forest shrubs”. <em>Plant Ecol</em> 212 (2011), pp. 1967-1978. doi: <a href="https://doi.org/10.1007%2Fs11258-011-9956-5">10.1007/s11258-011-9956-5</a>.</li>
<li>R. Kooyman, M. Rossetto, W. Cornwell, and M. Westoby. “Phylogenetic tests of community assembly across regional to continental scales in tropical and subtropical rain forests”. <em>Global Ecology and Biogeography</em> 20 (2011), pp. 707-716. doi: <a href="https://doi.org/10.1111%2Fj.1466-8238.2010.00641.x">10.1111/j.1466-8238.2010.00641.x</a>.</li>
<li>E. Laliberté, B. L. Turner, T. Costes, S. J. Pearse, K. Wyrwoll, G. Zemunik, and H. Lambers. “Experimental assessment of nutrient limitation along a 2-million-year dune chronosequence in the south-western Australia biodiversity hotspot”. <em>Journal of Ecology</em> 100 (2012), pp. 631-642. doi: <a href="https://doi.org/10.1111%2Fj.1365-2745.2012.01962.x">10.1111/j.1365-2745.2012.01962.x</a>.</li>
<li>B. B. Lamont, P. K. Groom, and R. M. Cowling. “High leaf mass per area of related species assemblages may reflect low rainfall and carbon isotope discrimination rather than low phosphorus and nitrogen concentrations”. <em>Functional Ecology</em> 16 (2002), pp. 403-412. doi: <a href="https://doi.org/10.1046%2Fj.1365-2435.2002.00631.x">10.1046/j.1365-2435.2002.00631.x</a>.</li>
<li>B. B. Lamont, P. K. Groom, M. Williams, and T. He. “LMA, density and thickness: recognizing different leaf shapes and correcting for their nonlaminarity”. <em>New Phytologist</em> 207.4 (2015), pp. 942-947. doi: <a href="https://doi.org/10.1111%2Fnph.13465">10.1111/nph.13465</a>.</li>
<li>M. J. Lawes, H. Adie, J. Russell-Smith, B. Murphy, and J. J. Midgley. “How do small savanna trees avoid stem mortality by fire? The roles of stem diameter, height and bark thickness”. <em>Ecosphere</em> 2.4 (2011), p. art42. doi: <a href="https://doi.org/10.1890%2Fes10-00204.1">10.1890/es10-00204.1</a>.</li>
<li>M. J. Lawes, A. Richards, J. Dathe, and J. J. Midgley. “Bark thickness determines fire resistance of selected tree species from fire-prone tropical savanna in north Australia”. <em>Plant Ecology</em> 212.12 (2011), pp. 2057-2069. doi: <a href="https://doi.org/10.1007%2Fs11258-011-9954-7">10.1007/s11258-011-9954-7</a>.</li>
<li>M. J. Lawes, J. J. Midgley, and P. J. Clarke. “Costs and benefits of relative bark thickness in relation to fire damage: A savanna/forest contrast”. <em>Journal of Ecology</em> 101 (2012), pp. 517-524. doi: <a href="https://doi.org/10.1111%2F1365-2745.12035">10.1111/1365-2745.12035</a>.</li>
<li>J. R. Lawson, K. A. Fryirs, and M. R. Leishman. <em>Data from: Hydrological conditions explain wood density in riparian plants of south-eastern Australia</em>. 2015. doi: <a href="https://doi.org/10.5061%2Fdryad.72h45">10.5061/dryad.72h45</a>.</li>
<li>E. Laxton. “Relationship between leaf traits, insect communities and resource availability”. PhD thesis. Macquarie University, 2005. url: <a href="http://hdl.handle.net/1959.14/483">http://hdl.handle.net/1959.14/483</a>.</li>
<li>M. R. Lee, J. R. Powell, B. Oberle, W. K. Cornwell, M. Lyons, J. L. Rigg, and A. E. Zanne. “Good neighbors aplenty: fungal endophytes rarely exhibit competitive exclusion patterns across a span of woody habitats”. <em>Ecology</em> 100.9 (2019). doi: <a href="https://doi.org/10.1002%2Fecy.2790">10.1002/ecy.2790</a>.</li>
<li>M. R. Leishman and M. Westoby. “Classifying plants into groups on the basis of associations of individual traits-Evidence from Australian semi-arid woodlands”. <em>Journal of Ecology</em> 80 (1992), p. 417. doi: <a href="https://doi.org/10.2307%2F2260687">10.2307/2260687</a>.</li>
<li>Westoby, M., Rice, B., Howell, and J. “Seed Size and Plant Growth Form as Factors in Dispersal Spectra”. <em>Ecology</em> 71 (1990), pp. 1307-1315. doi: <a href="https://doi.org/10.2307%2F1938268">10.2307/1938268</a>.</li>
<li>M. R. Leishman, M. Westoby, and E. Jurado. “Correlates of seed size variation: A comparison among five temperate floras”. <em>Journal of Ecology</em> 83 (1995), pp. 517-529. doi: <a href="https://doi.org/10.2307%2F2261604">10.2307/2261604</a>.</li>
<li>R. Lemmens and I. Soerjanegara. <em>PROSEA, Volume 5/1: Timber Trees - Major Commercial Timbers</em>. Pudoc/Prosea, 1993. ISBN: 9789022010334.</li>
<li>F. K. Lim, L. J. Pollock, and P. A. Vesk. “The role of plant functional traits in shrub distribution around alpine frost hollows”. <em>Journal of Vegetation Science</em> 28.3 (2017). Ed. by F. de Bello, pp. 585-594. doi: <a href="https://doi.org/10.1111%2Fjvs.12517">10.1111/jvs.12517</a>.</li>
<li>J. Lord, J. Egan, T. Clifford, E. Jurado, M. Leishman, D. Williams, and M. Westoby. “Larger seeds in tropical floras: Consistent patterns independent of growth form and dispersal mode”. <em>Journal of Biogeography</em> 24 (1997), pp. 205-211. doi: <a href="https://doi.org/10.1046%2Fj.1365-2699.1997.00126.x">10.1046/j.1365-2699.1997.00126.x</a>.</li>
<li>I. Lunt and J. Morgan. “Unpublished data: Trait data from various Morgan and Lunt projects 1990-2012”. 2012.</li>
<li>C. H. Lusk, Y. Onoda, R. Kooyman, and A. Guti&lt;U+FFFD&gt;rrez-Gir&lt;U+FFFD&gt;n. “Reconciling species-level vs plastic responses of evergreen leaf structure to light gradients: shade leaves punch above their weight”. <em>New Phytologist</em> 186.2 (2010), pp. 429-438. doi: <a href="https://doi.org/10.1111%2Fj.1469-8137.2010.03202.x">10.1111/j.1469-8137.2010.03202.x</a>.</li>
<li>C. H. Lusk, J. W. G. Kelly, and S. M. Gleason. “Light requirements of Australian tropical vs. cool-temperate rainforest tree species show different relationships with seedling growth and functional traits”. <em>Annals of Botany</em> 111 (2012), pp. 479-488. doi: <a href="https://doi.org/10.1093%2Faob%2Fmcs289">10.1093/aob/mcs289</a>.</li>
<li>C. H. Lusk, K. M. Sendall, and P. J. Clarke. “Seedling growth rates and light requirements of subtropical rainforest trees associated with basaltic and rhyolitic soils”. <em>Australian Journal of Botany</em> 62.1 (2014), pp. 48-55. doi: <a href="https://doi.org/10.1071%2Fbt13262">10.1071/bt13262</a>.</li>
<li>A. Manea. “Unpublished data: Dispersal appendage data for acacia via worldwide wattle”. 2011.</li>
<li>B. Maslin. <em>WATTLE Acacias of Australia</em>. 2012.</li>
<li>J. K. McCarthy, J. M. Dwyer, and K. Mokany. “A regional-scale assessment of using metabolic scaling theory to predict ecosystem properties”. <em>Proceedings of the Royal Society B: Biological Sciences</em> 286.1915 (2019), p. 20192221. doi: <a href="https://doi.org/10.1098%2Frspb.2019.2221">10.1098/rspb.2019.2221</a>.</li>
<li>E. Meier. <em>The wood database</em>. 2007. url: <a href="http://www.wood-database.com/">http://www.wood-database.com/</a>.</li>
<li>K. Mokany and J. Ash. “Are traits measured on pot grown plants representative of those in natural communities?”. <em>Journal of Vegetation Science</em> 19 (2008), pp. 119-126. doi: <a href="https://doi.org/10.3170%2F2007-8-18340">10.3170/2007-8-18340</a>.</li>
<li>A. T. Moles and M. Westoby. “Do small leaves expand faster than large leaves, and do shorter expansion times reduce herbivore damage?”. <em>OIKOS</em> 90.3 (2000), pp. 517-524. doi: <a href="https://doi.org/https%3A%2F%2Fdoi.org%2F10.1034%2Fj.1600-0706.2000.900310.x.">https://doi.org/10.1034/j.1600-0706.2000.900310.x.</a>.</li>
<li>A. T. Moles, D. I. Warton, and M. Westoby. “Seed size and survival in the soil in arid Australia”. <em>Austral Ecology</em> 28.5 (2003), pp. 575-585. doi: <a href="https://doi.org/doi.org%2F10.1046%2Fj.1442-9993.2003.01314.x">doi.org/10.1046/j.1442-9993.2003.01314.x</a>.</li>
<li>A. T. Moles, I. R. Wallis, W. J. Foley, D. I. Warton, J. C. Stegen, A. J. Bisigato, L. Cella-Pizarro, C. J. Clark, P. S. Cohen, and W. K. C. et al. “Putting plant resistance traits on the map: a test of the idea that plants are better defended at lower latitudes”. <em>New Phytologist</em> 191.3 (2011), pp. 777-788. doi: <a href="https://doi.org/doi.org%2F10.1111%2Fj.1469-8137.2011.03732.x">doi.org/10.1111/j.1469-8137.2011.03732.x</a>.</li>
<li>Moore, B.D., DeGabriel, and J.L. “Leaf traits of the Eucalyptus series Siderophloiae”. 2019.</li>
<li>J. Morgan. “Unpublished data: Grassy woodland traits”. 2011.</li>
<li>J. Morgan. “Unpublished data: Trait database”. 2011.</li>
<li>J. Morgan. “Unpublished data: Trait data from Australian plant species”. 2004.</li>
<li>H. Morgan. “Root system architecture, water use and rainfall repsonses of perennial species”. PhD thesis. Macquarie University, 2005.</li>
<li>Munroe, S.E.M., Guerin, G.R., and Sparrow, and B. “TERN Plant Photosynthetic Pathway Trait Dataset, version 1.0.”. <em>unknown</em> unknown.unknown (2019), p. unknown. doi: <a href="https://doi.org/unknown">unknown</a>.</li>
<li>National Herbarium of NSW. <em>Trait measurements for NSW rainforest species from PLantNET</em>. 2016. url: <a href="http://plantnet.rbgsyd.nsw.gov.au/">http://plantnet.rbgsyd.nsw.gov.au/</a>.</li>
<li>Ü. Niinemets, I. J. Wright, and J. R. Evans. “Leaf mesophyll diffusion conductance in 35 Australian sclerophylls covering a broad range of foliage structural and physiological variation”. <em>Journal of Experimental Botany</em> 60 (2009), pp. 2433-2449. doi: <a href="https://doi.org/10.1093%2Fjxb%2Ferp045">10.1093/jxb/erp045</a>.</li>
<li>Northern Territory Herbarium. <em>Flora of the Darwin Region Online</em>. 2014. url: <a href="http://www.lrm.nt.gov.au/plants-and-animals/herbarium/darwin_flora_online">http://www.lrm.nt.gov.au/plants-and-animals/herbarium/darwin_flora_online</a>.</li>
<li>M. Ooi, P. Myerscough, and T. Auld. <em>Longterm seed mass field campaign</em>. 2007.</li>
<li>A. O’Reilly-Nugent, E. Wandrag, C. Catford, B. Gruber, D. Driscoll, and R. Duncan. <em>Measuring impact: Joint-species modelling of invaded plant communities</em>. 2018.</li>
<li>C. Osborne, A. Salomaa, T. Kluyver, V. Visser, E. Kellogg, O. Morrone, M. Vorontsova, W. Clayton, and D. Simpson. “A global database of C4 photosynthesis in grasses”. <em>New Phytologist</em> 114 (2014), pp. 441-446. doi: <a href="https://doi.org/10.1111%2Fnph.12942">10.1111/nph.12942</a>.</li>
<li>P. J. Peeters. “Correlations between leaf structural traits and the densities of herbivorous insect guilds”. <em>Biological Journal of the Linnean Society</em> 77 (2002), pp. 43-65. doi: <a href="https://doi.org/10.1046%2Fj.1095-8312.2002.00091.x">10.1046/j.1095-8312.2002.00091.x</a>.</li>
<li>B. K. Pekin, R. S. Wittkuhn, M. M. Boer, C. Macfarlane, and P. F. Grierson. “Plant functional traits along environmental gradients in seasonally dry and fire-prone ecosystem”. <em>Journal of Vegetation Science</em> 22 (2011), pp. 1009-1020. doi: <a href="https://doi.org/10.1111%2Fj.1654-1103.2011.01323.x">10.1111/j.1654-1103.2011.01323.x</a>.</li>
<li>M. Pickup. “wood density data from Kuring-gai National Park”. 2002.</li>
<li>M. PICKUP, M. WESTOBY, and A. BASDEN. “Dry mass costs of deploying leaf area in relation to leaf size”. <em>Functional Ecology</em> 19.1 (2005), pp. 88-97. doi: <a href="https://doi.org/10.1111%2Fj.0269-8463.2005.00927.x">10.1111/j.0269-8463.2005.00927.x</a>.</li>
<li>L. D. Prior, D. Eamus, and D. M. J. S. Bowman. “Leaf attributes in the seasonally dry tropics: A comparison of four habitats in northern Australia”. <em>Functional Ecology</em> 17 (2003), pp. 504-515. doi: <a href="https://doi.org/10.1046%2Fj.1365-2435.2003.00761.x">10.1046/j.1365-2435.2003.00761.x</a>.</li>
<li>L. D. PRIOR, D. M. J. S. BOWMAN, and D. EAMUS. “Seasonal differences in leaf attributes in Australian tropical tree species: family and habitat comparisons”. <em>Functional Ecology</em> 18.5 (2004), pp. 707-718. doi: <a href="https://doi.org/10.1111%2Fj.0269-8463.2004.00885.x">10.1111/j.0269-8463.2004.00885.x</a>.</li>
<li>Oxford Forestry Institute. <em>Prospect: The wood database</em>. 2009. url: <a href="http://dps.plants.ox.ac.uk/ofi/prospect/index.htm">http://dps.plants.ox.ac.uk/ofi/prospect/index.htm</a>.</li>
<li>Royal Botanic Gardens Kew. <em>Seed Information Database (SID)</em>. 2014. url: <a href="http://data.kew.org/sid/">http://data.kew.org/sid/</a>.</li>
<li>Royal Botanic Gardens Sydney. <em>PLantNET: NSW flora online</em>. 2014. url: <a href="http://plantnet.rbgsyd.nsw.gov.au/">http://plantnet.rbgsyd.nsw.gov.au/</a>.</li>
<li>Royal Botanic Gardens Sydney. <em>PLantNET</em>. 2014. url: <a href="http://plantnet.rbgsyd.nsw.gov.au/search/simple.htm">http://plantnet.rbgsyd.nsw.gov.au/search/simple.htm</a>.</li>
<li>B. Rice. “Unpublished data: Growth form of Kakadu plant species”. 1991.</li>
<li>A. Richards, A. Shapcott, J. Playford, B. Morrison, C. Critchley, and S. Schmidt. “Physiological profiles of restricted endemic plants and their widespread congenors in the North Queensland wet tropics, Australia”. <em>Biological Conservation</em> 111 (2003), pp. 41-52. doi: <a href="https://doi.org/10.1016%2Fs0006-3207%2802%2900245-8">10.1016/s0006-3207(02)00245-8</a>.</li>
<li>B. Roberts. “Unpublished data”. 2006.</li>
<li>B. Rye. “A revision of south-western Australian species of Micromyrtus (Myrtaceae) with five antisepalous ribs on the hypanthium.”. <em>Nuytsia</em> 15 (2002), pp. 101-122.</li>
<li>B. Rye. “A partial revision of the south-western Australian species of Micromyrtus (Myrtaceae: Chamelaucieae)”. <em>Nuytsia</em> 16 (2006), pp. 117-147.</li>
<li>B. Rye. “Reinstatement of the Western Australian genus Oxymyrrhine (Myrtaceae: Chamelaucieae) with three new species”. <em>Nuytsia</em> 19 (2009), pp. 149-165. url: <a href="https://florabase.dpaw.wa.gov.au/science/nuytsia/468.pdf">https://florabase.dpaw.wa.gov.au/science/nuytsia/468.pdf</a>.</li>
<li>B. L. Rye. “A revision of the Micromyrtus racemosa complex (Myrtaceae: Chamelaucieae) of south-western Australia”. <em>Nuytsia</em> 20 (2010), pp. 37-56.</li>
<li>B. Rye, P. Wilson, and G. Keighery. “A revision of the species of Hypocalymma (Myrtaceae: Chamelaucieae) with smooth or colliculate seeds”. <em>Nuytsia</em> 23 (2013), pp. 283-312. url: <a href="https://florabase.dpaw.wa.gov.au/science/nuytsia/672.pdf">https://florabase.dpaw.wa.gov.au/science/nuytsia/672.pdf</a>.</li>
<li>B. L. Rye. “An update to the taxonomy of some western Australian genera of Myrtaceae tribe Chamelaucieae. 1. Calytrix”. <em>Nuytsia</em> 23 (2013), pp. 483-501. url: <a href="https://florabase.dpaw.wa.gov.au/science/nuytsia/687.pdf">https://florabase.dpaw.wa.gov.au/science/nuytsia/687.pdf</a>.</li>
<li>B. L. Rye. “A revision of the south-western Australian genus Babingtonia (Myrtaceae: Chamelaucieae).”. <em>Nuytsia</em> 25 (2015), pp. 219-250. url: <a href="https://florabase.dpaw.wa.gov.au/science/nuytsia/758.pdf">https://florabase.dpaw.wa.gov.au/science/nuytsia/758.pdf</a>.</li>
<li>J. P. Jessop and H. R. Toelken. <em>Flora of South Australia, 4th edition, 4 vols</em>. Electronic version curated by J. Kellermann, State Herbarium of South Australia.. 1986.</li>
<li>S. SCHMIDT and G. STEWART. “Waterlogging and fire impacts on nitrogen availability and utilization in a subtropical wet heathland (wallum)”. <em>Plant, Cell and Environment</em> 20.10 (1997), pp. 1231-1241. doi: <a href="https://doi.org/10.1046%2Fj.1365-3040.1997.d01-20.x">10.1046/j.1365-3040.1997.d01-20.x</a>.</li>
<li>S. Schmidt and G. R. Stewart. “d15N values of tropical savanna and monsoon forest species reflect root specialisations and soil nitrogen status”. <em>Oecologia</em> 134 (2003), pp. 569-577. doi: <a href="https://doi.org/10.1007%2Fs00442-002-1150-y">10.1007/s00442-002-1150-y</a>.</li>
<li>E. -. Schulze, R. J. Williams, G. D. Farquhar, W. Schulze, J. Langridge, J. M. Miller, and B. H. Walker. “Carbon and nitrogen isotope discrimination and nitrogen nutrition of trees along a rainfall gradient in northern Australia”. <em>Australian Journal of Plant Physiology</em> 25.4 (1998), pp. 413-425. doi: <a href="https://doi.org/10.1071%2FPP97113">10.1071/PP97113</a>.</li>
<li>A. J. Scott. “Unpublished data: Honours Project at La Trobe University”. 2010.</li>
<li>K. M. Sendall, C. H. Lusk, and P. B. Reich. “Trade-offs in juvenile growth potential vs. shade tolerance among subtropical rain forest trees on soils of contrasting fertility”. <em>Functional Ecology</em> 30.6 (2015). Ed. by N. Norden, pp. 845-855. doi: <a href="https://doi.org/10.1111%2F1365-2435.12573">10.1111/1365-2435.12573</a>.</li>
<li>O. D. Seng. <em>Specific gravity of Indonesian Woods and its significance for practical use</em>. Tech. rep. FPRDC Forestry Department, Bogor, Indonesia, 1951.</li>
<li>S. Soliveres, D. J. Eldridge, F. Hemmings, and F. T. Maestre. “Nurse plant effects on plant species richness in drylands: The role of grazing, rainfall and species specificity”. <em>Perspectives in Plant Ecology, Evolution and Systematics</em> 14 (2012), pp. 402-410. doi: <a href="https://doi.org/10.1016%2Fj.ppees.2012.09.003">10.1016/j.ppees.2012.09.003</a>.</li>
<li>R. Standish. “unpublished jarrah forest dataset”. 2019.</li>
<li>I. R. Thompson. “Morphometric analysis and revision of eastern Australian Hovea (Brongniartieae-Fabaceae)”. <em>Australian Systematic Botany</em> 14 (2001), p. 1. doi: <a href="https://doi.org/10.1071%2Fsb00008">10.1071/sb00008</a>.</li>
<li>Tasmanian Herbarium. <em>Flora of Tasmania Online</em>. 2009. url: <a href="www.tmag.tas.gov.au/floratasmania">www.tmag.tas.gov.au/floratasmania</a>.</li>
<li>D. Y. P. Tng, G. J. Jordan, and D. M. J. S. Bowman. “Plant traits demonstrate that temperate and tropical giant Eucalypt forests are ecologically convergent with rainforest not savanna”. <em>PLoS ONE</em> 8 (2013), p. e84378. doi: <a href="https://doi.org/10.1371%2Fjournal.pone.0084378">10.1371/journal.pone.0084378</a>.</li>
<li>H. Toelken. “A revision of the genus Kunzea (Myrtaceae) I. The western Australian section Zeanuk”. <em>Journal of the Adelaide Botanic Garden</em> 17 (1996), pp. 29-106.</li>
<li>K. W. Tomlinson, L. Poorter, F. J. Sterck, F. Borghetti, D. Ward, S. de Bie, and F. van Langevelde. “Leaf adaptations of evergreen and deciduous trees of semi-arid and humid savannas on three continents”. <em>Journal of Ecology</em> 101 (2013), pp. 430-440. doi: <a href="https://doi.org/10.1111%2F1365-2745.12056">10.1111/1365-2745.12056</a>.</li>
<li>M. Trudgeon and B. Rye. “Astus, a new western Australian genus of Myrtaceae with heterocarpidic fruits”. <em>Nuytsia</em> 14 (2005), pp. 495-512.</li>
<li>M. Trudgen and B. Rye. “An update to the taxonomy of some western Australian genera of Myrtaceae tribe Chamelaucieae. 2. Cyathostemon”. <em>Nuytsia</em> 24 (2014), pp. 7-16. url: <a href="https://florabase.dpaw.wa.gov.au/science/nuytsia/692.pdf">https://florabase.dpaw.wa.gov.au/science/nuytsia/692.pdf</a>.</li>
<li>P. A. Vesk, M. R. Leishman, and M. Westoby. “Simple traits do not predict grazing response in Australian dry shrublands and woodlands”. <em>Journal of Applied Ecology</em> 41.1 (2004), pp. 22-31. doi: <a href="https://doi.org/10.1111%2Fj.1365-2664.2004.00857.x">10.1111/j.1365-2664.2004.00857.x</a>.</li>
<li>P. Vesk. “Unpublished data: Leaf traits for flora of Falls Creek Victoria”. 2007.</li>
<li>Western Australian Herbarium. <em>FloraBase: the western Australian flora</em>. 1998. url: <a href="https://florabase.dpaw.wa.gov.au/">https://florabase.dpaw.wa.gov.au/</a>.</li>
<li>Western Australian Herbarium. <em>FloraBase: the western Australian flora</em>. 2016. url: <a href="https://florabase.dpaw.wa.gov.au/">https://florabase.dpaw.wa.gov.au/</a>.</li>
<li>J. Wells. “Unpublished data: Wood density data for global wood density database”. 2009.</li>
<li>Westoby, M., Rice, B., Howell, and J. “Seed Size and Plant Growth Form as Factors in Dispersal Spectra”. <em>Ecology</em> 71 (1990), pp. 1307-1315. doi: <a href="https://doi.org/10.2307%2F1938268">10.2307/1938268</a>.</li>
<li>M. Westoby and I. J. Wright. “The leaf size - twig size spectrum and its relationship to other important spectra of variation among species”. <em>Oecologia</em> 135 (2003), pp. 621-628. doi: <a href="https://doi.org/10.1007%2Fs00442-003-1231-6">10.1007/s00442-003-1231-6</a>.</li>
<li>M. Westoby. “Unpublished data: Trait data for plant species at Mt Wellington and Sea Acres”. Collected by a research assistant Erica Williams while at Macquarie University. 2004.</li>
<li>S. M. Gleason, D. W. Butler, K. Zieminska, P. Waryszak, and M. Westoby. “Stem xylem conductivity is key to plant water balance across Australian angiosperm species”. <em>Functional Ecology</em> 26 (2012), pp. 343-352. doi: <a href="https://doi.org/10.1111%2Fj.1365-2435.2012.01962.x">10.1111/j.1365-2435.2012.01962.x</a>.</li>
<li>M. Westoby. “Unpublished data: SLA data for various plant species”. 1996.</li>
<li>J. R. Wheeler, N. G. Marchant, and M. Lewington. <em>Flora of the south west: Bunbury, Augusta, Denmark</em>. Canberra, A.C.T. : Crawley, W.A: Australian Biological Resources Study ; University of Western Australia Press, 2002. ISBN: 9780642568168.</li>
<li>J. Wills, J. Herbohn, J. Hu, S. Sohel, J. Baynes, and J. Firn. “Tree leaf trade-offs are stronger for sub-canopy trees: leaf traits reveal little about growth rates in canopy trees”. <em>Ecological Applications</em> 28.4 (2018), pp. 1116-1125. doi: <a href="https://doi.org/doi.org%2F10.1002%2Feap.1715">doi.org/10.1002/eap.1715</a>.</li>
<li>P. G. Wilson and R. Rowe. “A revision of the Indigofereae (Fabaceae) in Australia. 2. Indigofera species with trifoliolate and alternately pinnate leaves”. <em>Telopea</em> 12 (2008), pp. 293-307.</li>
<li>I. J. Wright, H. T. Clifford, R. Kidson, M. L. Reed, B. L. Rice, and M. Westoby. “A survey of seed and seedling characters in 1744 Australian dicotyledon species: Cross-species trait correlations and correlated trait-shifts within evolutionary lineages”. <em>Biological Journal of the Linnean Society</em> 69 (2000), pp. 521-547. doi: <a href="https://doi.org/10.1111%2Fj.1095-8312.2000.tb01222.x">10.1111/j.1095-8312.2000.tb01222.x</a>.</li>
<li>I. J. Wright. “Unpublished data: Seed mass reserve data for various species in NSW”. 2001.</li>
<li>I. J. Wright, M. Westoby, and P. B. Reich. “Convergence towards higher leaf mass per area in dry and nutrient-poor habitats has different consequences for leaf life span”. <em>Journal of Ecology</em> 90.3 (2002), pp. 534-543. doi: <a href="https://doi.org/10.1046%2Fj.1365-2745.2002.00689.x">10.1046/j.1365-2745.2002.00689.x</a>.</li>
<li>I. J. Wright and M. Westoby. “Leaves at low versus high rainfall: Coordination of structure, lifespan and physiology”. <em>New Phytologist</em> 155 (2002), pp. 403-416. doi: <a href="https://doi.org/10.1046%2Fj.1469-8137.2002.00479.x">10.1046/j.1469-8137.2002.00479.x</a>.</li>
<li>T. I. Lenz, I. J. Wright, and M. Westoby. “Interrelations among pressure-volume curve traits across species and water availability gradients”. <em>Physiologia Plantarum</em> 127.3 (2006), pp. 423-433. doi: <a href="https://doi.org/10.1111%2Fj.1399-3054.2006.00680.x">10.1111/j.1399-3054.2006.00680.x</a>.</li>
<li>I. J. Wright, P. B. Reich, and M. Westoby. “Strategy shifts in leaf physiology, structure and nutrient content between species of high- and low-rainfall and high- and low-nutrient habitats”. <em>Functional Ecology</em> 15.4 (2001), pp. 423-434. doi: <a href="https://doi.org/10.1046%2Fj.0269-8463.2001.00542.x">10.1046/j.0269-8463.2001.00542.x</a>.</li>
<li>I. J. Wright, P. B. Reich, M. Westoby, D. D. Ackerly, Z. Baruch, F. Bongers, J. Cavender-Bares, T. Chapin, J. H. C. Cornelissen, M. Diemer, J. Flexas, E. Garnier, P. K. Groom, J. Gulias, K. Hikosaka, B. B. Lamont, T. Lee, W. Lee, C. Lusk, J. J. Midgley, M. Navas, Ü. Niinemets, J. Oleksyn, N. Osada, H. Poorter, P. Poot, L. Prior, V. I. Pyankov, C. Roumet, S. C. Thomas, M. G. Tjoelker, E. J. Veneklaas, and R. Villar. “The worldwide leaf economics spectrum”. <em>Nature</em> 428 (2004), pp. 821-827. doi: <a href="https://doi.org/10.1038%2Fnature02403">10.1038/nature02403</a>.</li>
<li>I. J. Wright, D. S. Falster, M. Pickup, and M. Westoby. “Cross-species patterns in the coordination between leaf and stem traits, and their implications for plant hydraulics”. <em>Physiologia Plantarum</em> 127 (2006), pp. 445-456. doi: <a href="https://doi.org/10.1111%2Fj.1399-3054.2006.00699.x">10.1111/j.1399-3054.2006.00699.x</a>.</li>
<li>I. Wright. “Unpublished data: Trait data for Northern Territory savanna species”. 2008.</li>
<li>I. J. Wright, J. Cooke, L. A. Cernusak, L. B. Hutley, M. C. Scalon, W. C. Tozer, and C. E. R. Lehmann. “Stem diameter growth rates in a fire-prone savanna correlate with photosynthetic rate and branch-scale biomass allocation, but not specific leaf area”. <em>Austral Ecology</em> 44.2 (2018), pp. 339-350. doi: <a href="https://doi.org/10.1111%2Faec.12678">10.1111/aec.12678</a>.</li>
<li>A. Zanne. “Unpublished data: Hydraulic traits for NSW plant species from four field sites”. 2007.</li>
<li>A. E. Zanne, G. Lopez-Gonzalez, D. A. Coomes, J. Ilic, S. Jansen, S. L. Lewis, R. B. Miller, N. G. Swenson, M. C. Wiemann, and J. Chave. <em>Data from: Towards a worldwide wood economics spectrum</em>. 2009. doi: <a href="https://doi.org/10.5061%2Fdryad.234">10.5061/dryad.234</a>.</li>
<li>K. Zieminska, D. W. Butler, S. M. Gleason, I. J. Wright, and M. Westoby. “Fibre wall and lumen fractions drive wood density variation across 24 Australian angiosperms”. <em>AoB PLANTS</em> 5 (2013), p. plt046. doi: <a href="https://doi.org/10.1093%2Faobpla%2Fplt046">10.1093/aobpla/plt046</a>.</li>
</ul>
</div>
<div id="file-types" class="section level2">
<h2 class="hasAnchor">
<a href="#file-types" class="anchor"></a>File types</h2>
<div id="csv" class="section level3">
<h3 class="hasAnchor">
<a href="#csv" class="anchor"></a>CSV</h3>
</div>
<div id="yaml" class="section level3">
<h3 class="hasAnchor">
<a href="#yaml" class="anchor"></a>YAML files</h3>
<p>yml: The <code>yml</code> file extension (pronounced “YAML”) <a href="https://en.wikipedia.org/wiki/YAML">is a type structured data file</a>, that is both human and machine readable. The information in it appears under various labels, which are imported into AusTraits. You can edit it any text editor, or also in Rstudio. Generally, yml is used in situations where a table does not suit because of variable lengths and or nested structures. This document outlines how to describe metadata for any given study</p>
<p>The metadata is compiled in a <code>.yml</code> file, a structured data file where information is presented in a hierarchical format. It has the advantage over a spreadsheet in that the nested “headers” can have variable numbers of categories. The data under each of the hierarchical headings are easily extracted by R.</p>
<p>Assumed character unless specified otherwise</p>
<p>possible types:</p>
<ul>
<li>
<code>character</code>:</li>
<li>
<code>numeric</code>: must have fields <code>description</code>, <code>type</code>, <code>units</code>, <code>values</code> with filed <code>minimum</code> and <code>maximum</code>.</li>
<li>
<code>categorical</code>: Unlike character, only specific values are allowed. must have fields <code>description</code>, <code>type</code>,and <code>values</code>, with the latter including a list of possible values.</li>
<li>
<code>table</code>:</li>
<li>
<code>list</code>:</li>
<li>
<code>array</code>:</li>
</ul>
</div>
</div>
<div id="custom_R_code" class="section level2">
<h2 class="hasAnchor">
<a href="#custom_R_code" class="anchor"></a>Adding custom R code into metadata.yml</h2>
<p>Occasionally all the changes we want to make to dataset may not fit into the prescribed workflow used in AusTraits. For example, we assume each trait has a single unit. But there are a few datasets where data on different rows have different units. So we want to make to make some custom modifications to this particular dataset before the common pipeline of operations gets applied. To make this possible, the workflow allows for some custom R code to be run as a first step in the processing pipeline. That pipeline (in the function <a href="https://github.com/traitecoevo/austraits/blob/master/R/steps.R#L59"><code>read_data_study</code></a>) looks like:</p>
<div class="sourceCode" id="cb116"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb116-1" title="1">  data &lt;-<span class="st"> </span></a>
<a class="sourceLine" id="cb116-2" title="2"><span class="st">    </span><span class="kw">read_csv</span>(filename_data_raw, <span class="dt">col_types =</span> <span class="kw">cols</span>()) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb116-3" title="3"><span class="st">    </span><span class="kw">custom_manipulation</span>(metadata[[<span class="st">"config"</span>]][[<span class="st">"custom_R_code"</span>]])() <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb116-4" title="4"><span class="st">    </span><span class="kw">parse_data</span>(dataset_id, metadata) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb116-5" title="5"><span class="st">    </span>...</a></code></pre></div>
<p>Note the second line.</p>
<div id="example-problem" class="section level3">
<h3 class="hasAnchor">
<a href="#example-problem" class="anchor"></a>Example problem</h3>
<p>As an example, <code>Barlow_1981</code> has multiple units per traits. Check it out:</p>
<div class="sourceCode" id="cb117"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb117-1" title="1"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(readr)</a>
<a class="sourceLine" id="cb117-2" title="2"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(yaml)</a></code></pre></div>
<pre><code>## 
## Attaching package: 'yaml'</code></pre>
<pre><code>## The following objects are masked _by_ '.GlobalEnv':
## 
##     read_yaml, write_yaml</code></pre>
<p>Load the data</p>
<div class="sourceCode" id="cb120"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb120-1" title="1">key &lt;-<span class="st"> "Barlow_1981"</span></a>
<a class="sourceLine" id="cb120-2" title="2">data &lt;-<span class="st"> </span><span class="kw"><a href="https://readr.tidyverse.org/reference/read_delim.html">read_csv</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span>(<span class="st">"data"</span>, key, <span class="st">"data.csv"</span>), <span class="dt">col_types =</span> <span class="kw"><a href="https://readr.tidyverse.org/reference/cols.html">cols</a></span>(<span class="dt">.default =</span> <span class="st">"c"</span>))</a>
<a class="sourceLine" id="cb120-3" title="3">metadata &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/yaml/man/read_yaml.html">read_yaml</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span>(<span class="st">"data"</span>, key, <span class="st">"metadata.yml"</span>))</a></code></pre></div>
<p>Here’s the problem - note tat several traits have multiple units used:</p>
<div class="sourceCode" id="cb121"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb121-1" title="1"><span class="kw"><a href="https://rdrr.io/r/base/table.html">table</a></span>(data<span class="op">$</span>trait, data<span class="op">$</span>units)</a></code></pre></div>
<pre><code>##                       
##                          cm    m   mm
##   dispersal               0    0    0
##   flowering time end      0    0    0
##   flowering time start    0    0    0
##   leaf length maximum  2570    7  836
##   leaf length minimum  2489    7  815
##   leaf type               0    0    0
##   leaf width maximum   1050    0 1986
##   leaf width minimum   1005    0 1974
##   lifeform                0    0    0
##   longevity               0    0    0
##   plant height maximum  831 2371    6
##   plant height minimum  200  862    0
##   seed breadth maximum    5    0   72
##   seed breadth minimum    5    0   72
##   seed length maximum    33    0  718
##   seed length minimum    34    0  733
##   seed shape              0    0    0
##   seed width maximum     22    0  404
##   seed width minimum     22    0  407</code></pre>
<p>So you want to write some R code that fixes this and gets the dataset into the processing pipeline, satisfying all the assumptions.</p>
</div>
<div id="developing-solutions" class="section level3">
<h3 class="hasAnchor">
<a href="#developing-solutions" class="anchor"></a>Developing solutions</h3>
<p>We want to write some custom R code that will appear in the <code>metadata.yml</code> file for that study, under a title <code>config</code> -&gt; <code>custom_R_code</code>. E.g. see this example for <a href="https://github.com/traitecoevo/austraits/blob/master/data/Barlow_1981/metadata.yml">data/Barlow_1981/metadata.yml</a>.</p>
<p>Your code should assume a single object called <code>data</code>. And apply whatever fixes are needed. Also it should be</p>
<ul>
<li>fully self contained (we’re not going to use any of the other remake machinery here)</li>
<li>have semi colons <code>;</code> at the end of each line. This will be needed because we’re adding the code to the <code>metadata.yml</code> file and newlines get lost when reading in the file.</li>
</ul>
<p>The workflow is to first develop some code that applies a suitable fix. E.g. for Barlow_1981 here is the code we eventually applied to the object <code>data</code> loaded above:</p>
<div class="sourceCode" id="cb123"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb123-1" title="1">metadata[[<span class="st">"config"</span>]][[<span class="st">"custom_R_code"</span>]] <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span>(<span class="st">";"</span>, <span class="st">";</span><span class="ch">\n</span><span class="st">"</span>, .) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/writeLines.html">writeLines</a></span>()</a></code></pre></div>
<p>Running this removes the problem with multiple units:</p>
<div class="sourceCode" id="cb124"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb124-1" title="1"><span class="kw"><a href="https://rdrr.io/r/base/table.html">table</a></span>(data<span class="op">$</span>trait, data<span class="op">$</span>units)</a></code></pre></div>
<pre><code>##                       
##                          cm    m   mm
##   dispersal               0    0    0
##   flowering time end      0    0    0
##   flowering time start    0    0    0
##   leaf length maximum  2570    7  836
##   leaf length minimum  2489    7  815
##   leaf type               0    0    0
##   leaf width maximum   1050    0 1986
##   leaf width minimum   1005    0 1974
##   lifeform                0    0    0
##   longevity               0    0    0
##   plant height maximum  831 2371    6
##   plant height minimum  200  862    0
##   seed breadth maximum    5    0   72
##   seed breadth minimum    5    0   72
##   seed length maximum    33    0  718
##   seed length minimum    34    0  733
##   seed shape              0    0    0
##   seed width maximum     22    0  404
##   seed width minimum     22    0  407</code></pre>
<p>(we can ignore the NULLs here, these are when data and units are both NA. Those get pruged furtehr down the pipeline).</p>
<p>Once you have some working code, you then want to add it into your yml file under a group <code>config</code> -&gt; <code>custom_R_code</code>.</p>
<p>And then check it works.</p>
<p>Let’s assume you added it in, so we’ll load the metadata (and also reload the data)</p>
<p>In the build process we use the function <code>custom_manipulation</code> to create a function that accepts a data frame and modifies it according to the code in <code>txt</code></p>
<div class="sourceCode" id="cb126"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb126-1" title="1">custom_manipulation</a></code></pre></div>
<pre><code>function (txt) 
{
    if (!is.null(txt) &amp;&amp; !is.na(txt) &amp;&amp; nchar(txt) &gt; 0) {
        function(data) {
            eval(parse(text = txt), env = new.env())
        }
    }
    else {
        identity
    }
}</code></pre>
<p>So now lets use it to create a function</p>
<div class="sourceCode" id="cb128"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb128-1" title="1">f &lt;-<span class="st"> </span><span class="kw">custom_manipulation</span>(metadata[[<span class="st">"config"</span>]][[<span class="st">"custom_R_code"</span>]])</a>
<a class="sourceLine" id="cb128-2" title="2">f</a></code></pre></div>
<pre><code>## function (data) 
## {
##     eval(parse(text = txt), env = new.env())
## }
## &lt;environment: 0x7f890db759d8&gt;</code></pre>
<p>And finally we can apply the function to our data:</p>
<div class="sourceCode" id="cb130"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb130-1" title="1">data2 &lt;-<span class="st"> </span><span class="kw">f</span>(data)</a></code></pre></div>
<p>(If it fails at this point it won’t work in the build).</p>
<p>Now let’s compare it to our original data (the columns units and values should now differ)</p>
<div class="sourceCode" id="cb131"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb131-1" title="1"><span class="kw"><a href="https://rdrr.io/r/base/all.equal.html">all.equal</a></span>(data, data2)</a></code></pre></div>
<pre><code>## [1] "Different number of rows"</code></pre>
<p>And also see the units and traits:</p>
<div class="sourceCode" id="cb133"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb133-1" title="1"><span class="kw"><a href="https://rdrr.io/r/base/table.html">table</a></span>(data2<span class="op">$</span>trait, data2<span class="op">$</span>units)</a></code></pre></div>
<pre><code>##                       
##                          mm
##   dispersal               0
##   flowering_time          0
##   leaf length maximum  3413
##   leaf length minimum  3311
##   leaf type               0
##   leaf width maximum   3036
##   leaf width minimum   2979
##   leaf_phenology          0
##   leaf_type               0
##   life_history            0
##   lifeform                0
##   plant height maximum 3208
##   plant height minimum 1062
##   seed breadth maximum   77
##   seed breadth minimum   77
##   seed length maximum   751
##   seed length minimum   767
##   seed shape              0
##   seed width maximum    426
##   seed width minimum    429</code></pre>
<p>Finally, check it works in the context of loading the metadata:</p>
<div class="sourceCode" id="cb135"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb135-1" title="1">data2 &lt;-<span class="st"> </span><span class="kw">custom_manipulation</span>(metadata[[<span class="st">"config"</span>]][[<span class="st">"custom_R_code"</span>]])(data)</a>
<a class="sourceLine" id="cb135-2" title="2"><span class="kw"><a href="https://rdrr.io/r/base/all.equal.html">all.equal</a></span>(data, data2)</a></code></pre></div>
<pre><code>## [1] "Different number of rows"</code></pre>
<p>Now you’re ready to go.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#overview">Overview</a></li>
      <li>
<a href="#mission">Mission</a><ul class="nav nav-pills nav-stacked">
<li><a href="#license">License</a></li>
      <li><a href="#access">Access</a></li>
      <li><a href="#citations">Citation</a></li>
      <li><a href="#transparency">Transparency</a></li>
      <li><a href="#contributing">Contributing</a></li>
      <li><a href="#loading-data">Loading data</a></li>
      </ul>
</li>
      <li>
<a href="#structure">Structure of AusTraits</a><ul class="nav nav-pills nav-stacked">
<li><a href="#elements">Elements of AusTraits</a></li>
      <li><a href="#observation-ids">Observation IDs</a></li>
      <li><a href="#value_types">Values and Value types</a></li>
      <li><a href="#taxonomy">Species taxonomy</a></li>
      <li><a href="#sources-1">Sources</a></li>
      <li><a href="#trait_defs">Trait definitions</a></li>
      </ul>
</li>
      <li>
<a href="#building-and-contributing-to-austraits">Building and contributing to AusTraits</a><ul class="nav nav-pills nav-stacked">
<li><a href="#approach">Approach</a></li>
      <li><a href="#structure2">Structure of raw data files for AusTraits</a></li>
      <li><a href="#building">Rebuilding from scratch</a></li>
      </ul>
</li>
      <li>
<a href="#format">Format of raw data files for AusTraits</a><ul class="nav nav-pills nav-stacked">
<li><a href="#building2">Building &amp; rebuilding AusTraits</a></li>
      <li><a href="#adding">Adding data to AusTraits</a></li>
      <li><a href="#substitutions-1">Substitutions</a></li>
      <li><a href="#taxonomic-updates-2">Taxonomic Updates</a></li>
      <li><a href="#running_tests">Tests</a></li>
      <li><a href="#reports-quality-checks">Reports / quality checks</a></li>
      <li><a href="#pushing-to-github">Pushing to GitHub</a></li>
      </ul>
</li>
      <li>
<a href="#usage-examples">Usage examples</a><ul class="nav nav-pills nav-stacked">
<li><a href="#taking-subsets">Taking subsets</a></li>
      <li><a href="#extracting-a-particular-trait-or-study">Extracting a particular trait or study</a></li>
      <li><a href="#merging-with-taxonomic-data">Merging with taxonomic data</a></li>
      <li><a href="#merging-with-location-data">Merging with location data</a></li>
      <li><a href="#removing-suspected-duplicates">Removing suspected duplicates</a></li>
      <li><a href="#converting-from-long-to-wide-format">Converting from long to wide format</a></li>
      <li><a href="#categorical-vs-numerical-traits">Categorical vs numerical traits</a></li>
      </ul>
</li>
      <li>
<a href="#appendices">Appendices</a><ul class="nav nav-pills nav-stacked">
<li><a href="#acknowledgements">Acknowledgements</a></li>
      <li><a href="#primary-sources">Primary sources</a></li>
      <li><a href="#file-types">File types</a></li>
      <li><a href="#custom_R_code">Adding custom R code into metadata.yml</a></li>
      </ul>
</li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Daniel Falster.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
